---
"title: "SAS mutants time-course RNAseq analysis"
output: html_document
---
## revision history
* focused on using edgeRglm results
* using genotype correction ver1 (eliminating wrong genotypes): 2014 ASPB version
* using final reads count (081514 - )
* using variance stabilization transformed expression value (by Julin, Nov 2014) for SOM (122114 -)
* change format into Rmarkedown (122114 - )
* modified `r paste("on",date()) `
* move output file directory to prevent from reaching Bitbucket max volume (0601415)
* delete old Bitbucket repository and make new one (06??15)
* start working on SAS module network by multiple regression and move them to new Rmd file (111415)
* move mutant expression data transformation part for module network analysis into module network analysis scripts ("SAS_module_network_model6v2.Rmd") (040116)
* reorder of SOM clusters according to (a) upregulated, and (b) downregulated. (052316)
* change legend name in heatmaps (052616)
* use GOseq.overlapTable for all SOM cluster overlapTable (052616)
* needs to eliminate eliminate bsk5_1 (genotype15) and AT5G02540_1 (genotype 2) from the beggining?
* creating "SAS_mutants_RNAseq_analysis_document_v3.Rmd" (052616)
* replace "toc" in to "counts" (070516)
## To do
* remove all unused genoytpe (eg. NAM parents) informaiton (eg. conversion table (083016), DGE, etc)
* 
## outline of this scripts
* eliminating libraries with wrong genotypes
* eliminating outlier libraries
+ Use samples.nolow5.Rdata object for select libraries
* differentially expressed gene analysis
+ 1. DE (treatment, ie, shade responsiveness) at any time point within given genotype
+ 2. DE at any time point (shade)
* VST 
* SOM analysis with DE(1) genes (using shade-responsiveness (shade/sun))
* GOseq analysis of SOM clusters
* overlap between SOM clusters and misexpressed genes in mutants
* co-expression analysis by WGCNA (all Col samples)

## preparation
```{r load-data}
if(Sys.info()["user"]=="nozue") {
#homedir<-"/Volumes/Data8/NGS_related/Arabidopsis_analysis/Nozue2016_SAStranscriptome_finalfinal_newVST/Nozue2016_SAStranscriptome_scripts/SAS_muts_time_course_RNAseq_finalfinal"
# for temporary computer system (J's ex hard drive)
homedir<-"/Volumes/Data8new/Data8/NGS_related/Arabidopsis_analysis/Nozue2016_SAStranscriptome_finalfinal_newVST/Nozue2016_SAStranscriptome_scripts/SAS_muts_time_course_RNAseq_finalfinal"
} # customize homedir
# download "Nozue2016_SAStranscriptome_data" at the same level of "Nozue2016_SAStranscriptome_scripts"
# Atmosphere (Cyverse)
if(Sys.info()["user"]=="kazu") { 
homedir<-"/vol1/SAS_muts_time_course_RNAseq4/Nozue2016_SAStranscriptome_scripts/SAS_muts_time_course_RNAseq_finalfinal"
}
# Whitney
if(Sys.info()["nodename"]=="whitney") {
  homedir<-"~/NGS/Nozue2016_SAStranscriptome_finalfinal_newVST/Nozue2016_SAStranscriptome_scripts/SAS_muts_time_course_RNAseq_finalfinal"
} # use git repository
setwd(homedir)

# reading necessary funcitons and libraries
source("function_RNAseq_time_course.R")
```

## eliminating libraries with wrong genotypes
```{r eval=FALSE}
counts<-read.table("../../Nozue2016_SAStranscriptome_data/input/SAS_counts_merged_updated_final.bam.tsv",header=TRUE) # use SAS_counts_merged_updated.bam.tsv (081514)
# remove * row and move gene column into rowname
rownames(counts)<-counts$gene
counts<-counts[!counts$gene=="*",-1] 
#head(counts)
label<-names(counts)
# eliminating wrong libraries (see "samples" spread sheet in https://docs.google.com/spreadsheets/d/1-llHCbrDQfQhLXIlRsLhabN-ezBjwKvrdfjiNy36wCg/edit#gid=2102778022)
# updated 081514 based on final sequencing data 
eliminated.libraries<-c(
  "D2H16A.merged.bam", # AT5G02540_2
  "D4H4A.merged.bam",  # jar1-1
  "D4H16A.merged.bam", # jar1-1
  "D4H49A.merged.bam", # jar1-1
  "C6H1A.merged.bam", # phyB
  "C6H16A.merged.bam", # phyB  	
  "C7L4A.merged.bam", # pif4pi5
  "C7H16A.merged.bam", # pif4pif5	
  "C7H49A.merged.bam", # pif4pif5
  "C7L49A.merged.bam", # pif4pif5
  "C11H1A.merged.bam", # coi1_16
  "C11L1A.merged.bam", # coi1_16
  "C11H4A.merged.bam", # coi1_16
  "D11H4A.merged.bam", # coi1_16
  "C11L4A.merged.bam", # coi1_16
  "C11H16A.merged.bam", # coi1_16
  "C11L16A.merged.bam", # coi1_16
  "C11H25A.merged.bam", # coi1_16
  "C11L25A.merged.bam", # coi1_16
  "C11H49A.merged.bam", # coi1_16
  "D11H49A.merged.bam", # coi1_16
  "C11L49A.merged.bam", # coi1_16
  "C12H1A.merged.bam", # phyAB
  "C12L1A.merged.bam", # phyAB
  "D12L1A.merged.bam", # phyAB
  "C12H4A.merged.bam", # phyAB
  "C12L4A.merged.bam", # phyAB
  "C12H16A.merged.bam", # phyAB
  "C12L16A.merged.bam", # phyAB
  "C12L25A.merged.bam", # phyAB
  "C12H49A.merged.bam", # phyAB
  "C12L49A.merged.bam", # phyAB
  "C13H1A.merged.bam", # pif3_1
  "C13H16A.merged.bam", # pif3_1
  "E13H16A.merged.bam", # pif3_1
  "C13L16A.merged.bam", # pif3_1
  "E13L16A.merged.bam", # pif3_1
  "C13H25A.merged.bam", # pif3_1
  "C13L25A.merged.bam", # pif3_1
  "C13H49A.merged.bam", # pif3_1
  "D13H49A.merged.bam", # pif3_1
  "C13L49A.merged.bam", # pif3_1
  "D14L4A.merged.bam", # mida9_4
  "C14H25A.merged.bam", # mida9_4
  "D14L25A.merged.bam", # mida9_4
  "C16H1A.merged.bam", # sto
  "C16L1A.merged.bam", # sto
  "C16H4A.merged.bam", # sto
  "C16L4A.merged.bam", # sto
  "C16H16A.merged.bam", # sto
  "C16H25A.merged.bam", # sto
  "C16L25A.merged.bam", # sto
  "C16H49A.merged.bam", # sto
  "C16L49A.merged.bam", # sto
  "D16L49A.merged.bam") # sto
# eliminate 55 libraries (081514)
counts<-counts[,!names(counts) %in% eliminated.libraries]
#head(counts) 
dim(counts) #[1] 30533   749

samples <- data.frame(file=colnames(counts),
                      batch=factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)A+[[:print:]]+(.bam)","\\1",colnames(counts))),
                      trt=factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)A+[[:print:]]+(.bam)","\\3",colnames(counts))),
                      time=factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)A+[[:print:]]+(.bam)","\\4",colnames(counts))),
                      genotype=factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)A+[[:print:]]+(.bam)","\\2",colnames(counts)))
) # corrected
#
samples
counts.nolow <- counts[,colSums(counts,na.rm=TRUE) > 1000000]
samples.low<-samples[samples$file %in% names(counts)[colSums(counts,na.rm=TRUE) < 1000000],]
samples.low$counts<-colSums(counts,na.rm=TRUE)[colSums(counts,na.rm=TRUE) < 1000000]
samples.low$genotype2<-""
for(i in 1:27){
  samples.low[samples.low$genotype==conversion.table[i,1],"genotype2"]<-as.character(conversion.table[i,2])  
} 
names(counts.nolow)
counts.nolow[is.na(counts.nolow)]<-0
save(counts.nolow,file="../../Nozue2016_SAStranscriptome_output/output/counts.nolow.Rdata")

```

* eliminating outlier libraries
```{r, eval=FALSE}
# in whitney. Initial DE analysis
load("../../Nozue2016_SAStranscriptome_output/output/counts.nolow.Rdata")
samples.nolow <- data.frame(file=colnames(counts.nolow),
                            batch=factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)A+[[:print:]]+(.bam)","\\1",colnames(counts.nolow))),
                            trt=factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)A+[[:print:]]+(.bam)","\\3",colnames(counts.nolow))),
                            time=factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)A+[[:print:]]+(.bam)","\\4",colnames(counts.nolow))),
                            genotype=factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)A+[[:print:]]+(.bam)","\\2",colnames(counts.nolow)))
) # corrected
summary(counts.nolow)
Genotype<-levels(samples.nolow$genotype)
# Genotype = "11" and "16" had an error:
Genotype<-Genotype[!Genotype %in% c("11","16")]

## normalizing 
samples.nolow$group <- paste(samples.nolow$genotype,samples.nolow$trt,samples.nolow$time,sep=".") 
samples.nolow$genotype<-as.character(samples.nolow$genotype)
dge.nolow <- DGEList(counts=counts.nolow,group=samples.nolow$group)  
dge.nolow<-calcNormFactors(dge.nolow)

for(i in Genotype) { # error in i=11,16
  print(paste("Genotype is ",i))
  # subset genotype of interests from DGEList (with normalized). see ablove (081214)
  samples.nolow.s<-samples.nolow[samples.nolow$genotype==i,]
  dge.nolow.s<-dge.nolow[,samples.nolow$genotype==i]
  # calculating cpm
  temp<-cpm(dge.nolow.s)
  #save(temp,file=paste("../../Nozue2016_SAStranscriptome_output/output/genotype",i,"cpm.Rdata",sep=".")) # better to have mean value at each time point
  #save(temp,file=paste("../../Nozue2016_SAStranscriptome_output/output/genotype",i,"cpm.Rdata",sep="."))
  
  # eliminating low expressed genes in dge.nolow.s
  dge.nolow.s<-dge.nolow.s[rowSums(dge.nolow.s$counts > 5) >= 3,]
  
  design.nolow.s.batch<- model.matrix(~samples.nolow.s$time*samples.nolow.s$trt + samples.nolow.s$batch)
  colnames(design.nolow.s.batch) <- gsub("samples.nolow.s$","",colnames(design.nolow.s.batch),fixed=TRUE) # nicer column names
  colnames(design.nolow.s.batch)
  print(design.nolow.s.batch)
  dge.nolow.s.batch.glm <- estimateGLMCommonDisp(dge.nolow.s,design.nolow.s.batch, verbose=TRUE) 
  dge.nolow.s.batch.glm <- estimateGLMTrendedDisp(dge.nolow.s.batch.glm,design.nolow.s.batch, verbose=TRUE)
  dge.nolow.s.batch.glm <- estimateGLMTagwiseDisp(dge.nolow.s.batch.glm,design.nolow.s.batch) # why this is not included in original Julin's script?
  pdf(file=paste("../../Nozue2016_SAStranscriptome_output/output/BCV_MDS_with_genotype",i,".pdf",sep=""),width=11,height=8)
  par(mfrow=c(1,2))
  plotBCV(dge.nolow.s.batch.glm,main=paste("genotype",i))
  plotMDS(dge.nolow.s.batch.glm,main=paste("genotype",i))
  dev.off()
  
  dge.nolow.s.batch.glm.fit <- glmFit(dge.nolow.s.batch.glm,design.nolow.s.batch)
  dge.nolow.s.batch.glm.lrt <- glmLRT(dge.nolow.s.batch.glm.fit,coef=c("trtL","time4:trtL","time16:trtL","time25:trtL","time49:trtL"))
  DE.batch.anytime<-topTags(dge.nolow.s.batch.glm.lrt,n=2000)$table[topTags(dge.nolow.s.batch.glm.lrt,n=2000)$table$FDR<0.05,]
  dim(DE.batch.anytime) # 416 (Col, w/o eliminating low expressed genes). 518 (Col, after eliminating low expressed genes)
  save(DE.batch.anytime,file=paste("../../Nozue2016_SAStranscriptome_output/output/DE.batch.anytime_genotype",i,".Rdata",sep=""))    
}
# eliminate 16h and find outlier
dge.nolow4<-dge.nolow[,!samples.nolow$time==16]
samples.nolow4<-samples.nolow[!samples.nolow$time==16,]
samples.nolow4$time<-as.character(samples.nolow4$time)
# starting finalfinal scripts from samples.nolow4? (092116)
save(samples.nolow4,file="../../Nozue2016_SAStranscriptome_output/output/samples.nolow4.Rdata")
load("../../Nozue2016_SAStranscriptome_output/output/samples.nolow4.Rdata")

for(i in Genotype) { # error in i=11,16 (they has been eliminated in Genotype)
  print(paste("Genotype is ",i))
  # subset genotype of interests from DGEList (with normalized) ?? see ablove (081214)
  samples.nolow.s<-samples.nolow4[samples.nolow4$genotype==i,]
  dge.nolow.s<-dge.nolow4[,samples.nolow4$genotype==i]
  
  # eliminating low expressed genes in dge.nolow.s
  dge.nolow.s<-dge.nolow.s[rowSums(dge.nolow.s$counts > 5) >= 3,]
  
  design.nolow.s.batch<- model.matrix(~samples.nolow.s$time*samples.nolow.s$trt + samples.nolow.s$batch)
  colnames(design.nolow.s.batch) <- gsub("samples.nolow.s$","",colnames(design.nolow.s.batch),fixed=TRUE) # nicer column names
  colnames(design.nolow.s.batch)
  print(design.nolow.s.batch)
  dge.nolow.s.batch.glm <- estimateGLMCommonDisp(dge.nolow.s,design.nolow.s.batch, verbose=TRUE) 
  dge.nolow.s.batch.glm <- estimateGLMTrendedDisp(dge.nolow.s.batch.glm,design.nolow.s.batch, verbose=TRUE)
  dge.nolow.s.batch.glm <- estimateGLMTagwiseDisp(dge.nolow.s.batch.glm,design.nolow.s.batch) # why this is not included in original Julin's script?
  pdf(file=paste("../../Nozue2016_SAStranscriptome_output/output/BCV_MDS_with_genotype",i,".pdf",sep=""),width=11,height=8)
  par(mfrow=c(1,2))
  plotBCV(dge.nolow.s.batch.glm,main=paste("genotype",i))
  plotMDS(dge.nolow.s.batch.glm,main=paste("genotype",i))
  dev.off()
  
  dge.nolow.s.batch.glm.fit <- glmFit(dge.nolow.s.batch.glm,design.nolow.s.batch)
  dge.nolow.s.batch.glm.lrt <- glmLRT(dge.nolow.s.batch.glm.fit,coef=c("trtL","time4:trtL","time25:trtL","time49:trtL"))
  
  DE.batch.anytime<-topTags(dge.nolow.s.batch.glm.lrt,n=2000)$table[topTags(dge.nolow.s.batch.glm.lrt,n=2000)$table$FDR<0.05,]
  dim(DE.batch.anytime) # 416 (Col, w/o eliminating low expressed genes). 518 (Col, after eliminating low expressed genes)
  save(DE.batch.anytime,file=paste("../../Nozue2016_SAStranscriptome_output/output/DE.batch.anytime_genotype_wo16h",i,".Rdata",sep=""))  	
}

## check MDS plot and find outliner libraries (081514)
# eliminate following libraries due to outlier after looking MDS plot
# Col (genotype = 1) # 90 DE
Col_outlier<-"C1L4A.merged.bam" # after elimination; 80 DE
# PAR1_RNAi09 (genotype = 10) only 7 DE
PAR1_RNAi09_outlier<-c("E10H4A.merged.bam","D10H1A.merged.bam") # 6 DE
# coi1_16 (11) only 1 DE (did not work with edgeR possibly due to lots of library are eliminated by genotyping)
#phyAB, (12), 105 DE 
phyAB_outlier<-"C12H25A.merged.bam" #>>>  
#pif3 (13) 73 DE
# AT5G02760_2 (mida9_4) (genotype = 14) original DE (dim(DE.batch.anytime), 10 8
# AT5G59010_2 (15) # 46 DE (no need to eliminate)
# sto (genotype = 16) (eliminated entire sto libraries)
# aos (17) #52 (no need to eliminate)
# 18 (genotyping showed this is not mutant. Eliminate this genotype)
# argos_outlier<-"D18L1A.merged.bam" #11 DE
#co_9 (genotype=19) 20 DE 
# AT5G02540_1 (genotype = 2)
AT5G02540_1_outlier<-"D2H4A.merged.bam" # 23 DE
# Blh_1 (genotype = 20) DE 4
# ? (genotype = 21) DE 20 (no need to eliminate libraries)
#Shahdara (genotype =22) # DE 2
Shahdara_outlier<-"C22H25A.merged.bam" # 3 DE
# Col_0 (23) # 59
Col_0_outlier<-"E23H49A.merged.bam" # 58 DE
# Cvi_0 (24) # 43
# Bur_0 (25) # 53
Bur_0_outlier<-"E25H25A.merged.bam" # ** DE
# Oy_0 (26) , 2 DE 
Oy_0_outlier<-"E26L1A.merged.bam" # only * DE
# (27) only 3 DE genes, but MDS plot looks good 
# hy5, (genotype 3) only 0 DE genes (MDS plot looks good) probably no response to shade?
# jar1 (genotype 4) 30 DE (MDS plot looks good)
# kat1_2 (5) 117 DE (no need to eliminate)
# phyB (6) 25 DE
# pif4 pif5 (7) only 6 DE genes (MDS plot looks OK)
# spt_11 (8) 81 DE >> 
#yuc2589 (9), only 10 DE genes but MDS plot looks OK. >> after outliner elimination, ** DE
# combine outlier libraries
outlier<-c(
  Col_outlier,
  PAR1_RNAi09_outlier,
  phyAB_outlier,
  AT5G02540_1_outlier,
  Shahdara_outlier, 
  Col_0_outlier,
  Bur_0_outlier,
  Oy_0_outlier
)
# eliminate outlier libraries
system("mkdir ../../Nozue2016_SAStranscriptome_output/output/edgeR")
system("mkdir ../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h")
samples.nolow5<-samples.nolow4[!samples.nolow4$file %in% outlier,]
dge.nolow5<-dge.nolow4[,!samples.nolow4$file %in% outlier]
Genotype<-levels(as.factor(samples.nolow5$genotype))
Genotype<-Genotype[!(Genotype=="11"|Genotype=="12"|Genotype=="18"|Genotype=="16")]
save(dge.nolow5,file="../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/dge.nolow5.Rdata")
save(samples.nolow5,file="../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/samples.nolow5.Rdata")
```
## differentially expressed gene analysis
1.DE (treatment, ie, shade responsiveness) at any time point within given genotype
```{r eval=FALSE}
load("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/samples.nolow5.Rdata")
load("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/dge.nolow5.Rdata")
# needs to fix Genotype part (083016)
# Genotype<-levels(as.factor(samples.nolow5$genotype))
# Genotype<-Genotype[!(Genotype=="11"|Genotype=="12"|Genotype=="18"|Genotype=="16")]
SAS.expression.vst<-read.csv("../../Nozue2016_SAStranscriptome_data/input/kazu.SAS.expression.VST.082016.csv") # new transformed data by DEseq2 (082016)
rownames(SAS.expression.vst)<-SAS.expression.vst$X
Genotype<-levels(as.factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|25|49)(A.merged.bam)","\\2",names(SAS.expression.vst)[-1])))
#
for(i in Genotype) { # why I eliminate genotype 1 and 10???
  print(paste("Genotype is ",i))
  samples.nolow.s<-samples.nolow5[samples.nolow5$genotype==i,]
  dge.nolow.s<-dge.nolow5[,samples.nolow5$genotype==i]
#   # calculating cpm
#   temp<-cpm(dge.nolow.s)
#   save(temp,file=paste("../../Nozue2016_SAStranscriptome_output/output/genotype",i,"cpm.Rdata",sep=".")) # better to have mean value at each time point  
  # eliminating low expressed genes in dge.nolow.s
  dge.nolow.s<-dge.nolow.s[rowSums(dge.nolow.s$counts > 5) >= 3,]  
  design.nolow.s.batch<- model.matrix(~samples.nolow.s$time*samples.nolow.s$trt + samples.nolow.s$batch)
  colnames(design.nolow.s.batch) <- gsub("samples.nolow.s$","",colnames(design.nolow.s.batch),fixed=TRUE) # nicer column names
  colnames(design.nolow.s.batch)
  print(design.nolow.s.batch)
  dge.nolow.s.batch.glm <- estimateGLMCommonDisp(dge.nolow.s,design.nolow.s.batch, verbose=TRUE) 
  dge.nolow.s.batch.glm <- estimateGLMTrendedDisp(dge.nolow.s.batch.glm,design.nolow.s.batch, verbose=TRUE)
  dge.nolow.s.batch.glm <- estimateGLMTagwiseDisp(dge.nolow.s.batch.glm,design.nolow.s.batch) # why this is not included in original Julin's script?
  #pdf(file=paste("BCV_MDS_with_genotype_w_outlier_wo16h",i,".pdf",sep=""),width=11,height=8)
  pdf(file=paste("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/BCV_MDS_with_genotype_wo_outlier_wo16h",i,".pdf",sep=""),width=11,height=8) # fine name fixed (122214)
  par(mfrow=c(1,2))
  plotBCV(dge.nolow.s.batch.glm,main=paste("genotype",i))
  plotMDS(dge.nolow.s.batch.glm,main=paste("genotype",i))
  dev.off() 
  dge.nolow.s.batch.glm.fit <- glmFit(dge.nolow.s.batch.glm,design.nolow.s.batch)
  dge.nolow.s.batch.glm.lrt <- glmLRT(dge.nolow.s.batch.glm.fit,coef=c("trtL","time4:trtL","time25:trtL","time49:trtL"))
  DE.batch.anytime<-topTags(dge.nolow.s.batch.glm.lrt,n=2000)$table[topTags(dge.nolow.s.batch.glm.lrt,n=2000)$table$FDR<0.05,]
  dim(DE.batch.anytime) # 
  #save(DE.batch.anytime,file=paste("DE.batch.anytime_genotype_w_outlier_wo16h",i,".Rdata",sep=""))  
  save(DE.batch.anytime,file=paste("../../Nozue2016_SAStranscriptome_output/output/DE.batch.anytime_genotype_wo_outlier_wo16h",i,".Rdata",sep="")) # file name fixed (122114)    
}
```

# summary of DE genes 
* (053014 - 053114): revised version (081514 - );more revision (060415 - )
```{r eval=FALSE}
DEfiles<-list.files(path="../../Nozue2016_SAStranscriptome_output/output/",pattern="DE.batch.anytime")
DEfiles<-grep("DE.batch.anytime_genotype_wo_outlier_wo16h",DEfiles,value=TRUE)
setwd("../../Nozue2016_SAStranscriptome_output/output/")
load(DEfiles[1]) 
rownames(DE.batch.anytime)<-paste(rownames(DE.batch.anytime),gsub("(DE.batch.anytime_genotype_wo_outlier_wo16h)([[:print:]]+)(.Rdata)","\\2",DEfiles[1]),sep=".")
temp<-DE.batch.anytime

for(i in DEfiles[-1]) {
  load(i) 
  if(dim(DE.batch.anytime)[1]>1) {
    rownames(DE.batch.anytime)<-paste(rownames(DE.batch.anytime),gsub("(DE.batch.anytime_genotype_w_outlier_wo16h)([[:print:]]+)(.Rdata)","\\2",i),sep=".")
    temp<-rbind(temp,DE.batch.anytime)  }
  else next
}
DGEsummary.edgeR.response<-temp
tail(DGEsummary.edgeR.response)
dim(DGEsummary.edgeR.response) #755   8 for final sequence data with normalization (081514); 669 8 , why??? in edgeR DE loop I used "for(i in Genotype[-(1:2)]) {", but this should be "for(i in Genotype) {". After fixing this I got "755 8" . oK. (060515)

save(DGEsummary.edgeR.response,file="../../Nozue2016_SAStranscriptome_output/output/DGEsummary.edgeR.response.Rdata")
# # genotype conversion 
load("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/samples.nolow5.Rdata")
samples.nolow5
samples.nolow5$genotype2<-""
for(i in 1:13){ # only for Genotype
  samples.nolow5[samples.nolow5$genotype==conversion.table[i,1],"genotype2"]<-as.character(conversion.table[i,2])	
} 
## summary of DE gene numbers
load(DEfiles[1]) #"DE.batch.anytime_genotype_wo_outlier_wo16h1.Rdata"
DEgenes.summary<-data.frame(genotype=gsub("(DE.batch.anytime_genotype_wo_outlier_wo16h)([[:digit:]]+)(.Rdata)","\\2",DEfiles))
for(i in 1:length(DEfiles)) {
  load(DEfiles[i])
  DEgenes.summary$DEgenes_num[i]<-dim(DE.batch.anytime)[1]
}
DEgenes.summary$genotype2<-""
for(i in 1:13) {
  DEgenes.summary[DEgenes.summary$genotype==conversion.table[i,1],"genotype2"]<-as.character(conversion.table[i,2])
}
DEgenes.summary<-DEgenes.summary[!DEgenes.summary$genotype2==""|!is.na(DEgenes.summary$genotype2==""),]
DEgenes.summary # summary

setwd(homedir)
```
## select genes with largeCV and calculate responsiveness for SOM analysis
```{r eval=FALSE}
setwd(homedir)
# read vst transformed data
#SAS.expression.vst<-read.csv("../../Nozue2016_SAStranscriptome_data/input/kazu.SAS.expression.VST.070516.csv") # new transformed data by DEseq2 (070416)
SAS.expression.vst<-read.csv("../../Nozue2016_SAStranscriptome_data/input/kazu.SAS.expression.VST.082016.csv") # new transformed data by DEseq2 (082016)

rownames(SAS.expression.vst)<-SAS.expression.vst$X
# calculate mean value among replicates
load("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/samples.nolow5.Rdata")
samples.nolow5$sample<-with(samples.nolow5,paste(genotype,trt,time,sep=""))
samples.nolow5$sample<-factor(samples.nolow5$sample) #where this is used? (122214)
SAS.expression.vst.s<-SAS.expression.vst[,names(SAS.expression.vst) %in% samples.nolow5$file] 

Genotype<-levels(as.factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|25|49)(A.merged.bam)","\\2",names(SAS.expression.vst)[-1])))

temp<-SAS.expression.vst.s[,names(SAS.expression.vst.s) %in% samples.nolow5[samples.nolow5$genotype==1,"file"]]
summary.vst<-expression.mean(temp)ã€€# custom function

for(i in Genotype[-1]) {
  temp<-SAS.expression.vst.s[,names(SAS.expression.vst.s) %in% samples.nolow5[samples.nolow5$genotype==i,"file"]]
	temp<-as.data.frame(temp)
	temp2<-expression.mean(temp)
	summary.vst<-cbind(summary.vst,temp2)
}
#head(summary.vst,3)
rownames(summary.vst)<-gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",SAS.expression.vst$X)
save(summary.vst,file="../../Nozue2016_SAStranscriptome_output/output/summary.vst.Rdata") #
# relative value 
load("../../Nozue2016_SAStranscriptome_output/output/summary.vst.Rdata")
# note that summary.vst is log2 transformed
summary.vst.response<-as.data.frame((2^summary.vst[,gsub("(.+)(H|L)(1|4|16|25|49)","\\2",names(summary.vst))=="L"])/(2^summary.vst[,gsub("(.+)(H|L)(1|4|16|25|49)","\\2",names(summary.vst))=="H"])) # shade/sun 
rownames(summary.vst.response)<-rownames(summary.vst)
names(summary.vst.response)<-paste(gsub("(.+)(H|L)(1|4|16|25|49)","\\1_\\3",names(summary.vst.response)),"hrA",sep="") #
save(summary.vst.response,file="../../Nozue2016_SAStranscriptome_output/output/summary.vst.response.Rdata")
# select only Kazu's samples (eliminating Palmer's NAM parents)
SAS.expression.vst.s.kazu<-SAS.expression.vst.s[,gsub("([C|D|E])([[:digit:]]+)(H|L)(1|4|16|25|49)(A.merged.bam)","\\2",colnames(SAS.expression.vst.s)) %in% 1:19]
dim(SAS.expression.vst.s.kazu) # 18540   402
# select genes with higher CV
co.var.df <- function(x) ( 100*apply(x,1,sd)/rowMeans(x) )
SAS.expression.vst.s.kazu$cv <- co.var.df(SAS.expression.vst.s.kazu[,-1])
hist(SAS.expression.vst.s.kazu$cv)
sum(as.integer(SAS.expression.vst.s.kazu$cv>3))/dim(SAS.expression.vst.s.kazu)[1] #[1] 0.4011866 > 0.6019389
SAS.expression.vst.s.kazu.largeCV<-SAS.expression.vst.s.kazu[SAS.expression.vst.s.kazu$cv>3,]
dim(SAS.expression.vst.s.kazu.largeCV) #[1] 7438  403 > 18379   264
save(SAS.expression.vst.s.kazu.largeCV,file="../../Nozue2016_SAStranscriptome_output/output/SAS.expression.vst.s.kazu.largeCV.Rdata")

# response value of genes with largeCV 
summary.vst.response.kazu<-summary.vst.response
dim(summary.vst.response.kazu)
summary.vst.response.kazu.largeCV<-summary.vst.response.kazu[rownames(summary.vst.response.kazu) %in% gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",rownames(SAS.expression.vst.s.kazu.largeCV)),]

dim(summary.vst.response.kazu.largeCV) # 7438   60 > 18379    48
save(summary.vst.response.kazu.largeCV,file="../../Nozue2016_SAStranscriptome_output/output/summary.vst.response.kazu.largeCV.Rdata")
```
## select only DE genes
```{r eval=TRUE}
load("../../Nozue2016_SAStranscriptome_output/output/summary.vst.response.kazu.largeCV.Rdata")
summary.vst.response.kazu.largeCV[summary.vst.response.kazu.largeCV=="Inf"]<-0 # arbitral
summary.vst.response.kazu.largeCV[summary.vst.response.kazu.largeCV=="-Inf"]<-0 # arbitral
summary.vst.response.kazu.largeCV[is.na(summary.vst.response.kazu.largeCV)]<-0 # arbitral
load("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/DGEsummary.edgeR.response.Rdata") #"AT1G33925.1"
#DGEgenelist.edgeRglm<-unique(gsub("([[:print:]]+)(.[[:digit:]]+)(.DE.batch.anytime)([[:print:]]+)(.Rdata)","\\1",rownames(DGEsummary.edgeR.response))) # wait! this is wrong? (052916)
DGEgenelist.edgeRglm<-unique(gsub("([[:print:]]+)(.)([[:digit:]]+)(.DE.batch.anytime_genotype_wo_outlier_wo16h)([[:print:]]+)","\\1",rownames(DGEsummary.edgeR.response))) # 

length(DGEgenelist.edgeRglm) #369 > 373

# 
DGE.summary.vst.response.kazu.largeCV<-summary.vst.response.kazu.largeCV[rownames(summary.vst.response.kazu.largeCV) %in% DGEgenelist.edgeRglm,]
save(DGE.summary.vst.response.kazu.largeCV,file="../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/DGE.summary.vst.response.kazu.largeCV.Rdata") # updated 052916
DGE.summary.vst.response.kazu.largeCV[rownames(DGE.summary.vst.response.kazu.largeCV)=="AT4G16780",] #ATHB2 has to be induced in genotype 1 (= Col)
 
```
## SOM analysis (with Yasu's help)
```{r eval=TRUE}
library(ggplot2);library(class);library(MASS);library(kohonen);library(reshape2);library(plyr);
load("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/DGE.summary.vst.response.kazu.largeCV.Rdata") # this was updated... (052916; 082016)
genotypes<-levels(as.factor(gsub("(.+)(_)(1|4|16|25|49)(hrA)","\\1",as.character(names(DGE.summary.vst.response.kazu.largeCV))))) # 12? why not 13?? "3" is missing. fixed. (090116)


DGE.summary.vst.response.kazu.largeCV<-log2(DGE.summary.vst.response.kazu.largeCV) # log2 transformation of fold change
i<-1
temp<-DGE.summary.vst.response.kazu.largeCV[,gsub("(.+)(_)(1|4|16|25|49)(hrA)","\\1",names(DGE.summary.vst.response.kazu.largeCV))==genotypes[i]]
rownames(temp)<-paste(rownames(temp),genotypes[i],sep=".")
colnames(temp)<-paste(gsub("(.+)(_)(1|4|16|25|49)(hrA)","\\3",colnames(temp)),"h",sep="") ### fix (070914)
for(i in 2:length(genotypes)) {
  #temp2<-DGE.summary.vst.response.kazu.largeCV[,grep(genotypes[i],names(DGE.summary.vst.response.kazu.largeCV))]
  temp2<-DGE.summary.vst.response.kazu.largeCV[,gsub("(.+)(_)(1|4|16|25|49)(hrA)","\\1",names(DGE.summary.vst.response.kazu.largeCV))==genotypes[i]]
  rownames(temp2)<-paste(rownames(temp2),genotypes[i],sep=".")
  colnames(temp2)<-paste(gsub("(.+)(_)(1|4|16|25|49)(hrA)","\\3",colnames(temp2)),"h",sep="") ### fix (070914)  
  temp<-rbind(temp,temp2)
}
summary(temp)
head(temp)
data5<-temp
#### no scaling
## SOM
set.seed(2) # Set a random seed so that SOM results are reproducible
#set.seed(123) # I do not like to have cluster 5 mixed with constitutive and 4h peak in set.seed(2) 3x4 version
ssom1.7.noscaling <- som(as.matrix(data5), somgrid(1,7,"hexagonal")) 
plot(ssom1.7.noscaling, type = "codes") #
ssom2.4.noscaling <- som(as.matrix(data5), somgrid(2,4,"hexagonal")) 
plot(ssom2.4.noscaling, type = "codes") #
ssom3.3.noscaling <- som(as.matrix(data5), somgrid(3,3,"hexagonal")) 
plot(ssom3.3.noscaling,type="codes")
ssom2.5.noscaling <- som(as.matrix(data5), somgrid(2,5,"hexagonal")) 
plot(ssom2.5.noscaling, type = "codes") #
ssom3.4.noscaling <- som(as.matrix(data5), somgrid(3,4,"hexagonal")) 
plot(ssom3.4.noscaling) 
ssom4.4.noscaling <- som(as.matrix(data5), somgrid(4,4,"hexagonal")) 
plot(ssom4.4.noscaling) 

## Create and write-out master SOM file
newdata.vst<-cbind(data5,ssom3.4.noscaling$unit.classif,ssom3.4.noscaling$distances)
names(newdata.vst)[5]<-"ssom3.4.unit.classif"
names(newdata.vst)[6]<-"ssom3.4.unit.distances"
newdata.vst$gt<-as.factor(gsub("(AT[[:alnum:]]+)(\\.)([[:digit:]]+)","\\3",rownames(newdata.vst)))
newdata.vst$AGI<-as.factor(gsub("(AT[[:alnum:]]+)(\\.)([[:digit:]]+)","\\1",rownames(newdata.vst)))
head(newdata.vst)
## ATHB2
newdata.vst[rownames(newdata.vst)=="AT4G16780.1",]
save(newdata.vst,file="../../Nozue2016_SAStranscriptome_output/output/newdata.vst.Rdata")

#### making complete table (for SOM data) ###
load("../../Nozue2016_SAStranscriptome_output/output/newdata.vst.Rdata")

# replace gt num into gt name
### convert genotype number into actual genotype name
newdata.vst$gt2<-""
for(i in 1:13){ # new conversion table
  newdata.vst[newdata.vst$gt==conversion.table[i,1],"gt2"]<-as.character(conversion.table[i,2])
} 

# adding gene description
# TAIR10_gene_descriptions<-read.csv("../../Nozue2016_SAStranscriptome_data/input/TAIR10_functional_descriptions.csv") 
# head(TAIR10_gene_descriptions)
# TAIR10_gene_descriptions$AGI<-gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",TAIR10_gene_descriptions$"Model_name")
SOMclusters.description<-merge(newdata.vst[newdata.vst$gt==levels(newdata.vst$gt)[1],],TAIR10_gene_descriptions[!duplicated(TAIR10_gene_descriptions$AGI),c("AGI","Short_description","colE...col.F")],by="AGI")

for(x in levels(newdata.vst$gt)[-1]) {
temp<-merge(newdata.vst[newdata.vst$gt==x,],TAIR10_gene_descriptions[!duplicated(TAIR10_gene_descriptions$AGI),c("AGI","Short_description","colE...col.F")],by="AGI")
SOMclusters.description<-rbind(SOMclusters.description,temp)
}
write.csv(SOMclusters.description,"../../Nozue2016_SAStranscriptome_output/figs_tables/TableS3_SOMclusters.description.csv") #090116

```
## drawing bloxplot for each cluster
```{r eval=FALSE, ggplot, fig.width=11, fig.height=8}
load("../../Nozue2016_SAStranscriptome_output/output/newdata.vst.Rdata") # 122214, 090116
length(unique(newdata.vst$AGI)) # 261 genes
# boxplot
newdata.vstmelt<-melt(newdata.vst[,c("1h","4h","25h","49h","ssom3.4.unit.classif","gt","AGI")],id.var=c("gt","AGI","ssom3.4.unit.classif"))

names(newdata.vstmelt)[3]<-"ssom3.4.unit.classif"
newdata.vstmelt$ssom3.4.unit.classif<-factor(newdata.vstmelt$ssom3.4.unit.classif,levels=c(10, 7,12,1,11,4,2,8,5,9,6,3)) # use this factor for all heatmap. Needs to update.
newdata.vstmelt$variable<-factor(newdata.vstmelt$variable,levels=c("1h","4h","25h","49h")) # order in graphs was fixed (022115)
p <- ggplot(newdata.vstmelt,aes(x=variable,y=value,color=value))
p <- p + geom_point(position="jitter",size=1.5,alpha=0.3) + geom_boxplot(outlier.size=0,aes(fill=factor(ssom3.4.unit.classif),alpha=0.8)) + theme_bw() +xlab("sun/shade treatment (hr)") + scale_colour_gradient2(low="green1",mid="gray",high="magenta")
p <- p + facet_grid(ssom3.4.unit.classif~gt,scale="free",space="free") #+ facet_wrap(ssom3.4unit.classif~.)
p <- p + theme(strip.text.x = element_text(angle=90),axis.text.x=element_text(angle=90))
#p
ggsave("boxplot.node.genotype.3.4.noscaling.pdf",width=11,height=20,unit="in", path="../../Nozue2016_SAStranscriptome_output/figs_tables")
#ggsave("boxplot.node.genotype.png",width=11,height=20,unit="in",dpi=600)
newdata.vst[rownames(newdata.vst)=="AT4G16780.1",] # ATHB2 cluster 10 in Col
newdata.vst[rownames(newdata.vst)=="AT2G46970.1",] # PIL1 cluster 10 in Col
```
### landscape format (similar to one used in 2014 ASPB meeting)
```{r eval=FALSE, ggplot2, fig.width=11, fig.height=8}
## probably this is better (matching to PIL1 expressio patterns in all genotypes (see plot))
## Also number of genes in cluster 5 and 6 are increased which is consistent with mutant phenotype
## do with this. (070814)
load("../../Nozue2016_SAStranscriptome_output/output/newdata.vst.Rdata") # 122214
# boxplot
newdata.vstmelt<-melt(newdata.vst[,c("1h","25h","4h","49h","ssom3.4.unit.classif","gt","AGI")],id.var=c("gt","AGI","ssom3.4.unit.classif"))
names(newdata.vstmelt)[3]<-"ssom3.4.unit.classif"
newdata.vstmelt$variable<-factor(newdata.vstmelt$variable,levels=c("1h","4h","25h","49h")) # order in graphs was fixed (022115)
newdata.vstmelt$ssom3.4.unit.classif<-factor(newdata.vstmelt$ssom3.4.unit.classif,levels=c(10, 7,12,1,11,4,2,8,5,9,6,3)) # use this factor for all heatmap. Needs to update.
### convert genotype number into actual genotype name
conversion.table<-data.frame(num=1:27, genotype=c("Col","AT5G02540_1","hy5","jar1","kat1_2","phyB","pif45","spt_11","yuc2589","PAR1_RNAi09","coi1_16","phyAB","pif3","mida9_4","bsk5_1","sto","aos","argos","co_9","Blh_1","Jea","Shahdara","Col_0","Cvi_0","Bur_0","Oy_0","Ita_0"))
head(newdata.vstmelt)
for(i in 1:27){
  newdata.vstmelt[newdata.vstmelt$gt==conversion.table[i,1],"gt2"]<-as.character(conversion.table[i,2])
} 
# head(newdata.vstmelt)
#draw boxplot with 90 degree orientation for ppt presentation (2014 ASPB)
p <- ggplot(newdata.vstmelt,aes(x=variable,y=value,color=value))
p <- p + geom_hline(yintercept=0,colour="red",size=1) # adding red line at fold change = 0
p <- p + geom_point(position="jitter",size=1.5,alpha=0.3) 
p <- p + geom_boxplot(outlier.size=0,aes(fill=factor(ssom3.4.unit.classif),alpha=0.8)) + theme_bw() +xlab("sun/shade treatment (hr)") + scale_colour_gradient2(low="green1",mid="gray",high="magenta")
p <- p + facet_grid(gt2~ssom3.4.unit.classif,space="free") #+ facet_wrap(ssom3.4unit.classif~.)  ,scale="free" 
p <- p + theme(axis.text.x=element_text(angle=90),strip.text.y = element_text(angle=0),legend.position="none") #
p <- p + coord_flip()
p
# TO DO: draw y=0 line (thick red)
ggsave("boxplot.node.genotype3.4.noscaling_similarto2014ASPB.pdf",width=15,height=8,unit="in",path="../../Nozue2016_SAStranscriptome_output/figs_tables")
```
### only Col and yucQ
```{r eval=FALSE, ggplot3, fig.width=11, fig.height=4} 
### plot only Col and yucQ (071214)
load("../../Nozue2016_SAStranscriptome_output/output/newdata.vst.Rdata") # 122214
newdata.vst.Col.yucQ<-newdata.vst[newdata.vst$gt==1|newdata.vst$gt==9,c("1h","25h","4h","49h","ssom3.4.unit.classif","gt","AGI")]
newdata.vst.Col.yucQ.melt<-melt(newdata.vst.Col.yucQ,id.var=c("gt","AGI","ssom3.4.unit.classif"))
newdata.vst.Col.yucQ.melt$gt<-as.character(newdata.vst.Col.yucQ.melt$gt)
newdata.vst.Col.yucQ.melt$variable<-factor(newdata.vst.Col.yucQ.melt$variable, levels=c("1h","4h","25h","49h"))
conversion.table<-data.frame(num=1:27, genotype=c("Col","AT5G02540_1","hy5","jar1","kat1_2","phyB","pif45","spt_11","yuc2589","PAR1_RNAi09","coi1_16","phyAB","pif3","mida9_4","bsk5_1","sto","aos","argos","co_9","Blh_1","Jea","Shahdara","Col_0","Cvi_0","Bur_0","Oy_0","Ita_0"))
head(newdata.vst.Col.yucQ.melt)
for(i in 1:27){
  newdata.vst.Col.yucQ.melt[newdata.vst.Col.yucQ.melt$gt==conversion.table[i,1],"gt2"]<-as.character(conversion.table[i,2])
} 
head(newdata.vst.Col.yucQ.melt)
newdata.vst.Col.yucQ.melt$ssom3.4.unit.classif<-factor(newdata.vst.Col.yucQ.melt$ssom3.4.unit.classif,levels=c(10, 7,12,1,11,4,2,8,5,9,6,3)) # use this factor for all heatmap
names(newdata.vst.Col.yucQ.melt)[3]<-"ssom3.4.unit.classif"
p <- ggplot(newdata.vst.Col.yucQ.melt,aes(x=variable,y=value,color=value))
p <- p + geom_hline(yintercept=0,colour="red",size=1) # adding red line at fold change = 0 (060515)
p <- p + geom_point(position="jitter",size=1.5,alpha=0.3) 
p <- p + geom_boxplot(outlier.size=0,aes(fill=factor(ssom3.4.unit.classif),alpha=0.8)) + theme_bw() +xlab("sun/shade treatment (hr)") + scale_colour_gradient2(low="green1",mid="gray",high="magenta")
p <- p + facet_grid(gt2~ssom3.4.unit.classif,space="free") #+ facet_wrap(ssom3.4unit.classif~.)  ,scale="free"
p <- p + theme(axis.text.x=element_text(angle=90),strip.text.y = element_text(angle=0),legend.position="none")
p <- p + labs(x="Time (hr)", y="Log2 fold change (shade/sun)") #
p <- p + coord_flip()
p
ggsave("boxplot.node.genotype3.4.noscaling_vst_ColyucQ.pdf",width=11,height=5,unit="in",path="../../Nozue2016_SAStranscriptome_output/figs_tables")
```
### only Col (for explanation)
```{r eval=FALSE, ggplot4, fig.width=11, fig.height=2}
#only Col (for lab meeting, 091114 version; vst version)
load("../../Nozue2016_SAStranscriptome_output/output/newdata.vst.Rdata") # 122214
newdata.vst.Col<-newdata.vst[newdata.vst$gt==1,]
newdata.vst.Col.melt<-melt(newdata.vst.Col[,c("1h","25h","4h","49h","ssom3.4.unit.classif","gt","AGI")],id.var=c("gt","AGI","ssom3.4.unit.classif"))
conversion.table<-data.frame(num=1:27, genotype=c("Col","AT5G02540_1","hy5","jar1","kat1_2","phyB","pif45","spt_11","yuc2589","PAR1_RNAi09","coi1_16","phyAB","pif3","mida9_4","bsk5_1","sto","aos","argos","co_9","Blh_1","Jea","Shahdara","Col_0","Cvi_0","Bur_0","Oy_0","Ita_0"))
head(newdata.vst.Col.melt)
for(i in 1:13){ # new conversion.table (090116)
  newdata.vst.Col.melt[newdata.vst.Col.melt$gt==conversion.table[i,1],"gt2"]<-as.character(conversion.table[i,2])
} 
head(newdata.vst.Col.melt)
newdata.vst.Col.melt$ssom3.4.unit.classif<-factor(newdata.vst.Col.melt$ssom3.4.unit.classif,levels=c(10, 7,4,12,1,11,2,8,5,9,6,3)) # use this factor for all heatmap. 10,7,4 together because I grouped them as "auxin" in manuscript (092315)
newdata.vst.Col.melt$ssom3.4.unit.classif<-factor(newdata.vst.Col.melt$ssom3.4.unit.classif,levels=c(10, 7,4,12,11,9,6,5,2,1,3,8)) # use this factor for all heatmap. start from upregulated (10,7,4 together because I grouped them as "auxin" in manuscript) and then down-regulated as manuscript described (052316)


#newdata.vst.Col.melt$variable<-factor(newdata.vst.Col.melt$variable,levels=c("1h","4h","25h","49h"))
newdata.vst.Col.melt$variable<-factor(newdata.vst.Col.melt$variable,levels=c("49h","25h","4h","1h")) # Julin's request for manuscript fig (092215)
names(newdata.vst.Col.melt)[3]<-"ssom3.4.unit.classif"
p <- ggplot(newdata.vst.Col.melt,aes(x=variable,y=value,color=value))
p <- p + geom_hline(yintercept=0,colour="red",size=1) # adding red line at fold change = 0 (060515)
p <- p + geom_point(position="jitter",size=1.5,alpha=0.3) 
p <- p + geom_boxplot(outlier.size=0,aes(fill=factor(ssom3.4.unit.classif),alpha=0.8)) + theme_bw() +xlab("sun/shade treatment (hr)") + scale_colour_gradient2(low="green1",mid="gray",high="magenta")
p <- p + facet_grid(gt2~ssom3.4.unit.classif,space="free") #+ facet_wrap(ssom3.4unit.classif~.)  ,scale="free"
p <- p + theme(axis.text.x=element_text(angle=90),strip.text.y = element_text(angle=0),legend.position="none")
p <- p + labs(x="Time (hr)", y="Log2 fold change (shade/sun)") #
p <- p + coord_flip()
p 
ggsave("Fig2_Coexpression_clusters_of_shade-responsive_genes_090116.pdf",width=11,height=5,unit="in",path="../../Nozue2016_SAStranscriptome_output/figs_tables") # needs to change order of cluster (090116). File name changed (091016; 110716)

```

# GOseq anaysis of SOM clusters (022215)
```{r eval=FALSE}
load("../../Nozue2016_SAStranscriptome_output/output/newdata.vst.Rdata") # 122214
newdata.vst.Col<-newdata.vst[newdata.vst$gt==1,]
clusters<-levels(as.factor(newdata.vst.Col$ssom3.4.unit.classif))
# modify GOseq.ORA if there is no enriched GO ("return" instead of "stop") (070716)
for(x in 1:12) {  
  temp.GOseq<-GOseq.ORA(newdata.vst.Col[newdata.vst.Col$ssom3.4.unit.classif==x,"AGI"])
  if(temp.GOseq=="no enriched GO") {next} else {
     temp.genes<-genes.in.enriched.category(enrich.result=temp.GOseq,gene.list=newdata.vst.Col[newdata.vst.Col$ssom3.4.unit.classif==x,"AGI"],category.table=Atgo.list)
    save(temp.GOseq,temp.genes,file=paste("../../Nozue2016_SAStranscriptome_output/output/SOM.Col.cluster.",x,".GOseq.Rdata",sep=""))
  } 
}
```
# write summary of GOseq analsyis of SOM clusters (022215)
```{r eval=FALSE}
rm(list=ls())
SOM.Col.cluster.GOseq.files<-list.files(pattern="SOM.Col.cluster.",path="../../Nozue2016_SAStranscriptome_output/output/")
SOM.Col.cluster.GOseq.files<-grep(".GOseq.Rdata",SOM.Col.cluster.GOseq.files,value=TRUE)
load(paste("../../Nozue2016_SAStranscriptome_output/output/",SOM.Col.cluster.GOseq.files[1],sep=""))
temp.GOseq$clusters<-gsub("(SOM.Col.cluster.)([[:print:]]+)(.GOseq.Rdata)","\\2",SOM.Col.cluster.GOseq.files[1])
temp<-temp.GOseq
for(i in SOM.Col.cluster.GOseq.files[-1]) {
  load(paste("../../Nozue2016_SAStranscriptome_output/output/",i,sep=""))
  temp.GOseq$clusters<-gsub("(SOM.Col.cluster.)([[:print:]]+)(.GOseq.Rdata)","\\2",i)
  temp<-rbind(temp,temp.GOseq)
}
SOM.Col.cluster.GOseq.summary<-temp
write.csv(SOM.Col.cluster.GOseq.summary,file="../../Nozue2016_SAStranscriptome_output/figs_tables/TableS2_GO_enrichment_analysis_of_co-expressed_SOM_clusters_of_shade-responsive_genes.csv") # 091016


```


## differentially expressed gene analysis
2a.DE at any time point (sun)
```{r eval=FALSE}
# setwd("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/kazu/Documents/SASmutRNAseq_whitney/edgeR/outlier_wo16h/genotype_specific")
rm(list=ls())
system("mkdir ../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/genotype_specific5")
#setwd("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/genotype_specific5")
load("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/samples.nolow5.Rdata")
load("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/dge.nolow5.Rdata")

# Genotype<-levels(as.factor(samples.nolow5$genotype))
# Genotype<-Genotype[!(Genotype=="11"|Genotype=="12"|Genotype=="18"|Genotype=="16")]
# Use Genotype in new vst data. (082016)
#SAS.expression.vst<-read.csv("../../Nozue2016_SAStranscriptome_data/input/kazu.SAS.expression.VST.070516.csv") # new transformed data by DEseq2 (070416)
SAS.expression.vst<-read.csv("../../Nozue2016_SAStranscriptome_data/input/kazu.SAS.expression.VST.082016.csv") # new transformed data by DEseq2 (070416)

rownames(SAS.expression.vst)<-SAS.expression.vst$X
Genotype<-levels(as.factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|25|49)(A.merged.bam)","\\2",names(SAS.expression.vst)[-1]))) # why 12 ? one missing? 082016 version had 13 ("3" was missing in 070516 version. Replace them and reanalyze all related steps.)

for(i in Genotype[-1]) {# Genotype 17, 3,4,5,6,7,9 exceed 2000 genes. List up 5000 genes
  samples.nolow5.s<-samples.nolow5[samples.nolow5$genotype=="1"|samples.nolow5$genotype==i,]
  samples.nolow5.s<-samples.nolow5.s[samples.nolow5.s$trt=="H",] # only "sun" treatment
  samples.nolow5.s$genotype<-as.character(samples.nolow5.s$genotype)
  dge.nolow.s<-dge.nolow5[,colnames(dge.nolow5) %in% samples.nolow5.s$file]
  # eliminating low expressed genes in dge.nolow.s
  dge.nolow.s<-dge.nolow.s[rowSums(dge.nolow.s$counts > 5) >= 3,]
  design.nolow.s.gt.batch3<- model.matrix(~samples.nolow5.s$genotype*samples.nolow5.s$time  + samples.nolow5.s$batch) # new model to detect genes with gt effects (100314)  
  colnames(design.nolow.s.gt.batch3) <- gsub("samples.nolow5.s$","",colnames(design.nolow.s.gt.batch3),fixed=TRUE) # nicer column names
  colnames(design.nolow.s.gt.batch3)
  print(design.nolow.s.gt.batch3)
  dge.nolow.s.batch3.glm <- estimateGLMCommonDisp(dge.nolow.s,design.nolow.s.gt.batch3, verbose=TRUE)
  dge.nolow.s.batch3.glm <- estimateGLMTrendedDisp(dge.nolow.s.batch3.glm,design.nolow.s.gt.batch3, verbose=TRUE)
  dge.nolow.s.batch3.glm<-estimateGLMTagwiseDisp(dge.nolow.s.batch3.glm,design.nolow.s.gt.batch3)
  #pdf(file=paste("BCV_MDS_with_genotype.wo16h_gt",i,".pdf",sep=""),width=11,height=8)
  #par(mfrow=c(1,2))
  #plotBCV(dge.nolow.s.batch3.glm,main=paste("genotype",i))
  #plotMDS(dge.nolow.s.batch3.glm,main=paste("genotype",i))
  #dev.off()  
  dge.nolow.s.batch3.glm.fit <- glmFit(dge.nolow.s.batch3.glm,design.nolow.s.gt.batch3)
  genotype.alltime.coef<-c(paste("genotype",i,sep=""),paste("genotype",i,":time4",sep=""),paste("genotype",i,":time25",sep=""),paste("genotype",i,":time49",sep=""))  
  dge.nolow.s.batch3.glm.lrt <- glmLRT(dge.nolow.s.batch3.glm.fit,coef=genotype.alltime.coef)
  DE.batch3.gt<-topTags(dge.nolow.s.batch3.glm.lrt,n=10000)$table[topTags(dge.nolow.s.batch3.glm.lrt,n=10000)$table$FDR<0.001,] # 0.05 for genotype_specific2 which gave me 5000 genes in hy5 (genotype3)
  dim(DE.batch3.gt) #
  save(DE.batch3.gt,file=paste("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/genotype_specific5/DE.genotype.batch.sun.gt.wo16h",i,".Rdata",sep="")) 
  # save  dge.nolow.s.batch3.glm.lrt
    save(dge.nolow.s.batch3.glm.lrt,file=paste("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/genotype_specific5/dge.nolow.s.batch3.glm.lrt.sun",i,".Rdata",sep="")) # added this line (try to use this object for coef, 061715)
}
getwd()

# for(i in Genotype[-1]) {
#   load(paste("DE.genotype.batch.sun.gt.wo16h",i,".Rdata",sep=""))
#   ####### under construction ############# 
#   temp.data<-summary4 # needs to use absolute value (mean value); summary4? (080714); needs to make new version of mean expression data frame (060515)
#   
#   #target.genes<-c("AT3G01345.1","AT3G14210.1","AT3G30122.1","AT3G22231.1")
#   target.genes<-rownames(DE.batch.gt)[1:20]
#   temp.data<-temp.data[rownames(temp.data) %in% target.genes,]
#   temp.data$AGI<-rownames(temp.data)
#   temp.data.melt<-melt(temp.data,id.var="AGI")
#   temp.data.melt$gt<-gsub("([[:digit:]]+)(L|H)(1|4|16|25|49)","\\1",temp.data.melt$variable)
#   temp.data.melt$time<-factor(gsub("([[:digit:]]+)(L|H)(1|4|16|25|49)","\\3",temp.data.melt$variable),levels=c(1,4,25,49))
#   temp.data.melt$trt<-factor(gsub("([[:digit:]]+)(L|H)(1|4|16|25|49)","\\2",temp.data.melt$variable),levels=c("H","L"))
#   temp.data.melt<-temp.data.melt[temp.data.melt$gt %in% c(1:17,19),]
#   temp.data.melt$gt<-factor(temp.data.melt$gt,levels=c(1:17,19))
#   
#   ## change genotype name
#   temp.data.melt$gt2<-""
#   for(n in 1:27) {
#     temp.data.melt$gt2[temp.data.melt$gt==as.character(conversion.table[n,"num"])]<-as.character(conversion.table[n,"genotype"])
#   }
#   # drawing graph
#   q<-ggplot(temp.data.melt,aes(x=time,y=value,color=trt)) + geom_point(alpha = 0.5) + facet_grid(AGI~gt2,scale="free") + theme_bw() 
#   q<- q + theme(axis.text.x=element_text(angle=90),strip.text.x=element_text(angle=90),strip.text.y=element_text(angle=0))
#   ggsave(q,file=paste("DE.genotype.batch.gt.wo16h_normalized.",i,".expression.pattern.pdf",sep=""),width=8,height=20,path="../../Nozue2016_SAStranscriptome_output/figs_tables/")
# }
```
2b.DE at any time point (shade)
```{r eval=FALSE}
for(i in Genotype[-1]) {# Genotype 17, 3,4,5,6,7,9 exceed 2000 genes. List up 5000 genes
  samples.nolow5.s<-samples.nolow5[samples.nolow5$genotype=="1"|samples.nolow5$genotype==i,]
  samples.nolow5.s<-samples.nolow5.s[samples.nolow5.s$trt=="L",] # only "shade" treatment
  samples.nolow5.s$genotype<-as.character(samples.nolow5.s$genotype)
  dge.nolow.s<-dge.nolow5[,colnames(dge.nolow5) %in% samples.nolow5.s$file]
  # eliminating low expressed genes in dge.nolow.s
  dge.nolow.s<-dge.nolow.s[rowSums(dge.nolow.s$counts > 5) >= 3,]
  design.nolow.s.gt.batch3<- model.matrix(~samples.nolow5.s$genotype*samples.nolow5.s$time  + samples.nolow5.s$batch) # new model to detect genes with gt effects (100314)  
  colnames(design.nolow.s.gt.batch3) <- gsub("samples.nolow5.s$","",colnames(design.nolow.s.gt.batch3),fixed=TRUE) # nicer column names
  colnames(design.nolow.s.gt.batch3)
  print(design.nolow.s.gt.batch3)
  dge.nolow.s.batch3.glm <- estimateGLMCommonDisp(dge.nolow.s,design.nolow.s.gt.batch3, verbose=TRUE)
  dge.nolow.s.batch3.glm <- estimateGLMTrendedDisp(dge.nolow.s.batch3.glm,design.nolow.s.gt.batch3, verbose=TRUE)
  dge.nolow.s.batch3.glm<-estimateGLMTagwiseDisp(dge.nolow.s.batch3.glm,design.nolow.s.gt.batch3)
  #pdf(file=paste("BCV_MDS_with_genotype.wo16h_gt",i,".pdf",sep=""),width=11,height=8)
  #par(mfrow=c(1,2))
  #plotBCV(dge.nolow.s.batch3.glm,main=paste("genotype",i))
  #plotMDS(dge.nolow.s.batch3.glm,main=paste("genotype",i))
  #dev.off()  
  dge.nolow.s.batch3.glm.fit <- glmFit(dge.nolow.s.batch3.glm,design.nolow.s.gt.batch3)
  genotype.alltime.coef<-c(paste("genotype",i,sep=""),paste("genotype",i,":time4",sep=""),paste("genotype",i,":time25",sep=""),paste("genotype",i,":time49",sep=""))  
  dge.nolow.s.batch3.glm.lrt <- glmLRT(dge.nolow.s.batch3.glm.fit,coef=genotype.alltime.coef)
  
  DE.batch3.gt<-topTags(dge.nolow.s.batch3.glm.lrt,n=10000)$table[topTags(dge.nolow.s.batch3.glm.lrt,n=10000)$table$FDR<0.001,] # 0.05 for genotype_specific2 which gave me 5000 genes in hy5 (genotype3)
  dim(DE.batch3.gt) #
  save(DE.batch3.gt,file=paste("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/genotype_specific5/DE.genotype.batch.shade.gt.wo16h",i,".Rdata",sep=""))   
  save(dge.nolow.s.batch3.glm.lrt,file=paste("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/genotype_specific5/dge.nolow.s.batch3.glm.lrt.shade",i,".Rdata",sep="")) # added this line (try to use this object for coef, 061715)
}


# # calculate DE.batch3.gt (081716)
#   dge.nolow.s.batch3.glm.lrt.files<-list.files(pattern="dge.nolow.s.batch3.glm.lrt",path="../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/genotype_specific5/")
# setwd("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/genotype_specific5/")
# for(i in 1:44) {
#   load(dge.nolow.s.batch3.glm.lrt.files[i])
#   DE.batch3.gt<-topTags(dge.nolow.s.batch3.glm.lrt,n=10000)$table[topTags(dge.nolow.s.batch3.glm.lrt,n=10000)$table$FDR<0.001,] # 0.05 for genotype_specific2 which gave me 5000 genes in hy5 (genotype3)
#   dim(DE.batch3.gt) #  
#   save(DE.batch3.gt,file=paste("DE.genotype.batch.",gsub("(dge.nolow.s.batch3.glm.lrt.)(sun|shade)([[:digit:]]+)(.Rdata)","\\2",dge.nolow.s.batch3.glm.lrt.files[i]),".gt.wo16h",gsub("(dge.nolow.s.batch3.glm.lrt.)(sun|shade)([[:digit:]]+)(.Rdata)","\\3",dge.nolow.s.batch3.glm.lrt.files[i]),".Rdata",sep="")) # added this line (try to use this object for coef, 061715)
# }
setwd(homedir)

```




## overlap between Col modules and mutation affected genes
```{r eval=FALSE}
#rm(list=ls())
library(WGCNA);library(ggplot2);library(reshape2);library(scales)
conversion.table # has to be made in function_
# sun
setwd(homedir)
DE.genotype.batch.sun.files<-list.files(pattern="DE.genotype.batch.sun",path="../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/genotype_specific5/") # needs to select "sun" object (060615)
setwd("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/genotype_specific5")
DE.genotype.batch.sun.files.kazu<-DE.genotype.batch.sun.files[gsub("([[:print:]]+)(wo16h)([[:digit:]]+)(.Rdata)","\\3",DE.genotype.batch.sun.files) %in% 1:19]
#setwd("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/genotype_specific5")

load(DE.genotype.batch.sun.files.kazu[1])
DE.genotype.batch.sun.genes<-rownames(DE.batch3.gt)
for(xx in DE.genotype.batch.sun.files.kazu) {
  load(xx)
  DE.genotype.batch.sun.genes<-c(DE.genotype.batch.sun.genes,rownames(DE.batch3.gt))
}
# shade
#setwd("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/genotype_specific5")
DE.genotype.batch.shade.files<-list.files(pattern="DE.genotype.batch.shade")

DE.genotype.batch.shade.files.kazu<-DE.genotype.batch.shade.files[gsub("([[:print:]]+)(wo16h)([[:digit:]]+)(.Rdata)","\\3",DE.genotype.batch.shade.files) %in% 1:19]
load(DE.genotype.batch.shade.files.kazu[1])
DE.genotype.batch.shade.genes<-rownames(DE.batch3.gt)
for(xx in DE.genotype.batch.shade.files.kazu) {
  load(xx)
  DE.genotype.batch.shade.genes<-c(DE.genotype.batch.shade.genes,rownames(DE.batch3.gt))
}
######
DE.genotype.batch.genes.unique<-unique(c(DE.genotype.batch.sun.genes,DE.genotype.batch.shade.genes))
length(DE.genotype.batch.genes.unique) #3356
# 
DE.genotype.batch.genes.unique.DF<-data.frame(AGI=DE.genotype.batch.genes.unique)
# sun
for(x in DE.genotype.batch.sun.files.kazu) {
#setwd("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/genotype_specific5")
#setwd("/Volumes/Data8/NGS_related/Arabidopsis_analysis/SAS_muts_time_course_RNAseq2_mycomp/")
load(x)
temp<-as.integer(DE.genotype.batch.genes.unique %in% rownames(DE.batch3.gt)) 
temp[temp==1]<-rep(paste(conversion.table[conversion.table==gsub("(DE.genotype.batch.sun.gt.wo16h)([[:digit:]]+)(.Rdata)","\\2",x),2],"sun",sep="."),sum(temp))
DE.genotype.batch.genes.unique.DF<-cbind(DE.genotype.batch.genes.unique.DF,temp)                   
}
names(DE.genotype.batch.genes.unique.DF)[-1]<-paste(conversion.table[as.numeric(gsub("(DE.genotype.batch.sun.gt.wo16h)([[:digit:]]+)(.Rdata)","\\2",DE.genotype.batch.sun.files.kazu)),2],"sun",sep=".")
#head(DE.genotype.batch.genes.unique.DF)
# shade
for(x in DE.genotype.batch.shade.files.kazu) {
#setwd("/Volumes/Data8/NGS_related/Arabidopsis_analysis/SAS_muts_time_course_RNAseq2_mycomp/edgeR/outlier_wo16h/genotype_specific4")  
#setwd("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/genotype_specific5")  

load(x)
temp<-as.integer(DE.genotype.batch.genes.unique %in% rownames(DE.batch3.gt)) 
temp[temp==1]<-rep(paste(conversion.table[conversion.table==gsub("(DE.genotype.batch.shade.gt.wo16h)([[:digit:]]+)(.Rdata)","\\2",x),2],"shade",sep="."),sum(temp))
DE.genotype.batch.genes.unique.DF<-cbind(DE.genotype.batch.genes.unique.DF,temp)                   
}
head(DE.genotype.batch.genes.unique.DF)
dim(DE.genotype.batch.genes.unique.DF)

#### error #####
# names(DE.genotype.batch.genes.unique.DF)[-(1:15)]<-paste(conversion.table[as.numeric(gsub("(DE.genotype.batch.shade.gt.wo16h)([[:digit:]]+)(.Rdata)","\\2",DE.genotype.batch.shade.files.kazu)),2],"shade",sep=".")
# ## 
names(DE.genotype.batch.genes.unique.DF)
# 091016
DE.genotype<-as.numeric(gsub("(DE.genotype.batch.shade.gt.wo16h)([[:digit:]]+)(.Rdata)","\\2",DE.genotype.batch.shade.files.kazu))
names(DE.genotype.batch.genes.unique.DF)[-1]<-c(paste(vlookup(DE.genotype,conversion.table,2),"sun",sep="."),paste(vlookup(DE.genotype,conversion.table,2),"shade",sep=".")) # the first column is "AGI"
# the end of 091016 version

# remove "0"
#DE.genotype.batch.genes.unique.DF[DE.genotype.batch.genes.unique.DF==0]<-""
DE.genotype.batch.genes.unique.DF$AGI<-gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",DE.genotype.batch.genes.unique.DF$AGI)
setwd(homedir)
save(DE.genotype.batch.genes.unique.DF,file="../../Nozue2016_SAStranscriptome_output/output/DE.genotype.batch.genes.unique.DF.Rdata")


# SOM clusters
setwd(homedir)
load("../../Nozue2016_SAStranscriptome_output/output/DE.genotype.batch.genes.unique.DF.Rdata")
load("../../Nozue2016_SAStranscriptome_output/output/newdata.vst.Rdata") # 
newdata.vst.Col<-newdata.vst[newdata.vst$gt==1,]
# merge with SOM clusters
test<-merge(DE.genotype.batch.genes.unique.DF,newdata.vst.Col[,c("AGI","ssom3.4.unit.classif")],by="AGI",all=TRUE)
DE.genotype.batch.genes.SOMclusters<-test
DE.genotype.batch.genes.SOMclusters[is.na(DE.genotype.batch.genes.SOMclusters)]<-"0"

# write.csv(DE.genotype.batch.genes.SOMclusters,file="../../Nozue2016_SAStranscriptome_output/figs_tables/DE.genotype.batch.genes.SOMclusters.csv")
# DE.genotype.batch.genes.SOMclusters<-read.csv("../../Nozue2016_SAStranscriptome_output/figs_tables/DE.genotype.batch.genes.SOMclusters.csv")
# DE.genotype.batch.genes.SOMclusters<-DE.genotype.batch.genes.SOMclusters[,-1]
# adding gene description
TAIR10_gene_descriptions<-read.csv("../../Nozue2016_SAStranscriptome_data/input/TAIR10_functional_descriptions.csv") 
head(TAIR10_gene_descriptions)
TAIR10_gene_descriptions$AGI<-gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",TAIR10_gene_descriptions$"Model_name")
DE.genotype.batch.genes.SOMclusters.description<-merge(DE.genotype.batch.genes.SOMclusters,TAIR10_gene_descriptions[!duplicated(TAIR10_gene_descriptions$AGI),c("AGI","Short_description","colE...col.F")],by="AGI")
write.csv(DE.genotype.batch.genes.SOMclusters.description,"../../Nozue2016_SAStranscriptome_output/figs_tables/Dataset3_Complete_list_of_shade-responsive_genes_and_SOM_gene_clusters_and_misexpressed_genes.csv")
test<-DE.genotype.batch.genes.SOMclusters
# overlap table has to be against all genes? (071716)
# overlap table (revised to GOseq.overlapTable version; 082515)
i<-2
#overlap<-overlapTable(test[,30],test[,i])
overlap.GOseq<-GOseq.overlapTable(test[,"ssom3.4.unit.classif"],test[,i],genes=test$AGI)

overlap.GOseq.melt<-melt(overlap.GOseq$pTable,id=1) #
overlap.GOseq.melt$count<-melt(overlap.GOseq$countTable,id=1)$value
overlap.GOseq.melt.summary<-overlap.GOseq.melt[!overlap.GOseq.melt$Var2=="0",]
for(i in 3:dim(test)[2]) {
overlap.GOseq<-GOseq.overlapTable(test[,"ssom3.4.unit.classif"],test[,i],genes=test$AGI)
overlap.GOseq.melt<-melt(overlap.GOseq$pTable,id=1) #
overlap.GOseq.melt$count<-melt(overlap.GOseq$countTable,id=1)$value
overlap.GOseq.melt<-overlap.GOseq.melt[!overlap.GOseq.melt$Var2=="0",]
overlap.GOseq.melt.summary<-rbind(overlap.GOseq.melt.summary,overlap.GOseq.melt)
}
overlap.GOseq.melt.summary[,"value"]<--log10(overlap.GOseq.melt.summary[,"value"])
overlap.GOseq.melt.summary[overlap.GOseq.melt.summary$value>10,"value"]<-10
overlap.GOseq.melt.summary[overlap.GOseq.melt.summary$count==0,"count"]<-""
#table(overlap.melt.summary[,"value"])
save(overlap.GOseq.melt.summary, file="../../Nozue2016_SAStranscriptome_output/output/overlap.GOseq.melt.summary.SOM.DEgenes.110716.Rdata") 
load("../../Nozue2016_SAStranscriptome_output/output/overlap.GOseq.melt.summary.SOM.DEgenes.110716.Rdata")
# # graph
# overlap.GOseq.melt.summary.s<-overlap.GOseq.melt.summary[!overlap.GOseq.melt.summary$Var2 %in% c("bsk5_1.sun","bsk5_1.shade","AT5G02540_1.sun","AT5G02540_1.shade"),] # eliminate bsk5_1 (genotype15) and AT5G02540_1 (genotype 2)
overlap.GOseq.melt.summary.s<-overlap.GOseq.melt.summary[!is.na(overlap.GOseq.melt.summary$Var2),]
overlap.GOseq.melt.summary.ss<-overlap.GOseq.melt.summary.s[!overlap.GOseq.melt.summary.s$Var1==0,]
overlap.GOseq.melt.summary.ss$Var1<-factor(overlap.GOseq.melt.summary.ss$Var1,levels=c(10, 7,4,12,11,9,6,5,2,1,3,8)) # 052516
# needs to change Var1 levels (091016)

# 
# q4<-ggplot(overlap.GOseq.melt.summary.ss,aes(y=factor(Var1),x=factor(Var2))) 
# #q4 <- q4 + geom_tile(color="black",aes(fill=value))+scale_fill_gradient2(low=muted("green"), high=muted("magenta"),midpoint=2)
# q4 <- q4 + geom_tile(color="black",aes(fill=value))+scale_fill_gradient(low="white", high="magenta") # 061816
# #q4 <- q4 + theme(axis.text.x=element_text(angle=90),axis.title=element_blank())
# q4 <- q4 + theme(axis.text.x=element_text(angle=90,hjust=1),
#                  panel.grid.major = element_blank(),
#                  panel.grid.minor = element_blank(),
#                  panel.background = element_blank(),
#                  axis.ticks = element_line(size=0,color="white"))
# q4 <- q4 + geom_text(aes(label=count,size=1))
# q4 <- q4 + labs(x="genotype specific genes",y="SOM clusters in Col",title="SOM clusters affected in SAS mutants")
# #setwd("/Volumes/Data8/NGS_related/Arabidopsis_analysis/SAS_muts_time_course_RNAseq2_mycomp")
# ggsave(q4,file="Col.SOM.clusters.vs.GenotypeSpecific.genes.table.s.GOseq.newcolor.pdf",width=11,height=8,path="../../Nozue2016_SAStranscriptome_output/figs_tables")
# q4 # magenta indicates pvalue is smaller than 0.01.
# switching x and y axis
#overlap.GOseq.melt.summary.ss$Var1<-factor(overlap.GOseq.melt.summary.ss$Var1,levels=rev(c(3,6,9,5,8,2,4,11,1,12,7,10)))
overlap.GOseq.melt.summary.ss$Var1<-factor(overlap.GOseq.melt.summary.ss$Var1,levels=c(10, 7,4,12,11,9,6,5,2,1,3,8)) # 052516
overlap.GOseq.melt.summary.ss$Var2<-factor(overlap.GOseq.melt.summary.ss$Var2,levels=rev(levels(overlap.GOseq.melt.summary.ss$Var2))) 

q4b<-ggplot(overlap.GOseq.melt.summary.ss,aes(y=factor(Var2),x=factor(Var1))) 
#q4b <- q4b + geom_tile(color="black",aes(fill=value))+scale_fill_gradient2(low=muted("green"), high=muted("magenta"),midpoint=2)  #+ ,guide = guide_legend(title="-log10\n p-value")) 
q4b <- q4b + geom_tile(color="black",aes(fill=value))+scale_fill_gradient(low="white", high="magenta")  
#q4 <- q4 + theme(axis.text.x=element_text(angle=90),axis.title=element_blank())
q4b <- q4b + theme(axis.text.x=element_text(angle=90,hjust=1),
                 panel.grid.major = element_blank(),
                 panel.grid.minor = element_blank(),
                 panel.background = element_blank(),
                 axis.ticks = element_line(size=0,color="white"))
q4b <- q4b + geom_text(aes(label=count),size=5)
q4b <- q4b + labs(y="genotype specific genes",x="SOM clusters in Col",title="SOM clusters affected in SAS mutants",fill="-log10\n p-value") # chagne legend title (do no use gude_legend in scale_fill_gradient2)
setwd("../../Nozue2016_SAStranscriptome_output/figs_tables/")
ggsave(q4b,file="Fig4_Shade_avoidance_mutants_misexpress_specific_shade-response_SOM_cluster_110716.pdf",width=13,height=6,path="../../Nozue2016_SAStranscriptome_output/figs_tables") # running in whitney (110716)
q4b # almost the same with fisher's exact test version
# needs to remove Var2="0" (091016)

setwd(homedir)
```
# Overlap analysis with custome categories (SOM)
```{r eval=FALSE}
rm(list=ls())
source("function_RNAseq_time_course.R")
library(WGCNA);library(ggplot2);library(reshape2);library(scales)
options(stringsAsFactors = FALSE);
conversion.table
# custom categories
load("../../Nozue2016_SAStranscriptome_data/input/hormone.responsiveness4.Rdata") #new version. check if new version is identical to this. copy the file to this location
## change list into data.frame
attributes(hormone.responsiveness4)
hormone.responsiveness4.DF<-data.frame(t(sapply(hormone.responsiveness4,c)))
names(hormone.responsiveness4.DF)<-attributes(hormone.responsiveness4)$category
## convert into hormone.responsiveness3.DF (omit "PCC1", "GRF1_3")
hormone.responsiveness3.DF<-hormone.responsiveness4.DF[,!names(hormone.responsiveness4.DF) %in% c("GRF1_3target","PCC1up","PCC1down")]
summary(hormone.responsiveness3.DF)
hormone.responsiveness3.DF$AGI<-rownames(hormone.responsiveness3.DF)
# SOM clusters
load("../../Nozue2016_SAStranscriptome_output/output/newdata.vst.Rdata") # 022215

newdata.vst.Col<-newdata.vst[newdata.vst$gt==1,] 
# merge with SOM clusters
test<-merge(hormone.responsiveness3.DF,newdata.vst.Col[,c("AGI","ssom3.4.unit.classif")],by="AGI",all=TRUE)
hormone.responsiveness.SOMclusters<-test
hormone.responsiveness.SOMclusters[is.na(hormone.responsiveness.SOMclusters)]<-"0"
# write.csv(hormone.responsiveness.SOMclusters,file="../../Nozue2016_SAStranscriptome_output/figs_tables/TableS4_hormone.responsiveness.SOMclusters.csv")
# hormone.responsiveness.SOMclusters<-read.csv("../../Nozue2016_SAStranscriptome_output/figs_tables/TableS4_hormone.responsiveness.SOMclusters.csv")
#hormone.responsiveness.SOMclusters<-hormone.responsiveness.SOMclusters[,-1]
# addint gene description
# TAIR10_gene_descriptions<-read.csv("../../Nozue2016_SAStranscriptome_data/input/TAIR10_functional_descriptions.csv") 
# head(TAIR10_gene_descriptions)
# TAIR10_gene_descriptions$AGI<-gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",TAIR10_gene_descriptions$"Model_name")
hormone.responsiveness.SOMclusters.description<-merge(hormone.responsiveness.SOMclusters,TAIR10_gene_descriptions[!duplicated(TAIR10_gene_descriptions$AGI),c("AGI","Short_description","colE...col.F")],by="AGI")
write.csv(hormone.responsiveness.SOMclusters.description,file="../../Nozue2016_SAStranscriptome_output/figs_tables//Dataset2_Custom_enrichment_analysis_of_co-expressed_SOM_clusters_of_shade-responsive_genes.csv")
# in whitney (****16) for ORA
test<-read.csv("../../Nozue2016_SAStranscriptome_output/figs_tables/Dataset2_Custom_enrichment_analysis_of_co-expressed_SOM_clusters_of_shade-responsive_genes.csv") # 
test<-test[,-1]
# overlap table 
i<-2 
overlap<-GOseq.overlapTable(test[,"ssom3.4.unit.classif"],test[,i],genes=test$AGI,permutation=100)
overlap.melt<-melt(overlap$pTable,id=1) #
overlap.melt$count<-melt(overlap$countTable,id=1)$value
overlap.melt.summary<-overlap.melt[!overlap.melt$Var2=="0",]
for(i in 3:26) {
overlap<-GOseq.overlapTable(test[,"ssom3.4.unit.classif"],test[,i],genes=test$AGI,permutation=100)
overlap.melt<-melt(overlap$pTable,id=1) #
overlap.melt$count<-melt(overlap$countTable,id=1)$value
overlap.melt<-overlap.melt[!overlap.melt$Var2=="0",]
overlap.melt.summary<-rbind(overlap.melt.summary,overlap.melt)
}
overlap.melt.summary[,"value"]<--log10(overlap.melt.summary[,"value"])
overlap.melt.summary[overlap.melt.summary$value>10,"value"]<-10
overlap.melt.summary[overlap.melt.summary$count==0,"count"]<-""
save(overlap.melt.summary,file="../../Nozue2016_SAStranscriptome_output/output/overlap.melt.summary.SOM.custom.GOseq.110816.Rdata") # whitney. 
setwd(homedir)
load("../../Nozue2016_SAStranscriptome_output/output/overlap.melt.summary.SOM.custom.GOseq.110816.Rdata")
#load("overlap.melt.summary.SOM.custom.GOseq.091216.Rdata")
#setwd(homedir)
#table(overlap.melt.summary[,"value"])
# graph
#overlap.melt.summary.s<-overlap.melt.summary[!overlap.melt.summary$Var2 %in% c("bsk5_1.sun","bsk5_1.shade","AT5G02540_1.sun","AT5G02540_1.shade"),]
overlap.melt.summary.s<-overlap.melt.summary[!overlap.melt.summary$Var1==0,]
# overlap.melt.summary.s$Var1<-factor(overlap.melt.summary.s$Var1,levels=c(3,6,9,5,8,2,4,11,1,12,7,10))
overlap.melt.summary.s$Var1<-factor(overlap.melt.summary.s$Var1,levels=c(10, 7,4,12,11,9,6,5,2,1,3,8)) # 052516

overlap.melt.summary.s<-overlap.melt.summary.s[!overlap.melt.summary.s$Var2=="",]
overlap.melt.summary.s<-overlap.melt.summary.s[!overlap.melt.summary.s$Var2=="GRF1_3target",] # # remove "GRF1_3target" (052616)
overlap.melt.summary.s$Var2<-factor(overlap.melt.summary.s$Var2,levels=c("IAAup","IAAdown","BLup","BLdown","MJup","MJdown","GAup","GAdown","ABAup","ABAdown","ACCup","ACCdown","CKup","CKdown","SAup","SAdown","SARplus","SARminus","PIFtarget","MYC234up","MYC234down","NPR1up","NPR1down","WRKY33up","WRKY33down")) # remove "GRF1_3target" (052616)
# switching x and y axis
#overlap.melt.summary.s$Var1<-factor(overlap.melt.summary.s$Var1,levels=rev(c(3,6,9,5,8,2,4,11,1,12,7,10)))
overlap.melt.summary.s$Var1<-factor(overlap.melt.summary.s$Var1,levels=c(10, 7,4,12,11,9,6,5,2,1,3,8)) # 052616
overlap.melt.summary.s$Var2<-factor(overlap.melt.summary.s$Var2,levels=rev(levels(overlap.melt.summary.s$Var2)))                                    
q4b<-ggplot(overlap.melt.summary.s,aes(y=factor(Var2),x=factor(Var1))) 
#q4b <- q4b + geom_tile(color="black",aes(fill=value))+scale_fill_gradient2(low=muted("green"), high=muted("magenta"),midpoint=2)# ,guide = guide_legend(title="-log10 p-value"))
q4b <- q4b + geom_tile(color="black",aes(fill=value))+scale_fill_gradient(low="white", high="magenta")
#q4 <- q4 + theme(axis.text.x=element_text(angle=90),axis.title=element_blank())
q4b <- q4b + theme(axis.text.x=element_text(angle=90,hjust=1),
                 panel.grid.major = element_blank(),
                 panel.grid.minor = element_blank(),
                 panel.background = element_blank(),
                 axis.ticks = element_line(size=0,color="white"))
q4b <- q4b + geom_text(aes(label=count),size=5) 
q4b <- q4b + labs(y="custom categories",x="SOM clusters in Col",title="custom category enrichment",fill="-log10\n p-value")
#q4b
#ggsave(q4b,file="Col.SOM.clusters.vs.my.category.table.s.2.pdf",width=13,height=6,path="../../Nozue2016_SAStranscriptome_output/figs_tables")
ggsave(q4b,file="Fig3_Col.SOM.clusters.vs.my.category.table.s.2.GOseq.110816.pdf",width=13,height=6,path="../../Nozue2016_SAStranscriptome_output/figs_tables")
ggsave(q4b,file="Fig3_Col.SOM.clusters.vs.my.category.table.s.2.GOseq.110816.png",width=13,height=6,path="../../Nozue2016_SAStranscriptome_output/figs_tables") # testing this with Whitney (110816)

```

## co-expression analysis by WGCNA
```{r eval=FALSE}
library(WGCNA)
# load("/Volumes/Data8/NGS_related/Arabidopsis_analysis/SAS_muts_time_course_RNAseq2_mycomp/WGCNA/SASdata_voom/WGCNA_function.R")
# The following setting is important, do not omit.
options(stringsAsFactors = FALSE);
```
## Work in Whitney with multi-threds 
```{r eval=FALSE}
# Allow multi-threading within WGCNA. This helps speed up certain calculations.
# At present this call is necessary for the code to work.
# Any error here may be ignored but you may want to update WGCNA if you see one.
# Caution: skip this line if you run RStudio or other third-party R environments.
# See note above.
enableWGCNAThreads(10) # in Whitney (Maloof lab server)
#setwd("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/kazu/Documents/SASmutRNAseq_whitney/WGCNA/SASdata_vst") # replace "voom" into "vst"
```
## Co-expressed gene module detection by WGCNA
```{r eval=FALSE}
load("../../Nozue2016_SAStranscriptome_output/output/SAS.expression.vst.s.kazu.largeCV.Rdata") #  where this come from?
#setwd("/Volumes/Data8/NGS_related/Arabidopsis_analysis/SAS_muts_time_course_RNAseq2_mycomp")
load("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/samples.nolow5.Rdata")
# Genotype<-levels(as.factor(samples.nolow5$genotype))
#SAS.expression.vst<-read.csv("../../Nozue2016_SAStranscriptome_data/input/kazu.SAS.expression.VST.082016.csv") # new transformed data by DEseq2 (070416), fixed (082016)
#rownames(SAS.expression.vst)<-SAS.expression.vst$X
# # calculate mean value among replicates?
# load("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/samples.nolow5.Rdata")
# samples.nolow5$sample<-with(samples.nolow5,paste(genotype,trt,time,sep=""))
# samples.nolow5$sample<-factor(samples.nolow5$sample) #where this is used? (122214)
# Genotype used in SAS.expression.vst 
#Genotype<-levels(as.factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|25|49)(A.merged.bam)","\\2",names(SAS.expression.vst)[-1])))
Genotype<-levels(as.factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|25|49)(A.merged.bam)","\\2",names(SAS.expression.vst.s.kazu.largeCV)[-c(1,288)])))
# why 56 modules? (091716)
for(x in Genotype[-1]) {
  SAS.expression.vst.s.kazu.largeCV.x<-SAS.expression.vst.s.kazu.largeCV[,gsub("([C|D|E])([[:digit:]]+)(H|L)(1|4|16|25|49)(A.merged.bam)","\\2",colnames(SAS.expression.vst.s.kazu.largeCV))==x]
    dim(SAS.expression.vst.s.kazu.largeCV.x) #[1] 7438     21
  datExpr <-t(SAS.expression.vst.s.kazu.largeCV.x)
  # Choose a set of soft-thresholding powers
  powers = c(c(1:9), seq(from = 2, to=20, by=10))
  sft = pickSoftThreshold(datExpr, powerVector = powers, verbose = 5)  
  # Plot the results:
  #sizeGrWindow(9, 5)
  pdf(paste("../../Nozue2016_SAStranscriptome_output/output/all",x,".largeCV.softthresholding.pdf",sep=""),width=10,height=8)
  par(mfrow = c(1,2));
  cex1 = 0.9;
  # Scale-free topology fit index as a function of the soft-thresholding power
  plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
       xlab="Soft Threshold (power)",ylab="Scale Free Topology Model Fit,signed R^2",type="n",
       main = paste("Scale independence"));
  text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
       labels=powers,cex=cex1,col="red");
  # this line corresponds to using an R^2 cut-off of h
  abline(h=0.90,col="red")
  # Mean connectivity as a function of the soft-thresholding power
  plot(sft$fitIndices[,1], sft$fitIndices[,5],
       xlab="Soft Threshold (power)",ylab="Mean Connectivity", type="n",
       main = paste("Mean connectivity"))
  text(sft$fitIndices[,1], sft$fitIndices[,5], labels=powers, cex=cex1,col="red")
  dev.off()
  # 
  net = blockwiseModules(datExpr, power = 6,
                         TOMType = "unsigned", minModuleSize = 20,
                         reassignThreshold = 0, mergeCutHeight = 0.25,
                         numericLabels = TRUE, pamRespectsDendro = FALSE,
                         saveTOMs = TRUE,
                         saveTOMFileBase = paste("SAS.expression.vst.s.kazu.largeCV",x,".TOM",sep=""),
                         verbose = 3)
  save(net,file=paste("../../Nozue2016_SAStranscriptome_output/output/SAS.expression.vst.s.kazu.largeCV",x,".Rdata",sep=""))  
  # how to extract module info?
  # open a graphics window
  pdf(file=paste("../../Nozue2016_SAStranscriptome_output/output/all",x,".largeCV.dendrogram.pdf",sep=""),width=10,height=8)
  # Convert labels to colors for plotting
  mergedColors = labels2colors(net$colors)
  # Plot the dendrogram and the module colors underneath
  plotDendroAndColors(net$dendrograms[[1]], mergedColors[net$blockGenes[[1]]],
                      "Module colors",
                      dendroLabels = FALSE, hang = 0.03,
                      addGuide = TRUE, guideHang = 0.05)
  dev.off()
  # save parameters  
  moduleLabels = net$colors
  moduleColors = labels2colors(net$colors)
  MEs = net$MEs
  geneTree = net$dendrograms[[1]]
  save(MEs, moduleLabels, moduleColors, geneTree,file = paste("../../Nozue2016_SAStranscriptome_output/output/all",x,".largeCV.RData",sep=""))
} # recalculating 082016. Recalculating genotype "3" due to "3" was missing (091616)
# new WGCNA modules are too many (110716)
## use recutBlockwiseTrees()
?recutBlockwiseTrees
## for Col (Genotype=1)
x<-Genotype[1]
  SAS.expression.vst.s.kazu.largeCV.x<-SAS.expression.vst.s.kazu.largeCV[,gsub("([C|D|E])([[:digit:]]+)(H|L)(1|4|16|25|49)(A.merged.bam)","\\2",colnames(SAS.expression.vst.s.kazu.largeCV))==x]
    dim(SAS.expression.vst.s.kazu.largeCV.x) #[1] 17757     21
  datExpr <-t(SAS.expression.vst.s.kazu.largeCV.x)
  # Choose a set of soft-thresholding powers
  powers = c(c(1:9), seq(from = 2, to=20, by=10))
  sft = pickSoftThreshold(datExpr, powerVector = powers, verbose = 5)  
  # Plot the results:
  #sizeGrWindow(9, 5)
  pdf(paste("../../Nozue2016_SAStranscriptome_output/output/all",x,".largeCV.softthresholding.pdf",sep=""),width=10,height=8)
  par(mfrow = c(1,2));
  cex1 = 0.9;
  # Scale-free topology fit index as a function of the soft-thresholding power
  plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
       xlab="Soft Threshold (power)",ylab="Scale Free Topology Model Fit,signed R^2",type="n",
       main = paste("Scale independence"));
  text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
       labels=powers,cex=cex1,col="red");
  # this line corresponds to using an R^2 cut-off of h
  abline(h=0.90,col="red")
  # Mean connectivity as a function of the soft-thresholding power
  plot(sft$fitIndices[,1], sft$fitIndices[,5],
       xlab="Soft Threshold (power)",ylab="Mean Connectivity", type="n",
       main = paste("Mean connectivity"))
  text(sft$fitIndices[,1], sft$fitIndices[,5], labels=powers, cex=cex1,col="red")
  dev.off()
  # 
  net = blockwiseModules(datExpr, power = 6,
                         TOMType = "unsigned", minModuleSize = 20,
                         reassignThreshold = 0, mergeCutHeight = 0.25,
                         numericLabels = TRUE, pamRespectsDendro = FALSE,
                         saveTOMs = TRUE,
                         saveTOMFileBase = paste("SAS.expression.vst.s.kazu.largeCV",x,".TOM",sep=""),
                         verbose = 3)
  save(net,file=paste("../../Nozue2016_SAStranscriptome_output/output/SAS.expression.vst.s.kazu.largeCV",x,".Rdata",sep=""))  
  # how to extract module info?
  # open a graphics window
  pdf(file=paste("../../Nozue2016_SAStranscriptome_output/output/all",x,".largeCV.dendrogram.pdf",sep=""),width=10,height=8)
  # Convert labels to colors for plotting
  mergedColors = labels2colors(net$colors)
  # Plot the dendrogram and the module colors underneath
  plotDendroAndColors(net$dendrograms[[1]], mergedColors[net$blockGenes[[1]]],
                      "Module colors",
                      dendroLabels = FALSE, hang = 0.03,
                      addGuide = TRUE, guideHang = 0.05)
  dev.off()
  # save parameters  
  moduleLabels = net$colors
  moduleColors = labels2colors(net$colors)
  MEs = net$MEs
  geneTree = net$dendrograms[[1]]
  save(MEs, moduleLabels, moduleColors, geneTree,file = paste("../../Nozue2016_SAStranscriptome_output/output/all",x,".largeCV.RData",sep=""))

 # recalculating 110716. 
minNSamples<-10
minNGenes<-30
goodSampleGenes.result<-goodSamplesGenes(
  datExpr, 
  minFraction = 1/2, 
  minNSamples = minNSamples, 
  minNGenes = minNGenes, 
  tol = NULL,
  verbose = 1, indent = 0)
####
minModuleSize<-20
recut.net<-recutBlockwiseTrees(datExpr,
  goodSamples=goodSampleGenes.result$goodSamples, 
  goodGenes=goodSampleGenes.result$goodGenes,
  blocks=net$blocks,
  TOMFiles=net$TOMFiles,
  dendrograms=net$dendrograms,
  corType = "pearson",
  networkType = "unsigned",
  deepSplit = 2,
  detectCutHeight = 0.995, minModuleSize = 20,
  maxCoreScatter = NULL, minGap = NULL,
  maxAbsCoreScatter = NULL, minAbsGap = NULL,
  minSplitHeight = NULL, minAbsSplitHeight = NULL,

  useBranchEigennodeDissim = FALSE,
  minBranchEigennodeDissim = mergeCutHeight,

  pamStage = TRUE, pamRespectsDendro = TRUE,
  minCoreKME = 0.5, minCoreKMESize = minModuleSize/3,
  minKMEtoStay = 0.3,
  reassignThreshold = 1e-6,
  mergeCutHeight = 0.1, impute = TRUE,
  trapErrors = FALSE, numericLabels = FALSE,
  verbose = 0, indent = 0)
table(recut.et$color) # this is more cut

## try mergeCloseMOdulesl wutg higher height (0.6) for all Genotype
setwd(homedir)
rm(net)
for(x in Genotype) {
  SAS.expression.vst.s.kazu.largeCV.x<-SAS.expression.vst.s.kazu.largeCV[,gsub("([C|D|E])([[:digit:]]+)(H|L)(1|4|16|25|49)(A.merged.bam)","\\2",colnames(SAS.expression.vst.s.kazu.largeCV))==x]
    dim(SAS.expression.vst.s.kazu.largeCV.x) #[1] 17757     21
  datExpr <-t(SAS.expression.vst.s.kazu.largeCV.x)
  load(paste("../../Nozue2016_SAStranscriptome_output/output/SAS.expression.vst.s.kazu.largeCV",x,".Rdata",sep=""))
merge.net2<-mergeCloseModules(
     # input data
  exprData=datExpr, colors=net$colors,
  # Optional starting eigengenes
  MEs = NULL,
  # Optional restriction to a subset of all sets
  useSets = NULL,
  # If missing data are present, impute them?
  impute = TRUE,
  # Input handling options
  checkDataFormat = TRUE,
  unassdColor = if (is.numeric(colors)) 0 else "grey",
  # Options for eigengene network construction
  corFnc = cor, corOptions = list(use = 'p'),
  useAbs = FALSE,
  # Options for constructing the consensus
  equalizeQuantiles = FALSE,
  quantileSummary = "mean",
  consensusQuantile = 0,
  # Merging options
  cutHeight = 0.6,
  iterate = TRUE,
  # Output options
  relabel = FALSE,
  colorSeq = NULL,
  getNewMEs = TRUE,
  getNewUnassdME = TRUE,
  # Options controlling behaviour of the function
  trapErrors = FALSE,
  verbose = 1, indent = 0)
# save parameters  
  moduleLabels = merge.net2$colors
  moduleColors = labels2colors(merge.net2$colors)
  MEs = merge.net2$MEs
  geneTree = merge.net2$dendrograms[[1]]
  save(MEs, moduleLabels, moduleColors, geneTree,file = paste("../../Nozue2016_SAStranscriptome_output/output/all",x,".largeCV.merge.net2.RData",sep=""))
rm(net)
rm(merge.net2)
}

```

# drawing boxplot (WGCNA module) (061915; no use for SALMON module)
# ```{r eval=FALSE}
# # load("../../Nozue2016_SAStranscriptome_output/output/summary.vst.Col.absolute.SOM.Rdata") # 022115
load("../../Nozue2016_SAStranscriptome_output/output/SAS.expression.vst.s.kazu.largeCV.modules.Rdata") 
# #library(WGCNA)
# SAS.expression.vst.s.kazu.largeCV.modules$AGI<-gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",SAS.expression.vst.s.kazu.largeCV.modules$AGI) 
# 
# # load mean expression value 
# load("../../Nozue2016_SAStranscriptome_output/output/summary.vst.Rdata")
# summary.vst.kazu<-summary.vst[,gsub("([[:digit:]]+)(H|L)(1|4|16|25|49)","\\1",names(summary.vst)) %in% 1:19]
# summary.vst.kazu$AGI<-rownames(summary.vst.kazu)
# summary.vst.kazu.WGCNAmodules<-merge(summary.vst.kazu[,c("1H1","1H25","1H4","1H49","1L1","1L25","1L4","1L49","AGI")],SAS.expression.vst.s.kazu.largeCV.modules[,c("AGI","1")],by="AGI")
# # boxplot
# head(summary.vst.kazu.WGCNAmodules)
# names(summary.vst.kazu.WGCNAmodules)[names(summary.vst.kazu.WGCNAmodules)=="1"]<-"WGCNAmodule"
# summary.vst.kazu.WGCNAmodules.melt<-melt(summary.vst.kazu.WGCNAmodules,id.var=c("AGI","WGCNAmodule"))
# 
# p <- ggplot(summary.vst.kazu.WGCNAmodules.melt,aes(x=variable,y=value,color=value))
# p <- p + geom_point(position="jitter",size=1.5,alpha=0.3) + geom_boxplot(outlier.size=0,aes(fill=factor(WGCNAmodule),alpha=0.8)) + theme_bw() +xlab("sun/shade treatment (hr)") + scale_colour_gradient2(low="green1",mid="gray",high="magenta")
# p <- p + facet_grid(WGCNAmodule~.,scale="free",space="free") #+ facet_wrap(WGCNAmodule~.)
# p <- p + theme(strip.text.x = element_text(angle=90),axis.text.x=element_text(angle=90))
# p
# ## 
# summary.vst.kazu.WGCNAmodules.melt$variable<-factor(summary.vst.kazu.WGCNAmodules.melt$variable,levels=c("1H1","1H4","1H25","1H49","1L1","1L4","1L25","1L49"))
# summary.vst.kazu.WGCNAmodules.melt$light<-gsub("([[:digit:]]+)(H|L)(1|4|25|49)","\\2",summary.vst.kazu.WGCNAmodules.melt$variable)
# 
# p <- ggplot(summary.vst.kazu.WGCNAmodules.melt,aes(x=variable,y=value,color=factor(light)))
# p <- p + geom_point(position="jitter",size=1.5,alpha=0.3) 
# p <- p + geom_boxplot(outlier.size=0,alpha=0.8) + theme_bw() +xlab("sun/shade treatment (hr)") #+ scale_colour_gradient2(low="green1",mid="gray",high="magenta")
# p <- p + facet_grid(WGCNAmodule~.,space="free",scale="free") #+ facet_wrap(WGCNAmodule~.)  
# p <- p + theme(axis.text.x=element_text(angle=90),strip.text.y = element_text(angle=0),legend.position="none")
# p <- p + labs(x="Time (hr)", y="expression level") 
# p <- p + coord_flip()
# #p
# ggsave("boxplot.node.WGCNAmodule.pdf",width=11,height=30,unit="in",path="../../Nozue2016_SAStranscriptome_output/figs_tables")
# # salmon module is interesting pattern
# # only salmon module
# 
# p <- ggplot(summary.vst.kazu.WGCNAmodules.melt[summary.vst.kazu.WGCNAmodules.melt$WGCNAmodule=="salmon",],aes(x=variable,y=value,color=factor(light)))
# p <- p + geom_point(position="jitter",size=1.5,alpha=0.3) 
# p <- p + geom_boxplot(outlier.size=0,alpha=0.8) + theme_bw() +xlab("sun/shade treatment (hr)") #+ scale_colour_gradient2(low="green1",mid="gray",high="magenta")
# p <- p + facet_grid(WGCNAmodule~.,space="free",scale="free") #+ facet_wrap(WGCNAmodule~.)  
# p <- p + theme(axis.text.x=element_text(angle=90),strip.text.y = element_text(angle=0),legend.position="none")
# p <- p + labs(x="Time (hr)", y="expression level") 
# p <- p + coord_flip()
# #p
# ggsave("boxplot.node.WGCNAmodule.salmon.pdf",width=11,height=8,unit="in",path="../../Nozue2016_SAStranscriptome_output/figs_tables") # salmon module is interesting pattern
# 
# # change layout of salmon module boxplot (071915)
# summary.vst.kazu.WGCNAmodules.melt.salmon<-summary.vst.kazu.WGCNAmodules.melt[summary.vst.kazu.WGCNAmodules.melt$WGCNAmodule=="salmon",]
# summary.vst.kazu.WGCNAmodules.melt.salmon$time<-factor(gsub("(1)(H|L)(1|4|25|49)","\\3",summary.vst.kazu.WGCNAmodules.melt.salmon$variable),levels=c("1","4","25","49"))
# summary.vst.kazu.WGCNAmodules.melt.salmon$light<-factor(summary.vst.kazu.WGCNAmodules.melt.salmon$light)
# levels(summary.vst.kazu.WGCNAmodules.melt.salmon$light)<-c("high R/FR","low R/FR")
# p <- ggplot(summary.vst.kazu.WGCNAmodules.melt.salmon,aes(x=time,y=value,color=factor(light)))
# p <- p + geom_boxplot(outlier.size=0,alpha=0.8) + theme_bw() +xlab("sun/shade treatment (hr)") #+ scale_colour_gradient2(low="green1",mid="gray",high="magenta")
# p <- p + geom_point(position="jitter",size=1.5,alpha=0.3) 
# #p <- p + geom_boxplot(outlier.size=NULL,alpha=0.8) + theme_bw() +xlab("sun/shade treatment (hr)")
# p <- p + facet_grid(.~light,space="free",scale="free") #+ facet_wrap(WGCNAmodule~.)  
# p <- p + theme(axis.text.x=element_text(angle=90),strip.text.y = element_text(angle=0),legend.position="none")
# p <- p + labs(x="Time after light treatment (hr)", y="log2 expression level") 
# p
# ggsave("boxplot.node.WGCNAmodule.salmon2.png",width=4,height=3,unit="in",path="../../Nozue2016_SAStranscriptome_output/figs_tables") # salmon should be changed into another module (091616)

```

## overlap table between Col WGCNA modules and mutant modules
```{r eval=FALSE}
setwd(homedir)
setwd("../../Nozue2016_SAStranscriptome_output/output/")
moduleColors.files<-list.files(pattern=".largeCV.merge.net2.RData") # for  merged modules by mergeCloseModules function above for Col (110816)
# load("all1.largeCV.RData")
### use merged modules by mergeCloseModules function above for Col (22 modules; 110716)
load("all1.largeCV.merge.net2.RData")
########
moduleColors.allCol.largeCV<-moduleColors
temp<-moduleColors.allCol.largeCV
for(y in moduleColors.files[!moduleColors.files=="all1.largeCV.merge.net2.RData"]) {
  load(y)
  print(length(moduleColors)) # [1] 17757
  matched.moduleColors<-matchLabels(moduleColors,moduleColors.allCol.largeCV)
  temp<-cbind(temp, matched.moduleColors)
}   
length(moduleColors.files)
temp<-as.data.frame(temp)
names(temp)<-gsub("(all)([[:digit:]]+)(.largeCV.merge.net2.RData)","\\2",c("all1.largeCV.merge.net2.RData",moduleColors.files[!moduleColors.files=="all1.largeCV.merge.net2.RData"]))
SAS.expression.vst.s.kazu.largeCV.modules<-cbind(rownames(SAS.expression.vst.s.kazu.largeCV),temp)
names(SAS.expression.vst.s.kazu.largeCV.modules)[1]<-"AGI"
names(SAS.expression.vst.s.kazu.largeCV.modules)
head(SAS.expression.vst.s.kazu.largeCV.modules)
setwd(homedir)
save(SAS.expression.vst.s.kazu.largeCV.modules,file="../../Nozue2016_SAStranscriptome_output/output/SAS.expression.vst.s.kazu.largeCV.modules110816.Rdata") #110816
```

# overlap between WGCNA Col modules and mutation affected genes
```{r,eval=FALSE}
setwd(homedir)
### I did the same calculation above? (091616)
setwd(homedir)
load("../../Nozue2016_SAStranscriptome_output/output/DE.genotype.batch.genes.unique.DF.Rdata")
load("../../Nozue2016_SAStranscriptome_output/output/SAS.expression.vst.s.kazu.largeCV.modules110816.Rdata")
# 
#setwd(homedir)
# load("../../Nozue2016_SAStranscriptome_output/output/SAS.expression.vst.s.kazu.largeCV.modules.Rdata")
load("../../Nozue2016_SAStranscriptome_output/output/SAS.expression.vst.s.kazu.largeCV.modules110816.Rdata")
#library(WGCNA)
SAS.expression.vst.s.kazu.largeCV.modules$AGI<-gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",SAS.expression.vst.s.kazu.largeCV.modules$AGI)
# merge with WGCNA modules
test<-merge(DE.genotype.batch.genes.unique.DF,SAS.expression.vst.s.kazu.largeCV.modules[,c("AGI","1")],by="AGI",all=TRUE)
DE.genotype.batch.genes.WGCNAmodules<-test
DE.genotype.batch.genes.WGCNAmodules[is.na(DE.genotype.batch.genes.WGCNAmodules)]<-"0"
names(DE.genotype.batch.genes.WGCNAmodules)<-sub("1","WGCNAmodules",names(DE.genotype.batch.genes.WGCNAmodules))

write.csv(DE.genotype.batch.genes.WGCNAmodules,file="../../Nozue2016_SAStranscriptome_output/figs_tables/DE.genotype.batch.genes.WGCNAmodules.csv")
DE.genotype.batch.genes.WGCNAmodules<-read.csv("../../Nozue2016_SAStranscriptome_output/figs_tables/DE.genotype.batch.genes.WGCNAmodules.csv")
DE.genotype.batch.genes.WGCNAmodules<-DE.genotype.batch.genes.WGCNAmodules[,-1]

length(table(DE.genotype.batch.genes.WGCNAmodules$WGCNAmodules)) # 23 modules (110816)
# adding gene description
# TAIR10_gene_descriptions<-read.csv("../../Nozue2016_SAStranscriptome_data/input/TAIR10_functional_descriptions.csv") 
# TAIR10_gene_descriptions$AGI<-gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",TAIR10_gene_descriptions$"Model_name")
DE.genotype.batch.genes.WGCNAmodules.description<-merge(DE.genotype.batch.genes.WGCNAmodules,TAIR10_gene_descriptions[!duplicated(TAIR10_gene_descriptions$AGI),c("AGI","Short_description","colE...col.F")],by="AGI")
#names(DE.genotype.batch.genes.WGCNAmodules.description)<-sub("X1","WGCNAmodules",names(test))
write.csv(DE.genotype.batch.genes.WGCNAmodules.description,file="../../Nozue2016_SAStranscriptome_output/figs_tables/Dataset7_Complete_gene_list_of_genes_misexpressed_in_mutant_and_found_in_WGCNA_modules_visualized_in_Fig7.csv") #091616
# in whitney
test<-read.csv("../../Nozue2016_SAStranscriptome_output/figs_tables/Dataset7_Complete_gene_list_of_genes_misexpressed_in_mutant_and_found_in_WGCNA_modules_visualized_in_Fig7.csv")
test<-test[,-1]
dim(test) # 17985   28
#test<-DE.genotype.batch.genes.WGCNAmodules
# overlap table (revsised into GOseq.overlapTable; 052616)
# with ,permutation=100 (060316)
i<-2
overlap<-GOseq.overlapTable(test[,"WGCNAmodules"],test[,i],genes=test$AGI,permutation=100)
overlap.melt<-melt(overlap$pTable,id=1,permutation=100) #
overlap.melt$count<-melt(overlap$countTable,id=1)$value
overlap.melt.summary<-overlap.melt[!overlap.melt$Var2=="0",]
for(i in 3:29) {
overlap<-GOseq.overlapTable(test[,"WGCNAmodules"],test[,i],genes=test$AGI,permutation=100)
overlap.melt<-melt(overlap$pTable,id=1) #
overlap.melt$count<-melt(overlap$countTable,id=1)$value
overlap.melt<-overlap.melt[!overlap.melt$Var2=="0",]
overlap.melt.summary<-rbind(overlap.melt.summary,overlap.melt)
}
overlap.melt.summary[,"value"]<--log10(overlap.melt.summary[,"value"])
overlap.melt.summary[overlap.melt.summary$value>10,"value"]<-10
overlap.melt.summary[overlap.melt.summary$count==0,"count"]<-""
#table(overlap.melt.summary[,"value"])
# graph
#overlap.melt.summary.s<-overlap.melt.summary[!overlap.melt.summary$Var2 %in% c("bsk5_1.sun","bsk5_1.shade","AT5G02540_1.sun","AT5G02540_1.shade"),]
save(overlap.melt.summary,file="../../Nozue2016_SAStranscriptome_output/output/overlap.melt.summary.WGCNA.mut.GOseq.Rdata") # 110816

# reading and cleaning overlap table
# load("../../Nozue2016_SAStranscriptome_output/output/overlap.melt.summary.s.Rdata")
load("../../Nozue2016_SAStranscriptome_output/output/overlap.melt.summary.s.WGCNA.mut.GOseq.Rdata") # calculated in whitney (052616); (110816, testing with permutation =100)
overlap.melt.summary.s
overlap.melt.summary.ss<-overlap.melt.summary.s[!overlap.melt.summary.s$Var1=="0",]

q4<-ggplot(overlap.melt.summary.ss,aes(y=factor(Var1),x=factor(Var2))) 
#q4 <- q4 + geom_tile(color="black",aes(fill=value))+scale_fill_gradient2(low=muted("green"), high=muted("magenta"),midpoint=2) # this color coding is not good (061716)
q4 <- q4 + geom_tile(color="black",aes(fill=value))+scale_fill_gradient(low="white", high="magenta") # Is this better? (061716)

#q4 <- q4 + theme(axis.text.x=element_text(angle=90),axis.title=element_blank())
#q4 <- q4 +theme(legend.position="none") # this removed all legend
q4 <- q4 + theme(axis.text.x=element_text(angle=90,hjust=1),
                 axis.ticks = element_line(size=0,color="white"))
q4 <- q4 + geom_text(aes(label=count),size=5)
q4 <- q4 + labs(x="genotype specific genes",y="WGCNA modules in Col",title="modules affected in SAS mutants",fill="-log10\n pvalue")
#q4 <- q4 + theme(legend.title=element_blank())
#ggsave(q4,file="Col.WGCNA.modules.vs.GenotypeSpecific.genes.table.pdf",width=11,height=8)
# ggsave(q4,file="Col.WGCNA.modules.vs.GenotypeSpecific.genes.table.s.pdf",width=11,height=8,path="../../Nozue2016_SAStranscriptome_output/figs_tables")
# ggsave(q4,file="Col.WGCNA.modules.vs.GenotypeSpecific.genes.table.s.png",width=11,height=8,path="../../Nozue2016_SAStranscriptome_output/figs_tables")
ggsave(q4,file="Fig7_Col.WGCNA.modules.vs.GenotypeSpecific.genes.table.s.GOseq11081616.pdf",width=11,height=8,path="../../Nozue2016_SAStranscriptome_output/figs_tables")
# ggsave(q4,file="Fig7_Col.WGCNA.modules.vs.GenotypeSpecific.genes.table.s.GOseq061816.png",width=11,height=8,path="../../Nozue2016_SAStranscriptome_output/figs_tables")

q4 # magenta indicates pvalue is smaller than 0.01.
#### only "salmon","pink" (=JA),"purple" (=auxin) for job application 2015 - 2016 ###
overlap.melt.summary.ss.selected<-overlap.melt.summary.ss[overlap.melt.summary.ss$Var1 %in% c("salmon","purple","pink"),]
q5<-ggplot(overlap.melt.summary.ss.selected,aes(y=factor(Var1),x=factor(Var2))) 
q5 <- q5 + geom_tile(color="black",aes(fill=value))+scale_fill_gradient2(low=muted("green"), high=muted("magenta"),midpoint=2)
q5 <- q5 + theme(axis.text.x=element_text(angle=90,hjust=1),
                 panel.grid.major = element_blank(),
                 panel.grid.minor = element_blank(),
                 panel.background = element_blank(),
                 axis.ticks = element_line(size=0,color="white"))
q5 <- q5 + geom_text(aes(label=count),size=5,show_guide = FALSE)

q5 <- q5 + labs(x="genotype specific genes",y="WGCNA modules in Col",title="modules affected in SAS mutants",fill="-log10\n p-value")
q5 <- q5 + theme(legend.title=element_blank())
#ggsave(q5,file="Col.WGCNA.modules.vs.GenotypeSpecific.genes.table.s.selected.pdf",width=11,height=4,path="../../Nozue2016_SAStranscriptome_output/figs_tables")
ggsave(q5,file="Fig6_Col.WGCNA.modules.vs.GenotypeSpecific.genes.table.s.selected.GOseq.pdf",width=11,height=4,path="../../Nozue2016_SAStranscriptome_output/figs_tables")
```
# Number of Col genes in each module are: 
```{r eval=FALSE} 
table(SAS.expression.vst.s.kazu.largeCV.modules$"1") 
```
# The next question is what are these modules?  GOseq analysis of each cluster
```{r eval=FALSE}
setwd(homedir)
# load("../../Nozue2016_SAStranscriptome_output/output/SAS.expression.vst.s.kazu.largeCV.modules.Rdata")
load("../../Nozue2016_SAStranscriptome_output/output/SAS.expression.vst.s.kazu.largeCV.modules110816.Rdata")
setwd("../../Nozue2016_SAStranscriptome_output/output/")
SAS.expression.vst.s.kazu.largeCV.modules$AGI<-gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",SAS.expression.vst.s.kazu.largeCV.modules$AGI)
# prep for GO category
  ###Read in AtGO
  Atgo <- toTable(org.At.tairGO)
  #head(Atgo)
  BP <- TRUE #only keep BP go TERMS
  if (BP) Atgo <- Atgo[Atgo$Ontology=="BP",]
  #convert to list
  Atgo.list <- tapply(Atgo$go_id,Atgo$gene_id,c)
# 
modules<-levels(as.factor(SAS.expression.vst.s.kazu.largeCV.modules$"1"))
# no enriched GO; "grey60","lightyellow","magenta"
# Warning messages:1: In pcls(G) : initial point very close to some inequality constraints in "yellow"
for(x in modules) {
    temp.GOseq<-GOseq.ORA(SAS.expression.vst.s.kazu.largeCV.modules[SAS.expression.vst.s.kazu.largeCV.modules$"1"==x,"AGI"])
  if(temp.GOseq=="no enriched GO") {next} else {
  temp.genes<-genes.in.enriched.category(enrich.result=temp.GOseq,gene.list=SAS.expression.vst.s.kazu.largeCV.modules[SAS.expression.vst.s.kazu.largeCV.modules$"1"==x,"AGI"],category.table=Atgo.list)
  save(temp.GOseq,temp.genes,file=paste("WGCNA.Col.module.",x,".GOseq.Rdata",sep=""))
}
  } # 091716
```
## write summary of WGCNA.Col.module.GOseq 
```{r eval=FALSE}
rm(list=ls())
setwd("../../Nozue2016_SAStranscriptome_output/output/")
WGCNA.Col.module.GOseq.files<-list.files(pattern="^WGCNA.Col.module.")
WGCNA.Col.module.GOseq.files<-grep(".GOseq.Rdata",WGCNA.Col.module.GOseq.files,value=TRUE)
load(WGCNA.Col.module.GOseq.files[1])
temp.GOseq$modules<-gsub("(WGCNA.Col.module.)([[:print:]]+)(.GOseq.Rdata)","\\2",WGCNA.Col.module.GOseq.files[1])
temp<-temp.GOseq
 
for(i in WGCNA.Col.module.GOseq.files[-1]) {
  load(i)
  temp.GOseq$modules<-gsub("(WGCNA.Col.module.)([[:print:]]+)(.GOseq.Rdata)","\\2",i)
  temp<-rbind(temp,temp.GOseq)
}
WGCNA.Col.module.GOseq.summary<-temp
write.csv(WGCNA.Col.module.GOseq.summary,file="../../Nozue2016_SAStranscriptome_output/figs_tables/Dataset5_GO_enrichment_analysis_of_WGCNA_modules.csv") # Dataset 5. GO enrichment analysis of WGCNA gene co-expression modules (091716)
```
## What genes are found in given module with given GO category?
```{r eval=FALSE}
# "GO:0007623"" in "yellow"
load("WGCNA.Col.module.yellow.GOseq.Rdata")
# TAIR10_gene_descriptions<-read.csv("/Volumes/Data8/NGS_related/Arabidopsis_analysis/reference/TAIR10_functional_descriptions.csv") 
# head(TAIR10_gene_descriptions)
# TAIR10_gene_descriptions$AGI<-gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",TAIR10_gene_descriptions$"Model_name")

temp.genes.wdescrition<-merge(temp.genes,TAIR10_gene_descriptions,by="AGI")
dim(temp.genes)
temp.genes.wdescrition<-temp.genes.wdescrition[!duplicated(temp.genes.wdescrition$AGI),]
#temp.genes.wdescrition[temp.genes.wdescrition$"GO:0007623"==1,c("AGI","Short_description","colE...col.F")]
temp.genes.wdescrition[temp.genes.wdescrition$"GO:0007623"==1,c("AGI","Short_description")]
# GO:0009697, "salmon"
load("WGCNA.Col.module.salmon.GOseq.Rdata")
temp.genes.wdescrition<-merge(temp.genes,TAIR10_gene_descriptions,by="AGI")
temp.genes.wdescrition<-temp.genes.wdescrition[!duplicated(temp.genes.wdescrition$AGI),]
#temp.genes.wdescrition[temp.genes.wdescrition$"GO:0009697"==1,c("AGI","Short_description")]
temp.genes.wdescrition[temp.genes.wdescrition$"GO:0009697"==1,c("AGI","Short_description","colE...col.F")]
# GO:0009825, "royalblue"
load("WGCNA.Col.module.royalblue.GOseq.Rdata")
temp.genes.wdescrition<-merge(temp.genes,TAIR10_gene_descriptions,by="AGI")
temp.genes.wdescrition<-temp.genes.wdescrition[!duplicated(temp.genes.wdescrition$AGI),]
temp.genes.wdescrition[temp.genes.wdescrition$"GO:0009825"==1,c("AGI","Short_description","colE...col.F")]
#          AGI                                   Short_description
# 4  AT1G23080                 Auxin efflux carrier family protein
# 13 AT1G75780                                tubulin beta-1 chain
# 14 AT2G06850         xyloglucan endotransglucosylase/hydrolase 4
# 19 AT2G38120 Transmembrane amino acid transporter family protein
# 23 AT3G23050                              indole-3-acetic acid 7
# 41 AT5G15580                                         longifolia1
temp.genes.wdescrition[temp.genes.wdescrition$"GO:0000271"==1,c("AGI","Short_description","colE...col.F")]
# the same list

```

## GOseq analysis of each cluster (cellular component)
```{r eval=FALSE}
# reading necessary funcitons and libraries
source("/Volumes/Data8/NGS_related/Arabidopsis_analysis/SAS_muts_time_course_RNAseq2_mycomp/function_RNAseq_time_course.R")
# load("../../Nozue2016_SAStranscriptome_output/output/SAS.expression.vst.s.kazu.largeCV.modules.Rdata")
load("../../Nozue2016_SAStranscriptome_output/output/SAS.expression.vst.s.kazu.largeCV.modules110816.Rdata")
#library(WGCNA)
setwd("/Volumes/Data8/NGS_related/Arabidopsis_analysis/SAS_muts_time_course_RNAseq2_mycomp_output")
SAS.expression.vst.s.kazu.largeCV.modules$AGI<-gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",SAS.expression.vst.s.kazu.largeCV.modules$AGI)
# prep for GO category
  ###Read in AtGO
  Atgo <- toTable(org.At.tairGO)
  #head(Atgo)
  CC <- TRUE #only keep CC go TERMS
  if (CC) Atgo <- Atgo[Atgo$Ontology=="CC",] # cellular component
  #convert to list
  Atgo.list <- tapply(Atgo$go_id,Atgo$gene_id,c)
# 
modules<-levels(as.factor(SAS.expression.vst.s.kazu.largeCV.modules$"1"))
# no enriched GO; 
for(x in modules[!modules %in% c("darkred","greenyellow","grey","grey60","lightcyan","purple","salmon","yellow")]) { # 
  temp.GOseq<-GOseq.CC.ORA(SAS.expression.vst.s.kazu.largeCV.modules[SAS.expression.vst.s.kazu.largeCV.modules$"1"==x,"AGI"])
  temp.genes<-genes.in.enriched.category(enrich.result=temp.GOseq,gene.list=SAS.expression.vst.s.kazu.largeCV.modules[SAS.expression.vst.s.kazu.largeCV.modules$"1"==x,"AGI"],category.table=Atgo.list)
  save(temp.GOseq,temp.genes,file=paste("GOCC.WGCNA.Col.module.",x,".GOseq.Rdata",sep=""))
}
```
## write summary of WGCNA.Col.module.GOseq.CC 
```{r eval=FALSE}
setwd(homedir)
setwd("../../Nozue2016_SAStranscriptome_output/output/")
rm(list=ls())
GOCC.WGCNA.Col.module.GOseq.files<-list.files(pattern="GOCC.WGCNA.Col.module.")
GOCC.WGCNA.Col.module.GOseq.files<-grep(".GOseq.Rdata",GOCC.WGCNA.Col.module.GOseq.files,value=TRUE)
load(GOCC.WGCNA.Col.module.GOseq.files[1])
temp.GOseq$modules<-gsub("(GOCC.WGCNA.Col.module.)([[:print:]]+)(.GOseq.Rdata)","\\2",GOCC.WGCNA.Col.module.GOseq.files[1])
temp<-temp.GOseq
 
for(i in GOCC.WGCNA.Col.module.GOseq.files[-1]) {
  load(i)
  temp.GOseq$modules<-gsub("(GOCC.WGCNA.Col.module.)([[:print:]]+)(.GOseq.Rdata)","\\2",i)
  temp<-rbind(temp,temp.GOseq)
}
WGCNA.Col.module.GOseq.CC.summary<-temp
write.csv(WGCNA.Col.module.GOseq.CC.summary,file="../../Nozue2016_SAStranscriptome_output/figs_tables/WGCNA.Col.module.GOseq.CC.summary.csv") # not used
```
# overlap analysis with custome categories (WGCNA modules)
```{r eval=FALSE}
library(WGCNA);library(ggplot2);library(reshape2);library(scales)
# custom categories
load("../../Nozue2016_SAStranscriptome_data/input/hormone.responsiveness4.Rdata")
## or change list into data.frame
hormone.responsiveness4.DF<-data.frame(t(sapply(hormone.responsiveness4,c))) # modified for ver4
names(hormone.responsiveness4.DF)<-c("ABAup","ABAdown","ACCup","ACCdown","BLup","BLdown","IAAup","IAAdown","MJup","MJdown","CKup","CKdown","GAup","GAdown","SAup","SAdown","SARminus","SARplus","GRF1_3target","PIFtarget","MYC234up","MYC234down","NPR1up","NPR1down","WRKY33up","WRKY33down","PCC1up","PCC1down")
hormone.responsiveness4.DF$AGI<-rownames(hormone.responsiveness4.DF)

load("../../Nozue2016_SAStranscriptome_output/output/SAS.expression.vst.s.kazu.largeCV.modules.Rdata") # is this current version? (100716)
SAS.expression.vst.s.kazu.largeCV.modules$AGI<-gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",SAS.expression.vst.s.kazu.largeCV.modules$AGI)
# merge with WGCNA modules
test<-merge(hormone.responsiveness4.DF,SAS.expression.vst.s.kazu.largeCV.modules[,c("AGI","1")],by="AGI",all=TRUE)
names(test)[names(test)=="1"]<-"WGCNAmodules"
i <- sapply(test, is.factor)
test[i]<-lapply(test[i], as.character)
class(test[,2])
  
hormone.responsiveness.WGCNAmodules<-test
hormone.responsiveness.WGCNAmodules[hormone.responsiveness.WGCNAmodules==""]<-"0"
hormone.responsiveness.WGCNAmodules[is.na(hormone.responsiveness.WGCNAmodules)]<-"0"
head(hormone.responsiveness.WGCNAmodules)

# write.csv(hormone.responsiveness.WGCNAmodules,file="../../Nozue2016_SAStranscriptome_output/figs_tables/TableS9_hormone.responsiveness.WGCNAmodules.csv") #
# Plant Cell version
write.csv(hormone.responsiveness.WGCNAmodules,file="../../Nozue2016_SAStranscriptome_output/figs_tables/Supplemental_Dataset_6_Custom_category_enrichment_analysis_of_WGCNA_gene_co-expression_modules.csv")



setwd("../../Nozue2016_SAStranscriptome_output/output/")
hormone.responsiveness.WGCNAmodules<-hormone.responsiveness.WGCNAmodules[,-1]
# remove GRF1_3target (052616)
hormone.responsiveness.WGCNAmodules<-hormone.responsiveness.WGCNAmodules[,!names(hormone.responsiveness.WGCNAmodules)=="GRF1_3target"]
test<-hormone.responsiveness.WGCNAmodules
# overlap table (revised to GOseq.overlapTable model, 052616, but scripts stoped. I am trying method="sampling" in dgoseq, though it takes time).
# Error in pWNCHypergeo(num_de_incat, num_incat, num_genes - num_incat,  : 
#   Inconsistency. mean = 280, lower limit = 264, upper limit = 279 
i<-2
overlap<-GOseq.overlapTable(test[,"WGCNAmodules"],test[,i],genes=test$AGI,permutation=100) # default permutaiton is 2000 in goseq(method="Sampling")
overlap.melt<-melt(overlap$pTable,id=1) #
overlap.melt$count<-melt(overlap$countTable,id=1)$value
overlap.melt.summary<-overlap.melt[!overlap.melt$Var2=="",]
for(i in 3:28) {
overlap<-GOseq.overlapTable(test[,"WGCNAmodules"],test[,i],genes=test$AGI)
overlap.melt<-melt(overlap$pTable,id=1) #
overlap.melt$count<-melt(overlap$countTable,id=1)$value
overlap.melt<-overlap.melt[!overlap.melt$Var2=="",]
overlap.melt.summary<-rbind(overlap.melt.summary,overlap.melt)
}
overlap.melt.summary[,"value"]<--log10(overlap.melt.summary[,"value"])
overlap.melt.summary[overlap.melt.summary$value>10,"value"]<-10
overlap.melt.summary[overlap.melt.summary$count==0,"count"]<-""

# graph
custom.category.overlap.melt.summary.s<-overlap.melt.summary
custom.category.overlap.melt.summary.s<-custom.category.overlap.melt.summary.s[!custom.category.overlap.melt.summary.s$Var1=="0",]
custom.category.overlap.melt.summary.s<-custom.category.overlap.melt.summary.s[!custom.category.overlap.melt.summary.s$Var2=="0",]
save(custom.category.overlap.melt.summary.s,file="../../../Nozue2016_SAStranscriptome_output/output/custom.category.v4.overlap.overlap.melt.summary.s.GOseq.Rdata") # in whiteney
load("../../Nozue2016_SAStranscriptome_output/output/custom.category.v4.overlap.overlap.melt.summary.s.GOseq.Rdata")
q4<-ggplot(custom.category.overlap.melt.summary.s,aes(y=factor(Var1),x=factor(Var2))) 
#q4 <- q4 + geom_tile(color="black",aes(fill=value))+scale_fill_gradient2(low=muted("green"), high=muted("magenta"),midpoint=2)
q4 <- q4 + geom_tile(color="black",aes(fill=value))+scale_fill_gradient(low="white", high="magenta")
#q4 <- q4 + theme(axis.text.x=element_text(angle=90),axis.title=element_blank())
q4 <- q4 + theme(axis.text.x=element_text(angle=90,hjust=1),
                 panel.grid.major = element_blank(),
                 panel.grid.minor = element_blank(),
                 panel.background = element_blank(),
                 axis.ticks = element_line(size=0,color="white"))
q4 <- q4 + geom_text(aes(label=count),size=5)
q4 <- q4 + labs(x="custom category",y="WGCNA modules in Col",title="custom category enrichment",fill="-log10\n pvalue")
#ggsave(q4,file="Col.WGCNA.modules.vs.GenotypeSpecific.genes.table.pdf",width=11,height=8)
ggsave(q4,file="Fig6_Col.WGCNA.modules.vs.my.category.table.s.061816.pdf",width=11,height=8,path="../../Nozue2016_SAStranscriptome_output/figs_tables")
ggsave(q4,file="Fig6_Col.WGCNA.modules.vs.my.category.table.s.061816.png",width=11,height=8,path="../../Nozue2016_SAStranscriptome_output/figs_tables")

q4 # magenta indicates pvalue is smaller than 0.01.
```



# Checking expression pattern of gene of interests (absolute and shade responsiveness) (060515)
```{r}
rm(list=ls())
# absolute value
load("../../Nozue2016_SAStranscriptome_output/output/summary.vst.Rdata")
summary.vst.kazu<-summary.vst[,gsub("([[:digit:]]+)(H|L)(1|4|16|25|49)","\\1",names(summary.vst)) %in% 1:19] 
# plot
p<-expression.pattern.graph3(summary.vst.kazu,"AT2G46970") # PIL1
p
# fold chagne (shade responsiveness)
load("../../Nozue2016_SAStranscriptome_output/output/summary.vst.response.Rdata") # response. needs to fix colnames (060515)
summary.vst.response.kazu<-summary.vst.response[,gsub("([[:digit:]]+)(_)(1|4|16|25|49)(hrA)","\\1",colnames(summary.vst.response)) %in% 1:19]
dim(summary.vst.response.kazu)
head(summary.vst.response.kazu)
q<-expression.pattern.graph2b(summary.vst.response.kazu,"AT2G46970") # PIL1
q

```
# GOseq analysis of misexpressed geens in phyB (sun) and public available phytochrome mutnats (exposed to long light condition) (060815) 
```{r eval=FALSE}
# my timecourse RNAseq data
setwd("/Volumes/Data8/NGS_related/Arabidopsis_analysis/SAS_muts_time_course_RNAseq2_mycomp_output")
load("./edgeR/outlier_wo16h/genotype_specific5/DE.genotype.batch.sun.gt.wo16h6.Rdata") # sun, phyB vs Col any timepoints
# load functions for GOseq
source("/Volumes/Data8/NGS_related/Arabidopsis_analysis/SAS_muts_time_course_RNAseq2_mycomp/function_RNAseq_time_course.R")
###Read in AtGO
  Atgo <- toTable(org.At.tairGO)
  #head(Atgo)
  BP <- TRUE #only keep BP go TERMS
  if (BP) Atgo <- Atgo[Atgo$Ontology=="BP",]
  #convert to list
  Atgo.list <- tapply(Atgo$go_id,Atgo$gene_id,c)
#
temp.GOseq<-GOseq.ORA(gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",rownames(DE.batch3.gt))) # JA and SA pathways enriched, no auxin?! as I expected.
write.csv(temp.GOseq,file="../../Nozue2016_SAStranscriptome_output/figs_tables/TableS5_GOseq.phyB.sun.csv")
temp.genes<-genes.in.enriched.category(enrich.result=temp.GOseq,gene.list=gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",rownames(DE.batch3.gt)),category.table=Atgo.list) 

# predFC (061015)
predFC.phyB<-predFC(dge.nolow.s.batch3.glm,design.nolow.s.gt.batch3,prior.count=1,dispersion=0.05)
head(predFC.phyB[rownames(predFC.phyB) %in% rownames(DE.batch3.gt),]) #
tail(predFC.phyB[rownames(predFC.phyB) %in% rownames(DE.batch3.gt),]) #
hist(predFC.phyB[,1]) # I do not understand why there are -15.... 

# plot 
load("../../Nozue2016_SAStranscriptome_output/output/summary.vst.Rdata")
summary.vst.kazu<-summary.vst[,gsub("([[:digit:]]+)(H|L)(1|4|16|25|49)","\\1",names(summary.vst)) %in% 1:19]
# larger gene sets
SAS.expression.vst.large<-read.csv("/Volumes/Data8/NGS_related/Arabidopsis_analysis/SAS_muts_time_course_RNAseq2_mycomp/kazu.SAS.expression.VST.112614.csv") # new transformed data by DEseq2 (done by J, see VST.R)
rownames(SAS.expression.vst.large)<-SAS.expression.vst.large$X
# calculate mean value among replicates
load("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/samples.nolow5.Rdata")
samples.nolow5$sample<-with(samples.nolow5,paste(genotype,trt,time,sep=""))
samples.nolow5$sample<-factor(samples.nolow5$sample) #where this is used? (122214)
SAS.expression.vst.large.s<-SAS.expression.vst.large[,names(SAS.expression.vst.large) %in% samples.nolow5$file]
Genotype<-levels(as.factor(samples.nolow5$genotype))
Genotype<-Genotype[!(Genotype=="11"|Genotype=="12"|Genotype=="18"|Genotype=="16")]
temp<-SAS.expression.vst.large.s[,names(SAS.expression.vst.large.s) %in% samples.nolow5[samples.nolow5$genotype==1,"file"]]
summary.vst.large<-expression.mean(temp)ã€€# custom function
for(i in Genotype[-1]) {
  temp<-SAS.expression.vst.large.s[,names(SAS.expression.vst.large.s) %in% samples.nolow5[samples.nolow5$genotype==i,"file"]]
	temp<-as.data.frame(temp)
	temp2<-expression.mean(temp)
	summary.vst.large<-cbind(summary.vst.large,temp2)
}
#head(summary.vst,3)
rownames(summary.vst.large)<-gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",SAS.expression.vst.large$X)
save(summary.vst.large,file="../../Nozue2016_SAStranscriptome_output/output/summary.vst.large.Rdata") #
###
load("../../Nozue2016_SAStranscriptome_output/output/summary.vst.large.Rdata")
summary.vst.large.kazu<-summary.vst.large[,gsub("([[:digit:]]+)(H|L)(1|4|16|25|49)","\\1",names(summary.vst.large)) %in% 1:19]

p<-expression.pattern.graph3(summary.vst.large.kazu,gsub("([[:print:]]+)(.[[:digit:]])","\\1",rownames(DE.batch3.gt)[1:10]))
p
# downregulated in phyB
p<-expression.pattern.graph3(summary.vst.large.kazu,gsub("([[:print:]]+)(.[[:digit:]])","\\1",rownames(DE.batch3.gt[order(DE.batch3.gt[,1]),][1:20,]))) # downregulated in phyB at time=1h (logFC.genotype6<0)
p<-p + labs(title="phyB logFC.genotype6 <0 top 20")
#ggsave("phyB_down_expression_pattern.pdf",height=20,width=8,path="../../Nozue2016_SAStranscriptome_output/figs_tables")
# GOseq with down regulated in phyB
phyBdown.GOseq<-GOseq.ORA(gsub("([[:print:]]+)(.[[:digit:]])","\\1",rownames(DE.batch3.gt[DE.batch3.gt[,1]>1,]))) # JA and SA pathways enriched, no auxin?! as I expected.
write.csv(phyBdown.GOseq,file="../../Nozue2016_SAStranscriptome_output/figs_tables/TableS1_phyBdown.GOseq.csv")
# upregulated in phyB
q<-expression.pattern.graph3(summary.vst.large.kazu,gsub("([[:print:]]+)(.[[:digit:]])","\\1",rownames(DE.batch3.gt[order(DE.batch3.gt[,1],decreasing=TRUE),][1:20,])) # upregulated in phyB at time=1h
q
# GOseq up in phyB
phyBup.GOseq<-GOseq.ORA(gsub("([[:print:]]+)(.[[:digit:]])","\\1",rownames(DE.batch3.gt[DE.batch3.gt[,1]< -0.5,]))) # JA and SA pathways enriched, no auxin?! as I expected.
write.csv(phyBup.GOseq,file="../../Nozue2016_SAStranscriptome_output/figs_tables/TableS1_phyBup.GOseq.csv")
#
p<-expression.pattern.graph3(summary.vst.large.kazu,gsub("([[:print:]]+)(.[[:digit:]])","\\1",rownames(DE.batch3.gt[DE.batch3.gt[,"logCPM"]< -1,]))) # logCPM is "average log2-counts per million, the average taken over all libraries in y." (help file of glmLRT)
# high-order phytochrome muntat seedlings (microarray)
# download some cel file from E-GEOD-31587 - Expression data of Arabidopsis thaliana (Ler accession) phytochrome phyABCDE quintuple mutant and phyABDE quadruple mutant in response to red light, and their comparison to WT expression data (http://www.ebi.ac.uk/arrayexpress/experiments/E-GEOD-31587/samples/?query=phytochrome&sortby=organism&sortorder=ascending)
# under construction #
setwd("/Volumes/Data8/E-GEOD-31587_RAW")
library(affy)
#Data <- ReadAffy(widget=T) # error
CEL.filename<-list.files(pattern=".CEL$")
CEL.filename
Data <- ReadAffy(filenames=CEL.filename)
Data
eset <- rma(Data) # normalize chips 
affyData <- read.csv("/Volumes/Data6/bioconductor_R/affy_ATH1_array_elements-2010-12-20.csv") #read TAIR annotation data from file (TAIR10)
affyData2<-affyData[,c("array_element_name","is_control","locus","description","chromosome","start","stop")]
names(affyData2) <- c("element","is.control","locus","description","chromosome","start","stop")#get rid of underlines
head(affyData2)
genes <- data.frame(element=featureNames(eset)) #get affy gene names from dataset
affyMatch <- merge(affyData2,genes,by.x="element",by.y=1)# reorder the annotation data to match chip data
exp.values <- data.frame(exprs(eset)) #put normalized intensities in a data frame
head(exp.values)
exp.values <- cbind(affyMatch,exp.values) #add annotation to same frame as expression values
head(exp.values)

write.csv(exp.values,file="/Volumes/Data8/E-GEOD-31587_RAW/exp.values.csv")
######## RP analysis
library(RankProd)
exp.values<-exp.values[,-1]
names(exp.values)
names(exp.values)[c(8:12)]
RP.E_GEOD_31587.phyABDE <- RP(exp.values[,c(8:12)], rep(c(1,0),c(3,2)), num.perm=100) #  0 (class 1) = Col, 1 (class 2)= phyABDE
topgene.RP.E_GEOD_31587.phyABDE<-topGene(RP.E_GEOD_31587.phyABDE,cutoff=0.01,gene.names=exp.values$element,logged=TRUE,logbase=2)
#  repressed genes in phyABDE
topgene.RP.E_GEOD_31587.phyABDE.T2<-as.data.frame(topgene.RP.E_GEOD_31587.phyABDE$Table2)
topgene.RP.E_GEOD_31587.phyABDE.T2$element<-rownames(topgene.RP.E_GEOD_31587.phyABDE.T2)
topgene.RP.E_GEOD_31587.phyABDE.T2.m<-merge(topgene.RP.E_GEOD_31587.phyABDE.T2, affyData2,by="element")
head(topgene.RP.E_GEOD_31587.phyABDE.T2.m)
topgene.RP.E_GEOD_31587.phyABDE.T2.m.exp.value<-merge(topgene.RP.E_GEOD_31587.phyABDE.T2.m, exp.values[,c(1,8:12)],by="element")
topgene.RP.E_GEOD_31587.phyABDE.T2.m.exp.value<=topgene.RP.E_GEOD_31587.phyABDE.T2.m.exp.value[order(topgene.RP.E_GEOD_31587.phyABDE.T2.m.exp.value$pfp),]

#  induced genes in phyABDE
topgene.RP.E_GEOD_31587.phyABDE.T1<-as.data.frame(topgene.RP.E_GEOD_31587.phyABDE$Table1)
topgene.RP.E_GEOD_31587.phyABDE.T1$element<-rownames(topgene.RP.E_GEOD_31587.phyABDE.T1)
topgene.RP.E_GEOD_31587.phyABDE.T1.m<-merge(topgene.RP.E_GEOD_31587.phyABDE.T1, affyData2,by="element")
head(topgene.RP.E_GEOD_31587.phyABDE.T1.m)
topgene.RP.E_GEOD_31587.phyABDE.T1.m.exp.value<-merge(topgene.RP.E_GEOD_31587.phyABDE.T1.m, exp.values[,c(1,8:12)],by="element")
topgene.RP.E_GEOD_31587.phyABDE.T1.m.exp.value<-topgene.RP.E_GEOD_31587.phyABDE.T1.m.exp.value[order(topgene.RP.E_GEOD_31587.phyABDE.T1.m.exp.value$pfp),]

# summary
topgene.RP.RP.E_GEOD_31587.phyABDE.summary<-rbind(topgene.RP.E_GEOD_31587.phyABDE.T1.m.exp.value,topgene.RP.E_GEOD_31587.phyABDE.T2.m.exp.value)
write.csv(topgene.RP.RP.E_GEOD_31587.phyABDE.summary,file="/Volumes/Data8/E-GEOD-31587_RAW/topgene.RP.RP.E_GEOD_31587.phyABDE.summary.csv")
# GOseq
topgene.RP.RP.E_GEOD_31587.phyABDE.summary.locus<-topgene.RP.RP.E_GEOD_31587.phyABDE.summary[nchar(topgene.RP.RP.E_GEOD_31587.phyABDE.summary$locus)==9,"locus"]
phyABDE.seedlings.GOseq<-GOseq.ORA(topgene.RP.RP.E_GEOD_31587.phyABDE.summary.locus)
phyABDE.seedlings.GOseq.genes<-genes.in.enriched.category(enrich.result=temp.GOseq,gene.list=topgene.RP.RP.E_GEOD_31587.phyABDE.summary.locus],category.table=Atgo.list) #  no auxin?! as I expected.
```

# misc
```{r}
# SA related genes are shade-responsive? (071115)
load("../../Nozue2016_SAStranscriptome_output/output/newdata.vst.Rdata") 
# conversion.table<-data.frame(num=1:27, genotype=c("Col","AT5G02540_1","hy5","jar1","kat1_2","phyB","pif45","spt_11","yuc2589","PAR1_RNAi09","coi1_16","phyAB","pif3","AT5G02760_2","AT5G59010_2","sto","aos","argos","co_9","Blh_1","Jea","Shahdara","Col_0","Cvi_0","Bur_0","Oy_0","Ita_0"))
newdata.vst$gt2<-""

for(i in 1:27){
  newdata.vst[newdata.vst$gt==conversion.table[i,1],"gt2"]<-as.character(conversion.table[i,2])  
}
newdata.vst[newdata.vst$AGI %in% c("AT3G48090","AT4G35580","AT1G74710","AT5G08330","AT3G52430","AT3G26830","AT5G26920","AT2G46970"),c("1h","4h","25h","49h","ssom3.4.unit.classif","gt2","AGI")] 
# NTL9 (AT4G35580), CHE (AT5G08330), ICS1(AT1G74710, SID2), PAD4 (AT3G52430), PAD3 (AT3G26830), CBP60g (AT5G26920), PIL1 (AT2G46970) as a positive control for this script. 

```
# visualize hormone pathway related genes (1213-1214,2015). Updated (052616-)
# new FDR version (071716)
```{r}
setwd(homedir)
my.category.diff.heatmap.graph<-list()
categories <- read.csv("../../Nozue2016_SAStranscriptome_output/figs_tables//TableS7_test.wozero.melt.hormone.signifiant.s.csv",row.names=1,as.is=TRUE)
summary(categories)
head(categories)
categories$my.category<-factor(categories$my.category,levels=rev(c("IAAup","BLup","BLdown","PIFtarget","ACCup","ABAup","ABAdown","MJup","MJdown","MYC234up","SARplus","NPR1down","WRKY33up")))
# summary.vst.response.kazu
load("../../Nozue2016_SAStranscriptome_output/output/summary.vst.response.Rdata") # response. needs to fix colnames (060515)
summary.vst.response.kazu<-summary.vst.response[,gsub("([[:digit:]]+)(_)(1|4|16|25|49)(hrA)","\\1",colnames(summary.vst.response)) %in% 1:19]
dim(summary.vst.response.kazu)
head(summary.vst.response.kazu)
# for Fig5A, Col
my.category.heatmap.graph<-list()
for(i in c(1,17,19,3,4,5,14,10,6,13,7,8,9)) {
my.category.heatmap.graph[[i]]<-my.category.heatmap5(summary3.response=summary.vst.response.kazu,gt.num=conversion.table[i,1],gt2=conversion.table[i,2],newdata.Col.wGO.selected=categories,plotname="my.category.expression.pattern.v10.png",w=5,h=5*4/3, path="../../Nozue2016_SAStranscriptome_output/figs_tables/",save.plot=F,legend.fill=T)
}
# drawing heatmap graph (difference): Fig5B
my.category.diff.data<-list()

for(i in as.character(c(17,19,3,4,5,14,10,6,13,7,8,9))) { # c("aos","co_9","hy5","jar1","kat1_2","mida9_4","PAR1","phyb","pif3","pif45","spt_11","yucQ")
my.category.diff.data[[i]]<-my.category.diff(summary3.response=summary.vst.response.kazu,gt.num=conversion.table[i,1],gt2=conversion.table[i,2],newdata.Col.wGO.selected=categories,plotname="my.category.expression.pattern.v8.png",w=5,h=5*4/3, path="../../Nozue2016_SAStranscriptome_output/figs_tables/",save.plot=F,legend.fill=T) # make specialized for difference heatmap function
}
# need to create diff data frame for (mut - Col)
summary.vst.response.kazu.diffCol<-as.data.frame(my.category.diff.data)  # under construction
max.value<-max(summary.vst.response.kazu.diffCol[,grep(".value",colnames(summary.vst.response.kazu.diffCol))])
min.value<-min(summary.vst.response.kazu.diffCol[,grep(".value",colnames(summary.vst.response.kazu.diffCol))])
max.min<-c(round(min.value,1),round(max.value,1))
my.category.diff.heatmap.graph<-list()

for(i in c(17,19,3,4,5,14,10,6,13,7,8,9)) { # c("aos","co_9","hy5","jar1","kat1_2","mida9_4","PAR1","phyb","pif3","pif45","spt_11","yucQ")
my.category.diff.heatmap.graph[[i]]<-my.category.diff.heatmap1(summary3.response=summary.vst.response.kazu,gt.num=conversion.table[i,1],gt2=conversion.table[i,2],newdata.Col.wGO.selected=categories,plotname="my.category.expression.pattern.v8.png",w=5,h=5*4/3, path="../../Nozue2016_SAStranscriptome_output/figs_tables/",save.plot=F,legend.fill=T,max.min.value=max.min) # make specialized for difference heatmap function
}
# significance
significant.position<-data.frame(category=c("IAAup","BLup","BLdown","PIFtarget","ACCup","ABAup","ABAdown","MJup","MJdown","MYC234up","SARplus","NPR1down","WRKY33up"),x=0,y=1:13,symbol.gt="",symbol.gt.time="")
# write how to calculate "Fig5B_response_p_values.csv"

# read p value
response.p.values<-read.csv("../../Nozue2016_SAStranscriptome_output/output/Fig5B_response_p_values.csv") #
response.p.values.melt<-melt(response.p.values[,-1],id="gt") # this is wrong

for(i in c(17,19,3,4,5,14,10,6,13,7,8,9)) {
response.p.values.melt.gt<-response.p.values.melt[response.p.values.melt$gt==i,]  
significant.position$symbol.gt<-(as.numeric(as.character(vlookup(significant.position$category,response.p.values.melt.gt[,c(2,3)],2,F)))<0.0001)*1 # significant level 
  significant.position$symbol.gt<-gsub("1","*",significant.position$symbol.gt)
  significant.position$symbol.gt<-gsub("0","",significant.position$symbol.gt)

# add significance to heatmap
my.category.diff.heatmap.graph[[i]]<-my.category.diff.heatmap.graph[[i]] + geom_text(data=significant.position,aes(x=x-0.2,y=y,label=symbol.gt),size=20,color="red") #, remove lines for x and y axis. "gt" is red.
}
### use geom_blank() for blank plot
library(cowplot)
my.category.heatmap.all <- plot_grid(my.category.heatmap.graph[[1]],ggplot()+geom_blank(),ggplot()+geom_blank(),my.category.diff.heatmap.graph[[17]],my.category.diff.heatmap.graph[[19]],my.category.diff.heatmap.graph[[3]],my.category.diff.heatmap.graph[[4]],my.category.diff.heatmap.graph[[5]],my.category.diff.heatmap.graph[[14]],my.category.diff.heatmap.graph[[10]],my.category.diff.heatmap.graph[[6]],my.category.diff.heatmap.graph[[13]],my.category.diff.heatmap.graph[[7]],my.category.diff.heatmap.graph[[8]],my.category.diff.heatmap.graph[[9]],
                      labels=c("A","","","B","","","","","","","","","","",""),
                      vjust = 1.1, hjust=-0.5,
                      ncol=3,nrow=5,
                      label_size = 50)
setwd("../../Nozue2016_SAStranscriptome_output/figs_tables/")
save_plot("Fig5_all.my.category.diff.expression.pattern.v4.FDR0.0001.png", my.category.heatmap.all,
          ncol = 3, # we're saving a grid plot of 2 columns
          nrow = 5, # and 2 rows
          base_aspect_ratio = 2.5,
          base_height=6,
          limitsize = FALSE
)
# supplement Fig2 (absolute value)
# extract legend
# since Col with legend is different in size from others. I need to extract legend and plot it in different space.
# see http://stackoverflow.com/questions/12041042/how-to-plot-just-the-legends-in-ggplot2
library(gridExtra)
g_legend<-function(a.gplot){ 
  tmp <- ggplot_gtable(ggplot_build(a.gplot)) 
  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box") 
  legend <- tmp$grobs[[leg]] 
  return(legend)} 
legend <- g_legend(my.category.heatmap.graph[[1]]) 
#
my.category.heatmap.all <- plot_grid(my.category.heatmap.graph[[1]]+theme(legend.position="none"),legend ,ggplot()+geom_blank(),my.category.heatmap.graph[[17]],my.category.heatmap.graph[[19]],my.category.heatmap.graph[[3]],my.category.heatmap.graph[[4]],my.category.heatmap.graph[[5]],my.category.heatmap.graph[[14]],my.category.heatmap.graph[[10]],my.category.heatmap.graph[[6]],my.category.heatmap.graph[[13]],my.category.heatmap.graph[[7]],my.category.heatmap.graph[[8]],my.category.heatmap.graph[[9]],
                      labels=c("A","","","B","","","","","","","","","","",""),
                      vjust = 1.1, hjust=-0.5,
                      ncol=3,nrow=5,
                      label_size = 50)
#
setwd("../../Nozue2016_SAStranscriptome_output/figs_tables/")
save_plot("Fig_S2_all.my.category.expression.pattern.v10.png", my.category.heatmap.all,
          ncol = 3, # we're saving a grid plot of 2 columns
          nrow = 5, # and 2 rows
          base_aspect_ratio = 2.5,
          base_height=6,
          limitsize = FALSE
)
```

# Does each WGCNA module contain shade-responsive genes? (011216)
```{r}
SAS.expression.vst.s.kazu.largeCV.modules
# shade responsive genes 
load("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/DGEsummary.edgeR.response.Rdata") # this has all genotypes
head(DGEsummary.edgeR.response)
Col.DGE<-DGEsummary.edgeR.response[1:80,]
Col.DGE$AGI<-gsub("([[:print:]]+)(.[[:digit:]]+)(.[[:digit:]]+)","\\1",rownames(Col.DGE))
# mark Col.DGE genes in SAS.expression.vst.s.kazu.largeCV.modules
SAS.expression.vst.s.kazu.largeCV.modules$ColDGE<-(SAS.expression.vst.s.kazu.largeCV.modules$AGI %in% Col.DGE$AGI)*1
# 
overlap.WGCNAmodule.ColDGE<-overlapTable(SAS.expression.vst.s.kazu.largeCV.modules$"1",SAS.expression.vst.s.kazu.largeCV.modules$ColDGE)
save(overlap.WGCNAmodule.ColDGE,file="../../Nozue2016_SAStranscriptome_output/output/overlap.WGCNAmodule.ColDGE.Rdata")
temp<-merge(Col.DGE,TAIR10_gene_descriptions[,c("AGI","colE...col.F")],by="AGI")
Col.DGE.descripiton<-temp[!duplicated(temp$AGI),]
write.csv(Col.DGE.descripiton,file="../../Nozue2016_SAStranscriptome_output/output/Col.DGE.descripiton.csv")

```
# Are salmon genes modulated in Sessa (2007) 4d samples?
```{r}
Sessa.4d<-read.csv("../../Nozue2016_SAStranscriptome_data/input/sessa4days.csv")
Sessa.1h<-read.csv("../../Nozue2016_SAStranscriptome_data/input/sessa1hour.csv")
DE.genotype.batch.genes.WGCNAmodules<-read.csv("../../Nozue2016_SAStranscriptome_output/figs_tables/DE.genotype.batch.genes.WGCNAmodules.csv")
salmon<-DE.genotype.batch.genes.WGCNAmodules[DE.genotype.batch.genes.WGCNAmodules$X1=="salmon",]
salmon$AGI
dim(Sessa.4d[Sessa.4d$ID %in% salmon$AGI,])
Sessa.4d.salmon<-Sessa.4d[Sessa.4d$ID %in% salmon$AGI,]
write.csv(Sessa.4d.salmon,file="Sessa.4d.salmon.csv")
ConvictionsSessa.1h[Sessa.1h$ID %in% salmon$AGI,]
length(salmon$AGI)
Sessa.4d.s<-Sessa.4d[Sessa.4d$adj.P.Val<0.01,]
# Fisher's exact test (assuming total gene number is 25000)
Sessa.4d.salmon<-matrix(c(11,1941-11,112-11,25000-1941-(112-11)),nrow=2)
fisher.test(Sessa.4d.salmon) # p-value = 0.3768
```
# Salmon module genes are shade responsive in Col?
```{r}
DEfiles<-list.files(path="../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/",pattern="DE.batch.anytime")
setwd("../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/")
load(DEfiles[1]) 
rownames(DE.batch.anytime)<-paste(rownames(DE.batch.anytime),gsub("(DE.batch.anytime_genotype_wo_outlier_wo16h)([[:print:]]+)(.Rdata)","\\2",DEfiles[1]),sep=".")
DE.batch.anytime[sub("([[:print:]]+)(.[[:digit:]]+)","\\1",rownames(DE.batch.anytime)) %in% salmon$AGI,] # none
# Salmon module genes do not have any shade responsive genes in Col...

```
# Salmon module genes vs 49h shade responsive genes in Col
```{r}
DE.genotype.batch.genes.WGCNAmodules<-read.csv("../../Nozue2016_SAStranscriptome_output/figs_tables/DE.genotype.batch.genes.WGCNAmodules.csv")
salmon<-DE.genotype.batch.genes.WGCNAmodules[DE.genotype.batch.genes.WGCNAmodules$X1=="salmon",]
load("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/samples.nolow5.Rdata")
load("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/dge.nolow5.Rdata")
  samples.nolow.s<-samples.nolow5[samples.nolow5$genotype==1,]
  dge.nolow.s<-dge.nolow5[,samples.nolow5$genotype==1]
  dge.nolow.s<-dge.nolow.s[rowSums(dge.nolow.s$counts > 5) >= 3,] 
# subset only 49h
samples.nolow.s.49h<-samples.nolow.s[samples.nolow.s$time=="49",]
design.nolow.s.batch.fixedtime<- model.matrix(~samples.nolow.s.49h$trt + samples.nolow.s.49h$batch)
colnames(design.nolow.s.batch.fixedtime) <- gsub("samples.nolow.s.49h$","",colnames(design.nolow.s.batch.fixedtime),fixed=TRUE) # nicer column names
colnames(design.nolow.s.batch.fixedtime)
dge.nolow.s.fixedtime<-dge.nolow5[,samples.nolow5$file %in% samples.nolow.s.49h$file]
  print(design.nolow.s.batch.fixedtime)
  print(dge.nolow.s.fixedtime)
  dge.nolow.s.batch.fixedtime.glm <- estimateGLMCommonDisp(dge.nolow.s.fixedtime,design.nolow.s.batch.fixedtime, verbose=TRUE) 
  dge.nolow.s.batch.glm <- estimateGLMTrendedDisp(dge.nolow.s.batch.fixedtime.glm,design.nolow.s.batch.fixedtime, verbose=TRUE)
  dge.nolow.s.batch.fixedtime.glm.fit <- glmFit(dge.nolow.s.batch.fixedtime.glm,design.nolow.s.batch.fixedtime)
  dge.nolow.s.batch.fixedtime.glm.lrt <- glmLRT(dge.nolow.s.batch.fixedtime.glm.fit,coef="trtL")
  DE.batch.49h.Col<-topTags(dge.nolow.s.batch.fixedtime.glm.lrt,n=2000)$table[topTags(dge.nolow.s.batch.fixedtime.glm.lrt,n=2000)$table$FDR<0.05,]
# 
dim(DE.batch.49h.Col) # 278 5
DE.batch.49h.Col.salmon<-DE.batch.49h.Col[sub("([[:print:]]+)(.[[:digit:]]+)","\\1",rownames(DE.batch.49h.Col)) %in% salmon$AGI,]
setwd("../../Nozue2016_SAStranscriptome_output/output/")
write.csv(DE.batch.49h.Col.salmon,file="DE.batch.49h.Col.salmon.csv")
setwd("../../Nozue2016_SAStranscriptome_scripts/")
```
# appendix 1: differentially expressed geens used for ORA
```{r eval=FALSE}
# differential expression analysis with npr1-1
# http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE51626
# Singh M, Bag SK, Bhardwaj A, Ranjan A et al. Global nucleosome positioning 
# regulates salicylic acid mediated transcription in Arabidopsis thaliana. BMC Plant # Biol 2015 Jan 21;15:13. PMID: 25604550
###  The RNA of col-0 and npr1-1 mutant were isolated after 24hr water and SA treatment.The ATH1 gene chip was used for expression analysis.
###  GSM1249675  Col-0 water, biological replicate 1
###  GSM1249676  Col-0 water, biological replicate 2
###  GSM1249677  Col-0 SA, biological replicate 1
###  GSM1249678  Col-0 SA, biological replicate 2
###  GSM1249679	npr1-1 water, biological replicate 1
###  GSM1249680	npr1-1 water, biological replicate 2
###  GSM1249681	npr1-1 SA, biological replicate 1
###  GSM1249682	npr1-1 SA, biological replicate 2
setwd("../../Nozue2016_SAStranscriptome_data/input/GSE51626_RAW")
library(affy)
#Data <- ReadAffy(widget=T) # error
CEL.filename<-list.files(pattern=".CEL$")
CEL.filename
Data <- ReadAffy(filenames=CEL.filename)
Data
eset <- rma(Data) # normalize chips 
affyData <- read.csv("../../Nozue2016_SAStranscriptome_data/input/affy_ATH1_array_elements-2010-12-20.csv") #read TAIR annotation data from file (TAIR10)
affyData2<-affyData[,c("array_element_name","is_control","locus","description","chromosome","start","stop")]
names(affyData2) <- c("element","is.control","locus","description","chromosome","start","stop")#get rid of underlines
head(affyData2)
genes <- data.frame(element=featureNames(eset)) #get affy gene names from dataset
affyMatch <- merge(affyData2,genes,by.x="element",by.y=1)# reorder the annotation data to match chip data
exp.values <- data.frame(exprs(eset)) #put normalized intensities in a data frame
head(exp.values)
exp.values <- cbind(affyMatch,exp.values) #add annotation to same frame as expression values
head(exp.values)
write.csv(exp.values,file="../../Nozue2016_SAStranscriptome_data/input/GSE51626_RAW/exp.values.csv")
######## RP analysis
library(RankProd)
exp.values<-read.csv("../../Nozue2016_SAStranscriptome_data/inpu/GSE51626_RAW/exp.values.csv")
names(exp.values)
names(exp.values)[c(9:10,13:14)]
RP.GSE51626.npr1 <- RP(exp.values[,c(9:10,13:14)], rep(1:0,each=2), num.perm=100) # water control, 0 (class 1) = Col, 1 (class 2)= npr1-1 
topgene.RP.GSE51626.npr1<-topGene(RP.GSE51626.npr1,cutoff=0.01,gene.names=exp.values$element,logged=TRUE,logbase=2)
#  induced genes in npr1-1
topgene.RP.GSE51626.npr1.T2<-as.data.frame(topgene.RP.GSE51626.npr1$Table2)
topgene.RP.GSE51626.npr1.T2$element<-rownames(topgene.RP.GSE51626.npr1.T2)
topgene.RP.GSE51626.npr1.T2.m<-merge(topgene.RP.GSE51626.npr1.T2, affyData2,by="element")
head(topgene.RP.GSE51626.npr1.T2.m)
topgene.RP.GSE51626.npr1.T2.m.exp.value<-merge(topgene.RP.GSE51626.npr1.T2.m, exp.values[,c(2,9:10,13:14)],by="element")
topgene.RP.GSE51626.npr1.T2.m.exp.value<=topgene.RP.GSE51626.npr1.T2.m.exp.value[order(topgene.RP.GSE51626.npr1.T2.m.exp.value$pfp),]

#  repressed genes in npr1-1
topgene.RP.GSE51626.npr1.T1<-as.data.frame(topgene.RP.GSE51626.npr1$Table1)
topgene.RP.GSE51626.npr1.T1$element<-rownames(topgene.RP.GSE51626.npr1.T1)
topgene.RP.GSE51626.npr1.T1.m<-merge(topgene.RP.GSE51626.npr1.T1, affyData2,by="element")
head(topgene.RP.GSE51626.npr1.T1.m)
topgene.RP.GSE51626.npr1.T1.m.exp.value<-merge(topgene.RP.GSE51626.npr1.T1.m, exp.values[,c(2,9:10,13:14)],by="element")
topgene.RP.GSE51626.npr1.T1.m.exp.value<-topgene.RP.GSE51626.npr1.T1.m.exp.value[order(topgene.RP.GSE51626.npr1.T1.m.exp.value$pfp),]

# summary
topgene.RP.GSE51626.npr1.summary<-rbind(topgene.RP.GSE51626.npr1.T1.m.exp.value,topgene.RP.GSE51626.npr1.T2.m.exp.value)
write.csv(topgene.RP.GSE51626.npr1.summary,file="../../Nozue2016_SAStranscriptome_data/input/GSE51626_RAW/topgene.RP.GSE51626.npr1.summary.csv")

# ### other SA microarray
# http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE3984
# 3w to 4w old plants were sprayed with mock or SA with 0.01% Silwet. Leaves were collected 2h after treatment.
# Supplementary data files not provided
# use SOFT file (http://www2.warwick.ac.uk/fac/sci/moac/people/students/peter_cock/r/geo/)
source("http://www.bioconductor.org/biocLite.R")
setwd("/Volumes/Data8/")
biocLite("GEOquery")
library(Biobase)
library(GEOquery)
#Download GDS file, put it in the current directory, and load it:
#gds3984 <- getGEO('GSE3984', destdir=".")

#Or, open an existing GDS file (even if its compressed):
gse3984 <- getGEO(filename='GSE3984_family.soft.gz')
Meta(gds3984)
eset <- GDS2eSet(gds3984, do.log2=TRUE)
# error
# see  "4.4  Converting GSE to an ExpressionSet" in vignet of GEOquery
gsmplatforms <- lapply(GSMList(gse3984),function(x) {Meta(x)$platform})
Table(GSMList(gse3984)[[1]])[1:5,]
probesets <- Table(GPLList(gse3984)[[1]])$ID
data.matrix <- do.call('cbind',lapply(GSMList(gse3984),function(x) 
{tab <- Table(x) 
 mymatch <- match(probesets,tab$ID_REF) 
 return(tab$VALUE[mymatch])
}))
data.matrix <- apply(data.matrix,2,function(x) {as.numeric(as.character(x))})
hist(data.matrix) # log2 transformed data
#data.matrix <- log2(data.matrix) # is this already transformed?
data.matrix[1:5,]
rownames(data.matrix) <- probesets
colnames(data.matrix) <- names(GSMList(gse3984))
pdata <- data.frame(samples=names(GSMList(gse3984)))
rownames(pdata) <- names(GSMList(gse3984))
pheno <- as(pdata,"AnnotatedDataFrame")
eset2 <- new('ExpressionSet',exprs=data.matrix,phenoData=pheno)
eset2
genes <- data.frame(element=featureNames(eset2)) #get affy gene names from dataset
affyMatch <- merge(affyData2,genes,by.x="element",by.y=1)# reorder the annotation data to match chip data
exp.values <- data.frame(exprs(eset2)) #put normalized intensities in a data frame
head(exp.values)
exp.values <- cbind(affyMatch,exp.values) #add annotation to same frame as expression values
head(exp.values)
# GSM90867  2-hour mock-treated leaf material biological replicate 1
# GSM90868  2-hour mock-treated leaf material biological replicate 2
# GSM90869  2-hour mock-treated leaf material biological replicate 3
# GSM90870	leaf material harvested 2 hours after spraying of 1 mM salicylic acid biological replicate 1
# GSM90871	leaf material harvested 2 hours after spraying of 1 mM salicylic acid biological replicate 2
# GSM90872	leaf material harvested 2 hours after spraying of 1 mM salicylic acid biological replicate 3
# RP
library(RankProd)
RP.ThibaudNissen.SA <- RP(exp.values[,c(8:13)], rep(1:0,each=3), num.perm=100) # 0 (class 1) = mock, 1 (class 2)= SA
#  induced genes
topgene.RP.ThibaudNissen.SA<-topGene(RP.ThibaudNissen.SA,cutoff=0.01,gene.names=rownames(exp.values),logged=TRUE,logbase=2)
topgene.RP.ThibaudNissen.SA.T2<-as.data.frame(topgene.RP.ThibaudNissen.SA$Table2)
topgene.RP.ThibaudNissen.SA.T2$element<-rownames(topgene.RP.ThibaudNissen.SA.T2)
topgene.RP.ThibaudNissen.SA.T2.m<-merge(topgene.RP.ThibaudNissen.SA.T2, affyData2,by="element")
head(topgene.RP.ThibaudNissen.SA.T2.m)
#  repressed genes
topgene.RP.ThibaudNissen.SA.T1<-as.data.frame(topgene.RP.ThibaudNissen.SA$Table1)
topgene.RP.ThibaudNissen.SA.T1$element<-rownames(topgene.RP.ThibaudNissen.SA.T1)
topgene.RP.ThibaudNissen.SA.T1.m<-merge(topgene.RP.ThibaudNissen.SA.T1, affyData2,by="element")
head(topgene.RP.ThibaudNissen.SA.T1.m)

## limma (see limmar user guide, 091114)
design <- cbind(mock=c(1,1,1,0,0,0),SA=c(0,0,0,1,1,1))
# or
Group <- factor(c("mock","mock","mock","SA","SA","SA"), levels=c("mock","SA"))
design <- model.matrix(~0+Group)                
colnames(design) <- c("mock","SA")

fit <- lmFit(exp.values[,c(8:13)], design)
cont.matrix <- makeContrasts(mockvsSA=SA-mock, levels=design)
fit2 <- contrasts.fit(fit, cont.matrix)
fit2 <- eBayes(fit2,0.01) # 0.01 is used in GEO2R (http://www.ncbi.nlm.nih.gov/geo/geo2r/?acc=GSE3984)
topgene.limma.ThibaudNissen.SA<-topTable(fit2, adjust="fdr",sort.by="B",number=500) # this is different from GEO2R output ...
topgene.limma.ThibaudNissen.SA<-topgene.limma.ThibaudNissen.SA[topgene.limma.ThibaudNissen.SA$adj.P.Val<0.05,]

# overlap between RP and limma
topgene.RP.ThibaudNissen.SA.T1.m.limma<-topgene.RP.ThibaudNissen.SA.T1.m[topgene.RP.ThibaudNissen.SA.T1.m$element %in% rownames(topgene.limma.ThibaudNissen.SA),]
topgene.RP.ThibaudNissen.SA.T2.m.limma<-topgene.RP.ThibaudNissen.SA.T2.m[topgene.RP.ThibaudNissen.SA.T2.m$element %in% rownames(topgene.limma.ThibaudNissen.SA),]
setwd("/Volumes/Data8/NGS_related/Arabidopsis_analysis")
write.csv(topgene.RP.ThibaudNissen.SA.T1.m.limma,file="topgene.RP.ThibaudNissen.SA.T1.m.limma.csv")
write.csv(topgene.RP.ThibaudNissen.SA.T2.m.limma,file="topgene.RP.ThibaudNissen.SA.T2.m.limma.csv")
```
# 
```{r}
## start from this line ## 052616
# fold chagne (shade responsiveness)
load("../../Nozue2016_SAStranscriptome_output/output/summary.vst.response.Rdata") # response. needs to fix colnames (060515)
summary.vst.response.kazu<-summary.vst.response[,gsub("([[:digit:]]+)(_)(1|4|16|25|49)(hrA)","\\1",colnames(summary.vst.response)) %in% 1:19]
dim(summary.vst.response.kazu)
head(summary.vst.response.kazu)
# find overlap between shade responsive (in Col) and hormone responsive
test<-read.csv("../../Nozue2016_SAStranscriptome_output/figs_tables/TableS4_hormone.responsiveness.SOMclusters.csv")
test<-test[,-1]
# reading overlapTable
load("SAS_muts_time_course_RNAseq2_mycomp_output/overlap.melt.summary.SOM.custom.Rdata") 
overlap.melt.summary.s<-overlap.melt.summary[!overlap.melt.summary$Var1==0,]
#overlap.melt.summary.s$Var1<-factor(overlap.melt.summary.s$Var1,levels=c(3,6,9,5,8,2,4,11,1,12,7,10)) #
overlap.melt.summary.s$Var1<-factor(overlap.melt.summary.s$Var1,levels=c(10, 7,4,12,11,9,6,5,2,1,3,8)) #
overlap.melt.summary.s<-overlap.melt.summary.s[!overlap.melt.summary.s$Var2=="",]
overlap.melt.summary.s$Var2<-factor(overlap.melt.summary.s$Var2,levels=c("IAAup","IAAdown","BLup","BLdown","MJup","MJdown","GAup","GAdown","ABAup","ABAdown","ACCup","ACCdown","CKup","CKdown","SAup","SAdown","SARplus","SARminus","GRF1_3target","PIFtarget","MYC234up","MYC234down","NPR1up","NPR1down","WRKY33up","WRKY33down"))
# select log10>2 & !Var1==0 
overlap.melt.summary.significant<-overlap.melt.summary[overlap.melt.summary$value> -log10(0.05),] # pvalue<0.05
overlap.melt.summary.significant$category.cluster<-with(overlap.melt.summary.significant,paste(Var2,Var1,sep="_"))
# 
names(test)[2:27]<-c("ABAup","ABAdown","ACCup","ACCdown","BLup","BLdown","IAAup","IAAdown","MJup","MJdown","CKup","CKdown","GAup","GAdown","SAup","SAdown","SARminus","SARplus","GRF1_3target","PIFtarget","MYC234up","MYC234down","NPR1up","NPR1down","WRKY33up","WRKY33down")

test.wozero<-test[!test$ssom3.4.unit.classif=="0",]
head(test.wozero)
test.wozero.melt<-melt(test.wozero,id.var=c("AGI","ssom3.4.unit.classif"))
test.wozero.melt$category.cluster<-with(test.wozero.melt,paste(value,ssom3.4.unit.classif,sep="_"))

test.wozero.melt.hormone.signifiant<-test.wozero.melt[test.wozero.melt$category.cluster %in% overlap.melt.summary.significant$category.cluster,]
head(test.wozero.melt.hormone.signifiant)
# remove "GRF1_3target" and "GAdown"
test.wozero.melt.hormone.signifiant<-test.wozero.melt.hormone.signifiant[!(test.wozero.melt.hormone.signifiant$value %in% c("GRF1_3target","GAdown","IAAdown","ACCdown","MYC234down","SAup","GAup")),] # more than 3 genes
test.wozero.melt.hormone.signifiant$value<-as.character(test.wozero.melt.hormone.signifiant$value)
# test.wozero.melt.hormone.signifiant.s is my category marker gene list 
test.wozero.melt.hormone.signifiant.s<-test.wozero.melt.hormone.signifiant[,c("AGI","value")]
names(test.wozero.melt.hormone.signifiant.s)<-c("AGI","my.category")
test.wozero.melt.hormone.signifiant.s$my.category<-factor(test.wozero.melt.hormone.signifiant.s$my.category,levels=rev(c("IAAup","BLup","BLdown","PIFtarget","ACCup","ABAup","ABAdown","MJup","MJdown","MYC234up","SARplus","NPR1down","WRKY33up"))) # remove GAdown (052616)
write.csv(test.wozero.melt.hormone.signifiant.s,file="../../Nozue2016_SAStranscriptome_output/output/test.wozero.melt.hormone.signifiant.s.csv")
```
# appendix 2  "Should I filter probesets or genes?
*Probesets or genes may be filtered by mean expression or variance (or their robust analogs such as median and median absolute deviation, MAD) since low-expressed or non-varying genes usually represent noise. Whether it is better to filter by mean expression or variance is a matter of debate; both have advantages and disadvantages, but more importantly, they tend to filter out similar sets of genes since mean and variance are usually related.
*We do not recommend filtering genes by differential expression. WGCNA is designed to be an unsupervised analysis method that clusters genes based on their expression profiles. Filtering genes by differential expression will lead to a set of correlated genes that will essentially form a single (or a few highly correlated) modules. It also completely invalidates the scale-free topology assumption, so choosing soft thresholding power by scale-free topology fit will fail. "
### http://labs.genetics.ucla.edu/horvath/CoexpressionNetwork/Rpackages/WGCNA/faq.html
#### it is OK to filter genes by variance (= largeCV genes in my scripts)
# session info
```{r}
sessionInfo()
```
