---
"title: "SAS mutants time-course RNAseq analysis"
output: html_document
---
## revision history
* 
## To do
* 
## outline of this scripts
* eliminating libraries with wrong genotypes
* eliminating outlier libraries
+ Use samples.nolow5.Rdata object for select libraries
* differentially expressed gene analysis
+ 1. DE (treatment, ie, shade responsiveness) at any time point within given genotype
+ 2. DE at any time point (shade)
* VST 
* SOM analysis with DE(1) genes (using shade-responsiveness (shade/sun))
* GOseq analysis of SOM clusters
* overlap between SOM clusters and misexpressed genes in mutants
* co-expression analysis by WGCNA (all Col samples)
## revise history
* rewrite setting directory to be compatible to new RStudio
* Do not use setwd()
* SAS.expression.vst<-read.csv("../Nozue2016_SAStranscriptome_data/input/kazu.SAS.expression.VST.082016.csv") # new transformed data by DEseq2 (082016) is not reproducible (092617). Probably due to updated DESeq2?
* Use kazu.SAS.expression.VST.092517.csv (092617) for rest of analysis
* Replace "samples.nolow5" into "PLoSGenetics2015mutants.errrDE" (100717)
* Replace SOM analysis with BH-SNE anallysis (111517)
* Cleanup unused lines (113017)
* SAup_combined int SAup (020618)
## preparation
```{r}
knitr::opts_chunk$set(echo = TRUE)
homedir<-getwd()
homedir
# [1] "/Volumes/data_work/Data8/NGS_related/Arabidopsis_analysis/Nozue2016_SAStranscriptome_finalfinal_newVST/Nozue2016_SAStranscriptome_scripts/SAS_muts_time_course_RNAseq_finalfinal" # 
# reading necessary funcitons and libraries
source(file.path("function_RNAseq_time_course.R"))
```
## eliminating libraries with wrong genotypes
```{r eval=FALSE}
getwd()
counts<-read.table(file.path(homedir,"..","..","Nozue2016_SAStranscriptome_data/input/SAS_counts_merged_updated_final.bam.tsv"),header=TRUE) # use SAS_counts_merged_updated.bam.tsv (081514)
# remove * row and move gene column into rowname
rownames(counts)<-counts$gene
counts<-counts[!counts$gene=="*",-1] 
#head(counts)
label<-names(counts)
# eliminating wrong libraries (see "samples" spread sheet in
https://docs.google.com/spreadsheets/d/11F-0UrT9gGu7O-J--csjAKfpe2iN5JCOJEabIL2-EDg/edit#gid=759100153
# old URL is https://docs.google.com/spreadsheets/d/1-llHCbrDQfQhLXIlRsLhabN-ezBjwKvrdfjiNy36wCg/edit#gid=2102778022)
# updated 081514 based on final sequencing data 
eliminated.libraries<-c( # 
  "D2H16A.merged.bam", # AT5G02540_2
  "D4H4A.merged.bam",  # jar1-1
  "D4H16A.merged.bam", # jar1-1
  "D4H49A.merged.bam", # jar1-1
  "C6H1A.merged.bam", # phyB
  "C6H16A.merged.bam", # phyB  	
  "C7L4A.merged.bam", # pif4pi5
  "C7H16A.merged.bam", # pif4pif5	
  "C7H49A.merged.bam", # pif4pif5
  "C7L49A.merged.bam", # pif4pif5
  "C11H1A.merged.bam", # coi1_16
  "C11L1A.merged.bam", # coi1_16
  "C11H4A.merged.bam", # coi1_16
  "D11H4A.merged.bam", # coi1_16
  "C11L4A.merged.bam", # coi1_16
  "C11H16A.merged.bam", # coi1_16
  "C11L16A.merged.bam", # coi1_16
  "C11H25A.merged.bam", # coi1_16
  "C11L25A.merged.bam", # coi1_16
  "C11H49A.merged.bam", # coi1_16
  "D11H49A.merged.bam", # coi1_16
  "C11L49A.merged.bam", # coi1_16
  "C12H1A.merged.bam", # phyAB
  "C12L1A.merged.bam", # phyAB
  "D12L1A.merged.bam", # phyAB
  "C12H4A.merged.bam", # phyAB
  "C12L4A.merged.bam", # phyAB
  "C12H16A.merged.bam", # phyAB
  "C12L16A.merged.bam", # phyAB
  "C12L25A.merged.bam", # phyAB
  "C12H49A.merged.bam", # phyAB
  "C12L49A.merged.bam", # phyAB
  "C13H1A.merged.bam", # pif3_1
  "C13H16A.merged.bam", # pif3_1
  "E13H16A.merged.bam", # pif3_1
  "C13L16A.merged.bam", # pif3_1
  "E13L16A.merged.bam", # pif3_1
  "C13H25A.merged.bam", # pif3_1
  "C13L25A.merged.bam", # pif3_1
  "C13H49A.merged.bam", # pif3_1
  "D13H49A.merged.bam", # pif3_1
  "C13L49A.merged.bam", # pif3_1
  "D14L4A.merged.bam", # mida9_4
  "C14H25A.merged.bam", # mida9_4
  "D14L25A.merged.bam", # mida9_4
  "C16H1A.merged.bam", # sto
  "C16L1A.merged.bam", # sto
  "C16H4A.merged.bam", # sto
  "C16L4A.merged.bam", # sto
  "C16H16A.merged.bam", # sto
  "C16H25A.merged.bam", # sto
  "C16L25A.merged.bam", # sto
  "C16H49A.merged.bam", # sto
  "C16L49A.merged.bam", # sto
  "D16L49A.merged.bam") # sto
# eliminate 55 libraries (081514)
counts<-counts[,!names(counts) %in% eliminated.libraries]
#head(counts) 
dim(counts) #[1] 30533   749

samples <- data.frame(file=colnames(counts),
                      batch=factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)A+[[:print:]]+(.bam)","\\1",colnames(counts))),
                      trt=factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)A+[[:print:]]+(.bam)","\\3",colnames(counts))),
                      time=factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)A+[[:print:]]+(.bam)","\\4",colnames(counts))),
                      genotype=factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)A+[[:print:]]+(.bam)","\\2",colnames(counts)))
) # corrected
#
samples
counts.nolow <- counts[,colSums(counts,na.rm=TRUE) > 1000000]
samples.low<-samples[samples$file %in% names(counts)[colSums(counts,na.rm=TRUE) < 1000000],]
samples.low$counts<-colSums(counts,na.rm=TRUE)[colSums(counts,na.rm=TRUE) < 1000000]
names(counts.nolow)
counts.nolow[is.na(counts.nolow)]<-0
save(counts.nolow,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","counts.nolow.Rdata"))

```

* eliminating outlier libraries
```{r, eval=FALSE}
# in whitney. Initial DE analysis
load(file.path(homedir,"..","..",Nozue2016_SAStranscriptome_output","output","counts.nolow.Rdata"))
samples.nolow <- data.frame(file=colnames(counts.nolow),
                            batch=factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)A+[[:print:]]+(.bam)","\\1",colnames(counts.nolow))),
                            trt=factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)A+[[:print:]]+(.bam)","\\3",colnames(counts.nolow))),
                            time=factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)A+[[:print:]]+(.bam)","\\4",colnames(counts.nolow))),
                            genotype=factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)A+[[:print:]]+(.bam)","\\2",colnames(counts.nolow)))
) # corrected
summary(counts.nolow)
Genotype<-levels(samples.nolow$genotype)
# Genotype = "11" and "16" had an error in following DE analysis
Genotype<-Genotype[!Genotype %in% c("11","16")]

## normalizing 
samples.nolow$group <- paste(samples.nolow$genotype,samples.nolow$trt,samples.nolow$time,sep=".") 
samples.nolow$genotype<-as.character(samples.nolow$genotype)
dge.nolow <- DGEList(counts=counts.nolow,group=samples.nolow$group)  
dge.nolow<-calcNormFactors(dge.nolow)

for(i in Genotype) { # error in i=11,16
  print(paste("Genotype is ",i))
  # subset genotype of interests from DGEList (with normalized). see ablove (081214)
  samples.nolow.s<-samples.nolow[samples.nolow$genotype==i,]
  dge.nolow.s<-dge.nolow[,samples.nolow$genotype==i]
  # eliminating low expressed genes in dge.nolow.s
  dge.nolow.s<-dge.nolow.s[rowSums(dge.nolow.s$counts > 5) >= 3,]
  
  design.nolow.s.batch<- model.matrix(~samples.nolow.s$time*samples.nolow.s$trt + samples.nolow.s$batch)
  colnames(design.nolow.s.batch) <- gsub("samples.nolow.s$","",colnames(design.nolow.s.batch),fixed=TRUE) # nicer column names
  colnames(design.nolow.s.batch)
  print(design.nolow.s.batch)
  dge.nolow.s.batch.glm <- estimateGLMCommonDisp(dge.nolow.s,design.nolow.s.batch, verbose=TRUE) 
  dge.nolow.s.batch.glm <- estimateGLMTrendedDisp(dge.nolow.s.batch.glm,design.nolow.s.batch, verbose=TRUE)
  dge.nolow.s.batch.glm <- estimateGLMTagwiseDisp(dge.nolow.s.batch.glm,design.nolow.s.batch) # why this is not included in original Julin's script?
  # plot BCV and MDS plot  
    pdf(file=paste("../../Nozue2016_SAStranscriptome_output/output/BCV_MDS_with_genotype",i,".pdf",sep=""),width=11,height=8)
    par(mfrow=c(1,2))
    plotBCV(dge.nolow.s.batch.glm,main=paste("genotype",i))
    plotMDS(dge.nolow.s.batch.glm,main=paste("genotype",i))
    dev.off()
  # 
  dge.nolow.s.batch.glm.fit <- glmFit(dge.nolow.s.batch.glm,design.nolow.s.batch)
  dge.nolow.s.batch.glm.lrt <- glmLRT(dge.nolow.s.batch.glm.fit,coef=c("trtL","time4:trtL","time16:trtL","time25:trtL","time49:trtL"))
  DE.batch.anytime<-topTags(dge.nolow.s.batch.glm.lrt,n=2000)$table[topTags(dge.nolow.s.batch.glm.lrt,n=2000)$table$FDR<0.05,]
  dim(DE.batch.anytime) # 416 (Col, w/o eliminating low expressed genes). 518 (Col, after eliminating low expressed genes)
  save(DE.batch.anytime,file=paste("../../Nozue2016_SAStranscriptome_output/output/DE.batch.anytime_genotype",i,".Rdata",sep=""))    
}
# eliminate 16h and find outlier
dge.nolow4<-dge.nolow[,!samples.nolow$time==16]
samples.nolow4<-samples.nolow[!samples.nolow$time==16,]
samples.nolow4$time<-as.character(samples.nolow4$time)
# starting finalfinal scripts from samples.nolow4? (092116)
save(samples.nolow4,file="../../Nozue2016_SAStranscriptome_output/output/samples.nolow4.Rdata")
load("../../Nozue2016_SAStranscriptome_output/output/samples.nolow4.Rdata")

for(i in Genotype) { # error in i=11,16 (they has been eliminated in Genotype)
  print(paste("Genotype is ",i))
  # subset genotype of interests from DGEList (with normalized) ?? see ablove (081214)
  samples.nolow.s<-samples.nolow4[samples.nolow4$genotype==i,]
  dge.nolow.s<-dge.nolow4[,samples.nolow4$genotype==i]
  
  # eliminating low expressed genes in dge.nolow.s
  dge.nolow.s<-dge.nolow.s[rowSums(dge.nolow.s$counts > 5) >= 3,]
  
  design.nolow.s.batch<- model.matrix(~samples.nolow.s$time*samples.nolow.s$trt + samples.nolow.s$batch)
  colnames(design.nolow.s.batch) <- gsub("samples.nolow.s$","",colnames(design.nolow.s.batch),fixed=TRUE) # nicer column names
  colnames(design.nolow.s.batch)
  print(design.nolow.s.batch)
  dge.nolow.s.batch.glm <- estimateGLMCommonDisp(dge.nolow.s,design.nolow.s.batch, verbose=TRUE) 
  dge.nolow.s.batch.glm <- estimateGLMTrendedDisp(dge.nolow.s.batch.glm,design.nolow.s.batch, verbose=TRUE)
  dge.nolow.s.batch.glm <- estimateGLMTagwiseDisp(dge.nolow.s.batch.glm,design.nolow.s.batch) # why this is not included in original Julin's script?
  pdf(file=paste("../../Nozue2016_SAStranscriptome_output/output/BCV_MDS_with_genotype",i,".pdf",sep=""),width=11,height=8)
  par(mfrow=c(1,2))
  plotBCV(dge.nolow.s.batch.glm,main=paste("genotype",i))
  plotMDS(dge.nolow.s.batch.glm,main=paste("genotype",i))
  dev.off()
  
  dge.nolow.s.batch.glm.fit <- glmFit(dge.nolow.s.batch.glm,design.nolow.s.batch)
  dge.nolow.s.batch.glm.lrt <- glmLRT(dge.nolow.s.batch.glm.fit,coef=c("trtL","time4:trtL","time25:trtL","time49:trtL"))
  
  DE.batch.anytime<-topTags(dge.nolow.s.batch.glm.lrt,n=2000)$table[topTags(dge.nolow.s.batch.glm.lrt,n=2000)$table$FDR<0.05,]
  dim(DE.batch.anytime) # 416 (Col, w/o eliminating low expressed genes). 518 (Col, after eliminating low expressed genes)
  save(DE.batch.anytime,file=paste("../../Nozue2016_SAStranscriptome_output/output/DE.batch.anytime_genotype_wo16h",i,".Rdata",sep=""))  	
}

## check MDS plot and find outliner libraries (081514)
# eliminate following libraries due to outlier after looking MDS plot
# Col (genotype = 1) # 90 DE
Col_outlier<-"C1L4A.merged.bam" # after elimination; 80 DE
# PAR1_RNAi09 (genotype = 10) only 7 DE
PAR1_RNAi09_outlier<-c("E10H4A.merged.bam","D10H1A.merged.bam") # 6 DE
# coi1_16 (11) only 1 DE (did not work with edgeR possibly due to lots of library are eliminated by genotyping)
#phyAB, (12), 105 DE 
phyAB_outlier<-"C12H25A.merged.bam" #>>>  
#pif3 (13) 73 DE
# AT5G02760_2 (mida9_4) (genotype = 14) original DE (dim(DE.batch.anytime), 10 8
# AT5G59010_2 (15) # 46 DE (no need to eliminate)
# sto (genotype = 16) (eliminated entire sto libraries)
# aos (17) #52 (no need to eliminate)
# 18 (genotyping showed this is not mutant. Eliminate this genotype)
# argos_outlier<-"D18L1A.merged.bam" #11 DE
#co_9 (genotype=19) 20 DE 
# AT5G02540_1 (genotype = 2)
AT5G02540_1_outlier<-"D2H4A.merged.bam" # 23 DE
# Blh_1 (genotype = 20) DE 4
# ? (genotype = 21) DE 20 (no need to eliminate libraries)
#Shahdara (genotype =22) # DE 2
Shahdara_outlier<-"C22H25A.merged.bam" # 3 DE
# Col_0 (23) # 59
Col_0_outlier<-"E23H49A.merged.bam" # 58 DE
# Cvi_0 (24) # 43
# Bur_0 (25) # 53
Bur_0_outlier<-"E25H25A.merged.bam" # ** DE
# Oy_0 (26) , 2 DE 
Oy_0_outlier<-"E26L1A.merged.bam" # only * DE
# (27) only 3 DE genes, but MDS plot looks good 
# hy5, (genotype 3) only 0 DE genes (MDS plot looks good) probably no response to shade?
# jar1 (genotype 4) 30 DE (MDS plot looks good)
# kat1_2 (5) 117 DE (no need to eliminate)
# phyB (6) 25 DE
# pif4 pif5 (7) only 6 DE genes (MDS plot looks OK)
# spt_11 (8) 81 DE >> 
#yuc2589 (9), only 10 DE genes but MDS plot looks OK. >> after outliner elimination, ** DE
# combine outlier libraries
outlier<-c(
  Col_outlier,
  PAR1_RNAi09_outlier,
  phyAB_outlier,
  AT5G02540_1_outlier,
  Shahdara_outlier, 
  Col_0_outlier,
  Bur_0_outlier,
  Oy_0_outlier
)
# eliminate outlier libraries
## run once to create directories
## system("mkdir ../../Nozue2016_SAStranscriptome_output/output/edgeR") 
## system("mkdir ../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h") 

samples.nolow5<-samples.nolow4[!samples.nolow4$file %in% outlier,]
dge.nolow5<-dge.nolow4[,!samples.nolow4$file %in% outlier]
Genotype<-levels(as.factor(samples.nolow5$genotype))
Genotype<-Genotype[!(Genotype=="11"|Genotype=="12"|Genotype=="18"|Genotype=="16")]

#### 100717
# genotype used in PLoSGenetics2015
samples.PLoSGenetics2015mutants<-samples.nolow5[samples.nolow5$genotype %in% c("1","17","19","11","4","5","3","14","10","12","6","13","7","8","16","9"),]
# genotypes not working in edgeR 
errorDE<-c("11","12","16")
# combine
samples.PLoSGenetics2015mutants.errrDE<-samples.PLoSGenetics2015mutants[!samples.PLoSGenetics2015mutants$genotype %in% errorDE,]
save(samples.PLoSGenetics2015mutants.errrDE,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","samples.PLoSGenetics2015mutants.errrDE.Rdata")) # this has to be used below. Replace "samples.nolow5" into "sample.PLoSGenetics2015mutants.errrDE" 
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","samples.PLoSGenetics2015mutants.errrDE.Rdata"))
#load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","dge.nolow5.Rdata"))
# also change dge.nolow5.Rdata into "dge.PLoSGenetics2015mutants.errrDE.Rdata"

dge.PLoSGenetics2015mutants.errrDE<-dge.nolow5[,samples.PLoSGenetics2015mutants.errrDE$file]
save(dge.PLoSGenetics2015mutants.errrDE,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","dge.PLoSGenetics2015mutants.errrDE.Rdata")) 
```
## copied from vst.R (092517; 100717)
```{r}
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","samples.PLoSGenetics2015mutants.errrDE.Rdata"))
counts<-counts[,colnames(counts) %in% samples.PLoSGenetics2015mutants.errrDE$file]
counts[is.na(counts)] <- 0 # to avoid errors in DESeqDataSetFromMatrix
samples.PLoSGenetics2015mutants.errrDE$batch
dds <- DESeq2::DESeqDataSetFromMatrix(countData = counts, colData = samples.PLoSGenetics2015mutants.errrDE, design = ~ batch + genotype*trt*time) # modified (070616, 092517)
head(counts)
vsd <- DESeq2::varianceStabilizingTransformation(dds)
vstMat <- assay(vsd)
colnames(vstMat) <- colnames(counts)
vstMat[1:10,1:10]
write.csv(vstMat,file="../../Nozue2016_SAStranscriptome_data/input/kazu.SAS.expression.VST.092517.csv")
```
## differentially expressed gene analysis
1.DE (treatment, ie, shade responsiveness) at any time point within given genotype
```{r eval=FALSE}
#load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","samples.nolow5.Rdata"))
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","samples.PLoSGenetics2015mutants.errrDE.Rdata")) # 100717
#load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","dge.nolow5.Rdata"))
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","dge.PLoSGenetics2015mutants.errrDE.Rdata"))

#SAS.expression.vst<-read.csv(file.path("..","..","Nozue2016_SAStranscriptome_data","input","kazu.SAS.expression.VST.092517.csv"),row.names=1) # new transformed data by DEseq2 (082016), (092617). Only Genotype info is used here
Genotype<-levels(as.factor(samples.PLoSGenetics2015mutants.errrDE$genotype)) # 13 genotypes inclucing Col. better to use samples.nolow5.Rdata info (confusing). Use samples.PLoSGenetics2015mutants.errrDE$genotype (100717)
# 
dim(SAS.expression.vst) # [1] 30533   287
dim(samples.nolow5) #[1] 590   6
dim(samples.PLoSGenetics2015mutants.errrDE) #[1] 287   6. Reduced number of libraries!
for(i in Genotype) { # why I eliminate genotype 1 and 10???
  print(paste("Genotype is ",i))
  #samples.nolow.s<-samples.nolow5[samples.nolow5$genotype==i,]
  #dge.nolow.s<-dge.nolow5[,samples.nolow5$genotype==i]
  samples.nolow.s<-samples.PLoSGenetics2015mutants.errrDE[samples.PLoSGenetics2015mutants.errrDE$genotype==i,] # 100717
  dge.nolow.s<-dge.PLoSGenetics2015mutants.errrDE[,samples.PLoSGenetics2015mutants.errrDE$genotype==i] # 100717
  # eliminating low expressed genes in dge.nolow.s
  dge.nolow.s<-dge.nolow.s[rowSums(dge.nolow.s$counts > 5) >= 3,]  
  # recalculate normalization (100717)
  dge.nolow.s<-calcNormFactors(dge.nolow.s)
  # design model 
  design.nolow.s.batch<- model.matrix(~samples.nolow.s$time*samples.nolow.s$trt + samples.nolow.s$batch)
  colnames(design.nolow.s.batch) <- gsub("samples.nolow.s$","",colnames(design.nolow.s.batch),fixed=TRUE) # nicer column names
  colnames(design.nolow.s.batch)
  print(design.nolow.s.batch)
  dge.nolow.s.batch.glm <- estimateGLMCommonDisp(dge.nolow.s,design.nolow.s.batch, verbose=TRUE) 
  dge.nolow.s.batch.glm <- estimateGLMTrendedDisp(dge.nolow.s.batch.glm,design.nolow.s.batch, verbose=TRUE)
  dge.nolow.s.batch.glm <- estimateGLMTagwiseDisp(dge.nolow.s.batch.glm,design.nolow.s.batch) # why this is not included in original Julin's script?
  pdf(file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h",paste("BCV_MDS_with_genotype_wo_outlier_wo16h_PLoSGenetics2015mutants.errrDE",i,".pdf",sep="")),width=11,height=8) # fine name fixed (122214)
  par(mfrow=c(1,2))
  plotBCV(dge.nolow.s.batch.glm,main=paste("genotype",i))
  mds.dge<-plotMDS(dge.nolow.s.batch.glm,main=paste("genotype",i))
  mds.dge
  # for ggplot2 plotting of MDSplot (100417)
  save(mds.dge,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h",paste("MDS_with_genotype_wo_outlier_wo16h_PLoSGenetics2015mutants.errrDE",i,".RData",sep="")))
  dev.off() # the end of normal plotting
  #
  dge.nolow.s.batch.glm.fit <- glmFit(dge.nolow.s.batch.glm,design.nolow.s.batch)
  dge.nolow.s.batch.glm.lrt <- glmLRT(dge.nolow.s.batch.glm.fit,coef=c("trtL","time4:trtL","time25:trtL","time49:trtL"))
  DE.batch.anytime<-topTags(dge.nolow.s.batch.glm.lrt,n=2000)$table[topTags(dge.nolow.s.batch.glm.lrt,n=2000)$table$FDR<0.05,]
  dim(DE.batch.anytime) # 
    save(DE.batch.anytime,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output",paste("DE.batch.anytime_genotype_wo_outlier_wo16h.PLoSGenetics2015mutants.errrDE",i,".Rdata",sep=""))) # file name fixed (122114) , (092617)   
}
```
# MDS plot by ggplot2 (100417;100717)
```{r}
mds.files<-list.files(pattern="^MDS_with_genotype_wo_outlier_wo16hPLoSGenetics2015mutants.errrDE",path=file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h"))
for(i in mds.files) {
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h",i))
x.mds.dge.DF<-as.data.frame(mds.dge$x)
y.mds.dge.DF<-as.data.frame(mds.dge$y)
distance_matrix.mds.dge<-merge(x.mds.dge.DF,y.mds.dge.DF,by="row.names")
distance_matrix.mds.dge$group<-gsub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)(A.merged.bam)","\\2\\3",distance_matrix.mds.dge$Row.names)
distance_matrix.mds.dge$genotype<-gsub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)(A.merged.bam)","\\2",distance_matrix.mds.dge$Row.names)
distance_matrix.mds.dge$trt<-gsub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)(A.merged.bam)","\\3",distance_matrix.mds.dge$Row.names)
distance_matrix.mds.dge$time<-gsub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)(A.merged.bam)","\\4",distance_matrix.mds.dge$Row.names)
distance_matrix.mds.dge$batch<-gsub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)(A.merged.bam)","\\1",distance_matrix.mds.dge$Row.names)
p.distance_matrix.mds.dge <- ggplot(data = distance_matrix.mds.dge)+ geom_label(aes(mds.dge$x, mds.dge$y, color=factor(trt),fill=factor(time),label=paste(batch,time,sep="_"),size=0.3,alpha=0.5)) + labs(title=paste("genotype",vlookup(gsub("(MDS_with_genotype_wo_outlier_wo16h)([[:digit:]]+)(.RData)","\\2",i),conversion.table,2)))
ggsave(p.distance_matrix.mds.dge, filename =file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h",paste("p.distance_matrix.mds.dge.",vlookup(gsub("(MDS_with_genotype_wo_outlier_wo16hPLoSGenetics2015mutants.errrDE)([[:digit:]]+)","\\2",i),conversion.table,2),".png",sep="")), height = 11, width = 25) 
}

```
# summary of DE genes 
* (053014 - 053114): revised version (081514 - );more revision (060415 - ); more (092617); DE.batch.anytime_genotype_wo_outlier_wo16h.PLoSGenetics2015mutants.errrDE (100717)
```{r eval=FALSE}
#getwd()
DEfiles<-list.files(path=file.path("..","..","Nozue2016_SAStranscriptome_output","output"),pattern="DE.batch.anytime_genotype_wo_outlier_wo16h.PLoSGenetics2015mutants.errrDE") # works? 
#DEfiles<-grep("DE.batch.anytime_genotype_wo_outlier_wo16h.PLoSGenetics2015mutants.errrDE",DEfiles,value=TRUE)
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output",DEfiles[1]))
rownames(DE.batch.anytime)<-paste(rownames(DE.batch.anytime),gsub("(DE.batch.anytime_genotype_wo_outlier_wo16h.PLoSGenetics2015mutants.errrDE)([[:print:]]+)(.Rdata)","\\2",DEfiles[1]),sep=".")
temp<-DE.batch.anytime

for(i in DEfiles[-1]) {
  #load(i) 
  load(file.path("..","..","Nozue2016_SAStranscriptome_output","output",i))

  if(dim(DE.batch.anytime)[1]>1) {
    rownames(DE.batch.anytime)<-paste(rownames(DE.batch.anytime),gsub("(DE.batch.anytime_genotype_wo_outlier_wo16h.PLoSGenetics2015mutants.errrDE)([[:print:]]+)(.Rdata)","\\2",i),sep=".") # fixed, 112616
    temp<-rbind(temp,DE.batch.anytime)  }
  else next
}
DGEsummary.edgeR.response<-temp
tail(DGEsummary.edgeR.response)
dim(DGEsummary.edgeR.response) #755   8 for final sequence data with normalization (081514). This is whitney. but my computer is 505 8. with PLoSGenetics2015mutants.errrDE, it became 507 8 (100717)
save(DGEsummary.edgeR.response,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","DGEsummary.edgeR.response.PLoSGenetics2015mutants.errrDE.Rdata")) # 100717
# # genotype conversion 
#load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","samples.nolow5.Rdata"))
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","samples.PLoSGenetics2015mutants.errrDE.Rdata")) # 100717

samples.PLoSGenetics2015mutants.errrDE
samples.PLoSGenetics2015mutants.errrDE$genotype2<-""
for(i in 1:13){ # only for Genotype
  samples.PLoSGenetics2015mutants.errrDE[samples.PLoSGenetics2015mutants.errrDE$genotype==conversion.table[i,1],"genotype2"]<-as.character(conversion.table[i,2])	
} 

## summary of DE gene numbers
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output",DEfiles[1]))
DEgenes.summary<-data.frame(genotype=gsub("(DE.batch.anytime_genotype_wo_outlier_wo16h.PLoSGenetics2015mutants.errrDE)([[:digit:]]+)(.Rdata)","\\2",DEfiles))
for(i in 1:length(DEfiles)) {
  load(file.path("..","..","Nozue2016_SAStranscriptome_output","output",DEfiles[i]))

  DEgenes.summary$DEgenes_num[i]<-dim(DE.batch.anytime)[1]
}
DEgenes.summary$genotype2<-vlookup(DEgenes.summary$genotype,conversion.table,2)
DEgenes.summary<-DEgenes.summary[!is.na(DEgenes.summary$DEgenes_num)&!is.na(DEgenes.summary$genotype2),]
DEgenes.summary # summary
#setwd(homedir)
```
## select genes with largeCV and calculate responsiveness for SOM analysis, updated 082016, 092617 (no change in 100717)
## for t-SNE analysis
```{r eval=FALSE}
SAS.expression.vst<-read.csv(file.path("..","..","Nozue2016_SAStranscriptome_data","input","kazu.SAS.expression.VST.092517.csv"),row.names=1) # new transformed data by DEseq2 (092617)
# calculate mean value among replicates
Genotype<-levels(as.factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|25|49)(A.merged.bam)","\\2",names(SAS.expression.vst)[-1])))

temp<-SAS.expression.vst.s[,names(SAS.expression.vst.s) %in% samples.nolow5[samples.nolow5$genotype==1,"file"]]
summary.vst<-expression.mean(temp)ã€€# custom function

for(i in Genotype[-1]) {
  temp<-SAS.expression.vst.s[,names(SAS.expression.vst.s) %in% samples.nolow5[samples.nolow5$genotype==i,"file"]]
	temp<-as.data.frame(temp)
	temp2<-expression.mean(temp)
	summary.vst<-cbind(summary.vst,temp2)
}
rownames(summary.vst)<-gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",rownames(SAS.expression.vst))

save(summary.vst,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","summary.vst.Rdata")) #
# relative value 
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","summary.vst.Rdata"))
# note that summary.vst is log2 transformed
summary.vst.response<-as.data.frame((2^summary.vst[,gsub("(.+)(H|L)(1|4|16|25|49)","\\2",names(summary.vst))=="L"])/(2^summary.vst[,gsub("(.+)(H|L)(1|4|16|25|49)","\\2",names(summary.vst))=="H"])) # shade/sun 
rownames(summary.vst.response)<-rownames(summary.vst)
names(summary.vst.response)<-paste(gsub("(.+)(H|L)(1|4|16|25|49)","\\1_\\3",names(summary.vst.response)),"hrA",sep="") #
save(summary.vst.response,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","summary.vst.response.Rdata"))
# # rename as only Kazu's samples 
SAS.expression.vst.s.kazu<-SAS.expression.vst.s
dim(SAS.expression.vst.s.kazu) # 30533   287
# select genes with higher CV
co.var.df <- function(x) ( 100*apply(x,1,sd)/rowMeans(x) )
#SAS.expression.vst.s.kazu$cv <- co.var.df(SAS.expression.vst.s.kazu[,-1])
SAS.expression.vst.s.kazu$cv <- co.var.df(SAS.expression.vst.s.kazu)
a<-hist(SAS.expression.vst.s.kazu$cv,breaks=seq(0,76,1))
table(SAS.expression.vst.s.kazu$cv)

sum(as.integer(SAS.expression.vst.s.kazu$cv>3))/dim(SAS.expression.vst.s.kazu)[1] #[1] 0.4011866 > 0.6019389 > 0.5222546
sum(as.integer(SAS.expression.vst.s.kazu$cv>4))/dim(SAS.expression.vst.s.kazu)[1] #[1] 0.830806 > 0.3140864
sum(as.integer(SAS.expression.vst.s.kazu$cv>4.5))/dim(SAS.expression.vst.s.kazu)[1] #[1] 0.3621328 > 0.2300789
sum(as.integer(SAS.expression.vst.s.kazu$cv>5))/dim(SAS.expression.vst.s.kazu)[1] #[1] 0.1412242
#SAS.expression.vst.s.kazu.largeCV<-SAS.expression.vst.s.kazu[SAS.expression.vst.s.kazu$cv>3,] # 
SAS.expression.vst.s.kazu.largeCV<-SAS.expression.vst.s.kazu[SAS.expression.vst.s.kazu$cv>4.5,] 
dim(SAS.expression.vst.s.kazu.largeCV) #[1] 7438  403 > 18379   264 > 11057   288 > 7025 288
save(SAS.expression.vst.s.kazu.largeCV,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","SAS.expression.vst.s.kazu.largeCV.Rdata"))
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","SAS.expression.vst.s.kazu.largeCV.Rdata"))
dim(SAS.expression.vst.s.kazu.largeCV) # this is not response
# response value of genes with largeCV 
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","summary.vst.response.Rdata")) #summary.vst.response is object name
summary.vst.response.kazu<-summary.vst.response # where does this come from? 120116
dim(summary.vst.response.kazu) #30533    52
summary.vst.response.kazu.largeCV<-summary.vst.response.kazu[rownames(summary.vst.response.kazu) %in% gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",rownames(SAS.expression.vst.s.kazu.largeCV)),]
dim(summary.vst.response.kazu.largeCV) # 7438   60 > 18379    48 > 11057    52
save(summary.vst.response.kazu.largeCV,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","summary.vst.response.kazu.largeCV.Rdata"))
```
## select only DE genes (using PLoSGenetics2015mutants.errrDE, 100717)
```{r eval=TRUE}
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","summary.vst.response.kazu.largeCV.Rdata"))
summary.vst.response.kazu.largeCV[summary.vst.response.kazu.largeCV=="Inf"]<-0 # arbitral
summary.vst.response.kazu.largeCV[summary.vst.response.kazu.largeCV=="-Inf"]<-0 # arbitral
summary.vst.response.kazu.largeCV[is.na(summary.vst.response.kazu.largeCV)]<-0 # arbitral
#load("../../Nozue2016_SAStranscriptome_output/output/DGEsummary.edgeR.response.Rdata") #"AT1G33925
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","DGEsummary.edgeR.response.PLoSGenetics2015mutants.errrDE.Rdata"))

DGEgenelist.edgeRglm<-unique(gsub("([[:print:]]+)(.)([[:digit:]]+)(.)([[:print:]]+)","\\1",rownames(DGEsummary.edgeR.response))) # I missed the second (.)!
length(DGEgenelist.edgeRglm) # 308 genes (020718)
# 
DGE.summary.vst.response.kazu.largeCV<-summary.vst.response.kazu.largeCV[rownames(summary.vst.response.kazu.largeCV) %in% DGEgenelist.edgeRglm,]
save(DGE.summary.vst.response.kazu.largeCV,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","DGE.summary.vst.response.kazu.largeCV.PLoSGenetics2015mutants.errrDE.Rdata")) # updated 052916, 092517, 100717, 020718
dim(DGE.summary.vst.response.kazu.largeCV) # 276 genes used for t-SNE analysis (020718)
DGE.summary.vst.response.kazu.largeCV[rownames(DGE.summary.vst.response.kazu.largeCV)=="AT4G16780",] #ATHB2 has to be induced in genotype 1 (= Col)
 
```
## t-SNE analysis (111517)
## data prep
```{r}
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","DGE.summary.vst.response.kazu.largeCV.PLoSGenetics2015mutants.errrDE.Rdata")) # 100717, 020718
genotypes<-levels(as.factor(gsub("(.+)(_)(1|4|16|25|49)(hrA)","\\1",as.character(names(DGE.summary.vst.response.kazu.largeCV)))))
# combine all DE genes in each genotype (treat each gene in one genotype as different gene)
DGE.summary.vst.response.kazu.largeCV<-log2(DGE.summary.vst.response.kazu.largeCV) # log2 transformation of fold change. error (092617)
i<-1
temp<-DGE.summary.vst.response.kazu.largeCV[,gsub("(.+)(_)(1|4|16|25|49)(hrA)","\\1",names(DGE.summary.vst.response.kazu.largeCV))==genotypes[i]]
rownames(temp)<-paste(rownames(temp),genotypes[i],sep=".")
colnames(temp)<-paste(gsub("(.+)(_)(1|4|16|25|49)(hrA)","\\3",colnames(temp)),"h",sep="") ### fix (070914)
for(i in 2:length(genotypes)) {
   temp2<-DGE.summary.vst.response.kazu.largeCV[,gsub("(.+)(_)(1|4|16|25|49)(hrA)","\\1",names(DGE.summary.vst.response.kazu.largeCV))==genotypes[i]]
  rownames(temp2)<-paste(rownames(temp2),genotypes[i],sep=".")
  colnames(temp2)<-paste(gsub("(.+)(_)(1|4|16|25|49)(hrA)","\\3",colnames(temp2)),"h",sep="") ### fix (070914)  
  temp<-rbind(temp,temp2)
}
summary(temp)
head(temp)
data5<-temp

```



# ver3. for reproducibility and narrowed range (101917). Needs to run today (020718)
```{r}
 # Sets seed for reproducibility
for(n in 42:51) { # set
    set.seed(n)
    tsne_out <- Rtsne(as.matrix(data5),theta=0.3) # Run TSNE
    save(tsne_out,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap","tsne_out.Rdata"))
    
for(x in seq(3.1,3.35,0.05)) { # eps
  for(y in seq(13,18,1)) { # MinPts
    load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap","tsne_out.Rdata"))
    print(paste("eps is ",x,". MinPts is ",y,".",sep=""))
    dbscan_out<-dbscan(tsne_out$Y,eps=x,MinPts=y)
    #table(dbscan_out$cluster)
    plot(tsne_out$Y,col=dbscan_out$cluster) 
    if(length(table(dbscan_out$cluster))>5 && length(table(dbscan_out$cluster))<30 && sum((dbscan_out$cluster==0)*1) <2600) {
      save(dbscan_out,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap",paste("eps",x,"_MinPts",y,".set",n,".Rdata",sep="")))
    } else {next}
    }   
  }
}

```
# combine all into one data.frame
```{r}
tsne.dbscan.eps.MinPts.set.files<-list.files(file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap"),pattern="(eps)([[:print:]]+)(_MinPts)([[:digit:]]+)(\\.set)([[:digit:]]+)(\\.Rdata)")

#tsne.dbscan.eps.MinPts.set.files.list<-sapply(file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap",tsne.dbscan.eps.MinPts.set.files),function(x) {load(x);dbscan_out$cluster}) # does not work
i<-1
    load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap",tsne.dbscan.eps.MinPts.set.files[i]))
tsne.dbscan.eps.MinPts.set.files.DF<-data.frame(eps3.1_MinPts13.set42.Rdata=dbscan_out$cluster) # manually edit
for(i in 2:length(tsne.dbscan.eps.MinPts.set.files)) {
    load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap",tsne.dbscan.eps.MinPts.set.files[i]))
    tsne.dbscan.eps.MinPts.set.files.DF$temp<-dbscan_out$cluster
    colnames(tsne.dbscan.eps.MinPts.set.files.DF)<-gsub("temp",tsne.dbscan.eps.MinPts.set.files[i],colnames(tsne.dbscan.eps.MinPts.set.files.DF))
}

#tsne.dbscan.eps.MinPts.set.files.list.DF<-do.call("cbind",tsne.dbscan.eps.MinPts.set.files.list) 
tsne.dbscan.eps.MinPts.set.files.DF[,3]
colnames(tsne.dbscan.eps.MinPts.set.files.DF)
colnames(tsne.dbscan.eps.MinPts.set.files.DF)<-gsub("\\.Rdata","",colnames(tsne.dbscan.eps.MinPts.set.files.DF))

#colnames(tsne.dbscan.eps.MinPts.set.files.list.DF)<-gsub("(../../Nozue2016_SAStranscriptome_output/output/tsne_dbscan_overlapTable_heatmap/)(eps[[:print:]]+)(\\_MinPts[[:digit:]]+)(\\.)(set)([[:digit:]]+)(\\.Rdata\\.)([[:print:]]+)","\\2\\3",colnames(tsne.dbscan.eps.MinPts.set.files.list.DF)) # fix this line

tsne.dbscan.eps.MinPts.set.files.DF$AGI<-rownames(data5)
save(tsne.dbscan.eps.MinPts.set.files.DF,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap","tsne.dbscan.eps.MinPts.set.files.DF.Rdata"))
```
# Add custom category (020718)
```{r}
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap","tsne.dbscan.eps.MinPts.set.files.DF.Rdata"))
## making summary table/dataset (100717)
# custom categories
load(file.path("..","..","Nozue2016_SAStranscriptome_data","input","hormone.responsiveness6.DF.s.Rdata")) 
# add tsnedbscan clusters
tsne.dbscan.eps.MinPts.set.files.DF$gt<-gsub("(AT[[:alnum:]]+)(\\.)([[:digit:]]+)","\\3",tsne.dbscan.eps.MinPts.set.files.DF$AGI) 
tsne.dbscan.eps.MinPts.set.files.DF.Col<-tsne.dbscan.eps.MinPts.set.files.DF[tsne.dbscan.eps.MinPts.set.files.DF$gt=="1",]
dim(tsne.dbscan.eps.MinPts.set.files.DF.Col)
tsne.dbscan.eps.MinPts.set.files.DF.Col$AGI<-gsub("(AT[[:alnum:]]+)(\\.)([[:digit:]]+)","\\1",tsne.dbscan.eps.MinPts.set.files.DF.Col$AGI)
# merge with tsne clusters
hormone.responsiveness.tsnedbscanclusters<-merge(hormone.responsiveness6.DF.s,tsne.dbscan.eps.MinPts.set.files.DF.Col,by="AGI",all=TRUE)
hormone.responsiveness.tsnedbscanclusters[is.na(hormone.responsiveness.tsnedbscanclusters)]<-"0"
hormone.responsiveness.tsnedbscanclusters.description<-merge(hormone.responsiveness.tsnedbscanclusters,TAIR10_gene_descriptions[!duplicated(TAIR10_gene_descriptions$AGI),c("AGI","Short_description","colE...col.F")],by="AGI")
#write.csv(hormone.responsiveness.tsnedbscanclusters.description,file="../../Nozue2016_SAStranscriptome_output/figs_tables/Supplemental_Dataset1_Custom_enrichment_tsnedbscan_clusters_of_shade-responsive_genes110617.csv") 
write.csv(hormone.responsiveness.tsnedbscanclusters.description,file="../../Nozue2016_SAStranscriptome_output/figs_tables/Supplemental_Dataset4_Custom_enrichment_tsnedbscan_clusters_of_shade-responsive_genes020718.csv") 

```
# Run overlapTable (v2. with SAup or down combined; v3 with renamed SAup and SAdown)
```{r}
for(i in 2:20) { # 2:30
  for(n in grep("set",colnames(hormone.responsiveness.tsnedbscanclusters))) { # addition
  overlap<-overlapTable(hormone.responsiveness.tsnedbscanclusters[,n],hormone.responsiveness.tsnedbscanclusters[,i])
  overlap.melt<-melt(overlap$pTable,id=1) #
  overlap.melt$count<-melt(overlap$countTable,id=1)$value
  overlap.melt<-overlap.melt[!overlap.melt$Var2=="0",]
  sometype.setname<-colnames(hormone.responsiveness.tsnedbscanclusters)[n]
  names(hormone.responsiveness.tsnedbscanclusters)[n]
  save(overlap.melt,file=file.path(homedir,"..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap",paste("overlap.melt.tsnedbscan.customcat.PLoSGenetics2015mutants.errrDE.v3",sometype.setname,colnames(hormone.responsiveness.tsnedbscanclusters)[i],"Rdata",sep=".")))
  }
}
```
# summerize overlapTable (v2, 110617;v3, 020718)
```{r}
  overlap.melt.file<-list.files(file.path(homedir,"..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap"),pattern="overlap.melt.tsnedbscan.customcat.PLoSGenetics2015mutants.errrDE.v3.")
overlap.melt.file<-overlap.melt.file[grep("set.[[:digit:]].Rdata",overlap.melt.file,invert=TRUE)] #
table(gsub("(overlap.melt.tsnedbscan.customcat.PLoSGenetics2015mutants.errrDE.v3)(\\.)(eps[[:print:]]+)(\\_)(MinPts[[:digit:]]+)(\\.)(set)([[:digit:]]+)(\\.)([[:print:]]+)(\\.Rdata)","\\8",overlap.melt.file)) # should be 42 to 51 (=sets)

for(n in 42:51) {
  overlap.melt.file.set<-overlap.melt.file[gsub("(overlap.melt.tsnedbscan.customcat.PLoSGenetics2015mutants.errrDE.v3)(\\.)(eps[[:print:]]+)(\\_)(MinPts[[:digit:]]+)(\\.)(set)([[:digit:]]+)(\\.)([[:print:]]+)(\\.Rdata)","\\8",overlap.melt.file)==n] # revise
  eps.MinPts<-gsub("(overlap.melt.tsnedbscan.customcat.PLoSGenetics2015mutants.errrDE.v3)(\\.)(eps[[:print:]]+)(\\_)(MinPts[[:digit:]]+)(\\.)(set)([[:digit:]]+)(\\.)([[:print:]]+)(\\.Rdata)","\\3\\4\\5",overlap.melt.file.set) # revise
  
  eps.MinPts.levels<-levels(as.factor(eps.MinPts))
  for(x in eps.MinPts.levels) { # eps & MinPts
overlap.melt.file.set.eps.MinPts<-overlap.melt.file.set[eps.MinPts==x]
      overlap.melt.summary.list2<-sapply(file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap",overlap.melt.file.set.eps.MinPts),function(x) mget(load(x))) 
names(overlap.melt.summary.list2)
overlap.melt.summary<-do.call("rbind",overlap.melt.summary.list2) 
head(overlap.melt.summary) # make sure those are SOM.customcat files
rownames(overlap.melt.summary)<-1:nrow(overlap.melt.summary)
# format overlap.melt.summary
overlap.melt.summary[,"value"]<--log10(overlap.melt.summary[,"value"])
overlap.melt.summary[overlap.melt.summary$value>10,"value"]<-10
overlap.melt.summary[overlap.melt.summary$count==0,"count"]<-""

save(overlap.melt.summary, file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap",paste("overlap.melt.tsnedbscan.customcat.PLoSGenetics2015mutants.errrDE.summary.v3.",x,".set.",n,".Rdata",sep="")))
  }
}  
```
# draw heatmap (v3, 020618)
```{r}
  overlap.melt.tsnedbscan.customcat.PLoSGenetics2015mutants.errrDE.summary.files<-list.files(file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap"),pattern="overlap.melt.tsnedbscan.customcat.PLoSGenetics2015mutants.errrDE.summary.v3")

for(y in overlap.melt.tsnedbscan.customcat.PLoSGenetics2015mutants.errrDE.summary.files) 
  {
  load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap",y))
# graph
overlap.melt.summary.s<-overlap.melt.summary[!overlap.melt.summary$Var1==0,]
#overlap.melt.summary.s$Var1<-factor(overlap.melt.summary.s$Var1,levels=c(3,2,6,4,12,9,7,1,5,10,8,11))# 100717
overlap.melt.summary.s<-overlap.melt.summary.s[!overlap.melt.summary.s$Var2=="",]
overlap.melt.summary.s<-overlap.melt.summary.s[!overlap.melt.summary.s$Var2=="GRF1_3target",] # # remove "GRF1_3target" (052616)
table(overlap.melt.summary.s$Var2)
overlap.melt.summary.s$Var2<-factor(overlap.melt.summary.s$Var2,levels=c("IAAup","IAAdown","BLup","BLdown","MJup","MJdown","GAup","GAdown","ABAup","ABAdown","ACCup","ACCdown","CKup","CKdown","SAup","SAdown","PIFtarget","MYC234up","MYC234down")) # 110717, 020618 ("SAup" and "SAdown")
# switching x and y axis
overlap.melt.summary.s$Var2<-factor(overlap.melt.summary.s$Var2,levels=rev(levels(overlap.melt.summary.s$Var2)))     
overlap.melt.summary.s$Var1
# sort clusters according to mean expression value in each cluster (102317)
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap","tsne.dbscan.eps.MinPts.set.files.DF.Rdata"))
tsne.dbscan.eps.MinPts.set.files.DF
tsne.dbscan.eps.MinPts.set.data5<-cbind(data5,tsne.dbscan.eps.MinPts.set.files.DF) # what is "data5"?
tsne.dbscan.eps.MinPts.set.data5$gt<-gsub("(AT[[:alnum:]]+)(\\.)([[:digit:]]+)","\\3",tsne.dbscan.eps.MinPts.set.data5$AGI)
tsne.dbscan.eps.MinPts.set.data5.Col<-tsne.dbscan.eps.MinPts.set.data5[tsne.dbscan.eps.MinPts.set.data5$gt==1,]
title<-gsub("(overlap.melt.tsnedbscan.customcat.PLoSGenetics2015mutants.errrDE.summary.v3)(\\.)(eps[[:print:]]+)(\\_)(MinPts[[:digit:]]+)(\\.)(set)(\\.)([[:digit:]]+)(\\.Rdata)","\\3\\4\\5\\6\\7\\9",y)
tsne.dbscan.eps.MinPts.set.data5.Col.melt<-melt(tsne.dbscan.eps.MinPts.set.data5.Col[,c("1h","4h","25h","49h",title,"gt","AGI")],id.var=c("gt","AGI",title))
names(tsne.dbscan.eps.MinPts.set.data5.Col.melt)<-gsub(title,"tsne.cluster",names(tsne.dbscan.eps.MinPts.set.data5.Col.melt))
head(tsne.dbscan.eps.MinPts.set.data5.Col.melt) 
mean.cluster<-tapply(tsne.dbscan.eps.MinPts.set.data5.Col.melt$value,tsne.dbscan.eps.MinPts.set.data5.Col.melt$tsne.cluster,mean)
mean.cluster
mean.cluster<-mean.cluster[!names(mean.cluster)=="0"]
# select clusters only found in Col
overlap.melt.summary.ss<-overlap.melt.summary.s[overlap.melt.summary.s$Var1 %in% names(mean.cluster),]
overlap.melt.summary.ss$Var1<-factor(overlap.melt.summary.ss$Var1,levels=names(sort(mean.cluster,decreasing=T))) # sort clusters according to mean responsiveness
#overlap.melt.summary.s$Var1<-droplevels(overlap.melt.summary.s$Var1)
### end sort clusters
q4b<-ggplot(overlap.melt.summary.ss,aes(y=factor(Var2),x=Var1)) 
q4b <- q4b + geom_tile(color="black",aes(fill=value))+scale_fill_gradient2(low="white", mid="white",midpoint=2,high="magenta") # cut off is p value 0.01
#q4 <- q4 + theme(axis.text.x=element_text(angle=90),axis.title=element_blank())
q4b <- q4b + theme(axis.text.x=element_text(angle=90,hjust=1),
                 panel.grid.major = element_blank(),
                 panel.grid.minor = element_blank(),
                 panel.background = element_blank(),
                 axis.ticks = element_line(size=0,color="white"))
q4b <- q4b + geom_text(aes(label=count),size=5) 
q4b <- q4b + labs(y="custom categories",x="tsnedbscan clusters in Col",title=title,fill="-log10\n p-value")
ggsave(q4b,file=paste("Supplemental_Dataset5_Col.tsnedbscan.clusters.vs.my.category.v3.set.",title,".pdf",sep=""),width=13,height=6,path=file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap")) # 
}
# archive
# system(paste("zip ",file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Supplemental_Dataset5_ORA_custome_category_t-SNE_heatmap.zip")," ", file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap","Supplemental_Dataset5_Col.tsnedbscan.clusters.vs.my.category.v3.set.*"),sep=""))
# Select all those pdfs and finder > serve > select "combinePDFs2" (made by automator, https://www.orbitsit.co.uk/2013/06/combining-pdfs-using-automator/) named as "Supplemental_Dataset5_ORA_custom_category__tSNE_all.pdf"
# rename and transfer representative one
## eps3.3_MinPts13.set47 (11**17)
## eps3.25_MinPts14.set45 (020818)
system(paste("cp ",file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap","Supplemental_Dataset5_Col.tsnedbscan.clusters.vs.my.category.v3.set.eps3.25_MinPts14.set45.pdf")," ", file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Supplemental_Fig3_Col.tsnedbscan.clusters.vs.my.category.v3.set.eps3.25_MinPts14.set45.pdf"),sep=""))

```
# 
```{r}
GOseq.tsne<-function(tsne.set="eps3.3_MinPts16.set51") {
# load SOM cluster summary and format it
  load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap","tsne.dbscan.eps.MinPts.set.files.DF.Rdata"))
tsne.dbscan.eps.MinPts.set.files.DF$gt<-gsub("(AT[[:alnum:]]+)(\\.)([[:digit:]]+)","\\3",tsne.dbscan.eps.MinPts.set.files.DF$AGI)
tsne.dbscan.eps.MinPts.set.files.DF.Col<-tsne.dbscan.eps.MinPts.set.files.DF[tsne.dbscan.eps.MinPts.set.files.DF$gt==1,]
tsne.dbscan.eps.MinPts.set.files.DF.Col$AGI<-gsub("(AT[[:alnum:]]+)(\\.)([[:digit:]]+)","\\1",tsne.dbscan.eps.MinPts.set.files.DF.Col$AGI)
# select column and calculate cluster numbers
clusters<-as.numeric(levels(as.factor(tsne.dbscan.eps.MinPts.set.files.DF.Col[,colnames(tsne.dbscan.eps.MinPts.set.files.DF.Col)==tsne.set] ))) # for some reasosnes, colnames(tsne.dbscan.eps.MinPts.set.files.DF.Col)[141]==paste(tsne,".set",set,sep="") is false
print(clusters)  

# calculate GO ORA using GOseq
for(x in clusters) {  
  temp.GOseq<-GOseq.ORA(tsne.dbscan.eps.MinPts.set.files.DF.Col[tsne.dbscan.eps.MinPts.set.files.DF.Col[colnames(tsne.dbscan.eps.MinPts.set.files.DF.Col)==tsne.set]==x,"AGI"])
  if(temp.GOseq=="no enriched GO") {next} else {
     temp.genes<-genes.in.enriched.category(enrich.result=temp.GOseq,gene.list=tsne.dbscan.eps.MinPts.set.files.DF.Col[tsne.dbscan.eps.MinPts.set.files.DF.Col[,colnames(tsne.dbscan.eps.MinPts.set.files.DF.Col)==tsne.set]==x,"AGI"],category.table=Atgo.list)
  save(temp.GOseq,temp.genes,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap",paste(tsne.set,"tsne.Col.cluster",x,"GOseq.PLoSGenetics2015mutants.errrDE.Rdata",sep=".")))
  } 
}
# summary of GO ORA
tsne.Col.cluster.GOseq.files<-list.files(pattern=paste(tsne.set,"tsne.Col.cluster",sep="."),path=file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap"))
tsne.Col.cluster.GOseq.files<-grep(".GOseq.PLoSGenetics2015mutants.errrDE.Rdata",tsne.Col.cluster.GOseq.files,value=TRUE)
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap",tsne.Col.cluster.GOseq.files[1]))
temp.GOseq$clusters<-gsub("(eps[[:print:]]+)(_MinPts[[:digit:]]+)(\\.)(set[[:digit:]]+)(\\.)(tsne.Col.cluster)(\\.)([[:print:]]+)(\\.)(GOseq.PLoSGenetics2015mutants.errrDE.Rdata)","\\8",tsne.Col.cluster.GOseq.files[1])

temp<-temp.GOseq
for(i in tsne.Col.cluster.GOseq.files[-1]) {
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap",i))
temp.GOseq$clusters<-gsub("(eps[[:print:]]+)(_MinPts[[:digit:]]+)(\\.)(set[[:digit:]]+)(\\.)(tsne.Col.cluster)(\\.)([[:print:]]+)(\\.)(GOseq.PLoSGenetics2015mutants.errrDE.Rdata)","\\8",i) 
  temp<-rbind(temp,temp.GOseq)
}
tsne.Col.cluster.GOseq.summary<-temp
write.csv(tsne.Col.cluster.GOseq.summary,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap",paste(tsne.set,"GO.csv",sep=".")) )

} # the end of function

```
# GO ORA
```{r}
# all sets
# run on my computer: 
  load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap","tsne.dbscan.eps.MinPts.set.files.DF.Rdata"))
for(x in grep("eps",colnames(tsne.dbscan.eps.MinPts.set.files.DF),value=T)) {
GOseq.tsne(x)
}
```


# summarize GO ORA (102017;020718)
```{r}
tsne.GO.list.files<-list.files(file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap"),pattern="(\\.GO.csv)")
tsne<-gsub("([[:print:]]+)(\\.set[[:digit:]]+)(\\.GO.csv)","\\1",tsne.GO.list.files)
for(x in tsne) {
  tsne.GO.list.files.s<-grep(x,tsne.GO.list.files,value=TRUE)
  temp.list<-lapply(file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap",tsne.GO.list.files.s),function(x) read.csv(file=x,row.names=1))
  names(temp.list)<-tsne.GO.list.files.s
    set<-gsub("([[:print:]]+)(\\.set)([[:digit:]]+)(\\.GO.csv)","\\3",tsne.GO.list.files.s)  
  for(i in 1:length(tsne.GO.list.files.s)) {
    temp.list[[tsne.GO.list.files.s[i]]]$"set"<-set[i]
  }
  temp.DF<-do.call(rbind,temp.list)
  write.csv(temp.DF,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap",paste(x,"sets.GO.csv",sep=".")))
}

```
# box plot or heatmap of each cluster (102319)
```{r}
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap","tsne.dbscan.eps.MinPts.set.files.DF.Rdata"))

tsne.dbscan.eps.MinPts.set.files.DF
tsne.dbscan.eps.MinPts.set.data5<-cbind(data5,tsne.dbscan.eps.MinPts.set.files.DF)
tsne.dbscan.eps.MinPts.set.data5$gt<-gsub("(AT[[:alnum:]]+)(\\.)([[:digit:]]+)","\\3",tsne.dbscan.eps.MinPts.set.data5$AGI)
tsne.dbscan.eps.MinPts.set.data5.Col<-tsne.dbscan.eps.MinPts.set.data5[tsne.dbscan.eps.MinPts.set.data5$gt==1,]
for(x in grep("eps",colnames(tsne.dbscan.eps.MinPts.set.data5),value=TRUE)) { # begin loop
  
tsne.dbscan.eps.MinPts.set.data5.Col.melt<-melt(tsne.dbscan.eps.MinPts.set.data5.Col[,c("1h","4h","25h","49h",x,"gt","AGI")],id.var=c("gt","AGI",x))
head(tsne.dbscan.eps.MinPts.set.data5.Col.melt) 
tsne.dbscan.eps.MinPts.set.data5.Col.melt$gt2<-vlookup(tsne.dbscan.eps.MinPts.set.data5.Col.melt$gt,conversion.table,2)
head(tsne.dbscan.eps.MinPts.set.data5.Col.melt)
#newdata.vst.Col.melt$ssom3.4.unit.classif<-factor(newdata.vst.Col.melt$ssom3.4.unit.classif,levels=c(3,2,6,4,12,9,7,1,5,10,8,11))
tsne.dbscan.eps.MinPts.set.data5.Col.melt$variable<-factor(tsne.dbscan.eps.MinPts.set.data5.Col.melt$variable,levels=c("49h","25h","4h","1h")) # Julin's request for manuscript fig (092215)
names(tsne.dbscan.eps.MinPts.set.data5.Col.melt)<-gsub(x,"tsne.cluster",names(tsne.dbscan.eps.MinPts.set.data5.Col.melt))
tsne.dbscan.eps.MinPts.set.data5.Col.melt$tsne.cluster<-as.factor(tsne.dbscan.eps.MinPts.set.data5.Col.melt$tsne.cluster)

p <- ggplot(tsne.dbscan.eps.MinPts.set.data5.Col.melt,aes(x=variable,y=value,color=value))
p <- p + geom_hline(yintercept=0,colour="red",size=1) # adding red line at fold change = 0 (060515)
p <- p + geom_point(position="jitter",size=1.5,alpha=0.3) 
p <- p + geom_boxplot(outlier.size=0,aes(fill=factor(tsne.cluster),alpha=0.8)) + theme_bw() +xlab("sun/shade treatment (hr)") + scale_colour_gradient2(low="green1",mid="gray",high="magenta")
p <- p + facet_grid(gt2~tsne.cluster,space="free") #+ facet_wrap(ssom3.4unit.classif~.)  ,scale="free"
p <- p + theme(axis.text.x=element_text(angle=90),strip.text.y = element_text(angle=0),legend.position="none")
p <- p + labs(x="Time (hr)", y="Log2 fold change (shade/sun)") #
p <- p + coord_flip()
p 
ggsave(paste("Supplemental_Fig2_tsne_Coexpression_clusters_of_shade-responsive_genes_boxplot.020718",x,".pdf",sep=""),width=11,height=8,unit="in",path=file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap"))

} # end loop

# To which clusters in mutants do genes in Col cluster 3 belong? (useful for looking transition?)ã€€(error: newdata.vsd not found, 020718)
newdata.vst$gt2<-vlookup(newdata.vst$gt,conversion.table,2)
cluster3.Col.AGI<-newdata.vst[newdata.vst$gt==1 & newdata.vst$ssom3.4.unit.classif==3,"AGI"]
ftable(newdata.vst[newdata.vst$AGI %in% cluster3.Col.AGI,c("gt2","ssom3.4.unit.classif")])
```
# heatmap version
```{r}
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap","tsne.dbscan.eps.MinPts.set.files.DF.Rdata"))

tsne.dbscan.eps.MinPts.set.files.DF
tsne.dbscan.eps.MinPts.set.data5<-cbind(data5,tsne.dbscan.eps.MinPts.set.files.DF)
tsne.dbscan.eps.MinPts.set.data5$gt<-gsub("(AT[[:alnum:]]+)(\\.)([[:digit:]]+)","\\3",tsne.dbscan.eps.MinPts.set.data5$AGI)
tsne.dbscan.eps.MinPts.set.data5.Col<-tsne.dbscan.eps.MinPts.set.data5[tsne.dbscan.eps.MinPts.set.data5$gt==1,]
for(x in grep("eps",colnames(tsne.dbscan.eps.MinPts.set.data5),value=TRUE)) { # begin loop
tsne.dbscan.eps.MinPts.set.data5.Col.melt<-melt(tsne.dbscan.eps.MinPts.set.data5.Col[,c("1h","4h","25h","49h",x,"gt","AGI")],id.var=c("gt","AGI",x))
head(tsne.dbscan.eps.MinPts.set.data5.Col.melt) 
tsne.dbscan.eps.MinPts.set.data5.Col.melt$gt2<-vlookup(tsne.dbscan.eps.MinPts.set.data5.Col.melt$gt,conversion.table,2)
head(tsne.dbscan.eps.MinPts.set.data5.Col.melt)
#newdata.vst.Col.melt$ssom3.4.unit.classif<-factor(newdata.vst.Col.melt$ssom3.4.unit.classif,levels=c(3,2,6,4,12,9,7,1,5,10,8,11))
tsne.dbscan.eps.MinPts.set.data5.Col.melt$variable<-factor(tsne.dbscan.eps.MinPts.set.data5.Col.melt$variable,levels=c("49h","25h","4h","1h")) # Julin's request for manuscript fig (092215)
names(tsne.dbscan.eps.MinPts.set.data5.Col.melt)<-gsub(x,"tsne.cluster",names(tsne.dbscan.eps.MinPts.set.data5.Col.melt))
#tsne.dbscan.eps.MinPts.set.data5.Col.melt$tsne.cluster<-as.factor(tsne.dbscan.eps.MinPts.set.data5.Col.melt$tsne.cluster)

# sort clusters according to mean expression value in each cluster
mean.cluster<-tapply(tsne.dbscan.eps.MinPts.set.data5.Col.melt$value,tsne.dbscan.eps.MinPts.set.data5.Col.melt$tsne.cluster,mean)
mean.cluster
tsne.dbscan.eps.MinPts.set.data5.Col.melt$tsne.cluster<-factor(tsne.dbscan.eps.MinPts.set.data5.Col.melt$tsne.cluster,levels=names(sort(mean.cluster,decreasing=T)))
# plot
  p.heatmap <- ggplot(tsne.dbscan.eps.MinPts.set.data5.Col.melt,aes(x=variable,y=tsne.cluster)) + geom_tile(size=0.3,colour="black",aes(fill=value)) 
  
  library(scales) # for muted
  #p.selected <- p.selected + scale_fill_gradient2(limits=c(-1,1),low=muted("green"), high=muted("magenta")) 
  p.heatmap <- p.heatmap + scale_fill_gradient2(limits=c(-2,2),low=muted("green"), high=muted("magenta")) #,guide=guide_legend("none")) 
  p.heatmap <- p.heatmap  + 
    theme(axis.text.x=element_text(size=30,angle=90),
          axis.text.y=element_text(size=30),
          axis.title=element_text(size=40),
          axis.ticks = element_blank(),
          panel.background = element_rect(fill = "white"),
          plot.title=element_text(size=40),
          axis.line=element_blank())
p.heatmap <- p.heatmap + labs(x="Time (hr)", y="cluster",fill="Log2 fold change\n (shade/sun)",title=x) #
p.heatmap <- p.heatmap + coord_flip()
p.heatmap
ggsave(paste("Supplemental_Fig2_tsne_Coexpression_clusters_of_shade-responsive_genes_020718",x,".pdf",sep=""),width=11,height=8,unit="in",path=file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap"))
} 
# transfer representative one (eps3.3_MinPts13.set42) for Supplemental Fig2
# eps3.25_MinPts14.set45 (020818)
system(paste("cp ",file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap","Supplemental_Fig2_tsne_Coexpression_clusters_of_shade-responsive_genes_020718eps3.25_MinPts14.set45.pdf")," ", file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Supplemental_Fig2_tsne_Coexpression_clusters_of_shade-responsive_genes_020718eps3.25_MinPts14.set45.pdf"),sep=""))

# archive (Supplemental Dataset1) (112917,020818)
# system(paste("zip ",file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Supplemental_Dataset1_t-SNE_cluster_responsiveness_heatmap.zip")," ", file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap","Supplemental_Fig2_tsne_Coexpression_clusters_of_shade-responsive_genes_020718eps*"),sep=""))
# use autometer service to combine all pdfs into "Supplemental_Dataset1_t-SNE_cluster_responsiveness_heatmap.pdf"

```
# custom category heatmap (Fig. 2)
# Making list of custom catetory genes enriched in shade responsive genes (Fig2)
```{r}
## start from this line ## 052616
# fold chagne (shade responsiveness)
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","summary.vst.response.Rdata") )
dim(summary.vst.response) #[1] 30533    52
head(summary.vst.response)
# find overlap between shade responsive (in Col) and hormone responsive
#test<-read.csv(file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","TableS4_hormone.responsiveness.SOMclusters100717.csv") ,row.names=1) # needs to use updated one (092917)
test<-read.csv(file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Supplemental_Dataset2_Custom_enrichment_tsnedbscan_clusters_of_shade-responsive_genes020718.csv") ,row.names=1) # tsnedbscan version
test<-test %>% dplyr::select(-c(Short_description,colE...col.F))
colnames(test)
test<-test %>% filter(gt==1) %>% dplyr::select(-gt) # select only found in Col clusters. remove gt column

# reading overlapTable
#load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","overlap.GOseq.melt.summary.SOM.customcat.100717.Rdata")) 
overlap.melt.tsnedbscan.customcat.PLoSGenetics2015mutants.errrDE.summary.files<-list.files(file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap"),pattern="overlap.melt.tsnedbscan.customcat.PLoSGenetics2015mutants.errrDE.summary.v3")

for(y in overlap.melt.tsnedbscan.customcat.PLoSGenetics2015mutants.errrDE.summary.files) 
  { # 140 of y
  load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap",y))
overlap.melt.summary.s<-overlap.melt.summary[!overlap.melt.summary$Var1==0,] # remove cluster "0"
overlap.melt.summary.s<-overlap.melt.summary.s[!overlap.melt.summary.s$Var2=="",]
overlap.melt.summary.s$Var2<-droplevels(overlap.melt.summary.s$Var2)
# select log10>2 & !Var1==0 
overlap.melt.summary.s.significant<-overlap.melt.summary.s[overlap.melt.summary.s$value> -log10(0.01),] # pvalue<0.01 (the same as Fig3)
overlap.melt.summary.s.significant$category.cluster<-with(overlap.melt.summary.s.significant,paste(Var2,Var1,sep="_")) # Var1 = cluster, Var2=category

title<-gsub("(overlap.melt.tsnedbscan.customcat.PLoSGenetics2015mutants.errrDE.summary.v3)(\\.)(eps[[:print:]]+)(\\_)(MinPts[[:digit:]]+)(\\.)(set)(\\.)([[:digit:]]+)(\\.Rdata)","\\3\\4\\5\\6\\7\\9",y)

test.wozero<-test[!test[,title]=="0",c(grep("eps",colnames(test),invert=TRUE,value=TRUE),title)] # remove cluster 0 in test. 
head(test.wozero)
test.wozero.melt<-melt(test.wozero,id.var=c("AGI",title))
colnames(test.wozero.melt)<-gsub(title,"cluster",colnames(test.wozero.melt))
head(test.wozero.melt)
dim(test.wozero.melt) #6061 4
test.wozero.melt<-test.wozero.melt %>% dplyr::select(-variable) %>% filter(!value=="")
dim(test.wozero.melt) #191 4
table(test.wozero.melt$AGI) # all Col with "title" 

# # remove small group (less than 3)
test.wozero.melt$category.cluster<-with(test.wozero.melt,paste(value,cluster,sep="_"))
# select (1) category morethan 3 within test.wozero.melt 
#         (2) 
# %>% filter(value %in% category.morethan3$Var2)
test.wozero.melt.hormone.significant<- test.wozero.melt %>% filter(category.cluster %in% overlap.melt.summary.s.significant$category.cluster)
test.wozero.melt.hormone.significant
category.significnat.morethan3<-test.wozero.melt.hormone.significant %>% dplyr::count(value) %>% filter(n>3) %>% as.data.frame()

if(dim(category.significnat.morethan3)[1]==0) {next} else {

overlap.melt.summary.s.significant$value<-as.character(overlap.melt.summary.s.significant$value)
# test.wozero.melt.hormone.significant.s is my category marker gene list 
test.wozero.melt.hormone.significant.s<-test.wozero.melt.hormone.significant[,c("AGI","value")]
names(test.wozero.melt.hormone.significant.s)<-c("AGI","my.category")
test.wozero.melt.hormone.significant.s.summary<-table(test.wozero.melt.hormone.significant.s)[,-1]
table(test.wozero.melt.hormone.significant.s$my.category)
test.wozero.melt.hormone.significant.s
# reformatting 
test2<-data.frame(AGI=unique(test.wozero.melt.hormone.significant.s$AGI))
my.category<-levels(as.factor(test.wozero.melt.hormone.significant.s$my.category))
for(n in 1:length(my.category)) {
  temp<-rep("",dim(test2)[1])
  names(temp)<-test2$AGI
test.wozero.melt.hormone.significant.s[test.wozero.melt.hormone.significant.s$my.category==my.category[n],"AGI"]
temp[test2$AGI %in% test.wozero.melt.hormone.significant.s[test.wozero.melt.hormone.significant.s$my.category==my.category[n],"AGI"]]<-my.category[n]
temp
test2<-cbind(test2, temp)
}
colnames(test2)[-1]<-my.category
# alternative way
# test3<-tidyr::unite(test2, my.category2, colnames(test2)[-1], sep="",remove=TRUE)
# category.morethan3.v2<-test3 %>% dplyr::count(my.category2) %>% filter(n>3) %>% as.data.frame()
# test.wozero.melt.hormone.significant.sã€€%>% dplyr::count(my.category)
save(test.wozero.melt.hormone.significant.s,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap",paste("test.wozero.melt.hormone.significant.s.v3.",title,".Rdata",sep=""))) # v2 110717; v3 020718
} # category.morethan3 > 0
} # 
```
# comparison of test.wozero.melt.hormone.significant.s gene list (102517;110717ï¼›020818)
```{r}
#test.wozero.melt.hormone.significant.s.files<-list.files(pattern="test.wozero.melt.hormone.significant.s",path=file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap"))
test.wozero.melt.hormone.significant.s.files<-list.files(pattern="test.wozero.melt.hormone.significant.s.v3",path=file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap")) # v3

test.wozero.melt.hormone.significant.s.list<- sapply(file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap",test.wozero.melt.hormone.significant.s.files),function(x) mget(load(x))) # mget 
# rename 
names(test.wozero.melt.hormone.significant.s.list)<-gsub("(../../Nozue2016_SAStranscriptome_output/output/tsne_dbscan_overlapTable_heatmap/test.wozero.melt.hormone.significant.s.v2.)(eps[[:print:]]+)(\\_)(MinPts[[:digit:]]+)(\\.)(set[[:digit:]]+)(.Rdata.test.wozero.melt.hormone.significant.s)","\\2\\3\\4\\5\\6",names(test.wozero.melt.hormone.significant.s.list))

## how to make it data frame and save them as csv file? (111617 for Table S*)
# test.wozero.melt.hormone.significant.s.list.DF<-data.frame(t(sapply(test.wozero.melt.hormone.significant.s.list[[]][1],c)))
# custom categories
load(file.path("..","..","Nozue2016_SAStranscriptome_data","input","hormone.responsiveness6.DF.s.Rdata")) 
x<-names(test.wozero.melt.hormone.significant.s.list)[1]
 temp <-(hormone.responsiveness6.DF.s$AGI %in% test.wozero.melt.hormone.significant.s.list[[x]][1]$AGI)*1
for(x in names(test.wozero.melt.hormone.significant.s.list)[-1]) {
  temp2<-(hormone.responsiveness6.DF.s$AGI %in% test.wozero.melt.hormone.significant.s.list[[x]][1]$AGI)*1
 names(temp2)<-x
temp<-cbind(temp,temp2)
}
   temp<-as.data.frame(temp)
  names(temp)<-names(test.wozero.melt.hormone.significant.s.list)
  hormone.responsiveness6.DF.s.biomarker<-cbind(hormone.responsiveness6.DF.s[,"AGI"],hormone.responsiveness6.DF.s[,-22],temp)
# save csv file
write.csv(hormone.responsiveness6.DF.s.biomarker,file=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Supplemental_Dataset6_source_of_custom_categories_part2.csv")) 
#
test.wozero.melt.hormone.significant.s.list2<-sapply(test.wozero.melt.hormone.significant.s.list,c)
test.wozero.melt.hormone.significant.s.list2["AGI",] # compare this with heatmap (Fig3) for double chekcing
test.wozero.melt.hormone.significant.s.list2["AGI",] %>% sapply(length) %>% tibble() %>% ggplot(aes(.)) + geom_histogram(binwidth=5)

```
## draw expression pattern of custom categories (significantly enriched); Fig2A
# visualize hormone pathway related genes (1213-1214,2015). Updated (052616-), v3 020717, With color legend for Plant Physiol (v4, 091118)
```{r}
#test.wozero.melt.hormone.significant.s.files<-list.files(pattern="test.wozero.melt.hormone.significant.s",path=file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap"))
test.wozero.melt.hormone.significant.s.files<-list.files(pattern="test.wozero.melt.hormone.significant.s.v3",path=file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap"))
test.wozero.melt.hormone.significant.s.files

 test.wozero.melt.hormone.significant.s.list<- sapply(file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap",test.wozero.melt.hormone.significant.s.files),function(x) mget(load(x))) # mget 
# rename 
names(test.wozero.melt.hormone.significant.s.list)<-gsub("(../../Nozue2016_SAStranscriptome_output/output/tsne_dbscan_overlapTable_heatmap/test.wozero.melt.hormone.significant.s.v3.)(eps[[:print:]]+)(\\_)(MinPts[[:digit:]]+)(\\.)(set[[:digit:]]+)(.Rdata.test.wozero.melt.hormone.significant.s)","\\2\\3\\4\\5\\6",names(test.wozero.melt.hormone.significant.s.list))

# summary.vst.response.kazu
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","summary.vst.response.Rdata")) # response (092917)
summary.vst.response.kazu<-summary.vst.response
summary.vst.response.kazu$AGI <- row.names(summary.vst.response.kazu)
response.melt <- melt(summary.vst.response.kazu,id.vars= "AGI",variable.name = "sample", value.name = "log2.dif") # this value is not log transformed. This is just ratio (responsiveness to shade; 101117)
hist(response.melt$log2.dif)
response.melt$log2.dif<-log2(response.melt$log2.dif)
head(response.melt)
response.melt$gt <- as.factor(regmatches(response.melt$sample,regexpr("^[0-9]{1,2}",response.melt$sample)))
response.melt$time <- as.factor(regmatches(response.melt$sample,regexpr("[0-9]{1,2}hr",response.melt$sample)))
response.melt$rep <- as.factor(regmatches(response.melt$sample,regexpr("[A-Z]$",response.melt$sample)))
head(response.melt)
tail(response.melt)
manuscript.genotypes <- as.character(c(1,3,4,5,6,7,8,9,10,13,14,17,19) )
response.melt <- response.melt[response.melt$gt %in% manuscript.genotypes,]
response.melt <- droplevels(response.melt) # droplevels is useful.
summary(response.melt)

# making loop or function for drawing Fig5 graph
for(n in 1:length(names(test.wozero.melt.hormone.significant.s.list))) { # all eps and MinPts with all sets

my.category.diff.heatmap.graph<-list()
#categories <- read.csv(file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","TableS6_test.wozero.melt.hormone.significant.ssss.101017.csv"),row.names=1,as.is=TRUE) #  re-normalize data after subsetting in edgeR DE analysis. specific to each category (101017)
categories<-test.wozero.melt.hormone.significant.s.list[[n]]
print(categories)
title<-names(test.wozero.melt.hormone.significant.s.list)[n]
print(title)
head(categories)
table(categories$my.category)
# select categories larger than three genes
categories %>% dplyr::count(my.category) %>% filter(n>3) %>% as.data.frame() -> selected.category
categories<-categories[categories$my.category %in% selected.category$my.category,]
categories$my.category<-factor(categories$my.category) # categories$my.category has to be factor

#categories$my.category<-factor(categories$my.category,levels=rev(c("IAAup","BLup","BLdown","PIFtarget","ACCup","ABAup","ABAdown","MJup","MJdown","SAup_24h"))) # 120116
#categories$my.category<-factor(categories$my.category,levels=rev(c("IAAup","MJup","ABAdown"))) # 120116
# # summary.vst.response.kazu
# load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","summary.vst.response.Rdata")) # response (092917)
summary.vst.response.kazu<-summary.vst.response
dim(summary.vst.response.kazu) # [1] 30533    52
head(summary.vst.response.kazu)
# for Fig5A, Col
my.category.heatmap.graph<-list()
# add legend to Col (Plant Physiol rev1 version)
my.category.heatmap.graph[[1]]<-my.category.heatmap6(summary3.response=summary.vst.response.kazu,gt.num=1,gt2=conversion.table[conversion.table$num==1,2],newdata.Col.wGO.selected=categories,plotname="my.category.expression.pattern.102617v1.png",w=5,h=5*4/3, path=file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap"),save.plot=F,legend.fill=T,file.name=paste("v3",title,sep=""),category.level=levels(categories$my.category))
for(i in c(17,19,3,4,5,14,10,6,13,7,8,9)) { # 
my.category.heatmap.graph[[i]]<-my.category.heatmap6(summary3.response=summary.vst.response.kazu,gt.num=i,gt2=conversion.table[conversion.table$num==i,2],newdata.Col.wGO.selected=categories,plotname="my.category.expression.pattern.102617v1.png",w=5,h=5*4/3, path=file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap"),save.plot=F,legend.fill=F,file.name=paste("v3",title,sep=""),category.level=levels(categories$my.category)) #  legend.fill=F to T for new Fig5 (all absolute value, previously called FigS2). mean expression value was saved as FigS2.absolute.summary.table.melt   save(summary.table.melt,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output",paste("FigS2.absolute.summary.table.melt",gt.num,".Rdata",sep="")))

}

# summerize mean expression data in each category in one data frame
FigS2.absolute.summary.table.melt.files<-list.files(path=file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap"),pattern="FigS2.absolute.summary.table.melt") # (102617)
FigS2.absolute.summary.table.melt.files<-grep(paste("v3",title,".Rdata",sep=""),FigS2.absolute.summary.table.melt.files,value=T) # v2

FigS2.absolute.summary.table.melt.list2<-sapply(file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap",FigS2.absolute.summary.table.melt.files),function(x) mget(load(x))) # mget will return the value of the object(or objects) in a list. see 
FigS2.absolute.summary.table.melt.DF<-do.call(cbind,FigS2.absolute.summary.table.melt.list2)
colnames(FigS2.absolute.summary.table.melt.DF)
# remove redundant columns (does work properly?, 020818)
colnames(FigS2.absolute.summary.table.melt.DF)<-gsub("(../../Nozue2016_SAStranscriptome_output/output/tsne_dbscan_overlapTable_heatmap/FigS2.absolute.summary.table.melt)([[:print:]]+)(.Rdata.summary.table.melt.)(my.category|variable|value)","\\2.\\4",colnames(FigS2.absolute.summary.table.melt.DF))
FigS2.absolute.summary.table.melt.DF<-FigS2.absolute.summary.table.melt.DF[,c(paste("1v3",title,".my.category",sep=""),paste("1v3",title,".variable",sep="") ,grep("value",colnames(FigS2.absolute.summary.table.melt.DF),value=TRUE))] # v2
write.csv(FigS2.absolute.summary.table.melt.DF,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap",paste("FigS2.absolute.summary.table.melt.DF.v3",title,".csv",sep="")))  # v3

# significance calculation 
summary(categories)
table(categories$my.category)
# select genes in categories
response.melt.merge <- merge(response.melt,categories,by="AGI",all=FALSE)
response.melt.merge$gt <- relevel(response.melt.merge$gt,ref="1")
table(response.melt.merge$gt,response.melt.merge$time,response.melt.merge$my.category)
mutants <- levels(response.melt.merge$gt)[-1] #1 is Col
category.names<- unique(response.melt.merge$my.category)
results <- lapply(mutants,function(gt) {
  sapply(category.names,function(cat.name) {
    tmp <- response.melt.merge[(response.melt.merge$gt=="1" | response.melt.merge$gt==gt) & response.melt.merge$my.category==cat.name,]
    lmer1 <- lmer(log2.dif ~ gt*time + (1|AGI), data=tmp)
    tmp.results <- anova(lmer1)[, "Pr(>F)"]
    print(row.names(anova(lmer1)))
    names(tmp.results) <- row.names(anova(lmer1))
    min(tmp.results[-2]) # we do not care about time
    print(min(tmp.results[-2]))
  })
} )
names(results) <- mutants
results
results.df <- ldply(results,.id="gt")
results.df
colnames(results.df)[-1]<-as.character(category.names)
results.df
# calclate adjusted p-value
results.df.adj <- matrix(
  p.adjust(unlist(results.df[,-1]),"fdr",n=2*length(unlist(results.df[,-1]))), #multiply by 2 because we took the min value of gt and gt*time
  nrow=nrow(results.df))
results.df.adj <- cbind(as.character(results.df[,1]),as.data.frame(results.df.adj))
colnames(results.df.adj) <- colnames(results.df)
results.df.adj

# add p value in heatmap
response.p.values<-results.df.adj # due to skipping writing csv (102617)
response.p.values.melt<-melt(response.p.values,id="gt") #
#my.categories<-grep("gt",colnames(response.p.values),invert=TRUE,value=TRUE)

# add significant to absolute value heatmap
for(i in c(17,19,3,4,5,14,10,6,13,7,8,9)) {
significant.position<-data.frame(category=levels(categories$my.category),x=0,y=1:length(levels(categories$my.category)),symbol.gt="",symbol.gt.time="")
## uncer construction. order of my.categories in significant.position should be the same as levels of category in heatmap.
## use category.level paramenter in heatmap function (102617)
response.p.values.melt.gt<-response.p.values.melt[response.p.values.melt$gt==i,]
significant.position$symbol.gt<-(as.numeric(as.character(vlookup(significant.position$category,response.p.values.melt.gt[,c(2,3)],2,F)))<0.01)*1 # significant level
  significant.position$symbol.gt<-gsub("1","*",significant.position$symbol.gt)
  significant.position$symbol.gt<-gsub("0","",significant.position$symbol.gt)

# add significance to heatmap
my.category.heatmap.graph[[i]]<-my.category.heatmap.graph[[i]] + geom_text(data=significant.position,aes(x=x-0.2,y=y,label=symbol.gt),size=20,color="red") #, remove lines for x and y axis. "gt" is red.
rm(significant.position)
}
# separate drawing of Fig2A (with color legend) and Fig2B 
my.category.heatmap.Fig2A <- plot_grid(my.category.heatmap.graph[[1]]+labs(fill="Log2 fold change\n (shade/sun)"),ggplot()+geom_blank(),  ggplot()+geom_blank(),                            
                      vjust = 1.1, hjust=-0.5,
                      ncol=3,nrow=1,
                      rel_widths=c(1.275,0.725,1),
                      label_size = 10
                      ) # no "A" and "B" because they overlap with graphs. Label 
my.category.heatmap.Fig2B<-plot_grid(my.category.heatmap.graph[[17]],my.category.heatmap.graph[[19]],my.category.heatmap.graph[[3]],my.category.heatmap.graph[[4]],my.category.heatmap.graph[[5]],my.category.heatmap.graph[[14]],my.category.heatmap.graph[[10]],my.category.heatmap.graph[[6]],my.category.heatmap.graph[[13]],my.category.heatmap.graph[[7]],my.category.heatmap.graph[[8]],my.category.heatmap.graph[[9]],
                      vjust = 1.1, hjust=-0.5,
                      ncol=3,nrow=4,
                      label_size = 10) 
my.category.heatmap.all2<-plot_grid(my.category.heatmap.Fig2A,my.category.heatmap.Fig2B,rel_heights=c(1,4),ncol=1,nrow=2)+ draw_text(title,x=0.75,y=0.95,size=50)
# no "A" and "B" because they overlap with graphs. Label them later in Illustlator (020818)
save_plot(file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap",paste("Supplemental_Dataset7_all.my.category.expression.pattern.v16.log2.signif.v4.",title,".png",sep="")), my.category.heatmap.all2,
          base_aspect_ratio = 2.5,
          base_height=20,
          limitsize = FALSE,
          ncol=1,nrow=2
)

} # loop for test.wozero.melt.hormone.significant.s.list. v2
# transfer eps3.3_MinPts13.set42 for Fig2
# eps3.25_MinPts14.set45 (020818; Plant Physiol version 091118)
# system(paste("cp ",file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap","Supplemental_Dataset7_all.my.category.expression.pattern.v16.log2.signif.v4.eps3.25_MinPts14.set45.png")," ", file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Fig2_all.my.category.expression.pattern.v16.log2.signif.v4.eps3.25_MinPts14.set45.png"),sep=""))
```
# archivre Supplemental_Dataset7 files 
```{r}
# zip files (not used because they were too big (~1G)
# system(paste("zip ",file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Supplemental_Dataset7v4.zip")," ", file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap","Supplemental_Dataset7_all.my.category.expression.pattern.v16.log2.signif.v4.*"),sep=""))
```
# convert into pdf and add parameter name onto graph, convert into png by sips (command line), convert into pdf, combine all (020918); see Rscripts for that (copy and paste them later)
```{r}
# 
system(paste("cp ",file.path("..","..","Nozue2016_SAStranscriptome_output","output","tsne_dbscan_overlapTable_heatmap","Supplemental_Dataset7_all.my.category.expression.pattern.v16.log2.signif.v4.*")," ", file.path("..","..","Nozue2016_SAStranscriptome_output","output","pngs"),sep=""))
#pdf.files<-list.files(path="../../Nozue2016_SAStranscriptome_output/output/converted.pdfs/")
png.files<-list.files(path="../../Nozue2016_SAStranscriptome_output/output/pngs/")

system("mkdir ../reduced.pngs")
system("mkdir ../reduced.pdfs")
system(paste("sips -z 600 800 -s format png ",file.path("..","..","Nozue2016_SAStranscriptome_output","output","pngs",png.files[2])," --out ", file.path("..","..","Nozue2016_SAStranscriptome_output","output","reduced.pngs",png.files[2]),sep=""))

# reduce size
for(i in 1:length(png.files)) {
system(paste("sips -z 600 800 -s format png ",file.path("..","..","Nozue2016_SAStranscriptome_output","output","pngs",png.files[i])," --out ", file.path("..","..","Nozue2016_SAStranscriptome_output","output","reduced.pngs",png.files[i]),sep=""))
}
# convert resultant png into pdf
for(i in 1:length(png.files)) {
system(paste("sips -s format pdf ",file.path("..","..","Nozue2016_SAStranscriptome_output","output","reduced.pngs",png.files[i])," --out ", file.path("..","..","Nozue2016_SAStranscriptome_output","output","reduced.pdfs",gsub(".png",".pdf",png.files[i])),sep=""))
}

# combine those pdf into one pdf using Autometor (22.4 M with -z 600 800 for png)


```


# the end of custom category heatmap (shade responsiveness) (Fig. 2 and Supplemental_Dataset7)


## differentially expressed gene analysis (092617)
2a.DE at any time point (sun)
```{r eval=FALSE}
#system("mkdir ../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/genotype_specific5") # one time
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","samples.PLoSGenetics2015mutants.errrDE.Rdata")) # 100717
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","dge.PLoSGenetics2015mutants.errrDE.Rdata"))

SAS.expression.vst<-read.csv(file.path("..","..","Nozue2016_SAStranscriptome_data","input","kazu.SAS.expression.VST.092517.csv"),row.names=1) # new transformed data by DEseq2 (092617) # not found? (092817)
dim(SAS.expression.vst) #[1] 30533   287
Genotype<-levels(as.factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|25|49)(A.merged.bam)","\\2",names(SAS.expression.vst)))) # 13 genotype

for(i in Genotype[-1]) {# Genotype 17, 3,4,5,6,7,9 exceed 2000 genes. List up 5000 genes
  samples.nolow5.s<-samples.PLoSGenetics2015mutants.errrDE[samples.PLoSGenetics2015mutants.errrDE$genotype=="1"|samples.PLoSGenetics2015mutants.errrDE$genotype==i,] # 100717
  samples.nolow5.s<-samples.nolow5.s[samples.nolow5.s$trt=="H",] # only "sun" treatment
  samples.nolow5.s$genotype<-as.character(samples.nolow5.s$genotype)
  dge.nolow.s<-dge.PLoSGenetics2015mutants.errrDE[,colnames(dge.PLoSGenetics2015mutants.errrDE) %in% samples.nolow5.s$file]
  # eliminating low expressed genes in dge.nolow.s
  dge.nolow.s<-dge.nolow.s[rowSums(dge.nolow.s$counts > 5) >= 3,]
   # recalculate normalization (100717)
  dge.nolow.s<-calcNormFactors(dge.nolow.s)
  # design model 
  design.nolow.s.gt.batch3<- model.matrix(~samples.nolow5.s$genotype*samples.nolow5.s$time  + samples.nolow5.s$batch) # new model to detect genes with gt effects (100314)  
  colnames(design.nolow.s.gt.batch3) <- gsub("samples.nolow5.s$","",colnames(design.nolow.s.gt.batch3),fixed=TRUE) # nicer column names
  colnames(design.nolow.s.gt.batch3)
  print(design.nolow.s.gt.batch3)
  dge.nolow.s.batch3.glm <- estimateGLMCommonDisp(dge.nolow.s,design.nolow.s.gt.batch3, verbose=TRUE)
  dge.nolow.s.batch3.glm <- estimateGLMTrendedDisp(dge.nolow.s.batch3.glm,design.nolow.s.gt.batch3, verbose=TRUE)
  dge.nolow.s.batch3.glm<-estimateGLMTagwiseDisp(dge.nolow.s.batch3.glm,design.nolow.s.gt.batch3)
  #pdf(file=paste("BCV_MDS_with_genotype.wo16h_gt",i,".pdf",sep=""),width=11,height=8)
  #par(mfrow=c(1,2))
  #plotBCV(dge.nolow.s.batch3.glm,main=paste("genotype",i))
  #plotMDS(dge.nolow.s.batch3.glm,main=paste("genotype",i))
  #dev.off()  
  dge.nolow.s.batch3.glm.fit <- glmFit(dge.nolow.s.batch3.glm,design.nolow.s.gt.batch3)
  genotype.alltime.coef<-c(paste("genotype",i,sep=""),paste("genotype",i,":time4",sep=""),paste("genotype",i,":time25",sep=""),paste("genotype",i,":time49",sep=""))  
  dge.nolow.s.batch3.glm.lrt <- glmLRT(dge.nolow.s.batch3.glm.fit,coef=genotype.alltime.coef)
  DE.batch3.gt<-topTags(dge.nolow.s.batch3.glm.lrt,n=10000)$table[topTags(dge.nolow.s.batch3.glm.lrt,n=10000)$table$FDR<0.001,] # 0.05 for genotype_specific2 which gave me 5000 genes in hy5 (genotype3)
  dim(DE.batch3.gt) #
  save(DE.batch3.gt,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h","genotype_specific5",paste("DE.genotype.batch.sun.gt.wo16h.PLoSGenetics2015mutants.errrDE",i,".Rdata",sep=""))) 
  # save  dge.nolow.s.batch3.glm.lrt
    save(dge.nolow.s.batch3.glm.lrt,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h","genotype_specific5",paste("dge.nolow.s.batch3.glm.lrt.sun.PLoSGenetics2015mutants.errrDE",i,".Rdata",sep=""))) # added this line (try to use this object for coef, 061715)
}
```
2b.DE at any time point (shade)
```{r eval=FALSE}
for(i in Genotype[-1]) {# Genotype 17, 3,4,5,6,7,9 exceed 2000 genes. List up 5000 genes
  samples.nolow5.s<-samples.PLoSGenetics2015mutants.errrDE[samples.PLoSGenetics2015mutants.errrDE$genotype=="1"|samples.PLoSGenetics2015mutants.errrDE$genotype==i,] # 100717
  samples.nolow5.s<-samples.nolow5.s[samples.nolow5.s$trt=="L",] # only "shade" treatment
  samples.nolow5.s$genotype<-as.character(samples.nolow5.s$genotype)
  dge.nolow.s<-dge.PLoSGenetics2015mutants.errrDE[,colnames(dge.PLoSGenetics2015mutants.errrDE) %in% samples.nolow5.s$file] # 100717
  # eliminating low expressed genes in dge.nolow.s
  dge.nolow.s<-dge.nolow.s[rowSums(dge.nolow.s$counts > 5) >= 3,]
    # recalculate normalization (100717)
  dge.nolow.s<-calcNormFactors(dge.nolow.s)
  # design model 
  design.nolow.s.gt.batch3<- model.matrix(~samples.nolow5.s$genotype*samples.nolow5.s$time  + samples.nolow5.s$batch) # new model to detect genes with gt effects (100314)  
  colnames(design.nolow.s.gt.batch3) <- gsub("samples.nolow5.s$","",colnames(design.nolow.s.gt.batch3),fixed=TRUE) # nicer column names
  colnames(design.nolow.s.gt.batch3)
  print(design.nolow.s.gt.batch3)
  dge.nolow.s.batch3.glm <- estimateGLMCommonDisp(dge.nolow.s,design.nolow.s.gt.batch3, verbose=TRUE)
  dge.nolow.s.batch3.glm <- estimateGLMTrendedDisp(dge.nolow.s.batch3.glm,design.nolow.s.gt.batch3, verbose=TRUE)
  dge.nolow.s.batch3.glm<-estimateGLMTagwiseDisp(dge.nolow.s.batch3.glm,design.nolow.s.gt.batch3)
  #pdf(file=paste("BCV_MDS_with_genotype.wo16h_gt",i,".pdf",sep=""),width=11,height=8)
  #par(mfrow=c(1,2))
  #plotBCV(dge.nolow.s.batch3.glm,main=paste("genotype",i))
  #plotMDS(dge.nolow.s.batch3.glm,main=paste("genotype",i))
  #dev.off()  
  dge.nolow.s.batch3.glm.fit <- glmFit(dge.nolow.s.batch3.glm,design.nolow.s.gt.batch3)
  genotype.alltime.coef<-c(paste("genotype",i,sep=""),paste("genotype",i,":time4",sep=""),paste("genotype",i,":time25",sep=""),paste("genotype",i,":time49",sep=""))  
  dge.nolow.s.batch3.glm.lrt <- glmLRT(dge.nolow.s.batch3.glm.fit,coef=genotype.alltime.coef)
  
  DE.batch3.gt<-topTags(dge.nolow.s.batch3.glm.lrt,n=10000)$table[topTags(dge.nolow.s.batch3.glm.lrt,n=10000)$table$FDR<0.001,] # 0.05 for genotype_specific2 which gave me 5000 genes in hy5 (genotype3)
  dim(DE.batch3.gt) #
  save(DE.batch3.gt,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h","genotype_specific5",paste("DE.genotype.batch.shade.gt.wo16h.PLoSGenetics2015mutants.errrDE",i,".Rdata",sep="")))  
  save(dge.nolow.s.batch3.glm.lrt,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h","genotype_specific5",paste("dge.nolow.s.batch3.glm.lrt.shade.PLoSGenetics2015mutants.errrDE",i,".Rdata",sep=""))) # 100717
}
```
## overlap between Col modules and mutation affected genes
### summerize mutation affected genes
```{r eval=FALSE}
conversion.table # has to be made in function_
# sun
DE.genotype.batch.sun.files.kazu<-list.files(pattern="DE.genotype.batch.sun.gt.wo16h.PLoSGenetics2015mutants.errrDE",path="../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/genotype_specific5/") # needs to select "sun" object (060615)
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h","genotype_specific5",DE.genotype.batch.sun.files.kazu[1]))

DE.genotype.batch.sun.genes<-rownames(DE.batch3.gt)
for(xx in DE.genotype.batch.sun.files.kazu[-1]) {
  load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h","genotype_specific5",xx))
  DE.genotype.batch.sun.genes<-c(DE.genotype.batch.sun.genes,rownames(DE.batch3.gt))
}
# shade
DE.genotype.batch.shade.files.kazu<-list.files(pattern="DE.genotype.batch.shade.gt.wo16h.PLoSGenetics2015mutants.errrDE",path="../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/genotype_specific5/")
## load the first file
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h","genotype_specific5",DE.genotype.batch.shade.files.kazu[1]))
DE.genotype.batch.shade.genes<-rownames(DE.batch3.gt)
## load following files
for(xx in DE.genotype.batch.shade.files.kazu[-1]) {
  load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h","genotype_specific5",xx))
  DE.genotype.batch.shade.genes<-c(DE.genotype.batch.shade.genes,rownames(DE.batch3.gt))
}
# get unique gene list for making data frame
DE.genotype.batch.genes.unique<-unique(c(DE.genotype.batch.sun.genes,DE.genotype.batch.shade.genes))
length(DE.genotype.batch.genes.unique) #3356 > 3349
# convert to data frame version (092617)
DE.genotype.batch.genes.unique.DF<-data.frame(AGI=DE.genotype.batch.genes.unique)
# sun
for(x in DE.genotype.batch.sun.files.kazu) {
    load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h","genotype_specific5",x))
  temp<-as.integer(DE.genotype.batch.genes.unique %in% rownames(DE.batch3.gt)) 
  temp[temp==1]<-rep(paste(conversion.table[conversion.table==gsub("(DE.genotype.batch.sun.gt.wo16h.PLoSGenetics2015mutants.errrDE)([[:digit:]]+)(.Rdata)","\\2",x),2],"sun",sep="."),sum(temp))
  DE.genotype.batch.genes.unique.DF<-cbind(DE.genotype.batch.genes.unique.DF,temp)          
}
# shade
for(x in DE.genotype.batch.shade.files.kazu) {
  load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h","genotype_specific5",x))
  temp<-as.integer(DE.genotype.batch.genes.unique %in% rownames(DE.batch3.gt)) 
  temp[temp==1]<-rep(paste(conversion.table[conversion.table==gsub("(DE.genotype.batch.shade.gt.wo16h.PLoSGenetics2015mutants.errrDE)([[:digit:]]+)(.Rdata)","\\2",x),2],"shade",sep="."),sum(temp))
  DE.genotype.batch.genes.unique.DF<-cbind(DE.genotype.batch.genes.unique.DF,temp)                 
}
head(DE.genotype.batch.genes.unique.DF)
dim(DE.genotype.batch.genes.unique.DF)
# ## 
names(DE.genotype.batch.genes.unique.DF)
# 091016, 092617
DE.genotype.sun<-gsub("(DE.genotype.batch.sun.gt.wo16h.PLoSGenetics2015mutants.errrDE)([[:digit:]]+)(.Rdata)","\\2",DE.genotype.batch.sun.files.kazu) # 100717
DE.genotype.shade<-gsub("(DE.genotype.batch.shade.gt.wo16h.PLoSGenetics2015mutants.errrDE)([[:digit:]]+)(.Rdata)","\\2",DE.genotype.batch.shade.files.kazu) # 100717
names(DE.genotype.batch.genes.unique.DF)[-1]<-c(paste(vlookup(DE.genotype.sun,conversion.table,2),"sun",sep="."),paste(vlookup(DE.genotype.shade,conversion.table,2),"shade",sep=".")) # the first column is "AGI"
names(DE.genotype.batch.genes.unique.DF)
# the end of 092617 version

# remove "0"?
# remove splicing variant info in AGI
#DE.genotype.batch.genes.unique.DF[DE.genotype.batch.genes.unique.DF==0]<-""
DE.genotype.batch.genes.unique.DF$AGI<-gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",DE.genotype.batch.genes.unique.DF$AGI)
dim(DE.genotype.batch.genes.unique.DF) #3360 25 (100717)
save(DE.genotype.batch.genes.unique.DF,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","DE.genotype.batch.genes.unique.DF.Rdata")) 
```
## co-expression analysis by WGCNA
```{r eval=FALSE}
# The following setting is important, do not omit.
options(stringsAsFactors = FALSE)
```
## Work in Whitney with multi-threds 
```{r eval=FALSE}
# Allow multi-threading within WGCNA. This helps speed up certain calculations.
# At present this call is necessary for the code to work.
# Any error here may be ignored but you may want to update WGCNA if you see one.
# Caution: skip this line if you run RStudio or other third-party R environments.
# See note above.
if(Sys.info()["nodename"]=="whitney") {
  enableWGCNAThreads(10) # in Whitney (Maloof lab server) 
  }
```
## Co-expressed gene module detection by WGCNA
```{r eval=FALSE}
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","SAS.expression.vst.s.kazu.largeCV.Rdata")) 
#load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","samples.nolow5.Rdata"))
Genotype<-levels(as.factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|25|49)(A.merged.bam)","\\2",names(SAS.expression.vst.s.kazu.largeCV)[-c(1,288)])))

for(x in Genotype) {
  SAS.expression.vst.s.kazu.largeCV.x<-SAS.expression.vst.s.kazu.largeCV[,gsub("([C|D|E])([[:digit:]]+)(H|L)(1|4|16|25|49)(A.merged.bam)","\\2",colnames(SAS.expression.vst.s.kazu.largeCV))==x]
    dim(SAS.expression.vst.s.kazu.largeCV.x) #[1] 7438     21
  datExpr <-t(SAS.expression.vst.s.kazu.largeCV.x)
  # Choose a set of soft-thresholding powers
  powers = c(c(1:9), seq(from = 2, to=20, by=10))
  sft = pickSoftThreshold(datExpr, powerVector = powers, verbose = 5)  
  # Plot the results:
  #sizeGrWindow(9, 5)
  pdf(paste("../../Nozue2016_SAStranscriptome_output/output/all",x,".largeCV.softthresholding.pdf",sep=""),width=10,height=8)
  par(mfrow = c(1,2));
  cex1 = 0.9;
  # Scale-free topology fit index as a function of the soft-thresholding power
  plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
       xlab="Soft Threshold (power)",ylab="Scale Free Topology Model Fit,signed R^2",type="n",
       main = paste("Scale independence"));
  text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
       labels=powers,cex=cex1,col="red");
  # this line corresponds to using an R^2 cut-off of h
  abline(h=0.90,col="red")
  # Mean connectivity as a function of the soft-thresholding power
  plot(sft$fitIndices[,1], sft$fitIndices[,5],
       xlab="Soft Threshold (power)",ylab="Mean Connectivity", type="n",
       main = paste("Mean connectivity"))
  text(sft$fitIndices[,1], sft$fitIndices[,5], labels=powers, cex=cex1,col="red")
  dev.off()
  # 
  net = blockwiseModules(datExpr, power = 6,
                         TOMType = "unsigned", minModuleSize = 20,
                         reassignThreshold = 0, mergeCutHeight = 0.25,
                         numericLabels = TRUE, pamRespectsDendro = FALSE,
                         saveTOMs = TRUE,
                         saveTOMFileBase = paste("SAS.expression.vst.s.kazu.largeCV",x,".TOM",sep=""),
                         verbose = 3)
  save(net,file=paste("../../Nozue2016_SAStranscriptome_output/output/SAS.expression.vst.s.kazu.largeCV",x,".Rdata",sep=""))  # confusing name dim(net) [1] 17757
  # open a graphics window
  pdf(file=paste("../../Nozue2016_SAStranscriptome_output/output/all",x,".largeCV.dendrogram.pdf",sep=""),width=10,height=8)
  # Convert labels to colors for plotting
  mergedColors = labels2colors(net$colors)
  # Plot the dendrogram and the module colors underneath
  plotDendroAndColors(net$dendrograms[[1]], mergedColors[net$blockGenes[[1]]],
                      "Module colors",
                      dendroLabels = FALSE, hang = 0.03,
                      addGuide = TRUE, guideHang = 0.05)
  dev.off()
  # save parameters  
  moduleLabels = net$colors
  moduleColors = labels2colors(net$colors)
  MEs = net$MEs
  geneTree = net$dendrograms[[1]]
  save(MEs, moduleLabels, moduleColors, geneTree,file = paste("../../Nozue2016_SAStranscriptome_output/output/all",x,".largeCV.RData",sep=""))
} # recalculating 082016. Recalculating genotype "3" due to "3" was missing (091616), (092817)

# new WGCNA modules are too many (110716)
# how many modules?
  table(net$colors);length(table(net$colors)) #
## recutBlockwiseTrees() is too divide modules further
## try mergeCloseMOdulesl with higher height (0.6) for all Genotype
## resultant "tan" module is SA/JA module but it is too big. (110916)
## try 0.55
##  cutHeight = 0.55, 27 modules, yellow = SA is too large
## try 0.4, 37 modules, gray60 that PCC1 belongs is slight large (594)
## try 0.35 (merge.net35) 44 modules (JA and SA GO terms were found in the same module; gray60)
### all PCC1, XBAT34, AIG1, ACD6 belong to "gray60"
## try 0.3 (merge.net30) 50 modules (JA (gray60) and SA (orange) GO terms were separated). PCC1 (  AT3G22231; orange, XBAT34 (AT4G14365, orange), but AIG1 (thistle2;  AT1G33960), ACD6 (?)
## try 0.32; all four genes belong to "orange"

load("../../Nozue2016_SAStranscriptome_output/output/SAS.expression.vst.s.kazu.largeCV.Rdata") #  where 
load("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/samples.nolow5.Rdata")
Genotype<-levels(as.factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|25|49)(A.merged.bam)","\\2",names(SAS.expression.vst.s.kazu.largeCV)[-c(1,288)])))

rm(net)
for(x in Genotype) {
  SAS.expression.vst.s.kazu.largeCV.x<-SAS.expression.vst.s.kazu.largeCV[,gsub("([C|D|E])([[:digit:]]+)(H|L)(1|4|16|25|49)(A.merged.bam)","\\2",colnames(SAS.expression.vst.s.kazu.largeCV))==x]
    dim(SAS.expression.vst.s.kazu.largeCV.x) #[1] 17757     21
  datExpr <-t(SAS.expression.vst.s.kazu.largeCV.x)
  load(paste("../../Nozue2016_SAStranscriptome_output/output/SAS.expression.vst.s.kazu.largeCV",x,".Rdata",sep=""))
# merge.net2<-mergeCloseModules(
#      # input data
#   exprData=datExpr, colors=net$colors,
#   # Optional starting eigengenes
#   MEs = NULL,
#   # Optional restriction to a subset of all sets
#   useSets = NULL,
#   # If missing data are present, impute them?
#   impute = TRUE,
#   # Input handling options
#   checkDataFormat = TRUE,
#   unassdColor = if (is.numeric(colors)) 0 else "grey",
#   # Options for eigengene network construction
#   corFnc = cor, corOptions = list(use = 'p'),
#   useAbs = FALSE,
#   # Options for constructing the consensus
#   equalizeQuantiles = FALSE,
#   quantileSummary = "mean",
#   consensusQuantile = 0,
#   # Merging options
#   cutHeight = 0.6,
#   iterate = TRUE,
#   # Output options
#   relabel = FALSE,
#   colorSeq = NULL,
#   getNewMEs = TRUE,
#   getNewUnassdME = TRUE,
#   # Options controlling behaviour of the function
#   trapErrors = FALSE,
#   verbose = 1, indent = 0)
# # save parameters  
#   moduleLabels = merge.net2$colors
#   moduleColors = labels2colors(merge.net2$colors)
#   MEs = merge.net2$MEs
#   geneTree = merge.net2$dendrograms[[1]]
#   save(MEs, moduleLabels, moduleColors, geneTree,file = paste("../../Nozue2016_SAStranscriptome_output/output/all",x,".largeCV.merge.net2.RData",sep="")) 
# rm(net)
# rm(merge.net2)
# } # resultant "tan" module is SA/JA module but it is too big. (110916)
merge.net32<-mergeCloseModules(
     # input data
  exprData=datExpr, colors=net$colors,
  # Optional starting eigengenes
  MEs = NULL,
  # Optional restriction to a subset of all sets
  useSets = NULL,
  # If missing data are present, impute them?
  impute = TRUE,
  # Input handling options
  checkDataFormat = TRUE,
  unassdColor = if (is.numeric(colors)) 0 else "grey",
  # Options for eigengene network construction
  corFnc = cor, corOptions = list(use = 'p'),
  useAbs = FALSE,
  # Options for constructing the consensus
  equalizeQuantiles = FALSE,
  quantileSummary = "mean",
  consensusQuantile = 0,
  # Merging options
  cutHeight = 0.32, # need to optimize 
  iterate = TRUE,
  # Output options
  relabel = FALSE,
  colorSeq = NULL,
  getNewMEs = TRUE,
  getNewUnassdME = TRUE,
  # Options controlling behaviour of the function
  trapErrors = FALSE,
  verbose = 1, indent = 0)
# save parameters  
  moduleLabels = merge.net32$colors
  moduleColors = labels2colors(merge.net32$colors)
  #MEs = merge.net32$MEs # null (120916)
  MEs = merge.net32$newMEs # new one? (120916)
  geneTree = merge.net32$dendrograms[[1]]
  save(MEs, moduleLabels, moduleColors, geneTree,file = paste("../../Nozue2016_SAStranscriptome_output/output/all",x,".largeCV.merge.net32.RData",sep="")) # "salmon" fixed (121216)
rm(net)
rm(merge.net32)
} # resultant ***

# check new module numbers
moduleColors.files<-list.files(pattern=".largeCV.merge.net32.RData",path=file.path("..","..","Nozue2016_SAStranscriptome_output","output")) # for  merged modules by mergeCloseModules function above for Col (110816); newer (110916)
### use merged modules by mergeCloseModules function above for Col (22 modules; 110716; ** modules 092817)
#load("all1.largeCV.merge.net32.RData")
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","all1.largeCV.merge.net32.RData")) 
#load("/Volumes/data_work/Data8/NGS_related/Arabidopsis_analysis/Nozue2016_SAStranscriptome_finalfinal_newVST/Nozue2016_SAStranscriptome_output/output/all1.largeCV.merge.net32.RData")
########
moduleColors.allCol.largeCV<-moduleColors
table(moduleColors.allCol.largeCV);length(table(moduleColors.allCol.largeCV)) #
## check module character by GO ORA below
## back to single thred in whitney
if(Sys.info()["nodename"]=="whitney") {
  enableWGCNAThreads(2) # in Whitney (Maloof lab server) 
  }
```
## summary of WGCNA modules in mutants (092817 running with error due to using old SAS.expression.vst.s.kazu.largeCV.Rdata)
```{r eval=FALSE}
#setwd(homedir)
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","SAS.expression.vst.s.kazu.largeCV.Rdata")) # 7025 

moduleColors.files<-list.files(path=file.path("..","..","Nozue2016_SAStranscriptome_output","output"),pattern=".largeCV.merge.net32.RData") # for  merged modules by mergeCloseModules function above for Col (110816); newer (110916)
# 092717
### use merged modules by mergeCloseModules function above for Col (22 modules; 110716); (**modules; 092817)
## start with Genotype1
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","all1.largeCV.merge.net32.RData"))
moduleColors.allCol.largeCV<-moduleColors
table(moduleColors.allCol.largeCV)
temp<-moduleColors.allCol.largeCV
## adding more genotype
for(y in moduleColors.files[!moduleColors.files=="all1.largeCV.merge.net32.RData"]) {
  load(file.path("..","..","Nozue2016_SAStranscriptome_output","output",y))
  print(length(moduleColors)) # [1] 17757
  # switch "orange" and "salmon" (to keep "salomn" module name)
#moduleColors.temp<-moduleColors
#moduleColors.temp<-gsub("^orange","salmon",moduleColors)
#moduleColors.temp<-gsub("salmon4","orange4",moduleColors.temp)
#moduleColors<-moduleColors.temp
# the end of switch
  matched.moduleColors<-matchLabels(moduleColors,moduleColors.allCol.largeCV)
  temp<-cbind(temp, matched.moduleColors)
}   
length(moduleColors.files)
temp<-as.data.frame(temp)

names(temp)<-gsub("(all)([[:digit:]]+)(.largeCV.merge.net32.RData)","\\2",c("all1.largeCV.merge.net32.RData",moduleColors.files[!moduleColors.files=="all1.largeCV.merge.net32.RData"]))
SAS.expression.vst.s.kazu.largeCV.modules<-cbind(rownames(SAS.expression.vst.s.kazu.largeCV),temp)

names(SAS.expression.vst.s.kazu.largeCV.modules)[1]<-"AGI"
names(SAS.expression.vst.s.kazu.largeCV.modules)
table(SAS.expression.vst.s.kazu.largeCV.modules$"1")
# look modules of PCC1, XBAT34, AIG1 (PCC1 (  AT3G22231; orange, XBAT34 (AT4G14365, orange), but AIG1 (thistle2;  AT1G33960), ACD6 (AT4G14400, ?) > all became "salmon" in threshold 0.32
SAS.expression.vst.s.kazu.largeCV.modules[ gsub("([[:digit:]]+)(\\.)([[:digit:]]+)","\\1",SAS.expression.vst.s.kazu.largeCV.modules$AGI) %in% c("AT3G22231","AT4G14365","AT1G33960","AT4G14400"),]
# all of them belong to "purple" (092817)
# switch "purple" and "salmon" (092817)
SAS.expression.vst.s.kazu.largeCV.modules[SAS.expression.vst.s.kazu.largeCV.modules=="salmon"]<-"temp"
SAS.expression.vst.s.kazu.largeCV.modules[SAS.expression.vst.s.kazu.largeCV.modules=="purple"]<-"salmon"
SAS.expression.vst.s.kazu.largeCV.modules[SAS.expression.vst.s.kazu.largeCV.modules=="temp"]<-"purple"
# 
SAS.expression.vst.s.kazu.largeCV.modules[ gsub("([[:digit:]]+)(\\.)([[:digit:]]+)","\\1",SAS.expression.vst.s.kazu.largeCV.modules$AGI) %in% c("AT3G22231","AT4G14365","AT1G33960","AT4G14400"),] # OK
head(SAS.expression.vst.s.kazu.largeCV.modules)
table(SAS.expression.vst.s.kazu.largeCV.modules$"1") # OK purple and salmon switched

save(SAS.expression.vst.s.kazu.largeCV.modules,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","SAS.expression.vst.s.kazu.largeCV.modules092817.Rdata"))

```

# overlap between WGCNA Col modules and mutation affected genes 
```{r,eval=FALSE}
#setwd(homedir)
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","DE.genotype.batch.genes.unique.DF.Rdata")) #[1] 3350   25 > 3360 25 (100717)
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","SAS.expression.vst.s.kazu.largeCV.modules092817.Rdata"))

dim(SAS.expression.vst.s.kazu.largeCV.modules)# [1] 7025   14
table(SAS.expression.vst.s.kazu.largeCV.modules$"1") # OK purple and salmon switched
SAS.expression.vst.s.kazu.largeCV.modules[ gsub("([[:digit:]]+)(\\.)([[:digit:]]+)","\\1",SAS.expression.vst.s.kazu.largeCV.modules$AGI) %in% c("AT3G22231","AT4G14365","AT1G33960","AT4G14400"),]
# MYC2 (AT1G32640), MYC3 (AT5G46760), MYC4 (AT4G17880)
SAS.expression.vst.s.kazu.largeCV.modules[ gsub("([[:digit:]]+)(\\.)([[:digit:]]+)","\\1",SAS.expression.vst.s.kazu.largeCV.modules$AGI) %in% c("AT1G32640","AT5G46760","AT4G17880"),] # only MYC2
# six MYBs that regulate indolic glucosinolates (GSLs): MYB34 AT5G60890, MYB51 AT1G18570, MYB122 AT1G74080, MYB28 AT5G61420, MYB29 AT5G07690, MYB76      AT5G07700
SAS.expression.vst.s.kazu.largeCV.modules[ gsub("([[:digit:]]+)(\\.)([[:digit:]]+)","\\1",SAS.expression.vst.s.kazu.largeCV.modules$AGI) %in% c("AT5G60890","AT1G18570","AT1G74080","AT5G61420","AT5G07690","AT5G07700"),]

#library(WGCNA)
SAS.expression.vst.s.kazu.largeCV.modules$AGI<-gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",SAS.expression.vst.s.kazu.largeCV.modules$AGI)
# merge with WGCNA modules
test<-merge(DE.genotype.batch.genes.unique.DF,SAS.expression.vst.s.kazu.largeCV.modules[,c("AGI","1")],by="AGI",all=TRUE)
DE.genotype.batch.genes.WGCNAmodules<-test
DE.genotype.batch.genes.WGCNAmodules[is.na(DE.genotype.batch.genes.WGCNAmodules)]<-"0"
names(DE.genotype.batch.genes.WGCNAmodules)<-sub("1$","WGCNAmodules",names(DE.genotype.batch.genes.WGCNAmodules)) # 
table(DE.genotype.batch.genes.WGCNAmodules$WGCNAmodules) # "salmon" should be 155, "purple" should be 169
length(table(DE.genotype.batch.genes.WGCNAmodules$WGCNAmodules)) #
# adding gene description
DE.genotype.batch.genes.WGCNAmodules.description<-merge(DE.genotype.batch.genes.WGCNAmodules,TAIR10_gene_descriptions[!duplicated(TAIR10_gene_descriptions$AGI),c("AGI","Short_description","colE...col.F")],by="AGI")
# Supplemental Dataset6
write.csv(DE.genotype.batch.genes.WGCNAmodules.description,file=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Supplemental_Dataset10_Complete_gene_list_misexpressed_in_mutant_and_WGCNA_modules_in_Fig4.csv")) #112817

# in whitney
test<-read.csv(file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Supplemental_Dataset10_Complete_gene_list_misexpressed_in_mutant_and_WGCNA_modules_in_Fig4.csv"),row.names=1) # does  work (100717)
## checking test object
test[ gsub("([[:digit:]]+)(\\.)([[:digit:]]+)","\\1",test$AGI) %in% c("AT3G22231","AT4G14365","AT1G33960","AT4G14400"),] # OK they are salmon

# myc2 (AT1G32640), myc3 (AT5G46760), myc4 (AT4G17880)
test[ gsub("([[:digit:]]+)(\\.)([[:digit:]]+)","\\1",test$AGI) %in% c("AT1G32640","AT5G46760","AT4G17880"),] # only MYC2 found in this list

# six MYBs that regulate indolic glucosinolates (GSLs): MYB34 AT5G60890, MYB51 AT1G18570, MYB122 AT1G74080, MYB28 AT5G61420, MYB29 AT5G07690, MYB76      AT5G07700
test[ gsub("([[:digit:]]+)(\\.)([[:digit:]]+)","\\1",test$AGI) %in% c("AT5G60890","AT1G18570","AT1G74080","AT5G61420","AT5G07690","AT5G07700"),]
# WRKY
grep("WRKY",test[test$WGCNAmodules=="salmon","Short_description"],value=TRUE)

dim(test) #  8150 28 (100717)
table(test$WGCNAmodules)
#             0         black          blue         brown     darkgreen      darkgrey    darkorange       darkred 
#          1125           247           826           338            54            46            35            63 
# darkturquoise         green   greenyellow          grey        grey60     lightcyan   lightyellow       magenta 
#            52           277           148           706            93           100            65           281 
#  midnightblue paleturquoise        purple           red     royalblue   saddlebrown        salmon       skyblue 
#           100            23           169           260            64            27           155            34 
#     steelblue           tan     turquoise         white        yellow 
#            23           142          2343            34           320 
# overlap table (revsised into GOseq.overlapTable; 052616; 092817; 092917)
for(i in 2:25) { #
  overlap.GOseq<-GOseq.overlapTable(test[,"WGCNAmodules"],test[,i],genes=test$AGI)
  overlap.GOseq.melt<-melt(overlap.GOseq$pTable,id=1) #
  overlap.GOseq.melt$count<-melt(overlap.GOseq$countTable,id=1)$value
  overlap.GOseq.melt<-overlap.GOseq.melt[!overlap.GOseq.melt$Var2=="0",]
  #overlap.melt.summary<-rbind(overlap.melt.summary,overlap.melt)
  save(overlap.GOseq.melt,file=file.path(homedir,"..","..","Nozue2016_SAStranscriptome_output","output",paste("overlap.GOseq.melt.WGCNA.mut.PLoSGenetics2015mutants.errrDE.",i,".Rdata",sep="")))
}
```
# combine overlapTables
```{r}
overlap.GOseq.melt.WGCNA.mut.files<-list.files(path=file.path("..","..","Nozue2016_SAStranscriptome_output","output"),pattern="overlap.GOseq.melt.WGCNA.mut.PLoSGenetics2015mutants.errrDE")
overlap.GOseq.melt.summary.list2<-sapply(file.path("..","..","Nozue2016_SAStranscriptome_output","output",overlap.GOseq.melt.WGCNA.mut.files),function(x) mget(load(x))) # mget will return the value of the object(or objects) in a list. see https://stackoverflow.com/questions/29398630/load-data-frames-into-list
names(overlap.GOseq.melt.summary.list2)
overlap.GOseq.melt.summary<-do.call("rbind",overlap.GOseq.melt.summary.list2) 
head(overlap.GOseq.melt.summary) # make sure those are SOM.customcat files
rownames(overlap.GOseq.melt.summary)<-1:nrow(overlap.GOseq.melt.summary)
# format overlap.GOseq.melt.summary
overlap.GOseq.melt.summary[,"value"]<--log10(overlap.GOseq.melt.summary[,"value"])
overlap.GOseq.melt.summary[overlap.GOseq.melt.summary$value>10,"value"]<-10
overlap.GOseq.melt.summary[overlap.GOseq.melt.summary$count==0,"count"]<-""
save(overlap.GOseq.melt.summary, file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","overlap.GOseq.melt.summary.WGCNA.mut.100717.Rdata") )
```
# reading and cleaning overlap table and graph (WGCNA vs mut; 100817 updated)
```{r}
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","overlap.GOseq.melt.summary.WGCNA.mut.100717.Rdata"))
overlap.GOseq.melt.summary
overlap.GOseq.melt.summary.ss<-overlap.GOseq.melt.summary[!overlap.GOseq.melt.summary$Var1=="0",]
overlap.GOseq.melt.summary.ss$Var1<-as.factor(overlap.GOseq.melt.summary.ss$Var1)
overlap.GOseq.melt.summary.ss$Var2<-factor(overlap.GOseq.melt.summary.ss$Var2,levels=paste(rep(conversion.table[-1,"genotype"],2),rep(c("sun","shade"),each=12),sep=".")) # sort Var2 (muts)
# check data
table(overlap.GOseq.melt.summary.ss$Var1)
# plot
q4<-ggplot(overlap.GOseq.melt.summary.ss,aes(y=factor(Var1),x=factor(Var2))) 
#q4 <- q4 + geom_tile(color="black",aes(fill=value))+scale_fill_gradient2(low=muted("green"), high=muted("magenta"),midpoint=2) # this color coding is not good (061716)
q4 <- q4 + geom_tile(color="black",aes(fill=value))+scale_fill_gradient2(low="white", mid="white",high="magenta",midpoint=2) # Is this better? (061716)
#q4 <- q4 +theme(legend.position="none") # this removed all legend
q4 <- q4 + theme(axis.text.x=element_text(angle=90,hjust=1),
                 axis.ticks = element_line(size=0,color="white"),
                 axis.line=element_line(size=1)) # adjust axis line width
q4 <- q4 + geom_text(aes(label=count),size=4)
q4 <- q4 + labs(x="genotype specific genes",y="WGCNA modules in Col",title="modules affected in SAS mutants",fill="-log10\n pvalue")
#q4 <- q4 + theme(legend.title=element_blank())
ggsave(q4,file="Fig4_Col.WGCNA.modules.vs.GenotypeSpecific.genes.table.s.GOseq100717.pdf",width=11,height=8,path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables"))
ggsave(q4,file="Fig4_Col.WGCNA.modules.vs.GenotypeSpecific.genes.table.s.GOseq100717.png",width=11,height=8,path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables"))
q4 # magenta indicates pvalue is smaller than 0.01.
# Modify the pdf file in Illustrator (remove ".sun" and ".shade", change color of genotypes, highlighting genotypes with red or brown colors. 091318) 
```
# Number of Col genes in each module are: 
```{r eval=FALSE} 
table(SAS.expression.vst.s.kazu.largeCV.modules$"1")  # 28 modules
```
# The next question is what are these modules?  GOseq analysis of each module
```{r eval=FALSE}
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","SAS.expression.vst.s.kazu.largeCV.modules092817.Rdata")) # 092717
SAS.expression.vst.s.kazu.largeCV.modules$AGI<-gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",SAS.expression.vst.s.kazu.largeCV.modules$AGI)
modules<-levels(as.factor(SAS.expression.vst.s.kazu.largeCV.modules$"1"))
for(x in modules) {
    temp.GOseq<-GOseq.ORA(SAS.expression.vst.s.kazu.largeCV.modules[SAS.expression.vst.s.kazu.largeCV.modules$"1"==x,"AGI"])
  if(temp.GOseq=="no enriched GO") {next} else {
  temp.genes<-genes.in.enriched.category(enrich.result=temp.GOseq,gene.list=SAS.expression.vst.s.kazu.largeCV.modules[SAS.expression.vst.s.kazu.largeCV.modules$"1"==x,"AGI"],category.table=Atgo.list)
  save(temp.GOseq,temp.genes,file=paste("../../Nozue2016_SAStranscriptome_output/output/WGCNA.Col.module.",x,".GOseq092717.Rdata",sep=""))
}
  } # 110916, 121316, 091817,092817
```
## write summary of WGCNA.Col.module.GOseq (modified 092717; no need to change 100717)
```{r eval=FALSE}
WGCNA.Col.module.GOseq.files<-list.files(path=file.path("..","..","Nozue2016_SAStranscriptome_output","output"),pattern="^WGCNA.Col.module.")
WGCNA.Col.module.GOseq.files<-grep(".GOseq092717.Rdata",WGCNA.Col.module.GOseq.files,value=TRUE)

load(file.path("..","..","Nozue2016_SAStranscriptome_output","output",WGCNA.Col.module.GOseq.files[1])) # using file.path 091817, 092717

temp.GOseq$modules<-gsub("(WGCNA.Col.module.)([[:print:]]+)(.GOseq092717.Rdata)","\\2",WGCNA.Col.module.GOseq.files[1])
temp<-temp.GOseq
 
for(i in WGCNA.Col.module.GOseq.files[-1]) {
  load(file.path("..","..","Nozue2016_SAStranscriptome_output","output",i)) # 091817
  temp.GOseq$modules<-gsub("(WGCNA.Col.module.)([[:print:]]+)(.GOseq092717.Rdata)","\\2",i)
  temp<-rbind(temp,temp.GOseq)
}
WGCNA.Col.module.GOseq.summary<-temp
write.csv(WGCNA.Col.module.GOseq.summary,file=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Supplemental_Dataset8_GO_WGCNA_modules092817.csv")) # Dataset 5. GO enrichment analysis of WGCNA gene co-expression modules (091716, salmon4<>orange 121316, rerun 091817; 092817)
```
## What genes are found in given module with given GO category? (no need to change 100717)
```{r eval=FALSE}
# ""GO:0009507" in "salmon" (chloroplast)
load("../../Nozue2016_SAStranscriptome_output/output/GOCC.WGCNA.Col.module.salmon.GOseq.Rdata") # I did not run GOCC
temp.genes.wdescrition<-merge(temp.genes,TAIR10_gene_descriptions,by="AGI")
dim(temp.genes)
temp.genes.wdescrition<-temp.genes.wdescrition[!duplicated(temp.genes.wdescrition$AGI),]
temp.genes.wdescrition[temp.genes.wdescrition$"GO:0009507"==1,c("AGI","Short_description")]
# GO:0009697, "salmon" (old version)
load("../Nozue2016_SAStranscriptome_output/output/WGCNA.Col.module.salmon.GOseq.Rdata")
temp.genes.wdescrition<-merge(temp.genes,TAIR10_gene_descriptions,by="AGI")
temp.genes.wdescrition<-temp.genes.wdescrition[!duplicated(temp.genes.wdescrition$AGI),]
#temp.genes.wdescrition[temp.genes.wdescrition$"GO:0009697"==1,c("AGI","Short_description")]
temp.genes.wdescrition[temp.genes.wdescrition$"GO:0009697"==1,c("AGI","Short_description","colE...col.F")]
```

<!-- ## GOseq analysis of each cluster (cellular component) (no need to change, 100717) -->
<!-- ```{r eval=FALSE} -->
<!-- #load("../Nozue2016_SAStranscriptome_output/output/SAS.expression.vst.s.kazu.largeCV.module121316.Rdata") -->
<!-- load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","SAS.expression.vst.s.kazu.largeCV.modules092817.Rdata")) # 092717 -->

<!-- #library(WGCNA) -->
<!-- SAS.expression.vst.s.kazu.largeCV.modules$AGI<-gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",SAS.expression.vst.s.kazu.largeCV.modules$AGI) -->
<!-- modules<-levels(as.factor(SAS.expression.vst.s.kazu.largeCV.modules$"1")) -->
<!-- #setwd(homedir) -->
<!-- #setwd("../Nozue2016_SAStranscriptome_output/output/") -->
<!-- # needs to fix for new GOseq.CC.ORA (did I fixed it?; 110816) -->
<!-- for(x in modules) { #  -->
<!--   temp.GOseq<-GOseq.CC.ORA(SAS.expression.vst.s.kazu.largeCV.modules[SAS.expression.vst.s.kazu.largeCV.modules$"1"==x,"AGI"]) -->
<!--   if(temp.GOseq=="no enriched GO") {next} else { -->
<!--   temp.GOseq<-GOseq.CC.ORA(SAS.expression.vst.s.kazu.largeCV.modules[SAS.expression.vst.s.kazu.largeCV.modules$"1"==x,"AGI"]) -->
<!--   temp.genes<-genes.in.enriched.category(enrich.result=temp.GOseq,gene.list=SAS.expression.vst.s.kazu.largeCV.modules[SAS.expression.vst.s.kazu.largeCV.modules$"1"==x,"AGI"],category.table=Atgo.list) -->
<!--   save(temp.GOseq,temp.genes,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output",paste("GOCC.WGCNA.Col.module.",x,".GOseq.Rdata",sep=""))) -->
<!-- } -->
<!-- }  -->
<!-- ``` -->
# overlap analysis with custome categories (WGCNA modules): Fig6  (no need to change, 100717). using SAup_combined & SAdown_combined (110917). rename SAup_combined & SAdown_combined into SAup and sadown (020818)
```{r eval=FALSE}
# custom categories
load(file.path("..","..","Nozue2016_SAStranscriptome_data","input","hormone.responsiveness6.DF.s.Rdata")) # using SAup_combined & SAdown_combined
hormone.responsiveness6.DF.s
hormone.responsiveness6.DF.s$AGI
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","SAS.expression.vst.s.kazu.largeCV.modules092817.Rdata") ) #
SAS.expression.vst.s.kazu.largeCV.modules$AGI<-gsub("([[:print:]]+)(\\.[[:digit:]]+)","\\1",SAS.expression.vst.s.kazu.largeCV.modules$AGI)
test<-merge(hormone.responsiveness6.DF.s,SAS.expression.vst.s.kazu.largeCV.modules[,c("AGI","1")],by="AGI",all=TRUE)
# merge with WGCNA modules
names(test)[names(test)=="1"]<-"WGCNAmodules"
i <- sapply(test, is.factor)
test[i]<-lapply(test[i], as.character)
class(test[,2])
  
hormone.responsiveness.WGCNAmodules<-test
hormone.responsiveness.WGCNAmodules[hormone.responsiveness.WGCNAmodules==""]<-"0"
hormone.responsiveness.WGCNAmodules[is.na(hormone.responsiveness.WGCNAmodules)]<-"0"
head(hormone.responsiveness.WGCNAmodules)
# Plant Cell version
# remove GRF1_3target (052616)
hormone.responsiveness.WGCNAmodules<-hormone.responsiveness.WGCNAmodules[,!names(hormone.responsiveness.WGCNAmodules)=="GRF1_3target"]
# write in csv file (Supplemental Dataset6)
write.csv(hormone.responsiveness.WGCNAmodules,file=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Supplemental_Dataset9_Custom_category_enrichment_analysis_of_WGCNA_gene_co-expression_modules020818.csv"))


# salmon genes and SA, npr1 responsiveness? (PCC1, XBAT34, AIG1, ACD6,SARD1,cbp60g)
hormone.responsiveness.WGCNAmodules[hormone.responsiveness.WGCNAmodules$WGCNAmodules %in% "salmon" & hormone.responsiveness.WGCNAmodules$AGI %in% c("AT3G22231","AT4G14365","AT1G33960","AT4G14400","AT5G26920","AT1G73805","AT5G26920"), c("AGI","IAAup","IAAdown","SARminus","SARplus","NPR1down","WRKY33down","WRKY33up","SAup_24h","SAdown_24h","NPR1up_uponSA","NPR1down_uponSA","WGCNAmodules")]

# salmon genes and SA, npr1 responsiveness? (PCC1, XBAT34, AIG1, ACD6,SARD1,cbp60g)
hormone.responsiveness.WGCNAmodules[hormone.responsiveness.WGCNAmodules$WGCNAmodules %in% "salmon" & hormone.responsiveness.WGCNAmodules$AGI %in% c("AT3G22231","AT4G14365","AT1G33960","AT4G14400","AT5G26920","AT1G73805","AT5G26920"), c("AGI","IAAup","IAAdown","SAup_combined","SAdown_combined","WGCNAmodules")]

# OK. They belong to "salmon" and not IAA responsive.
# IAA responsive & green
hormone.responsiveness.WGCNAmodules[hormone.responsiveness.WGCNAmodules$IAAup %in% "IAAup" & hormone.responsiveness.WGCNAmodules$WGCNAmodules=="green", c("AGI","IAAup","IAAdown","SAup_combined","SAdown_combined","WGCNAmodules")]

```
# overlapTable for WGCNA module vs custome categories (100417)  (no need to change, 100717)(use updated csv file, 020818)
```{r}
hormone.responsiveness.WGCNAmodules<-read.csv(file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Supplemental_Dataset5_Custom_category_enrichment_analysis_of_WGCNA_gene_co-expression_modules020818.csv"),row.names=1)
hormone.responsiveness.WGCNAmodules[hormone.responsiveness.WGCNAmodules$WGCNAmodules %in% "salmon" & hormone.responsiveness.WGCNAmodules$AGI %in% c("AT3G22231","AT4G14365","AT1G33960","AT4G14400"), c("AGI","IAAup","IAAdown","SAup","SAdown","WGCNAmodules")]
# OK, Salmon module and SAup
test<-hormone.responsiveness.WGCNAmodules
for(i in 19:20) { # 2:32 for 092817.csv version. 19 and 20 are new (SAup_combined and SAdown_combined)
  overlap.GOseq<-GOseq.overlapTable(test[,"WGCNAmodules"],test[,i],genes=test$AGI) # all genes
  overlap.GOseq.melt<-melt(overlap.GOseq$pTable,id=1) #
  overlap.GOseq.melt$count<-melt(overlap.GOseq$countTable,id=1)$value
  overlap.GOseq.melt<-overlap.GOseq.melt[!overlap.GOseq.melt$Var2=="0",]
  #overlap.melt.summary<-rbind(overlap.melt.summary,overlap.melt)
  save(overlap.GOseq.melt,file=file.path(homedir,"..","..","Nozue2016_SAStranscriptome_output","output",paste("overlap.GOseq.melt.WGCNA.customcat.110917.",i,".Rdata",sep="")))
} # renamed
```
# combine files (WGCMA vs customcat): updated 110917 for SAup_combined and SAdown_combined
```{r}
# read files
overlap.GOseq.melt.WGCNA.customcat.files<-list.files(path=file.path("..","..","Nozue2016_SAStranscriptome_output","output"),pattern="overlap.GOseq.melt.WGCNA.customcat")
select.files<-gsub("(overlap.GOseq.melt.WGCNA.customcat)(\\.)([[:print:]]+)(\\.Rdata)","\\3",overlap.GOseq.melt.WGCNA.customcat.files)
overlap.GOseq.melt.WGCNA.customcat.files<-overlap.GOseq.melt.WGCNA.customcat.files[select.files %in% c("2","3","4","5","6","7","8","9","10","11","12","13","14","15","26","25","110917.19","110917.20")]
# select 2:18 from "overlap.GOseq.melt.WGCNA.customcat", 19:20 for "overlap.GOseq.melt.WGCNA.customcat".110917, 26, 25 for WRKY33up and down

# remove unnecessary files
overlap.GOseq.melt.WGCNA.customcat.files<-overlap.GOseq.melt.WGCNA.customcat.files[!overlap.GOseq.melt.WGCNA.customcat.files %in% c("overlap.GOseq.melt.WGCNA.customcat.27.Rdata","overlap.GOseq.melt.WGCNA.customcat.28.Rdata")]

overlap.GOseq.melt.summary.list2<-sapply(file.path("..","..","Nozue2016_SAStranscriptome_output","output",overlap.GOseq.melt.WGCNA.customcat.files),function(x) mget(load(x))) # mget will return the value of the object(or objects) in a list. see https://stackoverflow.com/questions/29398630/load-data-frames-into-list
names(overlap.GOseq.melt.summary.list2)
overlap.GOseq.melt.summary<-do.call("rbind",overlap.GOseq.melt.summary.list2) 
head(overlap.GOseq.melt.summary) # make sure those are SOM.customcat files
rownames(overlap.GOseq.melt.summary)<-1:nrow(overlap.GOseq.melt.summary)
# format overlap.GOseq.melt.summary
overlap.GOseq.melt.summary[,"value"]<--log10(overlap.GOseq.melt.summary[,"value"])
overlap.GOseq.melt.summary[overlap.GOseq.melt.summary$value>10,"value"]<-10
overlap.GOseq.melt.summary[overlap.GOseq.melt.summary$count==0,"count"]<-""
save(overlap.GOseq.melt.summary, file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","overlap.GOseq.melt.summary.WGCNA.customcat.110917.Rdata") )
```
## cleaning and plotting  (no need to change, 100717); (changed 110917 due to "SAup_combined" and "SAdown_combined")
```{r}
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","overlap.GOseq.melt.summary.WGCNA.customcat.110917.Rdata"))
custom.category.overlap.melt.summary.s<-overlap.GOseq.melt.summary
table(custom.category.overlap.melt.summary.s$Var2)
custom.category.overlap.melt.summary.s<-custom.category.overlap.melt.summary.s[!custom.category.overlap.melt.summary.s$Var1=="0",]
custom.category.overlap.melt.summary.s<-custom.category.overlap.melt.summary.s[!custom.category.overlap.melt.summary.s$Var2=="0",]
custom.category.overlap.melt.summary.s<-custom.category.overlap.melt.summary.s[!is.na(custom.category.overlap.melt.summary.s$Var2),]
custom.category.overlap.melt.summary.s[is.na(custom.category.overlap.melt.summary.s$Var2),]
table(custom.category.overlap.melt.summary.s$Var2)
custom.category.overlap.melt.summary.s$Var2<-factor(as.character(custom.category.overlap.melt.summary.s$Var2),levels=c("IAAup","IAAdown","BLup","BLdown","MJup","MJdown","GAup","GAdown","ABAup","ABAdown","ACCup","ACCdown","CKup","CKdown","SAup_combined","SAdown_combined","MYC234up","MYC234down","WRKY33up","WRKY33down")) # (110917)

table(as.character(custom.category.overlap.melt.summary.s$Var2))
save(custom.category.overlap.melt.summary.s,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","custom.category.v4.overlap.overlap.melt.summary.s.GOseq110917.Rdata"))

```
# graph  (no need to change, 100717); chagned 110917 due to "SAup_combined" and "SAdown_combined"
```{r}
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","custom.category.v4.overlap.overlap.melt.summary.s.GOseq110917.Rdata")) # 11091717
# rename (020818)
custom.category.overlap.melt.summary.s$Var2<-gsub("_combined","",custom.category.overlap.melt.summary.s$Var2)
# 
f6<-ggplot(custom.category.overlap.melt.summary.s,aes(y=factor(Var1),x=factor(Var2))) 
f6 <- f6 + geom_tile(color="black",aes(fill=value))+scale_fill_gradient2(low="white", mid="white",high="magenta",midpoint=2)
f6 <- f6 + theme(axis.text.x=element_text(angle=90,hjust=1),
                 panel.grid.major = element_blank(),
                 panel.grid.minor = element_blank(),
                 panel.background = element_blank(),
                 axis.ticks = element_line(size=0,color="white"),
                 axis.line=element_line(size=1)) # adjust axis line width
f6 <- f6 + geom_text(aes(label=count),size=4)
f6 <- f6 + labs(x="custom category",y="WGCNA modules in Col",title="custom category enrichment",fill="-log10\n pvalue")
#ggsave(f6,file="Fig3_Col.WGCNA.modules.vs.my.category.table.s.110917.pdf",width=11,height=8,path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables"))
#ggsave(f6,file="Fig3_Col.WGCNA.modules.vs.my.category.table.s.110917.png",width=11,height=8,path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables"))
ggsave(f6,file="Fig3_Col.WGCNA.modules.vs.my.category.table.s.020818.png",width=11,height=8,path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables"))

f6 # magenta indicates pvalue is smaller than 0.01.
```
# Checking expression pattern of gene of interests (absolute and shade responsiveness) (060515)
```{r}
rm(list=ls())
# absolute value
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","summary.vst.Rdata"))
summary.vst.kazu<-summary.vst[,gsub("([[:digit:]]+)(H|L)(1|4|16|25|49)","\\1",names(summary.vst)) %in% 1:19] 
# plot
p<-expression.pattern.graph3(summary.vst.kazu,"AT2G46970") # PIL1
p
p.Col<-expression.pattern.graph3(summary.vst.kazu[,c("1H1","1H25","1H4","1H49","1L1","1L25","1L4","1L49")],"AT2G46970") #PIL1, Col

# 100717
SAS.expression.vst<-read.csv(file.path("..","..","Nozue2016_SAStranscriptome_data","input","kazu.SAS.expression.VST.092517.csv"),row.names=1) # new transformed data by DEseq2 (092617) # not found? (092817)
SAS.expression.vst.Col<-SAS.expression.vst[,samples.PLoSGenetics2015mutants.errrDE[samples.PLoSGenetics2015mutants.errrDE$genotype=="1","file"]]
# remove splicing variant info in rownames
rownames(SAS.expression.vst.Col)<-gsub("([[:print:]]+)(.)([[:digit:]]+)","\\1",rownames(SAS.expression.vst.Col))
p.Col<-expression.pattern.graph3b(data.mean=summary.vst.kazu[,c("1H1","1H25","1H4","1H49","1L1","1L25","1L4","1L49")],data.expression=SAS.expression.vst.Col,target.genes="AT2G46970") #PIL1, Col

p.Col<-expression.pattern.graph3b(data.mean=summary.vst.kazu[,c("1H1","1H25","1H4","1H49","1L1","1L25","1L4","1L49")],data.expression=SAS.expression.vst.Col,target.genes=c("AT2G46970","AT4G16780","AT4G32280")) #Col, PIL1 ("AT2G46970"), ATHB2("AT4G16780"), IAA29 ("AT4G32280") # barplot and jitter overlapped. looks good
```
# phyABDE seedlings DE analysis (microarray) for Supplemental Figure S1 (public available phytochrome mutnats (exposed to long light condition)(100517)
```{r}
# high-order phytochrome muntat seedlings (microarray)
# download some cel file from E-GEOD-31587 - Expression data of Arabidopsis thaliana (Ler accession) phytochrome phyABCDE quintuple mutant and phyABDE quadruple mutant in response to red light, and their comparison to WT expression data (http://www.ebi.ac.uk/arrayexpress/experiments/E-GEOD-31587/samples/?query=phytochrome&sortby=organism&sortorder=ascending)
# under construction #
#setwd("/Volumes/Data8/E-GEOD-31587_RAW")
library(affy)
#Data <- ReadAffy(widget=T) # error
CEL.filename<-list.files(pattern=".CEL$",path=file.path("/Volumes","data_work","Data8","E-GEOD-31587_RAW"))
CEL.filename
Data <- ReadAffy(filenames=file.path("/Volumes","data_work","Data8","E-GEOD-31587_RAW",CEL.filename))
Data
eset <- rma(Data) # normalize chips 
affyData <- read.csv(file.path("/Volumes","data_work","Data6","bioconductor_R","affy_ATH1_array_elements-2010-12-20.csv")) #read TAIR annotation data from file (TAIR10)
affyData2<-affyData[,c("array_element_name","is_control","locus","description","chromosome","start","stop")]
names(affyData2) <- c("element","is.control","locus","description","chromosome","start","stop")#get rid of underlines
head(affyData2)
genes <- data.frame(element=featureNames(eset)) #get affy gene names from dataset
affyMatch <- merge(affyData2,genes,by.x="element",by.y=1)# reorder the annotation data to match chip data
exp.values <- data.frame(exprs(eset)) #put normalized intensities in a data frame
head(exp.values)
exp.values <- cbind(affyMatch,exp.values) #add annotation to same frame as expression values
head(exp.values)

write.csv(exp.values,file=file.path("/Volumes","data_work","Data8","E-GEOD-31587_RAW","exp.values.csv"))
######## RP analysis
library(RankProd)
exp.values<-read.csv(file.path("/Volumes","data_work","Data8","E-GEOD-31587_RAW","exp.values.csv"),row.names=1)
#exp.values<-exp.values[,-1]
names(exp.values)
names(exp.values)[c(8:12)] # should be six "CEL" file names
RP.E_GEOD_31587.phyABDE <- RP(exp.values[,c(8:12)], rep(c(1,0),c(3,2)), num.perm=100) #  0 (class 1) = Col, 1 (class 2)= phyABDE
topgene.RP.E_GEOD_31587.phyABDE<-topGene(RP.E_GEOD_31587.phyABDE,cutoff=0.01,gene.names=exp.values$element,logged=TRUE,logbase=2)
#  repressed genes in phyABDE
topgene.RP.E_GEOD_31587.phyABDE.T2<-as.data.frame(topgene.RP.E_GEOD_31587.phyABDE$Table2)
topgene.RP.E_GEOD_31587.phyABDE.T2$element<-rownames(topgene.RP.E_GEOD_31587.phyABDE.T2)
topgene.RP.E_GEOD_31587.phyABDE.T2.m<-merge(topgene.RP.E_GEOD_31587.phyABDE.T2, affyData2,by="element")
head(topgene.RP.E_GEOD_31587.phyABDE.T2.m)
topgene.RP.E_GEOD_31587.phyABDE.T2.m.exp.value<-merge(topgene.RP.E_GEOD_31587.phyABDE.T2.m, exp.values[,c(1,8:12)],by="element")
topgene.RP.E_GEOD_31587.phyABDE.T2.m.exp.value<=topgene.RP.E_GEOD_31587.phyABDE.T2.m.exp.value[order(topgene.RP.E_GEOD_31587.phyABDE.T2.m.exp.value$pfp),]

#  induced genes in phyABDE
topgene.RP.E_GEOD_31587.phyABDE.T1<-as.data.frame(topgene.RP.E_GEOD_31587.phyABDE$Table1)
topgene.RP.E_GEOD_31587.phyABDE.T1$element<-rownames(topgene.RP.E_GEOD_31587.phyABDE.T1)
topgene.RP.E_GEOD_31587.phyABDE.T1.m<-merge(topgene.RP.E_GEOD_31587.phyABDE.T1, affyData2,by="element")
head(topgene.RP.E_GEOD_31587.phyABDE.T1.m)
topgene.RP.E_GEOD_31587.phyABDE.T1.m.exp.value<-merge(topgene.RP.E_GEOD_31587.phyABDE.T1.m, exp.values[,c(1,8:12)],by="element")
topgene.RP.E_GEOD_31587.phyABDE.T1.m.exp.value<-topgene.RP.E_GEOD_31587.phyABDE.T1.m.exp.value[order(topgene.RP.E_GEOD_31587.phyABDE.T1.m.exp.value$pfp),]

# summary
topgene.RP.RP.E_GEOD_31587.phyABDE.summary<-rbind(topgene.RP.E_GEOD_31587.phyABDE.T1.m.exp.value,topgene.RP.E_GEOD_31587.phyABDE.T2.m.exp.value)
write.csv(topgene.RP.RP.E_GEOD_31587.phyABDE.summary,file=file.path("/Volumes","data_work","Data8","E-GEOD-31587_RAW","topgene.RP.RP.E_GEOD_31587.phyABDE.summary.csv"))
# GOseq
topgene.RP.RP.E_GEOD_31587.phyABDE.summary.locus<-topgene.RP.RP.E_GEOD_31587.phyABDE.summary[nchar(topgene.RP.RP.E_GEOD_31587.phyABDE.summary$locus)==9,"locus"]
phyABDE.seedlings.GOseq<-GOseq.ORA(topgene.RP.RP.E_GEOD_31587.phyABDE.summary.locus)
# write
write.csv(phyABDE.seedlings.GOseq,file=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Supplemental_Table1_phyABDE.GOseq.summary.csv_phyABDE.GOseq.summary.csv"))
phyABDE.seedlings.GOseq.genes<-genes.in.enriched.category(enrich.result=temp.GOseq,gene.list=topgene.RP.RP.E_GEOD_31587.phyABDE.summary.locus,category.table=Atgo.list) #  no auxin?! as I expected.
# Is auxin responsive genes enriched?
topgene.RP.RP.E_GEOD_31587.phyABDE.summary<-read.csv(file.path("/Volumes","data_work","Data8","E-GEOD-31587_RAW","topgene.RP.RP.E_GEOD_31587.phyABDE.summary.csv"),row.names = 1)
topgene.RP.RP.E_GEOD_31587.phyABDE.summary
# load custom categories
load(file.path("..","..","Nozue2016_SAStranscriptome_data","input","hormone.responsiveness6.DF.s.Rdata")) # using SAup_combined & SAdown_combined
hormone.responsiveness6.DF.s
hormone.responsiveness6.DF.s$AGI
# ORA (use overlapTable because of microarray data)
names(topgene.RP.RP.E_GEOD_31587.phyABDE.summary)<-gsub("locus","AGI",names(topgene.RP.RP.E_GEOD_31587.phyABDE.summary))
temp1<-merge(hormone.responsiveness6.DF.s,topgene.RP.RP.E_GEOD_31587.phyABDE.summary[topgene.RP.RP.E_GEOD_31587.phyABDE.summary$FC..class1.class2.<1,c("AGI","FC..class1.class2.")],by="AGI",all=TRUE) # upregulated in phyABDE
temp2<-merge(temp1,topgene.RP.RP.E_GEOD_31587.phyABDE.summary[topgene.RP.RP.E_GEOD_31587.phyABDE.summary$FC..class1.class2.>1,c("AGI","FC..class1.class2.")],by="AGI",all=TRUE)
  names(temp2)<-gsub("FC..class1.class2..x","phyABDE_up",names(temp2))
    names(temp2)<-gsub("FC..class1.class2..y","phyABDE_down",names(temp2))
temp2[!is.na(temp2$"phyABDE_up"),"phyABDE_up"]<-"phyABDE_up"
temp2[!is.na(temp2$"phyABDE_down"),"phyABDE_down"]<-"phyABDE_down"
temp2[is.na(temp2)]<-""
temp2$phyABDE<-paste(temp2$phyABDE_up,temp2$phyABDE_down)
#temp2<-temp2[temp2$AGI %in% names(bias2),]
test<-temp2
#hormone.responsiveness.E_GEOD_31587.phyABDE
for(i in 2:20) { # 2:32 for 092817.csv version. 19 and 20 are new (SAup_combined and SAdown_combined)
  overlap<-overlapTable(test[,"phyABDE"],test[,i]) # all genes
  overlap.melt<-melt(overlap$pTable,id=1) #
  overlap.melt$count<-melt(overlap$countTable,id=1)$value
  overlap.melt<-overlap.melt[!overlap.melt$Var2=="0",]
  #overlap.melt.summary<-rbind(overlap.melt.summary,overlap.melt)
  save(overlap.melt,file=file.path(homedir,"..","..","Nozue2016_SAStranscriptome_output","output",paste("overlap.melt.phyABDE.customcat.120717.",i,".Rdata",sep="")))
} # renamed
```
# overlapTable with Sessa 1h (as positive control) & our phyB (as another positive control for adult phase (033018))
```{r}
Sessa.1h<-read.csv("../../Nozue2016_SAStranscriptome_data/input/sessa1hour.csv")
Sessa.exp.value<-read.csv("../../Nozue2016_SAStranscriptome_data/input/exp.values.Sessa.csv")
names(Sessa.1h)<-gsub("ID","AGI",names(Sessa.1h))
names(Sessa.exp.value)<-gsub("locus","AGI",names(Sessa.exp.value))
Sessa.1h.exp.value<-merge(Sessa.1h,Sessa.exp.value[,c("AGI",grep("At_seedling",names(Sessa.exp.value),value=T))],by ="AGI")
Sessa.1h.exp.value$Sessa.1h.up<-""
Sessa.1h.exp.value[Sessa.1h.exp.value$logFC>1,] # up in shade 1h
Sessa.1h.exp.value[Sessa.1h.exp.value$logFC>1,"Sessa.1h.up"]<-"Sessa.1h.up"
Sessa.1h.exp.value$Sessa.1h.down<-""
Sessa.1h.exp.value[Sessa.1h.exp.value$logFC<1,"Sessa.1h.down"]<-"Sessa.1h.down"
Sessa.1h.exp.value$Sessa.1h<-paste(Sessa.1h.exp.value$Sessa.1h.up,Sessa.1h.exp.value$Sessa.1h.down)
Sessa.1h.exp.value$Sessa.1h<-gsub(" ","",Sessa.1h.exp.value$Sessa.1h)
temp3<-merge(temp2,Sessa.1h.exp.value[,c("AGI","Sessa.1h")],by="AGI",all=TRUE)
table(temp3$Sessa.1h)
temp3[is.na(temp3)]<-""
test<-temp3
#hormone.responsiveness.E_GEOD_31587.phyABDE
for(i in 2:20) { # 2:32 for 092817.csv version. 19 and 20 are new (SAup_combined and SAdown_combined)
  overlap<-overlapTable(test[,"Sessa.1h"],test[,i]) # all genes
  overlap.melt<-melt(overlap$pTable,id=1) #
  overlap.melt$count<-melt(overlap$countTable,id=1)$value
  overlap.melt<-overlap.melt[!overlap.melt$Var2=="0",]
  #overlap.melt.summary<-rbind(overlap.melt.summary,overlap.melt)
  save(overlap.melt,file=file.path(homedir,"..","..","Nozue2016_SAStranscriptome_output","output",paste("overlap.melt.Sessa1h.customcat.120817.",i,".Rdata",sep="")))
} # renamed
# add our phyB (sun); 033018
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h","genotype_specific5","DE.genotype.batch.sun.gt.wo16h6.Rdata"))# sun, phyB vs Col any timepoints
# separate upregulated and downregulated genes
# upregulated by PHYB (lower in phyB mutant) under 1h sun treatment
PHYB.adult.up<-rownames(DE.batch3.gt[DE.batch3.gt[,1]< -1,]) # two fold change at 1h low R/FR
## check expression pattern
p.up<-expression.pattern.graph3(summary.vst,gsub("([[:print:]]+)(.[[:digit:]])","\\1",PHYB.adult.up[1:10])) 
p.up # good
# downregulated by PHYB (up regulated in phyB mutant) under 1h sun treatment
PHYB.adult.down<-rownames(DE.batch3.gt[DE.batch3.gt[,1]>1,]) # two fold change
# adding DEG in phyB.adult.sun.1h column
temp3$phyB.adult.sun.1h<-""
temp3$phyB.adult.sun.1h[temp3$AGI %in% gsub("([[:print:]]+)(\\.)([[:digit:]]+)","\\1",PHYB.adult.up)]<-"phyB.adult.up"
# 
temp3$phyB.adult.sun.1h[temp3$AGI %in% gsub("([[:print:]]+)(\\.)([[:digit:]]+)","\\1",PHYB.adult.down)]<-"phyB.adult.down"
table(temp3$phyB.adult)
test<-temp3
#hormone.responsiveness.E_GEOD_31587.phyABDE
for(i in 2:20) { # 2:32 for 092817.csv version. 19 and 20 are new (SAup_combined and SAdown_combined)
  overlap<-overlapTable(test[,"phyB.adult.sun.1h"],test[,i]) # all genes
  overlap.melt<-melt(overlap$pTable,id=1) #
  overlap.melt$count<-melt(overlap$countTable,id=1)$value
  overlap.melt<-overlap.melt[!overlap.melt$Var2=="0",]
  #overlap.melt.summary<-rbind(overlap.melt.summary,overlap.melt)
  save(overlap.melt,file=file.path(homedir,"..","..","Nozue2016_SAStranscriptome_output","output",paste("overlap.melt.phyB.adult.sun.1h.customcat.120817.",i,".Rdata",sep="")))
} # renamed

```
# combine overlapTables (Sessa 1h, phyABDE, phyB.adult)
```{r}
# overlap.melt.phyABDE.custom.files<-list.files(path=file.path("..","..","Nozue2016_SAStranscriptome_output","output"),pattern="overlap.melt.phyABDE")
# overlap.melt.Sessa.custom.files<-list.files(path=file.path("..","..","Nozue2016_SAStranscriptome_output","output"),pattern="overlap.melt.Sessa")
# overlap.melt.phyB.adult.sun.1h.custom.files<-list.files(path=file.path("..","..","Nozue2016_SAStranscriptome_output","output"),pattern="overlap.melt.phyB.adult")
# # combine
# overlap.melt.summary.list2<-sapply(file.path("..","..","Nozue2016_SAStranscriptome_output","output",c(overlap.melt.phyABDE.custom.files,overlap.melt.Sessa.custom.files,overlap.melt.phyB.adult.sun.1h.custom.files)),function(x) mget(load(x))) # mget will return the value of the object(or objects) in a list. see https://stackoverflow.com/questions/29398630/load-data-frames-into-list
# names(overlap.melt.summary.list2)
# overlap.melt.summary<-do.call("rbind",overlap.melt.summary.list2) 
# head(overlap.melt.summary) # make sure those are file names
# rownames(overlap.melt.summary)<-1:nrow(overlap.melt.summary)
# # format overlap.GOseq.melt.summary
# overlap.melt.summary[,"value"]<--log10(overlap.melt.summary[,"value"])
# overlap.melt.summary[overlap.melt.summary$value>10,"value"]<-10
# overlap.melt.summary[overlap.melt.summary$count==0,"count"]<-""
# table(overlap.melt.summary$Var2)
# overlap.melt.summary<-overlap.melt.summary[!overlap.melt.summary$Var2=="",]
# 
# overlap.melt.summary$Var2<-factor(as.character(overlap.melt.summary$Var2),levels=c("IAAup","IAAdown","BLup","BLdown","MJup","MJdown","GAup","GAdown","ABAup","ABAdown","ACCup","ACCdown","CKup","CKdown","SAup","SAdown","MYC234up","MYC234down","PIFtarget")) # (110917)
# 
# save(overlap.melt.summary, file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","overlap.melt.summary.phyABDE.phyBadult.033018.Rdata") )
```
# reading and cleaning overlap table and graph (phyABDE, 120717, 020818,033118)
```{r}
# load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","overlap.melt.summary.phyABDE.phyBadult.033018.Rdata"))
# 
# overlap.melt.summary
# # rename SAup_combined and SAdown_combined (020818)
# overlap.melt.summary$Var2<-gsub("_combined","",overlap.melt.summary$Var2)
# # remove extra space in Var1
# overlap.melt.summary$Var1<-gsub(" ","",overlap.melt.summary$Var1)
# # remove no data in Var1
# overlap.melt.summary.ss<-overlap.melt.summary[!overlap.melt.summary$Var1==" ",]
# overlap.melt.summary.ss<-overlap.melt.summary.ss[!overlap.melt.summary.ss$Var1=="",]
# # count number of Var1
# names(table(overlap.melt.summary.ss$Var1))
# # cleanup Var2
# overlap.melt.summary.ss<-overlap.melt.summary.ss[!overlap.melt.summary.ss$Var2=="",]
# overlap.melt.summary.ss$Var1<-droplevels(overlap.melt.summary.ss$Var1)
# overlap.melt.summary.ss$Var2<-droplevels(overlap.melt.summary.ss$Var2)
# # check data
# table(overlap.melt.summary.ss$Var1)
# # plot
# q4<-ggplot(overlap.melt.summary.ss,aes(y=factor(overlap.melt.summary.ss$Var1,levels=rev(c("phyABDE_up","phyABDE_down","Sessa.1h.up","Sessa.1h.down","phyB.adult.up","phyB.adult.down"))),x=factor(Var2))) # sort row
# q4 <- q4 + geom_tile(color="black",aes(fill=value))+scale_fill_gradient2(low="white", mid="white",high="magenta",midpoint=2) # Is this better? (061716)
# #q4 <- q4 +theme(legend.position="none") # this removed all legend
# q4 <- q4 + theme(axis.text.x=element_text(angle=90,hjust=1),
#                  axis.ticks = element_line(size=0,color="white"),
#                  axis.line=element_line(size=1)) # adjust axis line width
# q4 <- q4 + geom_text(aes(label=count),size=4)
# q4 <- q4 + labs(x="custom category",y="",title="",fill="-log10\n pvalue")
# ggsave(q4,file="Supplemental_Fig1.phyABDE.Sessa1h.phyBadlut.vs.my.category.040618.pdf",width=11,height=5,path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables"))
# ggsave(q4,file="Supplemental_Fig1.phyABDE.Sessa1hvs.phyBadlut.vs.my.category.040618.png",width=11,height=5,path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables"))
# q4 # magenta indicates pvalue is smaller than 0.01.
```
# Separate heatmap (Sessa1h and phyABDE, and only phyB adult)
```{r}
# Sessa1h and phyABDE (Supplemental Fig 1A)
overlap.melt.phyABDE.custom.files<-list.files(path=file.path("..","..","Nozue2016_SAStranscriptome_output","output"),pattern="overlap.melt.phyABDE")
overlap.melt.Sessa.custom.files<-list.files(path=file.path("..","..","Nozue2016_SAStranscriptome_output","output"),pattern="overlap.melt.Sessa")
# combine
overlap.melt.summary.list2<-sapply(file.path("..","..","Nozue2016_SAStranscriptome_output","output",c(overlap.melt.phyABDE.custom.files,overlap.melt.Sessa.custom.files)),function(x) mget(load(x))) # mget will return the value of the object(or objects) in a list. see https://stackoverflow.com/questions/29398630/load-data-frames-into-list
names(overlap.melt.summary.list2)
overlap.melt.summary<-do.call("rbind",overlap.melt.summary.list2) 
head(overlap.melt.summary) # make sure those are file names
rownames(overlap.melt.summary)<-1:nrow(overlap.melt.summary)
# format overlap.GOseq.melt.summary
overlap.melt.summary[,"value"]<--log10(overlap.melt.summary[,"value"])
overlap.melt.summary[overlap.melt.summary$value>10,"value"]<-10
overlap.melt.summary[overlap.melt.summary$count==0,"count"]<-""
table(overlap.melt.summary$Var2)
overlap.melt.summary<-overlap.melt.summary[!overlap.melt.summary$Var2=="",]
overlap.melt.summary$Var2<-factor(as.character(overlap.melt.summary$Var2),levels=c("IAAup","IAAdown","BLup","BLdown","MJup","MJdown","GAup","GAdown","ABAup","ABAdown","ACCup","ACCdown","CKup","CKdown","SAup","SAdown","MYC234up","MYC234down","PIFtarget")) # (110917)
save(overlap.melt.summary, file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","overlap.melt.summary.phyABDE.120717.Rdata") )
# load overlap.melt.summary
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","overlap.melt.summary.phyABDE.120717.Rdata"))
overlap.melt.summary
# rename SAup_combined and SAdown_combined (020818)
overlap.melt.summary$Var2<-gsub("_combined","",overlap.melt.summary$Var2)
# remove extra space in Var1
overlap.melt.summary$Var1<-gsub(" ","",overlap.melt.summary$Var1)
# remove no data in Var1
overlap.melt.summary.ss<-overlap.melt.summary[!overlap.melt.summary$Var1==" ",]
overlap.melt.summary.ss<-overlap.melt.summary.ss[!overlap.melt.summary.ss$Var1=="",]
# count number of Var1
names(table(overlap.melt.summary.ss$Var1))
# cleanup Var2
overlap.melt.summary.ss<-overlap.melt.summary.ss[!overlap.melt.summary.ss$Var2=="",]
overlap.melt.summary.ss<-overlap.melt.summary.ss[!is.na(overlap.melt.summary.ss$Var2),]
# check data
table(overlap.melt.summary.ss$Var1)
table(overlap.melt.summary.ss$Var2)
# plot
SupF1A<-ggplot(overlap.melt.summary.ss,aes(y=factor(overlap.melt.summary.ss$Var1,levels=rev(c("phyABDE_up","phyABDE_down","Sessa.1h.up","Sessa.1h.down"))),x=factor(Var2))) # sort row
SupF1A <- SupF1A + geom_tile(color="black",aes(fill=value))+scale_fill_gradient2(low="white", mid="white",high="magenta",midpoint=2) # Is this better? (061716)
#q4 <- q4 +theme(legend.position="none") # this removed all legend
SupF1A <- SupF1A + theme(axis.text.x=element_text(angle=90,hjust=1),
                 axis.ticks = element_line(size=0,color="white"),
                 axis.line=element_line(size=1)) # adjust axis line width
SupF1A <- SupF1A + geom_text(aes(label=count),size=4)
SupF1A <- SupF1A + labs(x="custom category",y="",title="",fill="-log10\n pvalue")
#q4 <- q4 + theme(legend.title=element_blank())
# ggsave(q4,file="Supplemental_Fig1b.phyABDE.Sessa1hvs.my.category.040618.pdf",width=11,height=4,path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables"))
# ggsave(q4,file="Supplemental_Fig1b.phyABDE.Sessa1hvs.my.category.040618.png",width=11,height=4,path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables"))

```
# only phyB adult (Supplemental Fig1B)
```{r}
overlap.melt.phyB.adult.sun.1h.custom.files<-list.files(path=file.path("..","..","Nozue2016_SAStranscriptome_output","output"),pattern="overlap.melt.phyB.adult")
overlap.melt.summary.list2<-sapply(file.path("..","..","Nozue2016_SAStranscriptome_output","output",overlap.melt.phyB.adult.sun.1h.custom.files),function(x) mget(load(x))) # 
names(overlap.melt.summary.list2)
overlap.melt.summary<-do.call("rbind",overlap.melt.summary.list2) 
head(overlap.melt.summary) # make sure those are file names
rownames(overlap.melt.summary)<-1:nrow(overlap.melt.summary)
# format overlap.GOseq.melt.summary
overlap.melt.summary[,"value"]<--log10(overlap.melt.summary[,"value"])
overlap.melt.summary[overlap.melt.summary$value>10,"value"]<-10
overlap.melt.summary[overlap.melt.summary$count==0,"count"]<-""
table(overlap.melt.summary$Var2)
overlap.melt.summary<-overlap.melt.summary[!overlap.melt.summary$Var2=="",]
overlap.melt.summary$Var2<-factor(as.character(overlap.melt.summary$Var2),levels=c("IAAup","IAAdown","BLup","BLdown","MJup","MJdown","GAup","GAdown","ABAup","ABAdown","ACCup","ACCdown","CKup","CKdown","SAup","SAdown","MYC234up","MYC234down","PIFtarget")) # (110917)
save(overlap.melt.summary, file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","overlap.melt.summary.phyB.adult.040618.Rdata") )
# load overlap.melt.summary
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","overlap.melt.summary.phyB.adult.040618.Rdata"))
overlap.melt.summary
# rename SAup_combined and SAdown_combined (020818)
overlap.melt.summary$Var2<-gsub("_combined","",overlap.melt.summary$Var2)
# remove extra space in Var1
overlap.melt.summary$Var1<-gsub(" ","",overlap.melt.summary$Var1)
# remove no data in Var1
overlap.melt.summary.ss<-overlap.melt.summary[!overlap.melt.summary$Var1==" ",]
overlap.melt.summary.ss<-overlap.melt.summary.ss[!overlap.melt.summary.ss$Var1=="",]
# count number of Var1
names(table(overlap.melt.summary.ss$Var1))
# cleanup Var2
overlap.melt.summary.ss<-overlap.melt.summary.ss[!overlap.melt.summary.ss$Var2=="",]
overlap.melt.summary.ss<-overlap.melt.summary.ss[!is.na(overlap.melt.summary.ss$Var2),]
# check data
table(overlap.melt.summary.ss$Var1)
table(overlap.melt.summary.ss$Var2)
# plot
SupF1B<-ggplot(overlap.melt.summary.ss,aes(y=factor(Var1),x=factor(Var2))) # sort row
SupF1B <- SupF1B + geom_tile(color="black",aes(fill=value))+scale_fill_gradient2(low="white", mid="white",high="magenta",midpoint=2) # Is this better? (061716)
SupF1B <- SupF1B + theme(axis.text.x=element_text(angle=90,hjust=1),
                 axis.ticks = element_line(size=0,color="white"),
                 axis.line=element_line(size=1)) # adjust axis line width
SupF1B <- SupF1B + geom_text(aes(label=count),size=4)
SupF1B <- SupF1B + labs(x="custom category",y="",title="",fill="-log10\n pvalue")
#q4 <- q4 + theme(legend.title=element_blank())
# ggsave(SupF1B,file="Supplemental_FigX.phyB.adult.my.category.040618.pdf",width=11,height=4,path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables"))
# ggsave(SupF1B,file="Supplemental_FigX.phyB.adult.my.category.040618.png",width=11,height=4,path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables"))
```
# composit two heatmaps (Supplemental Fig 1)
```{r}
library(cowplot)
SupF1.AandB<-plot_grid(SupF1A,SupF1B,labels=c("A","B"),nrow=2)
save_plot("Supplemental_Fig1AB.phyABDE.Sessa1h.phyB.adult.vs.my.category.040618.pdf",SupF1.AandB,
          ncol=1,nrow=2,base_aspect_ratio = 2,
          path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables")
          )
```

# Number of Col genes in each module are: 
```{r eval=FALSE} 
table(SAS.expression.vst.s.kazu.largeCV.modules$"1")  # 28 modules
```
# Sessa.1h, phyABDE seedlings vs WGCNGA module ORA (Supplemental Fig 5, 040618)
```{r}
# load WGCNA modules
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","SAS.expression.vst.s.kazu.largeCV.modules092817.Rdata")) # 092717
SAS.expression.vst.s.kazu.largeCV.modules$AGI<-gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",SAS.expression.vst.s.kazu.largeCV.modules$AGI)
modules<-levels(as.factor(SAS.expression.vst.s.kazu.largeCV.modules$"1"))
# add Sessa.1h and phyABDE
test3
test<-merge(SAS.expression.vst.s.kazu.largeCV.modules[,c("AGI","1")],temp3[,c("AGI","Sessa.1h","phyABDE")],by="AGI",all=TRUE)
test[is.na(test)]<-""
table(test$Sessa.1h)
table(test$phyABDE)
table(test$`1`)s
# overlapTable
for(i in 3:4) { # 
  overlap<-overlapTable(test[,"1"],test[,i]) # all genes
  overlap.melt<-melt(overlap$pTable,id=1) #
  overlap.melt$count<-melt(overlap$countTable,id=1)$value
  overlap.melt<-overlap.melt[!overlap.melt$Var2=="0",]
  #overlap.melt.summary<-rbind(overlap.melt.summary,overlap.melt)
  save(overlap.melt,file=file.path(homedir,"..","..","Nozue2016_SAStranscriptome_output","output",paste("overlap.melt.Sessa1h.phyABDE.WGCNA.033118.",i,".Rdata",sep="")))
} # renamed
```
# combine overlapTables
```{r}
overlap.melt.Sessa1h.phyABDE.WGCNA.files<-list.files(path=file.path("..","..","Nozue2016_SAStranscriptome_output","output"),pattern="overlap.melt.Sessa1h.phyABDE.WGCNA.033118")
# combine
overlap.melt.summary.list2<-sapply(file.path("..","..","Nozue2016_SAStranscriptome_output","output",overlap.melt.Sessa1h.phyABDE.WGCNA.files),function(x) mget(load(x))) # mget will return the value of the object(or objects) in a list. see https://stackoverflow.com/questions/29398630/load-data-frames-into-list
names(overlap.melt.summary.list2)
overlap.melt.summary<-do.call("rbind",overlap.melt.summary.list2) 
head(overlap.melt.summary) # make sure those are file names
rownames(overlap.melt.summary)<-1:nrow(overlap.melt.summary)
# format overlap.GOseq.melt.summary
overlap.melt.summary[,"value"]<--log10(overlap.melt.summary[,"value"])
overlap.melt.summary[overlap.melt.summary$value>10,"value"]<-10
overlap.melt.summary[overlap.melt.summary$count==0,"count"]<-""
table(overlap.melt.summary$Var2)
overlap.melt.summary<-overlap.melt.summary[!overlap.melt.summary$Var2=="",]

save(overlap.melt.summary, file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","overlap.melt.summary.Sessa1h.phyABDE.WGCNA.033118.Rdata") )
```
# heatmap
```{r}
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","overlap.melt.summary.Sessa1h.phyABDE.WGCNA.033118.Rdata"))
overlap.melt.summary
# rename SAup_combined and SAdown_combined (020818)
overlap.melt.summary$Var2<-gsub("_combined","",overlap.melt.summary$Var2)
# 
overlap.melt.summary.ss<-overlap.melt.summary[!overlap.melt.summary$Var1==" ",]
overlap.melt.summary.ss<-overlap.melt.summary.ss[!overlap.melt.summary.ss$Var1=="",]
names(table(overlap.melt.summary.ss$Var1))
overlap.melt.summary.ss$Var1<-as.factor(overlap.melt.summary.ss$Var1)
overlap.melt.summary.ss$Var2<-gsub(" ", "",overlap.melt.summary.ss$Var2)

overlap.melt.summary.ss<-overlap.melt.summary.ss[!overlap.melt.summary.ss$Var2=="",]
overlap.melt.summary.ss$Var1<-droplevels(overlap.melt.summary.ss$Var1)
overlap.melt.summary.ss$Var2<-droplevels(overlap.melt.summary.ss$Var2)
names(table(overlap.melt.summary.ss$Var1))
names(table(overlap.melt.summary.ss$Var2))

#overlap.melt.summary.ss$Var2<-factor(overlap.melt.summary.ss$Var2,levels=paste(rep(conversion.table[-1,"genotype"],2),rep(c("sun","shade"),each=12),sep=".")) # sort Var2 (muts)
# check data
table(overlap.melt.summary.ss$Var1)
# plot
q5<-ggplot(overlap.melt.summary.ss,aes(y=Var2,x=factor(Var1))) 
q5 <- q5 + geom_tile(color="black",aes(fill=value))+scale_fill_gradient2(low="white", mid="white",high="magenta",midpoint=2) 
q5 <- q5 + theme(axis.text.x=element_text(angle=90,hjust=1),
                 axis.ticks = element_line(size=0,color="white"),
                 axis.line=element_line(size=1)) # adjust axis line width
q5 <- q5 + geom_text(aes(label=count),size=4)
q5 <- q5 + labs(x="WGCNA category",y="",title="",fill="-log10\n pvalue")
## 
ggsave(q5,file="Supplemental_Fig6.phyABDE.Sessa1h.vs.WGCNA.040618.pdf",width=11,height=4,path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables"))
ggsave(q5,file="Supplemental_Fig6.phyABDE.Sessa1h.vs.WGCNA.040618.png",width=11,height=4,path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables"))
q5

```

# Does each WGCNA module contain shade-responsive genes? (011216; 100417)
```{r}
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","SAS.expression.vst.s.kazu.largeCV.modules092817.Rdata") ) #
SAS.expression.vst.s.kazu.largeCV.modules
SAS.expression.vst.s.kazu.largeCV.modules$AGI<-gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",SAS.expression.vst.s.kazu.largeCV.modules$AGI)
# shade responsive genes 
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","DGEsummary.edgeR.response.Rdata")) # this has all genotypes
head(DGEsummary.edgeR.response)
DGEsummary.edgeR.response$AGI<-gsub("([[:print:]]+)(.[[:digit:]]+)(.)([[:digit:]]+)","\\1",rownames(DGEsummary.edgeR.response))
DGEsummary.edgeR.response$gt<-gsub("([[:print:]]+)(.[[:digit:]]+)(.)([[:digit:]]+)","\\4",rownames(DGEsummary.edgeR.response))
Col.DGE<-DGEsummary.edgeR.response[DGEsummary.edgeR.response$gt=="1",]

# mark Col.DGE genes in SAS.expression.vst.s.kazu.largeCV.modules
SAS.expression.vst.s.kazu.largeCV.modules$ColDGE<-(SAS.expression.vst.s.kazu.largeCV.modules$AGI %in% Col.DGE$AGI)*1
# 
overlap.WGCNAmodule.ColDGE<-overlapTable(SAS.expression.vst.s.kazu.largeCV.modules$"1",SAS.expression.vst.s.kazu.largeCV.modules$ColDGE)
save(overlap.WGCNAmodule.ColDGE,file=file.path("..","..","Nozue2016_SAStranscriptome_output/output/overlap.WGCNAmodule.ColDGE.Rdata"))

# add WGCNAmodule
temp<-merge(Col.DGE,TAIR10_gene_descriptions[,c("AGI","colE...col.F")],by="AGI")
#SAS.expression.vst.s.kazu.largeCV.modules$AGI<-gsub("([[:print:]]+)(.[[:digit:]]+)(.)([[:digit:]]+)","\\1",SAS.expression.vst.s.kazu.largeCV.modules$AGI)
temp2<-merge(SAS.expression.vst.s.kazu.largeCV.modules[,c("AGI","1")],temp)
Col.DGE.descripiton<-temp2[!duplicated(temp2$AGI),]
colnames(Col.DGE.descripiton)<-gsub("1","WGCNAmodule",colnames(Col.DGE.descripiton))
write.csv(Col.DGE.descripiton,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","Col.DGE.descripiton.csv"))

```
# Are salmon genes modulated in Sessa (2007) 4d samples?
```{r}
Sessa.4d<-read.csv("../../Nozue2016_SAStranscriptome_data/input/sessa4days.csv")
Sessa.1h<-read.csv("../../Nozue2016_SAStranscriptome_data/input/sessa1hour.csv")
DE.genotype.batch.genes.WGCNAmodules<-read.csv("../../Nozue2016_SAStranscriptome_output/figs_tables/DE.genotype.batch.genes.WGCNAmodules.csv")
salmon<-DE.genotype.batch.genes.WGCNAmodules[DE.genotype.batch.genes.WGCNAmodules$X1=="salmon",]
salmon$AGI
dim(Sessa.4d[Sessa.4d$ID %in% salmon$AGI,])
Sessa.4d.salmon<-Sessa.4d[Sessa.4d$ID %in% salmon$AGI,]
write.csv(Sessa.4d.salmon,file="Sessa.4d.salmon.csv")
ConvictionsSessa.1h[Sessa.1h$ID %in% salmon$AGI,]
length(salmon$AGI)
Sessa.4d.s<-Sessa.4d[Sessa.4d$adj.P.Val<0.01,]
# Fisher's exact test (assuming total gene number is 25000)
Sessa.4d.salmon<-matrix(c(11,1941-11,112-11,25000-1941-(112-11)),nrow=2)
fisher.test(Sessa.4d.salmon) # p-value = 0.3768
```
# Salmon module genes are shade responsive in Col? (100717)
```{r}
#setwd(homedir)
DEfiles<-list.files(path=file.path("..","..","Nozue2016_SAStranscriptome_output","output"),pattern="DE.batch.anytime")
DEfiles<-grep("DE.batch.anytime_genotype_wo_outlier_wo16h.PLoSGenetics2015mutants.errrDE",DEfiles,value=TRUE)
#setwd("../../Nozue2016_SAStranscriptome_output/output/")
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output",DEfiles[1])) 
rownames(DE.batch.anytime)<-paste(rownames(DE.batch.anytime),gsub("(DE.batch.anytime_genotype_wo_outlier_wo16h.PLoSGenetics2015mutants.errrDE)([[:print:]]+)(.Rdata)","\\2",DEfiles[1]),sep=".")
# salmon
#test<-read.csv("../Nozue2016_SAStranscriptome_output/figs_tables/Dataset7.csv")
DE.genotype.batch.genes.WGCNAmodules.description<-read.csv(file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Supplemental_Dataset4_Custom_category_enrichment_analysis_of_WGCNA_gene_co-expression_modules092817.csv"),row.names=1)
dim(DE.genotype.batch.genes.WGCNAmodules.description) # 17985   28 > 33602 31
salmon<-DE.genotype.batch.genes.WGCNAmodules.description[DE.genotype.batch.genes.WGCNAmodules.description$WGCNAmodules=="salmon",]
salmon$AGI
#
DE.batch.anytime[gsub("([[:print:]]+)(.)([[:digit:]]+)(.1)","\\1",rownames(DE.batch.anytime)) %in% salmon$AGI,] # none
# Salmon module genes do not have any shade responsive genes in Col...

```
# Salmon module genes vs 49h shade responsive genes in Col (old version; 100717s)
```{r}
DE.genotype.batch.genes.WGCNAmodules<-read.csv("../../Nozue2016_SAStranscriptome_output/figs_tables/DE.genotype.batch.genes.WGCNAmodules.csv")
salmon<-DE.genotype.batch.genes.WGCNAmodules[DE.genotype.batch.genes.WGCNAmodules$X1=="salmon",]
load("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/samples.nolow5.Rdata")
load("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/dge.nolow5.Rdata")
  samples.nolow.s<-samples.nolow5[samples.nolow5$genotype==1,]
  dge.nolow.s<-dge.nolow5[,samples.nolow5$genotype==1]
  dge.nolow.s<-dge.nolow.s[rowSums(dge.nolow.s$counts > 5) >= 3,] 
# subset only 49h
samples.nolow.s.49h<-samples.nolow.s[samples.nolow.s$time=="49",]
design.nolow.s.batch.fixedtime<- model.matrix(~samples.nolow.s.49h$trt + samples.nolow.s.49h$batch)
colnames(design.nolow.s.batch.fixedtime) <- gsub("samples.nolow.s.49h$","",colnames(design.nolow.s.batch.fixedtime),fixed=TRUE) # nicer column names
colnames(design.nolow.s.batch.fixedtime)
dge.nolow.s.fixedtime<-dge.nolow5[,samples.nolow5$file %in% samples.nolow.s.49h$file]
  print(design.nolow.s.batch.fixedtime)
  print(dge.nolow.s.fixedtime)
  dge.nolow.s.batch.fixedtime.glm <- estimateGLMCommonDisp(dge.nolow.s.fixedtime,design.nolow.s.batch.fixedtime, verbose=TRUE) 
  dge.nolow.s.batch.glm <- estimateGLMTrendedDisp(dge.nolow.s.batch.fixedtime.glm,design.nolow.s.batch.fixedtime, verbose=TRUE)
  dge.nolow.s.batch.fixedtime.glm.fit <- glmFit(dge.nolow.s.batch.fixedtime.glm,design.nolow.s.batch.fixedtime)
  dge.nolow.s.batch.fixedtime.glm.lrt <- glmLRT(dge.nolow.s.batch.fixedtime.glm.fit,coef="trtL")
  DE.batch.49h.Col<-topTags(dge.nolow.s.batch.fixedtime.glm.lrt,n=2000)$table[topTags(dge.nolow.s.batch.fixedtime.glm.lrt,n=2000)$table$FDR<0.05,]
# 
dim(DE.batch.49h.Col) # 278 5
DE.batch.49h.Col.salmon<-DE.batch.49h.Col[sub("([[:print:]]+)(.[[:digit:]]+)","\\1",rownames(DE.batch.49h.Col)) %in% salmon$AGI,]
setwd("../../Nozue2016_SAStranscriptome_output/output/")
write.csv(DE.batch.49h.Col.salmon,file="DE.batch.49h.Col.salmon.csv")
setwd("../../Nozue2016_SAStranscriptome_scripts/")
```
# appendix 1: differentially expressed geens used for ORA
```{r eval=FALSE}
# differential expression analysis with npr1-1
# http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE51626
# Singh M, Bag SK, Bhardwaj A, Ranjan A et al. Global nucleosome positioning 
# regulates salicylic acid mediated transcription in Arabidopsis thaliana. BMC Plant # Biol 2015 Jan 21;15:13. PMID: 25604550
# 24 h SA or mock treatment (thnaks for a Plant Cell reviewer)
###  The RNA of col-0 and npr1-1 mutant were isolated after 24hr water and SA treatment.The ATH1 gene chip was used for expression analysis.
###  GSM1249675  Col-0 water, biological replicate 1
###  GSM1249676  Col-0 water, biological replicate 2
###  GSM1249677  Col-0 SA, biological replicate 1
###  GSM1249678  Col-0 SA, biological replicate 2
###  GSM1249679	npr1-1 water, biological replicate 1
###  GSM1249680	npr1-1 water, biological replicate 2
###  GSM1249681	npr1-1 SA, biological replicate 1
###  GSM1249682	npr1-1 SA, biological replicate 2
setwd(hormedir)
setwd("../../Nozue2016_SAStranscriptome_data/input/GSE51626_RAW")
library(affy)
#Data <- ReadAffy(widget=T) # error
CEL.filename<-list.files(pattern=".CEL$")
CEL.filename
Data <- ReadAffy(filenames=CEL.filename)
Data
eset <- rma(Data) # normalize chips 
affyData <- read.csv("../affy_ATH1_array_elements-2010-12-20.csv") #read TAIR annotation data from file (TAIR10)
affyData2<-affyData[,c("array_element_name","is_control","locus","description","chromosome","start","stop")]
names(affyData2) <- c("element","is.control","locus","description","chromosome","start","stop")#get rid of underlines
head(affyData2)
genes <- data.frame(element=featureNames(eset)) #get affy gene names from dataset
affyMatch <- merge(affyData2,genes,by.x="element",by.y=1)# reorder the annotation data to match chip data
exp.values <- data.frame(exprs(eset)) #put normalized intensities in a data frame
head(exp.values)
exp.values <- cbind(affyMatch,exp.values) #add annotation to same frame as expression values
head(exp.values)
setwd(homedir)
write.csv(exp.values,file="../../Nozue2016_SAStranscriptome_data/input/GSE51626_RAW/exp.values.csv")
######## RP analysis
library(RankProd)
exp.values<-read.csv("../../Nozue2016_SAStranscriptome_data/input/GSE51626_RAW/exp.values.csv")
names(exp.values)
## Col (water) vs npr1 (water)
names(exp.values)[c(9:10,13:14)]
RP.GSE51626.npr1 <- RP(exp.values[,c(9:10,13:14)], rep(1:0,each=2), num.perm=100) # water control, 0 (class 1) = Col, 1 (class 2)= npr1-1 
topgene.RP.GSE51626.npr1<-topGene(RP.GSE51626.npr1,cutoff=0.01,gene.names=exp.values$element,logged=TRUE,logbase=2)
#  induced genes in npr1-1
topgene.RP.GSE51626.npr1.T2<-as.data.frame(topgene.RP.GSE51626.npr1$Table2)
topgene.RP.GSE51626.npr1.T2$element<-rownames(topgene.RP.GSE51626.npr1.T2)
topgene.RP.GSE51626.npr1.T2.m<-merge(topgene.RP.GSE51626.npr1.T2, affyData2,by="element")
head(topgene.RP.GSE51626.npr1.T2.m)
topgene.RP.GSE51626.npr1.T2.m.exp.value<-merge(topgene.RP.GSE51626.npr1.T2.m, exp.values[,c(2,9:10,13:14)],by="element")
topgene.RP.GSE51626.npr1.T2.m.exp.value<-topgene.RP.GSE51626.npr1.T2.m.exp.value[order(topgene.RP.GSE51626.npr1.T2.m.exp.value$pfp),]

#  repressed genes in npr1-1
topgene.RP.GSE51626.npr1.T1<-as.data.frame(topgene.RP.GSE51626.npr1$Table1)
topgene.RP.GSE51626.npr1.T1$element<-rownames(topgene.RP.GSE51626.npr1.T1)
topgene.RP.GSE51626.npr1.T1.m<-merge(topgene.RP.GSE51626.npr1.T1, affyData2,by="element")
head(topgene.RP.GSE51626.npr1.T1.m)
topgene.RP.GSE51626.npr1.T1.m.exp.value<-merge(topgene.RP.GSE51626.npr1.T1.m, exp.values[,c(2,9:10,13:14)],by="element")
topgene.RP.GSE51626.npr1.T1.m.exp.value<-topgene.RP.GSE51626.npr1.T1.m.exp.value[order(topgene.RP.GSE51626.npr1.T1.m.exp.value$pfp),]

# summary
topgene.RP.GSE51626.npr1.summary<-rbind(topgene.RP.GSE51626.npr1.T1.m.exp.value,topgene.RP.GSE51626.npr1.T2.m.exp.value)
write.csv(topgene.RP.GSE51626.npr1.summary,file="../../Nozue2016_SAStranscriptome_data/input/GSE51626_RAW/topgene.RP.GSE51626.npr1.summary.csv")

# npr1 responsive (24h SA treated) (110916)
names(exp.values)[c(11:12,15:16)] # 
RP.GSE51626.SA.npr1 <- RP(exp.values[,c(11:12,15:16)], rep(1:0,each=2), num.perm=100) # water control, 0 (class 1) = SA + Col, 1 (class 2)= SA + npr1
topgene.RP.GSE51626.SA.npr1<-topGene(RP.GSE51626.SA.npr1,cutoff=0.01,gene.names=exp.values$element,logged=TRUE,logbase=2)
#  induced genes by SA in Col
topgene.RP.GSE51626.SA.npr1.T2<-as.data.frame(topgene.RP.GSE51626.SA.npr1$Table2)
topgene.RP.GSE51626.SA.npr1.T2$element<-rownames(topgene.RP.GSE51626.SA.npr1.T2)
topgene.RP.GSE51626.SA.npr1.T2.m<-merge(topgene.RP.GSE51626.SA.npr1.T2, affyData2,by="element")
head(topgene.RP.GSE51626.SA.npr1.T2.m)
topgene.RP.GSE51626.SA.npr1.T2.m.exp.value<-merge(topgene.RP.GSE51626.SA.npr1.T2.m, exp.values[,c(2,11:12,15:16)],by="element")
topgene.RP.GSE51626.SA.npr1.T2.m.exp.value<-topgene.RP.GSE51626.SA.npr1.T2.m.exp.value[order(topgene.RP.GSE51626.SA.npr1.T2.m.exp.value$pfp),]

#  repressed genes by SA
topgene.RP.GSE51626.SA.npr1.T1<-as.data.frame(topgene.RP.GSE51626.SA.npr1$Table1)
topgene.RP.GSE51626.SA.npr1.T1$element<-rownames(topgene.RP.GSE51626.SA.npr1.T1)
topgene.RP.GSE51626.SA.npr1.T1.m<-merge(topgene.RP.GSE51626.SA.npr1.T1, affyData2,by="element")
head(topgene.RP.GSE51626.SA.npr1.T1.m)
topgene.RP.GSE51626.SA.npr1.T1.m.exp.value<-merge(topgene.RP.GSE51626.SA.npr1.T1.m, exp.values[,c(2,11:12,15:16)],by="element")
topgene.RP.GSE51626.SA.npr1.T1.m.exp.value<-topgene.RP.GSE51626.SA.npr1.T1.m.exp.value[order(topgene.RP.GSE51626.SA.npr1.T1.m.exp.value$pfp),]

# summary
topgene.RP.GSE51626.SA.npr1.summary<-rbind(topgene.RP.GSE51626.SA.npr1.T1.m.exp.value,topgene.RP.GSE51626.SA.npr1.T2.m.exp.value)
write.csv(topgene.RP.GSE51626.SA.npr1.summary,file="../../Nozue2016_SAStranscriptome_data/input/GSE51626_RAW/topgene.RP.GSE51626.SA.npr1.summary.csv")

# SA responsive
## Col (water) vs Col (SA)
names(exp.values)[c(9:12)]
RP.GSE51626.SA.Col <- RP(exp.values[,c(9:12)], rep(1:0,each=2), num.perm=100) # water control, 0 (class 1) = Col, 1 (class 2)= SA Col
topgene.RP.GSE51626.SA.Col<-topGene(RP.GSE51626.SA.Col,cutoff=0.01,gene.names=exp.values$element,logged=TRUE,logbase=2)
#  induced genes in npr1-1
topgene.RP.GSE51626.SA.Col.T2<-as.data.frame(topgene.RP.GSE51626.SA.Col$Table2)
topgene.RP.GSE51626.SA.Col.T2$element<-rownames(topgene.RP.GSE51626.SA.Col.T2)
topgene.RP.GSE51626.SA.Col.T2.m<-merge(topgene.RP.GSE51626.SA.Col.T2, affyData2,by="element")
head(topgene.RP.GSE51626.SA.Col.T2.m)
topgene.RP.GSE51626.SA.Col.T2.m.exp.value<-merge(topgene.RP.GSE51626.SA.Col.T2.m, exp.values[,c(2,9:12)],by="element")
topgene.RP.GSE51626.SA.Col.T2.m.exp.value<-topgene.RP.GSE51626.SA.Col.T2.m.exp.value[order(topgene.RP.GSE51626.SA.Col.T2.m.exp.value$pfp),]

#  SA repressed genes
topgene.RP.GSE51626.SA.Col.T1<-as.data.frame(topgene.RP.GSE51626.SA.Col$Table1)
topgene.RP.GSE51626.SA.Col.T1$element<-rownames(topgene.RP.GSE51626.SA.Col.T1)
topgene.RP.GSE51626.SA.Col.T1.m<-merge(topgene.RP.GSE51626.SA.Col.T1, affyData2,by="element")
head(topgene.RP.GSE51626.SA.Col.T1.m)
topgene.RP.GSE51626.SA.Col.T1.m.exp.value<-merge(topgene.RP.GSE51626.SA.Col.T1.m, exp.values[,c(2,9:12)],by="element")
topgene.RP.GSE51626.SA.Col.T1.m.exp.value<-topgene.RP.GSE51626.SA.Col.T1.m.exp.value[order(topgene.RP.GSE51626.SA.Col.T1.m.exp.value$pfp),]

# summary
topgene.RP.GSE51626.SA.Col.summary<-rbind(topgene.RP.GSE51626.SA.Col.T1.m.exp.value,topgene.RP.GSE51626.SA.Col.T2.m.exp.value)
write.csv(topgene.RP.GSE51626.SA.Col.summary,file="../../Nozue2016_SAStranscriptome_data/input/GSE51626_RAW/topgene.RP.GSE51626.SA.Col.summary.csv")

# VennDiagram (111116) see more in https://rstudio-pubs-static.s3.amazonaws.com/13301_6641d73cfac741a59c0a851feb99e98b.html. That is Ruijuan is using for her color VennDiagram.
library(VennDiagram)
setwd(homedir)
# reading DE gene list tables
GSE51626.Venn<-data.frame(element=affyData$array_element_name,AGI=affyData$locus,npr1=" ",SA.npr1=" ",SA.Col=" ")
topgene.RP.GSE51626.npr1.summary<-read.csv("../../Nozue2016_SAStranscriptome_data/input/GSE51626_RAW/topgene.RP.GSE51626.npr1.summary.csv");topgene.RP.GSE51626.npr1.summary<-topgene.RP.GSE51626.npr1.summary[,-1]
topgene.RP.GSE51626.SA.npr1.summary<-read.csv("../../Nozue2016_SAStranscriptome_data/input/GSE51626_RAW/topgene.RP.GSE51626.SA.npr1.summary.csv");topgene.RP.GSE51626.SA.npr1.summary<-topgene.RP.GSE51626.SA.npr1.summary[,-1]
topgene.RP.GSE51626.SA.Col.summary<-read.csv("../../Nozue2016_SAStranscriptome_data/input/GSE51626_RAW/topgene.RP.GSE51626.SA.Col.summary.csv");topgene.RP.GSE51626.SA.Col.summary<-topgene.RP.GSE51626.SA.Col.summary[,-1]

GSE51626.Venn[GSE51626.Venn$element %in% topgene.RP.GSE51626.npr1.summary$element,"npr1"]<-"npr1"
GSE51626.Venn[GSE51626.Venn$element %in% topgene.RP.GSE51626.SA.npr1.summary$element,"SA.npr1"]<-"SA.npr1"
GSE51626.Venn[GSE51626.Venn$element %in% topgene.RP.GSE51626.SA.Col.summary$element,"SA.Col"]<-"SA.Col"
# gene of interest in the VennDiagram  PCC1, XBAT34, AIG1 (PCC1 (  AT3G22231; orange, XBAT34 (AT4G14365, orange), but AIG1 (thistle2;  AT1G33960), ACD6 (AT4G14400, ?)
GSE51626.Venn[GSE51626.Venn$AGI %in% c("AT3G22231","AT4G14365","AT1G33960","AT4G14400"),] #
GSE51626.Venn[GSE51626.Venn$npr1=="npr1",]
# update hormone responsiveness data set (111116)

# no need to make data frame! Use use element name for each set
setwd("../../Nozue2016_SAStranscriptome_output/output/")
venn.plot<-venn.diagram(list(npr1=topgene.RP.GSE51626.npr1.summary$element,SA.npr1=topgene.RP.GSE51626.SA.npr1.summary$element,SA.Col=topgene.RP.GSE51626.SA.Col.summary$element),filename="Venn_npr1.png")
# up and down
setwd("/Volumes/data_work/Data8/NGS_related/Arabidopsis_analysis") # for temporal system
group_comparison.Ver5<-read.csv("group_comparison.Ver5.csv") 
# target categories
categories<-c("IAAup","IAAdown","BLup","BLdown","GAinga1.3UP","GAinga1.3DOWN","BLup","BLdown","MJup","MJdown","ABAup","ABAdown","ACCup","ACCdown","CKup","CKdown","SA_UP","SA_DOWN",
              "SARplus","SARminus","GRF1_3target","PIFtarget","MYC234_up","MYC234_down","NPR1_up","NPR1_down","WRKY33up","WRKY33down","PCC1up","PCC1down","SAup_24h","SAdown_24h","NPR1up_uponSA","NPR1down_uponSA")
# rename categories
group_comparison.Ver5.selected<-group_comparison.Ver5[,names(group_comparison.Ver5[-c(1:3),]) %in% categories]
#names(group_comparison.Ver5.selected)<-c("ABAup","ABAdown","ACCup","ACCdown","BLup","BLdown","IAAup","IAAdown","MJup","MJdown","CKup","CKdown","GAup","GAdown",
                                          "SAup_2h","SAdown_2h","SARminus","SARplus","GRF1_3target","PIFtarget","MYC234up","MYC234down","NPR1up","NPR1down","WRKY33up","WRKY33down","PCC1up","PCC1down","SAup_24h","SAdown_24h","NPR1up_uponSA","NPR1down_uponSA")
# "WRKY33up" and "WRKY33donw" are reversed! (correct group_comparison.Ver5.csv)
names(group_comparison.Ver5.selected)<-gsub("MYC234_","MYC234",names(group_comparison.Ver5.selected))
names(group_comparison.Ver5.selected)<-gsub("GAinga1.3","GA",names(group_comparison.Ver5.selected))
names(group_comparison.Ver5.selected)<-gsub("SA_UP","SAup_2h",names(group_comparison.Ver5.selected))
names(group_comparison.Ver5.selected)<-gsub("SA_DOWN","SAdown_2h",names(group_comparison.Ver5.selected))
names(group_comparison.Ver5.selected)<-gsub("UP","up",names(group_comparison.Ver5.selected))
names(group_comparison.Ver5.selected)<-gsub("DOWN","down",names(group_comparison.Ver5.selected))
names(group_comparison.Ver5.selected)<-gsub("NPR1_","NPR1",names(group_comparison.Ver5.selected))
names(group_comparison.Ver5.selected)
setwd(homedir)
setwd("../../Nozue2016_SAStranscriptome_output/output/")

temp.list<-list(NPR1.up=group_comparison.Ver5.selected$NPR1up[-(1:3)],NPR1.down=group_comparison.Ver5.selected$NPR1down[-(1:3)],NPR1up_uponSA=group_comparison.Ver5.selected$NPR1up_uponSA[-(1:3)],NPR1down_uponSA=group_comparison.Ver5.selected$NPR1down_uponSA[-(1:3)])
temp.list2<-lapply(temp.list,as.character)
venn.plot2<-venn.diagram(temp.list2,filename="Venn_npr2.png")
######


# ### other SA microarray
# http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE3984
# 3w to 4w old plants were sprayed with mock or SA with 0.01% Silwet. Leaves were collected 2h after treatment.
# Supplementary data files not provided
# use SOFT file (http://www2.warwick.ac.uk/fac/sci/moac/people/students/peter_cock/r/geo/)
source("http://www.bioconductor.org/biocLite.R")
setwd("/Volumes/Data8/")
biocLite("GEOquery")
library(Biobase)
library(GEOquery)
#Download GDS file, put it in the current directory, and load it:
#gds3984 <- getGEO('GSE3984', destdir=".")

#Or, open an existing GDS file (even if its compressed):
gse3984 <- getGEO(filename='GSE3984_family.soft.gz')
Meta(gds3984)
eset <- GDS2eSet(gds3984, do.log2=TRUE)
# error
# see  "4.4  Converting GSE to an ExpressionSet" in vignet of GEOquery
gsmplatforms <- lapply(GSMList(gse3984),function(x) {Meta(x)$platform})
Table(GSMList(gse3984)[[1]])[1:5,]
probesets <- Table(GPLList(gse3984)[[1]])$ID
data.matrix <- do.call('cbind',lapply(GSMList(gse3984),function(x) 
{tab <- Table(x) 
 mymatch <- match(probesets,tab$ID_REF) 
 return(tab$VALUE[mymatch])
}))
data.matrix <- apply(data.matrix,2,function(x) {as.numeric(as.character(x))})
hist(data.matrix) # log2 transformed data
#data.matrix <- log2(data.matrix) # is this already transformed?
data.matrix[1:5,]
rownames(data.matrix) <- probesets
colnames(data.matrix) <- names(GSMList(gse3984))
pdata <- data.frame(samples=names(GSMList(gse3984)))
rownames(pdata) <- names(GSMList(gse3984))
pheno <- as(pdata,"AnnotatedDataFrame")
eset2 <- new('ExpressionSet',exprs=data.matrix,phenoData=pheno)
eset2
genes <- data.frame(element=featureNames(eset2)) #get affy gene names from dataset
affyMatch <- merge(affyData2,genes,by.x="element",by.y=1)# reorder the annotation data to match chip data
exp.values <- data.frame(exprs(eset2)) #put normalized intensities in a data frame
head(exp.values)
exp.values <- cbind(affyMatch,exp.values) #add annotation to same frame as expression values
head(exp.values)
# GSM90867  2-hour mock-treated leaf material biological replicate 1
# GSM90868  2-hour mock-treated leaf material biological replicate 2
# GSM90869  2-hour mock-treated leaf material biological replicate 3
# GSM90870	leaf material harvested 2 hours after spraying of 1 mM salicylic acid biological replicate 1
# GSM90871	leaf material harvested 2 hours after spraying of 1 mM salicylic acid biological replicate 2
# GSM90872	leaf material harvested 2 hours after spraying of 1 mM salicylic acid biological replicate 3
# RP
library(RankProd)
RP.ThibaudNissen.SA <- RP(exp.values[,c(8:13)], rep(1:0,each=3), num.perm=100) # 0 (class 1) = mock, 1 (class 2)= SA
#  induced genes
topgene.RP.ThibaudNissen.SA<-topGene(RP.ThibaudNissen.SA,cutoff=0.01,gene.names=rownames(exp.values),logged=TRUE,logbase=2)
topgene.RP.ThibaudNissen.SA.T2<-as.data.frame(topgene.RP.ThibaudNissen.SA$Table2)
topgene.RP.ThibaudNissen.SA.T2$element<-rownames(topgene.RP.ThibaudNissen.SA.T2)
topgene.RP.ThibaudNissen.SA.T2.m<-merge(topgene.RP.ThibaudNissen.SA.T2, affyData2,by="element")
head(topgene.RP.ThibaudNissen.SA.T2.m)
#  repressed genes
topgene.RP.ThibaudNissen.SA.T1<-as.data.frame(topgene.RP.ThibaudNissen.SA$Table1)
topgene.RP.ThibaudNissen.SA.T1$element<-rownames(topgene.RP.ThibaudNissen.SA.T1)
topgene.RP.ThibaudNissen.SA.T1.m<-merge(topgene.RP.ThibaudNissen.SA.T1, affyData2,by="element")
head(topgene.RP.ThibaudNissen.SA.T1.m)

## limma (see limmar user guide, 091114)
design <- cbind(mock=c(1,1,1,0,0,0),SA=c(0,0,0,1,1,1))
# or
Group <- factor(c("mock","mock","mock","SA","SA","SA"), levels=c("mock","SA"))
design <- model.matrix(~0+Group)                
colnames(design) <- c("mock","SA")

fit <- lmFit(exp.values[,c(8:13)], design)
cont.matrix <- makeContrasts(mockvsSA=SA-mock, levels=design)
fit2 <- contrasts.fit(fit, cont.matrix)
fit2 <- eBayes(fit2,0.01) # 0.01 is used in GEO2R (http://www.ncbi.nlm.nih.gov/geo/geo2r/?acc=GSE3984)
topgene.limma.ThibaudNissen.SA<-topTable(fit2, adjust="fdr",sort.by="B",number=500) # this is different from GEO2R output ...
topgene.limma.ThibaudNissen.SA<-topgene.limma.ThibaudNissen.SA[topgene.limma.ThibaudNissen.SA$adj.P.Val<0.05,]

# overlap between RP and limma
topgene.RP.ThibaudNissen.SA.T1.m.limma<-topgene.RP.ThibaudNissen.SA.T1.m[topgene.RP.ThibaudNissen.SA.T1.m$element %in% rownames(topgene.limma.ThibaudNissen.SA),]
topgene.RP.ThibaudNissen.SA.T2.m.limma<-topgene.RP.ThibaudNissen.SA.T2.m[topgene.RP.ThibaudNissen.SA.T2.m$element %in% rownames(topgene.limma.ThibaudNissen.SA),]
setwd("/Volumes/Data8/NGS_related/Arabidopsis_analysis")
write.csv(topgene.RP.ThibaudNissen.SA.T1.m.limma,file="topgene.RP.ThibaudNissen.SA.T1.m.limma.csv")
write.csv(topgene.RP.ThibaudNissen.SA.T2.m.limma,file="topgene.RP.ThibaudNissen.SA.T2.m.limma.csv")
```
# appendix 2 summary of custom category table (Supplemental Table 3) 
## rename 
```{r}
group_comparison.Ver5<-read.csv(file.path("","Volumes","data_work","Data8","NGS_related","Arabidopsis_analysis","group_comparison.Ver5.csv")) #
# target categories (renewd 111417)
categories<-c("IAAup","IAAdown","BLup","BLdown","GAinga1.3UP","GAinga1.3DOWN","BLup","BLdown","MJup","MJdown","ABAup","ABAdown","ACCup","ACCdown","CKup","CKdown","SA_UP","SA_DOWN",
              "SARplus","SARminus","PIFtarget","MYC234_up","MYC234_down","NPR1_up","NPR1_down","WRKY33up","WRKY33down","SAup_24h","SAdown_24h","NPR1up_uponSA","NPR1down_uponSA")

# rename categories
group_comparison.Ver5.selected<-group_comparison.Ver5[,names(group_comparison.Ver5[-c(1:3),]) %in% categories]
names(group_comparison.Ver5.selected)<-gsub("MYC234_","MYC234",names(group_comparison.Ver5.selected))
names(group_comparison.Ver5.selected)<-gsub("GAinga1.3","GA",names(group_comparison.Ver5.selected))
names(group_comparison.Ver5.selected)<-gsub("SA_UP","SAup_2h",names(group_comparison.Ver5.selected))
names(group_comparison.Ver5.selected)<-gsub("SA_DOWN","SAdown_2h",names(group_comparison.Ver5.selected))
names(group_comparison.Ver5.selected)<-gsub("UP","up",names(group_comparison.Ver5.selected))
names(group_comparison.Ver5.selected)<-gsub("DOWN","down",names(group_comparison.Ver5.selected))
names(group_comparison.Ver5.selected)<-gsub("NPR1_","NPR1",names(group_comparison.Ver5.selected))
names(group_comparison.Ver5.selected)

dummy.cat3<-as.data.frame(matrix(nrow=length(names(bias)),ncol=length(names(group_comparison.Ver5.selected))))

dummy.cat3<-as.data.frame(matrix(nrow=length(names(bias)),ncol=1))
rownames(dummy.cat3)<-names(bias) # use "bias" object from function_RNAseq_time_course.R 
for(i in names(group_comparison.Ver5.selected)) {# 
  print(i)
  x.locus<-group_comparison.Ver5.selected[-c(1:3),names(group_comparison.Ver5.selected)==i]
  # clean up x.locus
  x.locus<-x.locus[!x.locus==""]  
  # only single AGI name is extracted (no probe naem, no multiple AGI name, such as "AT1G23900 AT1G23940") 
  dummy.cat3[rownames(dummy.cat3) %in% toupper(x.locus),i]<-rep(i,sum(as.integer(rownames(dummy.cat3) %in% toupper(x.locus))))
}
head(dummy.cat3)
dummy.cat3<-dummy.cat3[,-1]
# clean up <NA>
dummy.cat3[is.na(dummy.cat3)]<-""
# 
### convert into list
temp<-list()
for(i in 1:dim(dummy.cat3)[1]) {
  temp[[i]]<-paste(dummy.cat3[i,])
}
names(temp)<-rownames(dummy.cat3)   
hormone.responsiveness6b<-temp
head(hormone.responsiveness6b)
table(hormone.responsiveness6b[1:10])
attributes(hormone.responsiveness6b)$category<-names(group_comparison.Ver5.selected)
## change list into data.frame
attributes(hormone.responsiveness6b)
hormone.responsiveness6.DF<-data.frame(t(sapply(hormone.responsiveness6b,c)))
names(hormone.responsiveness6.DF)<-attributes(hormone.responsiveness6b)$category
summary(hormone.responsiveness6.DF)
hormone.responsiveness6.DF$AGI<-rownames(hormone.responsiveness6.DF)
# should I combine SA-related categories? (111417)
#hormone.responsiveness6.DF$SAup_combined<-with(hormone.responsiveness6.DF,paste(SAup_2h,SARplus,NPR1down,SAup_24h,NPR1up_uponSA),sep="")
hormone.responsiveness6.DF$SAup<-with(hormone.responsiveness6.DF,paste(SAup_2h,SARplus,NPR1down,SAup_24h,NPR1up_uponSA),sep="")

#hormone.responsiveness6.DF$SAup_combined[!hormone.responsiveness6.DF$SAup_combined=="    "]<-"SAup_combined"
hormone.responsiveness6.DF$SAup[!hormone.responsiveness6.DF$SAup=="    "]<-"SAup" # 020618
hormone.responsiveness6.DF$SAup[hormone.responsiveness6.DF$SAup=="    "]<-""
table(hormone.responsiveness6.DF$SAup)
# SAdown_combined
# hormone.responsiveness6.DF$SAdown_combined<-with(hormone.responsiveness6.DF,paste(SAdown_2h,SARminus,NPR1up,SAdown_24h,NPR1down_uponSA),sep="")
hormone.responsiveness6.DF$SAdown<-with(hormone.responsiveness6.DF,paste(SAdown_2h,SARminus,NPR1up,SAdown_24h,NPR1down_uponSA),sep="")

# hormone.responsiveness6.DF$SAdown_combined[!hormone.responsiveness6.DF$SAdown_combined=="    "]<-"SAdown_combined"
hormone.responsiveness6.DF$SAdown[!hormone.responsiveness6.DF$SAdown=="    "]<-"SAdown" #020618
hormone.responsiveness6.DF$SAdown[hormone.responsiveness6.DF$SAdown=="    "]<-""
table(hormone.responsiveness6.DF$SAdown)
# 
hormone.responsiveness6.DF.s<-hormone.responsiveness6.DF[,c("ABAup","ABAdown","ACCup","ACCdown","BLup","BLdown","IAAup","IAAdown","MJup","MJdown","CKup","CKdown","GAup","GAdown","PIFtarget","MYC234up","MYC234down","SAup","SAdown","WRKY33up","WRKY33down","AGI")] # 110617, 020618

summary(hormone.responsiveness6.DF.s)
lapply(hormone.responsiveness6.DF.s[,grep("AGI",colnames(hormone.responsiveness6.DF.s),invert=T)],table)
## for Supplemental Table 5. Supplemental Datasets for custom categories (110817)
save(hormone.responsiveness6.DF.s,file=file.path("..","..","Nozue2016_SAStranscriptome_data","input","hormone.responsiveness6.DF.s.Rdata"))
load(file.path("..","..","Nozue2016_SAStranscriptome_data","input","hormone.responsiveness6.DF.s.Rdata"))
# save hormone.responsiveness6.DF.s in csv file format as Table S5
hormone.responsiveness6.DF.s.list<-as.list(hormone.responsiveness6.DF.s)
hormone.responsiveness6.DF.s.list2<-lapply(hormone.responsiveness6.DF.s.list,function(x) hormone.responsiveness6.DF.s.list[["AGI"]][!x==""])
hormone.responsiveness6.DF.s.list2.DF<-plyr::ldply(hormone.responsiveness6.DF.s.list2, rbind) # works
hormone.responsiveness6.DF.s.list2.DF2<-as.data.frame(t(hormone.responsiveness6.DF.s.list2.DF[1:21,])) # eliminate "AGI" row (22)
hormone.responsiveness6.DF.s.list2.DF2[is.na(hormone.responsiveness6.DF.s.list2.DF2)]<-""
hormone.responsiveness6.DF.s.list2.DF2<-hormone.responsiveness6.DF.s.list2.DF2[,-1]
write.csv(hormone.responsiveness6.DF.s.list2.DF2,file=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Supplemental_Table3b_custom_categories.csv")) # remove 1st column and 1st row in excel. Add data source table to make Supplemental Table3
```
# appendix 3  "Should I filter probesets or genes for WGCNA analysis?
*Probesets or genes may be filtered by mean expression or variance (or their robust analogs such as median and median absolute deviation, MAD) since low-expressed or non-varying genes usually represent noise. Whether it is better to filter by mean expression or variance is a matter of debate; both have advantages and disadvantages, but more importantly, they tend to filter out similar sets of genes since mean and variance are usually related.
*We do not recommend filtering genes by differential expression. WGCNA is designed to be an unsupervised analysis method that clusters genes based on their expression profiles. Filtering genes by differential expression will lead to a set of correlated genes that will essentially form a single (or a few highly correlated) modules. It also completely invalidates the scale-free topology assumption, so choosing soft thresholding power by scale-free topology fit will fail. "
### http://labs.genetics.ucla.edu/horvath/CoexpressionNetwork/Rpackages/WGCNA/faq.html
#### it is OK to filter genes by variance (= largeCV genes in my scripts)
# misc
```{r}
# chloroplast localized genes in orange module (112916)
Atgo.CC.orange<-Atgo.list[names(Atgo.list) %in% SAS.expression.vst.s.kazu.largeCV.modules[SAS.expression.vst.s.kazu.largeCV.modules$"1"=="orange","AGI"]] #"GO:0009507" is chloroplast
Atgo.CC.orange.chl<-Atgo.CC.orange[grep("GO:0009507",Atgo.CC.orange)]
Atgo.CC.orange.chl.description<-TAIR10_gene_descriptions[TAIR10_gene_descriptions$AGI %in% names(Atgo.CC.orange.chl),]
write.csv(Atgo.CC.orange.chl.description,file="Atgo.CC.orange.chl.description.csv")
```
# misc
```{r}
# SA related genes are shade-responsive? (071115)
load("../../Nozue2016_SAStranscriptome_output/output/newdata.vst.Rdata") 
newdata.vst$gt2<-""
for(i in 1:27){
  newdata.vst[newdata.vst$gt==conversion.table[i,1],"gt2"]<-as.character(conversion.table[i,2])
}
newdata.vst[newdata.vst$AGI %in% c("AT3G48090","AT4G35580","AT1G74710","AT5G08330","AT3G52430","AT3G26830","AT5G26920","AT2G46970"),c("1h","4h","25h","49h","ssom3.4.unit.classif","gt2","AGI")] 
# NTL9 (AT4G35580), CHE (AT5G08330), ICS1(AT1G74710, SID2), PAD4 (AT3G52430), PAD3 (AT3G26830), CBP60g (AT5G26920), PIL1 (AT2G46970) as a positive control for this script. 

```

# session info
```{r}
sessionInfo()
```

# omake
```{r}
summary.vst.response.kazu %>% rownames_to_column() %>% group_by(rowname)
# usage of dplyr::gather (instead of melt)
summary.vst.response.kazu %>% rownames_to_column() %>% 
  gather(sample, # the column name of the new column that will contain the key 
         value,# the column name of the column that will contain the observations
         -rowname # the column names that we want to gather.
         ) %>% 
  group_by(rowname) %>% summarise(max=max(value),min=min(value),mean=mean(value)) %>%  ggplot(aes(max)) + geom_histogram(binwidth=0.2)

```