---
"title: "SAS mutants time-course RNAseq analysis"
output: html_document
---
## revision history
* 
## To do
* 
## outline of this scripts
* eliminating libraries with wrong genotypes
* eliminating outlier libraries
+ Use samples.nolow5.Rdata object for select libraries
* differentially expressed gene analysis
+ 1. DE (treatment, ie, shade responsiveness) at any time point within given genotype
+ 2. DE at any time point (shade)
* VST 
* SOM analysis with DE(1) genes (using shade-responsiveness (shade/sun))
* GOseq analysis of SOM clusters
* overlap between SOM clusters and misexpressed genes in mutants
* co-expression analysis by WGCNA (all Col samples)
* rewrite setting directory to be compatible to new RStudio
* Do not use setwd()
* SAS.expression.vst<-read.csv("../Nozue2016_SAStranscriptome_data/input/kazu.SAS.expression.VST.082016.csv") # new transformed data by DEseq2 (082016) is not reproducible (092617). Probably due to updated DESeq2?
* Use kazu.SAS.expression.VST.092517.csv (092617) for rest of analysis
* Replace "samples.nolow5" into "PLoSGenetics2015mutants.errrDE" (100717)
* 


## preparation
```{r}
knitr::opts_chunk$set(echo = TRUE)
homedir<-getwd()
homedir
# [1] "/Volumes/data_work/Data8/NGS_related/Arabidopsis_analysis/Nozue2016_SAStranscriptome_finalfinal_newVST/Nozue2016_SAStranscriptome_scripts/SAS_muts_time_course_RNAseq_finalfinal" # 
# reading necessary funcitons and libraries
source(file.path("function_RNAseq_time_course.R"))
```

## eliminating libraries with wrong genotypes
```{r eval=FALSE}
getwd()
counts<-read.table(file.path(homedir,"..","..","Nozue2016_SAStranscriptome_data/input/SAS_counts_merged_updated_final.bam.tsv"),header=TRUE) # use SAS_counts_merged_updated.bam.tsv (081514)
# remove * row and move gene column into rowname
rownames(counts)<-counts$gene
counts<-counts[!counts$gene=="*",-1] 
#head(counts)
label<-names(counts)
# eliminating wrong libraries (see "samples" spread sheet in https://docs.google.com/spreadsheets/d/1-llHCbrDQfQhLXIlRsLhabN-ezBjwKvrdfjiNy36wCg/edit#gid=2102778022)
# updated 081514 based on final sequencing data 
eliminated.libraries<-c(
  "D2H16A.merged.bam", # AT5G02540_2
  "D4H4A.merged.bam",  # jar1-1
  "D4H16A.merged.bam", # jar1-1
  "D4H49A.merged.bam", # jar1-1
  "C6H1A.merged.bam", # phyB
  "C6H16A.merged.bam", # phyB  	
  "C7L4A.merged.bam", # pif4pi5
  "C7H16A.merged.bam", # pif4pif5	
  "C7H49A.merged.bam", # pif4pif5
  "C7L49A.merged.bam", # pif4pif5
  "C11H1A.merged.bam", # coi1_16
  "C11L1A.merged.bam", # coi1_16
  "C11H4A.merged.bam", # coi1_16
  "D11H4A.merged.bam", # coi1_16
  "C11L4A.merged.bam", # coi1_16
  "C11H16A.merged.bam", # coi1_16
  "C11L16A.merged.bam", # coi1_16
  "C11H25A.merged.bam", # coi1_16
  "C11L25A.merged.bam", # coi1_16
  "C11H49A.merged.bam", # coi1_16
  "D11H49A.merged.bam", # coi1_16
  "C11L49A.merged.bam", # coi1_16
  "C12H1A.merged.bam", # phyAB
  "C12L1A.merged.bam", # phyAB
  "D12L1A.merged.bam", # phyAB
  "C12H4A.merged.bam", # phyAB
  "C12L4A.merged.bam", # phyAB
  "C12H16A.merged.bam", # phyAB
  "C12L16A.merged.bam", # phyAB
  "C12L25A.merged.bam", # phyAB
  "C12H49A.merged.bam", # phyAB
  "C12L49A.merged.bam", # phyAB
  "C13H1A.merged.bam", # pif3_1
  "C13H16A.merged.bam", # pif3_1
  "E13H16A.merged.bam", # pif3_1
  "C13L16A.merged.bam", # pif3_1
  "E13L16A.merged.bam", # pif3_1
  "C13H25A.merged.bam", # pif3_1
  "C13L25A.merged.bam", # pif3_1
  "C13H49A.merged.bam", # pif3_1
  "D13H49A.merged.bam", # pif3_1
  "C13L49A.merged.bam", # pif3_1
  "D14L4A.merged.bam", # mida9_4
  "C14H25A.merged.bam", # mida9_4
  "D14L25A.merged.bam", # mida9_4
  "C16H1A.merged.bam", # sto
  "C16L1A.merged.bam", # sto
  "C16H4A.merged.bam", # sto
  "C16L4A.merged.bam", # sto
  "C16H16A.merged.bam", # sto
  "C16H25A.merged.bam", # sto
  "C16L25A.merged.bam", # sto
  "C16H49A.merged.bam", # sto
  "C16L49A.merged.bam", # sto
  "D16L49A.merged.bam") # sto
# eliminate 55 libraries (081514)
counts<-counts[,!names(counts) %in% eliminated.libraries]
#head(counts) 
dim(counts) #[1] 30533   749

samples <- data.frame(file=colnames(counts),
                      batch=factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)A+[[:print:]]+(.bam)","\\1",colnames(counts))),
                      trt=factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)A+[[:print:]]+(.bam)","\\3",colnames(counts))),
                      time=factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)A+[[:print:]]+(.bam)","\\4",colnames(counts))),
                      genotype=factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)A+[[:print:]]+(.bam)","\\2",colnames(counts)))
) # corrected
#
samples
counts.nolow <- counts[,colSums(counts,na.rm=TRUE) > 1000000]
samples.low<-samples[samples$file %in% names(counts)[colSums(counts,na.rm=TRUE) < 1000000],]
samples.low$counts<-colSums(counts,na.rm=TRUE)[colSums(counts,na.rm=TRUE) < 1000000]
names(counts.nolow)
counts.nolow[is.na(counts.nolow)]<-0
save(counts.nolow,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","counts.nolow.Rdata"))

```

* eliminating outlier libraries
```{r, eval=FALSE}
# in whitney. Initial DE analysis
load(file.path(homedir,"..","..",Nozue2016_SAStranscriptome_output","output","counts.nolow.Rdata"))
samples.nolow <- data.frame(file=colnames(counts.nolow),
                            batch=factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)A+[[:print:]]+(.bam)","\\1",colnames(counts.nolow))),
                            trt=factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)A+[[:print:]]+(.bam)","\\3",colnames(counts.nolow))),
                            time=factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)A+[[:print:]]+(.bam)","\\4",colnames(counts.nolow))),
                            genotype=factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)A+[[:print:]]+(.bam)","\\2",colnames(counts.nolow)))
) # corrected
summary(counts.nolow)
Genotype<-levels(samples.nolow$genotype)
# Genotype = "11" and "16" had an error in following DE analysis
Genotype<-Genotype[!Genotype %in% c("11","16")]

## normalizing 
samples.nolow$group <- paste(samples.nolow$genotype,samples.nolow$trt,samples.nolow$time,sep=".") 
samples.nolow$genotype<-as.character(samples.nolow$genotype)
dge.nolow <- DGEList(counts=counts.nolow,group=samples.nolow$group)  
dge.nolow<-calcNormFactors(dge.nolow)

for(i in Genotype) { # error in i=11,16
  print(paste("Genotype is ",i))
  # subset genotype of interests from DGEList (with normalized). see ablove (081214)
  samples.nolow.s<-samples.nolow[samples.nolow$genotype==i,]
  dge.nolow.s<-dge.nolow[,samples.nolow$genotype==i]
  # calculating cpm
  temp<-cpm(dge.nolow.s)
  #save(temp,file=paste("../../Nozue2016_SAStranscriptome_output/output/genotype",i,"cpm.Rdata",sep=".")) # better to have mean value at each time point
  #save(temp,file=paste("../../Nozue2016_SAStranscriptome_output/output/genotype",i,"cpm.Rdata",sep="."))
  
  # eliminating low expressed genes in dge.nolow.s
  dge.nolow.s<-dge.nolow.s[rowSums(dge.nolow.s$counts > 5) >= 3,]
  
  design.nolow.s.batch<- model.matrix(~samples.nolow.s$time*samples.nolow.s$trt + samples.nolow.s$batch)
  colnames(design.nolow.s.batch) <- gsub("samples.nolow.s$","",colnames(design.nolow.s.batch),fixed=TRUE) # nicer column names
  colnames(design.nolow.s.batch)
  print(design.nolow.s.batch)
  dge.nolow.s.batch.glm <- estimateGLMCommonDisp(dge.nolow.s,design.nolow.s.batch, verbose=TRUE) 
  dge.nolow.s.batch.glm <- estimateGLMTrendedDisp(dge.nolow.s.batch.glm,design.nolow.s.batch, verbose=TRUE)
  dge.nolow.s.batch.glm <- estimateGLMTagwiseDisp(dge.nolow.s.batch.glm,design.nolow.s.batch) # why this is not included in original Julin's script?
  pdf(file=paste("../../Nozue2016_SAStranscriptome_output/output/BCV_MDS_with_genotype",i,".pdf",sep=""),width=11,height=8)
  par(mfrow=c(1,2))
  plotBCV(dge.nolow.s.batch.glm,main=paste("genotype",i))
  plotMDS(dge.nolow.s.batch.glm,main=paste("genotype",i))
  dev.off()
  
  dge.nolow.s.batch.glm.fit <- glmFit(dge.nolow.s.batch.glm,design.nolow.s.batch)
  dge.nolow.s.batch.glm.lrt <- glmLRT(dge.nolow.s.batch.glm.fit,coef=c("trtL","time4:trtL","time16:trtL","time25:trtL","time49:trtL"))
  DE.batch.anytime<-topTags(dge.nolow.s.batch.glm.lrt,n=2000)$table[topTags(dge.nolow.s.batch.glm.lrt,n=2000)$table$FDR<0.05,]
  dim(DE.batch.anytime) # 416 (Col, w/o eliminating low expressed genes). 518 (Col, after eliminating low expressed genes)
  save(DE.batch.anytime,file=paste("../../Nozue2016_SAStranscriptome_output/output/DE.batch.anytime_genotype",i,".Rdata",sep=""))    
}
# eliminate 16h and find outlier
dge.nolow4<-dge.nolow[,!samples.nolow$time==16]
samples.nolow4<-samples.nolow[!samples.nolow$time==16,]
samples.nolow4$time<-as.character(samples.nolow4$time)
# starting finalfinal scripts from samples.nolow4? (092116)
save(samples.nolow4,file="../../Nozue2016_SAStranscriptome_output/output/samples.nolow4.Rdata")
load("../../Nozue2016_SAStranscriptome_output/output/samples.nolow4.Rdata")

for(i in Genotype) { # error in i=11,16 (they has been eliminated in Genotype)
  print(paste("Genotype is ",i))
  # subset genotype of interests from DGEList (with normalized) ?? see ablove (081214)
  samples.nolow.s<-samples.nolow4[samples.nolow4$genotype==i,]
  dge.nolow.s<-dge.nolow4[,samples.nolow4$genotype==i]
  
  # eliminating low expressed genes in dge.nolow.s
  dge.nolow.s<-dge.nolow.s[rowSums(dge.nolow.s$counts > 5) >= 3,]
  
  design.nolow.s.batch<- model.matrix(~samples.nolow.s$time*samples.nolow.s$trt + samples.nolow.s$batch)
  colnames(design.nolow.s.batch) <- gsub("samples.nolow.s$","",colnames(design.nolow.s.batch),fixed=TRUE) # nicer column names
  colnames(design.nolow.s.batch)
  print(design.nolow.s.batch)
  dge.nolow.s.batch.glm <- estimateGLMCommonDisp(dge.nolow.s,design.nolow.s.batch, verbose=TRUE) 
  dge.nolow.s.batch.glm <- estimateGLMTrendedDisp(dge.nolow.s.batch.glm,design.nolow.s.batch, verbose=TRUE)
  dge.nolow.s.batch.glm <- estimateGLMTagwiseDisp(dge.nolow.s.batch.glm,design.nolow.s.batch) # why this is not included in original Julin's script?
  pdf(file=paste("../../Nozue2016_SAStranscriptome_output/output/BCV_MDS_with_genotype",i,".pdf",sep=""),width=11,height=8)
  par(mfrow=c(1,2))
  plotBCV(dge.nolow.s.batch.glm,main=paste("genotype",i))
  plotMDS(dge.nolow.s.batch.glm,main=paste("genotype",i))
  dev.off()
  
  dge.nolow.s.batch.glm.fit <- glmFit(dge.nolow.s.batch.glm,design.nolow.s.batch)
  dge.nolow.s.batch.glm.lrt <- glmLRT(dge.nolow.s.batch.glm.fit,coef=c("trtL","time4:trtL","time25:trtL","time49:trtL"))
  
  DE.batch.anytime<-topTags(dge.nolow.s.batch.glm.lrt,n=2000)$table[topTags(dge.nolow.s.batch.glm.lrt,n=2000)$table$FDR<0.05,]
  dim(DE.batch.anytime) # 416 (Col, w/o eliminating low expressed genes). 518 (Col, after eliminating low expressed genes)
  save(DE.batch.anytime,file=paste("../../Nozue2016_SAStranscriptome_output/output/DE.batch.anytime_genotype_wo16h",i,".Rdata",sep=""))  	
}

## check MDS plot and find outliner libraries (081514)
# eliminate following libraries due to outlier after looking MDS plot
# Col (genotype = 1) # 90 DE
Col_outlier<-"C1L4A.merged.bam" # after elimination; 80 DE
# PAR1_RNAi09 (genotype = 10) only 7 DE
PAR1_RNAi09_outlier<-c("E10H4A.merged.bam","D10H1A.merged.bam") # 6 DE
# coi1_16 (11) only 1 DE (did not work with edgeR possibly due to lots of library are eliminated by genotyping)
#phyAB, (12), 105 DE 
phyAB_outlier<-"C12H25A.merged.bam" #>>>  
#pif3 (13) 73 DE
# AT5G02760_2 (mida9_4) (genotype = 14) original DE (dim(DE.batch.anytime), 10 8
# AT5G59010_2 (15) # 46 DE (no need to eliminate)
# sto (genotype = 16) (eliminated entire sto libraries)
# aos (17) #52 (no need to eliminate)
# 18 (genotyping showed this is not mutant. Eliminate this genotype)
# argos_outlier<-"D18L1A.merged.bam" #11 DE
#co_9 (genotype=19) 20 DE 
# AT5G02540_1 (genotype = 2)
AT5G02540_1_outlier<-"D2H4A.merged.bam" # 23 DE
# Blh_1 (genotype = 20) DE 4
# ? (genotype = 21) DE 20 (no need to eliminate libraries)
#Shahdara (genotype =22) # DE 2
Shahdara_outlier<-"C22H25A.merged.bam" # 3 DE
# Col_0 (23) # 59
Col_0_outlier<-"E23H49A.merged.bam" # 58 DE
# Cvi_0 (24) # 43
# Bur_0 (25) # 53
Bur_0_outlier<-"E25H25A.merged.bam" # ** DE
# Oy_0 (26) , 2 DE 
Oy_0_outlier<-"E26L1A.merged.bam" # only * DE
# (27) only 3 DE genes, but MDS plot looks good 
# hy5, (genotype 3) only 0 DE genes (MDS plot looks good) probably no response to shade?
# jar1 (genotype 4) 30 DE (MDS plot looks good)
# kat1_2 (5) 117 DE (no need to eliminate)
# phyB (6) 25 DE
# pif4 pif5 (7) only 6 DE genes (MDS plot looks OK)
# spt_11 (8) 81 DE >> 
#yuc2589 (9), only 10 DE genes but MDS plot looks OK. >> after outliner elimination, ** DE
# combine outlier libraries
outlier<-c(
  Col_outlier,
  PAR1_RNAi09_outlier,
  phyAB_outlier,
  AT5G02540_1_outlier,
  Shahdara_outlier, 
  Col_0_outlier,
  Bur_0_outlier,
  Oy_0_outlier
)
# eliminate outlier libraries
# system("mkdir ../../Nozue2016_SAStranscriptome_output/output/edgeR")
# system("mkdir ../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h")
samples.nolow5<-samples.nolow4[!samples.nolow4$file %in% outlier,]
dge.nolow5<-dge.nolow4[,!samples.nolow4$file %in% outlier]
Genotype<-levels(as.factor(samples.nolow5$genotype))
Genotype<-Genotype[!(Genotype=="11"|Genotype=="12"|Genotype=="18"|Genotype=="16")]

#save(dge.nolow5,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","dge.nolow5.Rdata")) # 092617
#save(samples.nolow5,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","samples.nolow5.Rdata")) # 092617
#### 100717
#load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","samples.nolow5.Rdata"))
# genotype used in PLoSGenetics2015
samples.PLoSGenetics2015mutants<-samples.nolow5[samples.nolow5$genotype %in% c("1","17","19","11","4","5","3","14","10","12","6","13","7","8","16","9"),]
# genotypes not working in edgeR 
errorDE<-c("11","12","16")
# combine
samples.PLoSGenetics2015mutants.errrDE<-samples.PLoSGenetics2015mutants[!samples.PLoSGenetics2015mutants$genotype %in% errorDE,]
save(samples.PLoSGenetics2015mutants.errrDE,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","samples.PLoSGenetics2015mutants.errrDE.Rdata")) # this has to be used below. Replace "samples.nolow5" into "sample.PLoSGenetics2015mutants.errrDE" 
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","samples.PLoSGenetics2015mutants.errrDE.Rdata"))
#load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","dge.nolow5.Rdata"))
# also change dge.nolow5.Rdata into "dge.PLoSGenetics2015mutants.errrDE.Rdata"

dge.PLoSGenetics2015mutants.errrDE<-dge.nolow5[,samples.PLoSGenetics2015mutants.errrDE$file]
save(dge.PLoSGenetics2015mutants.errrDE,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","dge.PLoSGenetics2015mutants.errrDE.Rdata")) 

```
## copied from vst.R (092517; 100717)
```{r}
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","samples.PLoSGenetics2015mutants.errrDE.Rdata"))
counts<-counts[,colnames(counts) %in% samples.PLoSGenetics2015mutants.errrDE$file]
counts[is.na(counts)] <- 0 # to avoid errors in DESeqDataSetFromMatrix
library(DESeq2)
PLoSGenetics2015mutants.errrDE$batch

dds <- DESeqDataSetFromMatrix(countData = counts, colData = PLoSGenetics2015mutants.errrDE, design = ~ batch + genotype*trt*time) # modified (070616, 092517)

head(counts)
vsd <- varianceStabilizingTransformation(dds)

vstMat <- assay(vsd)
colnames(vstMat) <- colnames(counts)
vstMat[1:10,1:10]
#write.csv(vstMat,file="../input/kazu.SAS.expression.VST.070516.csv")
#write.csv(vstMat,file="../input/kazu.SAS.expression.VST.082016.csv")
write.csv(vstMat,file="../../Nozue2016_SAStranscriptome_data/input/kazu.SAS.expression.VST.092517.csv")

```
# compare old version and new version (delete this chunk if I checked reproducibility), 092517
```{r}
setwd("../Nozue2016_SAStranscriptome_data/input")
system("sort kazu.SAS.expression.VST.082016.csv > file1.sorted")
system("sort kazu.SAS.expression.VST.092517.csv > file2.sorted")
system("diff file1.sorted file2.sorted") # ? different???
#
# ## prune low expressed genes first?
# samples <- names(counts)
# 
# sample.info <- data.frame(sample=samples,
#                           rep=substr(samples,1,1),
#                           gt=sub("[A-Z]([0-9]{1,2})[A-Z].*","\\1",samples),
#                           trt=sub("[A-Z][0-9]{1,2}([A-Z]).*","\\1",samples),
#                           time=sub("[A-Z][0-9]{1,2}[A-Z]([0-9]{1,2}).*","\\1",samples)
# )
# 
# counts.small <- counts[rowSums(counts > 5) > ncol(counts)/2,] #more than 5 counts in more than half the samples
# dim(counts.small)
# dim(counts)
# 
# dds.small <- DESeqDataSetFromMatrix(countData = counts.small, colData = sample.info, design = ~ rep + gt*trt*time)
# system.time(vsd.small <- varianceStabilizingTransformation(dds.small))
# vstMat.small <- assay(vsd.small)
# colnames(vstMat.small) <- colnames(data.small)
# 
# #write.csv(vstMat.small,"kazu.SAS.expression.VST.SMALL.112614.csv")
# setwd("../output/")
#write.csv(vstMat.small,"kazu.SAS.expression.VST.SMALL.052616.csv")
#write.csv(vstMat.small,"kazu.SAS.expression.VST.SMALL.092517.csv")

##


# compare old (082016) vs new (092517)
vstMat<-read.csv(file.path("..","..","Nozue2016_SAStranscriptome_data","input","kazu.SAS.expression.VST.092517.csv"))
rownames(vstMat)<-vstMat[,1]
SAS.expression.vst<-read.csv("kazu.SAS.expression.VST.082016.csv",row.names=1) # 092617
#rownames(SAS.expression.vst)<-SAS.expression.vst[,1]
#SAS.expression.vst.s<-SAS.expression.vst[,-1] # should be numeric for cor()
# the same columns
SAS.expression.vst.s<-SAS.expression.vst.s[,colnames(SAS.expression.vst.s) %in% colnames(vstMat)]
vstMat.s<-vstMat[,colnames(vstMat) %in% colnames(SAS.expression.vst.s)]
# the same genes
SAS.expression.vst.ss<-SAS.expression.vst.s[rownames(SAS.expression.vst.s) %in% rownames(vstMat.s),]
vstMat.ss<-vstMat.s[rownames(vstMat.s) %in% rownames(SAS.expression.vst.s),]
# cor with library
X<-sapply(1:ncol(SAS.expression.vst.ss),function(i) cor(SAS.expression.vst.ss[,i],vstMat.ss[,i]))
min(X) # 0.9921697 # library
# cor with AGI
Y<-rep("",nrow(SAS.expression.vst.ss))
Y[1:100]<-sapply(1:100,function(i) cor(t(SAS.expression.vst.ss)[,i],t(vstMat.ss)[,i]))
min(Y) # AGI ... highly correlated, but not identical... (092617)

```

## differentially expressed gene analysis
1.DE (treatment, ie, shade responsiveness) at any time point within given genotype
```{r eval=FALSE}
#load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","samples.nolow5.Rdata"))
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","samples.PLoSGenetics2015mutants.errrDE.Rdata")) # 100717
#load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","dge.nolow5.Rdata"))
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","dge.PLoSGenetics2015mutants.errrDE.Rdata"))

#SAS.expression.vst<-read.csv(file.path("..","..","Nozue2016_SAStranscriptome_data","input","kazu.SAS.expression.VST.092517.csv"),row.names=1) # new transformed data by DEseq2 (082016), (092617). Only Genotype info is used here
Genotype<-levels(as.factor(samples.PLoSGenetics2015mutants.errrDE$genotype)) # 13 genotypes inclucing Col. better to use samples.nolow5.Rdata info (confusing). Use samples.PLoSGenetics2015mutants.errrDE$genotype (100717)
# 
dim(SAS.expression.vst) # [1] 30533   287
dim(samples.nolow5) #[1] 590   6
dim(samples.PLoSGenetics2015mutants.errrDE) #[1] 287   6. Reduced number of libraries!
for(i in Genotype) { # why I eliminate genotype 1 and 10???
  print(paste("Genotype is ",i))
  #samples.nolow.s<-samples.nolow5[samples.nolow5$genotype==i,]
  #dge.nolow.s<-dge.nolow5[,samples.nolow5$genotype==i]
  samples.nolow.s<-samples.PLoSGenetics2015mutants.errrDE[samples.PLoSGenetics2015mutants.errrDE$genotype==i,] # 100717
  dge.nolow.s<-dge.PLoSGenetics2015mutants.errrDE[,samples.PLoSGenetics2015mutants.errrDE$genotype==i] # 100717
  # eliminating low expressed genes in dge.nolow.s
  dge.nolow.s<-dge.nolow.s[rowSums(dge.nolow.s$counts > 5) >= 3,]  
  # recalculate normalization (100717)
  dge.nolow.s<-calcNormFactors(dge.nolow.s)
  # design model 
  design.nolow.s.batch<- model.matrix(~samples.nolow.s$time*samples.nolow.s$trt + samples.nolow.s$batch)
  colnames(design.nolow.s.batch) <- gsub("samples.nolow.s$","",colnames(design.nolow.s.batch),fixed=TRUE) # nicer column names
  colnames(design.nolow.s.batch)
  print(design.nolow.s.batch)
  dge.nolow.s.batch.glm <- estimateGLMCommonDisp(dge.nolow.s,design.nolow.s.batch, verbose=TRUE) 
  dge.nolow.s.batch.glm <- estimateGLMTrendedDisp(dge.nolow.s.batch.glm,design.nolow.s.batch, verbose=TRUE)
  dge.nolow.s.batch.glm <- estimateGLMTagwiseDisp(dge.nolow.s.batch.glm,design.nolow.s.batch) # why this is not included in original Julin's script?
  pdf(file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h",paste("BCV_MDS_with_genotype_wo_outlier_wo16h_PLoSGenetics2015mutants.errrDE",i,".pdf",sep="")),width=11,height=8) # fine name fixed (122214)
  par(mfrow=c(1,2))
  plotBCV(dge.nolow.s.batch.glm,main=paste("genotype",i))
  mds.dge<-plotMDS(dge.nolow.s.batch.glm,main=paste("genotype",i))
  mds.dge
  # for ggplot2 plotting of MDSplot (100417)
  save(mds.dge,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h",paste("MDS_with_genotype_wo_outlier_wo16h_PLoSGenetics2015mutants.errrDE",i,".RData",sep="")))
  dev.off() # the end of normal plotting
  #
  dge.nolow.s.batch.glm.fit <- glmFit(dge.nolow.s.batch.glm,design.nolow.s.batch)
  dge.nolow.s.batch.glm.lrt <- glmLRT(dge.nolow.s.batch.glm.fit,coef=c("trtL","time4:trtL","time25:trtL","time49:trtL"))
  DE.batch.anytime<-topTags(dge.nolow.s.batch.glm.lrt,n=2000)$table[topTags(dge.nolow.s.batch.glm.lrt,n=2000)$table$FDR<0.05,]
  dim(DE.batch.anytime) # 
    save(DE.batch.anytime,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output",paste("DE.batch.anytime_genotype_wo_outlier_wo16h.PLoSGenetics2015mutants.errrDE",i,".Rdata",sep=""))) # file name fixed (122114) , (092617)   
}
```
# MDS plot by ggplot2 (100417;100717)
```{r}
mds.files<-list.files(pattern="^MDS_with_genotype_wo_outlier_wo16hPLoSGenetics2015mutants.errrDE",path=file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h"))
for(i in mds.files) {
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h",i))
x.mds.dge.DF<-as.data.frame(mds.dge$x)
y.mds.dge.DF<-as.data.frame(mds.dge$y)
distance_matrix.mds.dge<-merge(x.mds.dge.DF,y.mds.dge.DF,by="row.names")
distance_matrix.mds.dge$group<-gsub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)(A.merged.bam)","\\2\\3",distance_matrix.mds.dge$Row.names)
distance_matrix.mds.dge$genotype<-gsub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)(A.merged.bam)","\\2",distance_matrix.mds.dge$Row.names)
distance_matrix.mds.dge$trt<-gsub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)(A.merged.bam)","\\3",distance_matrix.mds.dge$Row.names)
distance_matrix.mds.dge$time<-gsub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)(A.merged.bam)","\\4",distance_matrix.mds.dge$Row.names)
distance_matrix.mds.dge$batch<-gsub("(C|D|E)([[:digit:]]+)(L|H)(1|4|16|25|49)(A.merged.bam)","\\1",distance_matrix.mds.dge$Row.names)
p.distance_matrix.mds.dge <- ggplot(data = distance_matrix.mds.dge)+ geom_label(aes(mds.dge$x, mds.dge$y, color=factor(trt),fill=factor(time),label=paste(batch,time,sep="_"),size=0.3,alpha=0.5)) + labs(title=paste("genotype",vlookup(gsub("(MDS_with_genotype_wo_outlier_wo16h)([[:digit:]]+)(.RData)","\\2",i),conversion.table,2)))
ggsave(p.distance_matrix.mds.dge, filename =file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h",paste("p.distance_matrix.mds.dge.",vlookup(gsub("(MDS_with_genotype_wo_outlier_wo16hPLoSGenetics2015mutants.errrDE)([[:digit:]]+)","\\2",i),conversion.table,2),".png",sep="")), height = 11, width = 25) 
}

```
# summary of DE genes 
* (053014 - 053114): revised version (081514 - );more revision (060415 - ); more (092617); DE.batch.anytime_genotype_wo_outlier_wo16h.PLoSGenetics2015mutants.errrDE (100717)
```{r eval=FALSE}
#getwd()
DEfiles<-list.files(path=file.path("..","..","Nozue2016_SAStranscriptome_output","output"),pattern="DE.batch.anytime_genotype_wo_outlier_wo16h.PLoSGenetics2015mutants.errrDE") # works? 
#DEfiles<-grep("DE.batch.anytime_genotype_wo_outlier_wo16h.PLoSGenetics2015mutants.errrDE",DEfiles,value=TRUE)
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output",DEfiles[1]))
rownames(DE.batch.anytime)<-paste(rownames(DE.batch.anytime),gsub("(DE.batch.anytime_genotype_wo_outlier_wo16h.PLoSGenetics2015mutants.errrDE)([[:print:]]+)(.Rdata)","\\2",DEfiles[1]),sep=".")
temp<-DE.batch.anytime

for(i in DEfiles[-1]) {
  #load(i) 
  load(file.path("..","..","Nozue2016_SAStranscriptome_output","output",i))

  if(dim(DE.batch.anytime)[1]>1) {
    rownames(DE.batch.anytime)<-paste(rownames(DE.batch.anytime),gsub("(DE.batch.anytime_genotype_wo_outlier_wo16h.PLoSGenetics2015mutants.errrDE)([[:print:]]+)(.Rdata)","\\2",i),sep=".") # fixed, 112616
    temp<-rbind(temp,DE.batch.anytime)  }
  else next
}
DGEsummary.edgeR.response<-temp
tail(DGEsummary.edgeR.response)
dim(DGEsummary.edgeR.response) #755   8 for final sequence data with normalization (081514). This is whitney. but my computer is 505 8. with PLoSGenetics2015mutants.errrDE, it became 507 8 (100717)
save(DGEsummary.edgeR.response,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","DGEsummary.edgeR.response.PLoSGenetics2015mutants.errrDE.Rdata")) # 100717
# # genotype conversion 
#load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","samples.nolow5.Rdata"))
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","samples.PLoSGenetics2015mutants.errrDE.Rdata")) # 100717

samples.PLoSGenetics2015mutants.errrDE
samples.PLoSGenetics2015mutants.errrDE$genotype2<-""
for(i in 1:13){ # only for Genotype
  samples.PLoSGenetics2015mutants.errrDE[samples.PLoSGenetics2015mutants.errrDE$genotype==conversion.table[i,1],"genotype2"]<-as.character(conversion.table[i,2])	
} 

## summary of DE gene numbers
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output",DEfiles[1]))
DEgenes.summary<-data.frame(genotype=gsub("(DE.batch.anytime_genotype_wo_outlier_wo16h.PLoSGenetics2015mutants.errrDE)([[:digit:]]+)(.Rdata)","\\2",DEfiles))
for(i in 1:length(DEfiles)) {
  load(file.path("..","..","Nozue2016_SAStranscriptome_output","output",DEfiles[i]))

  DEgenes.summary$DEgenes_num[i]<-dim(DE.batch.anytime)[1]
}
DEgenes.summary$genotype2<-vlookup(DEgenes.summary$genotype,conversion.table,2)
DEgenes.summary<-DEgenes.summary[!is.na(DEgenes.summary$DEgenes_num)&!is.na(DEgenes.summary$genotype2),]
DEgenes.summary # summary
#setwd(homedir)
```
## select genes with largeCV and calculate responsiveness for SOM analysis, updated 082016, 092617 (no change in 100717)
```{r eval=FALSE}
SAS.expression.vst<-read.csv(file.path("..","..","Nozue2016_SAStranscriptome_data","input","kazu.SAS.expression.VST.092517.csv"),row.names=1) # new transformed data by DEseq2 (092617)
# calculate mean value among replicates
#load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","samples.nolow5.Rdata"))

# samples.nolow5$sample<-with(samples.nolow5,paste(genotype,trt,time,sep=""))
# samples.nolow5$sample<-factor(samples.nolow5$sample) #where this is used? (122214)
# SAS.expression.vst.s<-SAS.expression.vst[,names(SAS.expression.vst) %in% samples.nolow5$file] 

Genotype<-levels(as.factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|25|49)(A.merged.bam)","\\2",names(SAS.expression.vst)[-1])))

temp<-SAS.expression.vst.s[,names(SAS.expression.vst.s) %in% samples.nolow5[samples.nolow5$genotype==1,"file"]]
summary.vst<-expression.mean(temp)ã€€# custom function

for(i in Genotype[-1]) {
  temp<-SAS.expression.vst.s[,names(SAS.expression.vst.s) %in% samples.nolow5[samples.nolow5$genotype==i,"file"]]
	temp<-as.data.frame(temp)
	temp2<-expression.mean(temp)
	summary.vst<-cbind(summary.vst,temp2)
}
rownames(summary.vst)<-gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",rownames(SAS.expression.vst))

save(summary.vst,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","summary.vst.Rdata")) #
# relative value 
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","summary.vst.Rdata"))
# note that summary.vst is log2 transformed
summary.vst.response<-as.data.frame((2^summary.vst[,gsub("(.+)(H|L)(1|4|16|25|49)","\\2",names(summary.vst))=="L"])/(2^summary.vst[,gsub("(.+)(H|L)(1|4|16|25|49)","\\2",names(summary.vst))=="H"])) # shade/sun 
rownames(summary.vst.response)<-rownames(summary.vst)
names(summary.vst.response)<-paste(gsub("(.+)(H|L)(1|4|16|25|49)","\\1_\\3",names(summary.vst.response)),"hrA",sep="") #
save(summary.vst.response,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","summary.vst.response.Rdata"))
# # rename as only Kazu's samples 
SAS.expression.vst.s.kazu<-SAS.expression.vst.s
dim(SAS.expression.vst.s.kazu) # 30533   287
# select genes with higher CV
co.var.df <- function(x) ( 100*apply(x,1,sd)/rowMeans(x) )
#SAS.expression.vst.s.kazu$cv <- co.var.df(SAS.expression.vst.s.kazu[,-1])
SAS.expression.vst.s.kazu$cv <- co.var.df(SAS.expression.vst.s.kazu)
a<-hist(SAS.expression.vst.s.kazu$cv,breaks=seq(0,76,1))
table(SAS.expression.vst.s.kazu$cv)

sum(as.integer(SAS.expression.vst.s.kazu$cv>3))/dim(SAS.expression.vst.s.kazu)[1] #[1] 0.4011866 > 0.6019389 > 0.5222546
sum(as.integer(SAS.expression.vst.s.kazu$cv>4))/dim(SAS.expression.vst.s.kazu)[1] #[1] 0.830806 > 0.3140864
sum(as.integer(SAS.expression.vst.s.kazu$cv>4.5))/dim(SAS.expression.vst.s.kazu)[1] #[1] 0.3621328 > 0.2300789
sum(as.integer(SAS.expression.vst.s.kazu$cv>5))/dim(SAS.expression.vst.s.kazu)[1] #[1] 0.1412242
#SAS.expression.vst.s.kazu.largeCV<-SAS.expression.vst.s.kazu[SAS.expression.vst.s.kazu$cv>3,] # 
SAS.expression.vst.s.kazu.largeCV<-SAS.expression.vst.s.kazu[SAS.expression.vst.s.kazu$cv>4.5,] 
dim(SAS.expression.vst.s.kazu.largeCV) #[1] 7438  403 > 18379   264 > 11057   288 > 7025 288
save(SAS.expression.vst.s.kazu.largeCV,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","SAS.expression.vst.s.kazu.largeCV.Rdata"))
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","SAS.expression.vst.s.kazu.largeCV.Rdata"))
dim(SAS.expression.vst.s.kazu.largeCV) # this is not response
# response value of genes with largeCV 
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","summary.vst.response.Rdata")) #summary.vst.response is object name
summary.vst.response.kazu<-summary.vst.response # where does this come from? 120116
dim(summary.vst.response.kazu) #30533    52
summary.vst.response.kazu.largeCV<-summary.vst.response.kazu[rownames(summary.vst.response.kazu) %in% gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",rownames(SAS.expression.vst.s.kazu.largeCV)),]
dim(summary.vst.response.kazu.largeCV) # 7438   60 > 18379    48 > 11057    52
save(summary.vst.response.kazu.largeCV,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","summary.vst.response.kazu.largeCV.Rdata"))
```
## select only DE genes (using PLoSGenetics2015mutants.errrDE, 100717)
```{r eval=TRUE}
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","summary.vst.response.kazu.largeCV.Rdata"))
summary.vst.response.kazu.largeCV[summary.vst.response.kazu.largeCV=="Inf"]<-0 # arbitral
summary.vst.response.kazu.largeCV[summary.vst.response.kazu.largeCV=="-Inf"]<-0 # arbitral
summary.vst.response.kazu.largeCV[is.na(summary.vst.response.kazu.largeCV)]<-0 # arbitral
#load("../../Nozue2016_SAStranscriptome_output/output/DGEsummary.edgeR.response.Rdata") #"AT1G33925
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","DGEsummary.edgeR.response.PLoSGenetics2015mutants.errrDE.Rdata"))

DGEgenelist.edgeRglm<-unique(gsub("([[:print:]]+)(.)([[:digit:]]+)([[:print:]]+)","\\1",rownames(DGEsummary.edgeR.response))) # 
length(DGEgenelist.edgeRglm) # 418; 363 (092817); 364 (100717)
# 
DGE.summary.vst.response.kazu.largeCV<-summary.vst.response.kazu.largeCV[rownames(summary.vst.response.kazu.largeCV) %in% DGEgenelist.edgeRglm,]
save(DGE.summary.vst.response.kazu.largeCV,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","DGE.summary.vst.response.kazu.largeCV.PLoSGenetics2015mutants.errrDE.Rdata")) # updated 052916, 092517, 100717
DGE.summary.vst.response.kazu.largeCV[rownames(DGE.summary.vst.response.kazu.largeCV)=="AT4G16780",] #ATHB2 has to be induced in genotype 1 (= Col)
 
```
## SOM analysis (with Yasu's help)
```{r eval=TRUE}
#library(ggplot2);library(class);library(MASS);library(kohonen);library(reshape2);library(plyr)
#load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","summary.vst.response.kazu.largeCV.Rdata")) # 092517
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","DGE.summary.vst.response.kazu.largeCV.PLoSGenetics2015mutants.errrDE.Rdata")) # 100717

genotypes<-levels(as.factor(gsub("(.+)(_)(1|4|16|25|49)(hrA)","\\1",as.character(names(DGE.summary.vst.response.kazu.largeCV)))))
DGE.summary.vst.response.kazu.largeCV<-log2(DGE.summary.vst.response.kazu.largeCV) # log2 transformation of fold change. error (092617)
i<-1
temp<-DGE.summary.vst.response.kazu.largeCV[,gsub("(.+)(_)(1|4|16|25|49)(hrA)","\\1",names(DGE.summary.vst.response.kazu.largeCV))==genotypes[i]]
rownames(temp)<-paste(rownames(temp),genotypes[i],sep=".")
colnames(temp)<-paste(gsub("(.+)(_)(1|4|16|25|49)(hrA)","\\3",colnames(temp)),"h",sep="") ### fix (070914)
for(i in 2:length(genotypes)) {
  #temp2<-DGE.summary.vst.response.kazu.largeCV[,grep(genotypes[i],names(DGE.summary.vst.response.kazu.largeCV))]
  temp2<-DGE.summary.vst.response.kazu.largeCV[,gsub("(.+)(_)(1|4|16|25|49)(hrA)","\\1",names(DGE.summary.vst.response.kazu.largeCV))==genotypes[i]]
  rownames(temp2)<-paste(rownames(temp2),genotypes[i],sep=".")
  colnames(temp2)<-paste(gsub("(.+)(_)(1|4|16|25|49)(hrA)","\\3",colnames(temp2)),"h",sep="") ### fix (070914)  
  temp<-rbind(temp,temp2)
}
summary(temp)
head(temp)
data5<-temp
#### no scaling
## SOM
set.seed(2) # Set a random seed so that SOM results are reproducible
#set.seed(123) # I do not like to have cluster 5 mixed with constitutive and 4h peak in set.seed(2) 3x4 version
ssom1.7.noscaling <- som(as.matrix(data5), somgrid(1,7,"hexagonal")) 
plot(ssom1.7.noscaling, type = "codes") #
ssom2.4.noscaling <- som(as.matrix(data5), somgrid(2,4,"hexagonal")) 
plot(ssom2.4.noscaling, type = "codes") #
ssom3.3.noscaling <- som(as.matrix(data5), somgrid(3,3,"hexagonal")) 
plot(ssom3.3.noscaling,type="codes")
ssom2.5.noscaling <- som(as.matrix(data5), somgrid(2,5,"hexagonal")) 
plot(ssom2.5.noscaling, type = "codes") #
ssom3.4.noscaling <- som(as.matrix(data5), somgrid(3,4,"hexagonal")) 
plot(ssom3.4.noscaling) 
ssom4.4.noscaling <- som(as.matrix(data5), somgrid(4,4,"hexagonal")) 
plot(ssom4.4.noscaling) 
# Selecting ssom3.4 because of (1) sufficient variation among clusters and (2) minimum redundancy among clusters judted from those plot 

## Create and write-out master SOM file
newdata.vst<-cbind(data5,ssom3.4.noscaling$unit.classif,ssom3.4.noscaling$distances)
names(newdata.vst)[5]<-"ssom3.4.unit.classif"
names(newdata.vst)[6]<-"ssom3.4.unit.distances"
newdata.vst$gt<-as.factor(gsub("(AT[[:alnum:]]+)(\\.)([[:digit:]]+)","\\3",rownames(newdata.vst)))
newdata.vst$AGI<-as.factor(gsub("(AT[[:alnum:]]+)(\\.)([[:digit:]]+)","\\1",rownames(newdata.vst)))
head(newdata.vst)
## ATHB2
newdata.vst[rownames(newdata.vst)=="AT4G16780.1",]
save(newdata.vst,file="../../Nozue2016_SAStranscriptome_output/output/newdata.vst.PLoSGenetics2015mutants.errrDE.Rdata") # 100717
#### making complete table (for SOM data) ###
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","newdata.vst.PLoSGenetics2015mutants.errrDE.Rdata")) # 100717
# replace gt num into gt name
### convert genotype number into actual genotype name
newdata.vst$gt2<-vlookup(newdata.vst$gt,conversion.table,2)
# adding gene description
SOMclusters.description<-merge(newdata.vst[newdata.vst$gt==levels(newdata.vst$gt)[1],],TAIR10_gene_descriptions[!duplicated(TAIR10_gene_descriptions$AGI),c("AGI","Short_description","colE...col.F")],by="AGI")

for(x in levels(newdata.vst$gt)[-1]) {
temp<-merge(newdata.vst[newdata.vst$gt==x,],TAIR10_gene_descriptions[!duplicated(TAIR10_gene_descriptions$AGI),c("AGI","Short_description","colE...col.F")],by="AGI")
SOMclusters.description<-rbind(SOMclusters.description,temp)
}
#write.csv(SOMclusters.description,"../../Nozue2016_SAStranscriptome_output/figs_tables/TableS2_SOMclusters.description092617.csv") #090116, 100517
write.csv(SOMclusters.description,"../../Nozue2016_SAStranscriptome_output/figs_tables/TableS2_SOMclusters.description100717.csv") #090116, 100517

```
## drawing bloxplot for each cluster
```{r eval=FALSE, ggplot, fig.width=11, fig.height=8}
#load("../../Nozue2016_SAStranscriptome_output/output/newdata.vst.Rdata") # 122214, 090116
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","newdata.vst.PLoSGenetics2015mutants.errrDE.Rdata")) # 100717

length(unique(newdata.vst$AGI)) # 333 genes (111816) > 213 (100717)
# boxplot
newdata.vstmelt<-melt(newdata.vst[,c("1h","4h","25h","49h","ssom3.4.unit.classif","gt","AGI")],id.var=c("gt","AGI","ssom3.4.unit.classif"))

names(newdata.vstmelt)[3]<-"ssom3.4.unit.classif"
newdata.vstmelt$ssom3.4.unit.classif<-factor(newdata.vstmelt$ssom3.4.unit.classif,levels=c(10, 7,12,1,11,4,2,8,5,9,6,3)) # use this factor for all heatmap. Needs to update.
newdata.vstmelt$variable<-factor(newdata.vstmelt$variable,levels=c("1h","4h","25h","49h")) # order in graphs was fixed (022115)
p <- ggplot(newdata.vstmelt,aes(x=variable,y=value,color=value))
p <- p + geom_point(position="jitter",size=1.5,alpha=0.3) + geom_boxplot(outlier.size=0,aes(fill=factor(ssom3.4.unit.classif),alpha=0.8)) + theme_bw() +xlab("sun/shade treatment (hr)") + scale_colour_gradient2(low="green1",mid="gray",high="magenta")
p <- p + facet_grid(ssom3.4.unit.classif~gt,scale="free",space="free") #+ facet_wrap(ssom3.4unit.classif~.)
p <- p + theme(strip.text.x = element_text(angle=90),axis.text.x=element_text(angle=90))
#p
ggsave("boxplot.node.genotype.3.4.noscaling.pdf",width=11,height=20,unit="in", path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables"))
#ggsave("boxplot.node.genotype.png",width=11,height=20,unit="in",dpi=600)
newdata.vst[rownames(newdata.vst)=="AT4G16780.1",] # ATHB2 cluster 10 in Col
newdata.vst[rownames(newdata.vst)=="AT2G46970.1",] # PIL1 cluster 10 in Col
```
### landscape format (similar to one used in 2014 ASPB meeting)
```{r eval=FALSE, ggplot2, fig.width=11, fig.height=8}
## probably this is better (matching to PIL1 expressio patterns in all genotypes (see plot))
## Also number of genes in cluster 5 and 6 are increased which is consistent with mutant phenotype
## do with this. (070814)
load("../../Nozue2016_SAStranscriptome_output/output/newdata.vst.Rdata") # 122214
# boxplot
newdata.vstmelt<-melt(newdata.vst[,c("1h","25h","4h","49h","ssom3.4.unit.classif","gt","AGI")],id.var=c("gt","AGI","ssom3.4.unit.classif"))
names(newdata.vstmelt)[3]<-"ssom3.4.unit.classif"
newdata.vstmelt$variable<-factor(newdata.vstmelt$variable,levels=c("1h","4h","25h","49h")) # order in graphs was fixed (022115)
newdata.vstmelt$ssom3.4.unit.classif<-factor(newdata.vstmelt$ssom3.4.unit.classif,levels=c(10, 7,12,1,11,4,2,8,5,9,6,3)) # use this factor for all heatmap. Needs to update.
### convert genotype number into actual genotype name
newdata.vstmelt$gt2<-vlookup(newdata.vstmelt$gt,conversion.table,2) 
# head(newdata.vstmelt)
#draw boxplot with 90 degree orientation for ppt presentation (2014 ASPB)
p <- ggplot(newdata.vstmelt,aes(x=variable,y=value,color=value))
p <- p + geom_hline(yintercept=0,colour="red",size=1) # adding red line at fold change = 0
p <- p + geom_point(position="jitter",size=1.5,alpha=0.3) 
p <- p + geom_boxplot(outlier.size=0,aes(fill=factor(ssom3.4.unit.classif),alpha=0.8)) + theme_bw() +xlab("sun/shade treatment (hr)") + scale_colour_gradient2(low="green1",mid="gray",high="magenta")
p <- p + facet_grid(gt2~ssom3.4.unit.classif,space="free") #+ facet_wrap(ssom3.4unit.classif~.)  ,scale="free" 
p <- p + theme(axis.text.x=element_text(angle=90),strip.text.y = element_text(angle=0),legend.position="none") #
p <- p + coord_flip()
p
ggsave("boxplot.node.genotype3.4.noscaling_similarto2014ASPB.pdf",width=15,height=8,unit="in",path="../../Nozue2016_SAStranscriptome_output/figs_tables")
```
### only Col and yucQ
```{r eval=FALSE, ggplot3, fig.width=11, fig.height=4} 
### plot only Col and yucQ (071214)
load("../../Nozue2016_SAStranscriptome_output/output/newdata.vst.Rdata") # 122214
newdata.vst.Col.yucQ<-newdata.vst[newdata.vst$gt==1|newdata.vst$gt==9,c("1h","25h","4h","49h","ssom3.4.unit.classif","gt","AGI")]
newdata.vst.Col.yucQ.melt<-melt(newdata.vst.Col.yucQ,id.var=c("gt","AGI","ssom3.4.unit.classif"))
newdata.vst.Col.yucQ.melt$gt<-as.character(newdata.vst.Col.yucQ.melt$gt)
newdata.vst.Col.yucQ.melt$variable<-factor(newdata.vst.Col.yucQ.melt$variable, levels=c("1h","4h","25h","49h"))
conversion.table<-data.frame(num=1:27, genotype=c("Col","AT5G02540_1","hy5","jar1","kat1_2","phyB","pif45","spt_11","yuc2589","PAR1_RNAi09","coi1_16","phyAB","pif3","mida9_4","bsk5_1","sto","aos","argos","co_9","Blh_1","Jea","Shahdara","Col_0","Cvi_0","Bur_0","Oy_0","Ita_0"))
head(newdata.vst.Col.yucQ.melt)
for(i in 1:27){
  newdata.vst.Col.yucQ.melt[newdata.vst.Col.yucQ.melt$gt==conversion.table[i,1],"gt2"]<-as.character(conversion.table[i,2])
} 
head(newdata.vst.Col.yucQ.melt)
newdata.vst.Col.yucQ.melt$ssom3.4.unit.classif<-factor(newdata.vst.Col.yucQ.melt$ssom3.4.unit.classif,levels=c(10, 7,12,1,11,4,2,8,5,9,6,3)) # use this factor for all heatmap
names(newdata.vst.Col.yucQ.melt)[3]<-"ssom3.4.unit.classif"
p <- ggplot(newdata.vst.Col.yucQ.melt,aes(x=variable,y=value,color=value))
p <- p + geom_hline(yintercept=0,colour="red",size=1) # adding red line at fold change = 0 (060515)
p <- p + geom_point(position="jitter",size=1.5,alpha=0.3) 
p <- p + geom_boxplot(outlier.size=0,aes(fill=factor(ssom3.4.unit.classif),alpha=0.8)) + theme_bw() +xlab("sun/shade treatment (hr)") + scale_colour_gradient2(low="green1",mid="gray",high="magenta")
p <- p + facet_grid(gt2~ssom3.4.unit.classif,space="free") #+ facet_wrap(ssom3.4unit.classif~.)  ,scale="free"
p <- p + theme(axis.text.x=element_text(angle=90),strip.text.y = element_text(angle=0),legend.position="none")
p <- p + labs(x="Time (hr)", y="Log2 fold change (shade/sun)") #
p <- p + coord_flip()
p
ggsave("boxplot.node.genotype3.4.noscaling_vst_ColyucQ.pdf",width=11,height=5,unit="in",path="../../Nozue2016_SAStranscriptome_output/figs_tables")
```
### only Col (for explanation) for manuscript
```{r eval=FALSE, ggplot4, fig.width=11, fig.height=2}
#only Col (for lab meeting, 091114 version; vst version)
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","newdata.vst.PLoSGenetics2015mutants.errrDE.Rdata")) # 122214, 092617
newdata.vst.Col<-newdata.vst[newdata.vst$gt==1,]
newdata.vst.Col.melt<-melt(newdata.vst.Col[,c("1h","25h","4h","49h","ssom3.4.unit.classif","gt","AGI")],id.var=c("gt","AGI","ssom3.4.unit.classif"))
head(newdata.vst.Col.melt) 
newdata.vst.Col.melt$gt2<-vlookup(newdata.vst.Col.melt$gt,conversion.table,2)
head(newdata.vst.Col.melt)
newdata.vst.Col.melt$ssom3.4.unit.classif<-factor(newdata.vst.Col.melt$ssom3.4.unit.classif,levels=c(3,2,6,4,12,9,7,1,5,10,8,11))
newdata.vst.Col.melt$variable<-factor(newdata.vst.Col.melt$variable,levels=c("49h","25h","4h","1h")) # Julin's request for manuscript fig (092215)
names(newdata.vst.Col.melt)[3]<-"ssom3.4.unit.classif"
p <- ggplot(newdata.vst.Col.melt,aes(x=variable,y=value,color=value))
p <- p + geom_hline(yintercept=0,colour="red",size=1) # adding red line at fold change = 0 (060515)
p <- p + geom_point(position="jitter",size=1.5,alpha=0.3) 
p <- p + geom_boxplot(outlier.size=0,aes(fill=factor(ssom3.4.unit.classif),alpha=0.8)) + theme_bw() +xlab("sun/shade treatment (hr)") + scale_colour_gradient2(low="green1",mid="gray",high="magenta")
p <- p + facet_grid(gt2~ssom3.4.unit.classif,space="free") #+ facet_wrap(ssom3.4unit.classif~.)  ,scale="free"
p <- p + theme(axis.text.x=element_text(angle=90),strip.text.y = element_text(angle=0),legend.position="none")
p <- p + labs(x="Time (hr)", y="Log2 fold change (shade/sun)") #
p <- p + coord_flip()
p 
ggsave("Fig2_Coexpression_clusters_of_shade-responsive_genes_PLoSGenetics2015mutants.errrDE_
100717.pdf",width=11,height=5,unit="in",path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables")) # needs to change order of cluster (112816). File name changed (10071716)

# To which clusters in mutants do genes in Col cluster 3 belong? (useful for looking transition?)
newdata.vst$gt2<-vlookup(newdata.vst$gt,conversion.table,2)
cluster3.Col.AGI<-newdata.vst[newdata.vst$gt==1 & newdata.vst$ssom3.4.unit.classif==3,"AGI"]
ftable(newdata.vst[newdata.vst$AGI %in% cluster3.Col.AGI,c("gt2","ssom3.4.unit.classif")])
```
# GOseq anaysis of SOM clusters (022215; 092617; 100717 with .PLoSGenetics2015mutants.errrDE)
```{r eval=FALSE}
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","newdata.vst.PLoSGenetics2015mutants.errrDE.Rdata")) #
newdata.vst.Col<-newdata.vst[newdata.vst$gt==1,]
clusters<-levels(as.factor(newdata.vst.Col$ssom3.4.unit.classif))
# modify GOseq.ORA if there is no enriched GO ("return" instead of "stop") (070716)
for(x in 1:12) {  
  temp.GOseq<-GOseq.ORA(newdata.vst.Col[newdata.vst.Col$ssom3.4.unit.classif==x,"AGI"])
  if(temp.GOseq=="no enriched GO") {next} else {
     temp.genes<-genes.in.enriched.category(enrich.result=temp.GOseq,gene.list=newdata.vst.Col[newdata.vst.Col$ssom3.4.unit.classif==x,"AGI"],category.table=Atgo.list)
    #save(temp.GOseq,temp.genes,file=paste("../../Nozue2016_SAStranscriptome_output/output/SOM.Col.cluster.",x,".GOseq.Rdata",sep=""))
  save(temp.GOseq,temp.genes,file=paste("../../Nozue2016_SAStranscriptome_output/output/SOM.Col.cluster.",x,".GOseq.PLoSGenetics2015mutants.errrDE.Rdata",sep=""))
  } 
}
```
# write summary of GOseq analsyis of SOM clusters (022215; 100617,100717)
```{r eval=FALSE}
#rm(list=ls())
SOM.Col.cluster.GOseq.files<-list.files(pattern="SOM.Col.cluster.",path=file.path("..","..","Nozue2016_SAStranscriptome_output","output"))
SOM.Col.cluster.GOseq.files<-grep(".GOseq.PLoSGenetics2015mutants.errrDE.Rdata",SOM.Col.cluster.GOseq.files,value=TRUE)
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output",SOM.Col.cluster.GOseq.files[1]))
temp.GOseq$clusters<-gsub("(SOM.Col.cluster.)([[:print:]]+)(.GOseq.PLoSGenetics2015mutants.errrDE.Rdata)","\\2",SOM.Col.cluster.GOseq.files[1])
temp<-temp.GOseq
for(i in SOM.Col.cluster.GOseq.files[-1]) {
  load(file.path("..","..","Nozue2016_SAStranscriptome_output","output",i))
  temp.GOseq$clusters<-gsub("(SOM.Col.cluster.)([[:print:]]+)(.GOseq.PLoSGenetics2015mutants.errrDE.Rdata)","\\2",i)
  temp<-rbind(temp,temp.GOseq)
}
SOM.Col.cluster.GOseq.summary<-temp
write.csv(SOM.Col.cluster.GOseq.summary,file=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","TableS3_GO_enrichment_analysis_of_co-expressed_SOM_clusters_of_shade-responsive_genes100717.csv")) # 091016,092617, 100617

```
# SOM analysis only with Col DE genes and non-scaling (112616)
```{r}
setwd(homedir)
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","DGEsummary.edgeR.response.Rdata"))
# select only Col DE genes
DGEsummary.edgeR.response$gt<-gsub("([[:print:]]+)(.)([[:digit:]]+)(.)([[:print:]]+)","\\5",rownames(DGEsummary.edgeR.response))
DGEsummary.edgeR.response.Col<-DGEsummary.edgeR.response[DGEsummary.edgeR.response$gt=="1",]
DGEsummary.edgeR.response.Col$AGI<-gsub("([[:print:]]+)(.)([[:digit:]]+)(.)([[:print:]]+)","\\1",rownames(DGEsummary.edgeR.response.Col))
# 
#save(summary.vst.response,file="../../Nozue2016_SAStranscriptome_output/output/summary.vst.response.Rdata")
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","summary.vst.response.Rdata")) # 
summary.vst.response
hist(summary.vst.response[,1]) # fold change
genotypes<-levels(as.factor(gsub("(.+)(_)(1|4|16|25|49)(hrA)","\\1",as.character(names(summary.vst.response)))))
DGE.summary.vst.response.Col<-summary.vst.response[rownames(summary.vst.response) %in% DGEsummary.edgeR.response.Col$AGI,]
DGE.summary.vst.response.Col<-log2(DGE.summary.vst.response.Col) # log2 transformation of fold change

i<-1
temp<-DGE.summary.vst.response.Col[,gsub("(.+)(_)(1|4|16|25|49)(hrA)","\\1",names(DGE.summary.vst.response.Col))==genotypes[i]]
rownames(temp)<-paste(rownames(temp),genotypes[i],sep=".")
colnames(temp)<-paste(gsub("(.+)(_)(1|4|16|25|49)(hrA)","\\3",colnames(temp)),"h",sep="") ### fix (070914)
summary(temp)
dim(temp) #[1] 80 4
head(temp)
data5<-temp

# under construction #
#### no scaling
## SOM
set.seed(2) # Set a random seed so that SOM results are reproducible
#set.seed(123) # I do not like to have cluster 5 mixed with constitutive and 4h peak in set.seed(2) 3x4 version
ssom1.7.noscaling <- som(as.matrix(data5), somgrid(1,7,"hexagonal")) 
plot(ssom1.7.noscaling, type = "codes") #
ssom2.4.noscaling <- som(as.matrix(data5), somgrid(2,4,"hexagonal")) 
plot(ssom2.4.noscaling, type = "codes") #
ssom3.3.noscaling <- som(as.matrix(data5), somgrid(3,3,"hexagonal")) 
plot(ssom3.3.noscaling,type="codes")
ssom2.5.noscaling <- som(as.matrix(data5), somgrid(2,5,"hexagonal")) 
plot(ssom2.5.noscaling, type = "codes") #
ssom3.4.noscaling <- som(as.matrix(data5), somgrid(3,4,"hexagonal")) 
plot(ssom3.4.noscaling) 
ssom4.4.noscaling <- som(as.matrix(data5), somgrid(4,4,"hexagonal")) 
plot(ssom4.4.noscaling) 
# Selecting ssom1.7 because of (1) sufficient variation among clusters and (2) minimum redundancy among clusters judted from those plot 

## Create and write-out master SOM file
SOM.onlyCol.112616<-cbind(data5,ssom1.7.noscaling$unit.classif,ssom1.7.noscaling$distances)

names(SOM.onlyCol.112616)[5]<-"ssom1.7.unit.classif"
names(SOM.onlyCol.112616)[6]<-"ssom1.7.unit.distances"
SOM.onlyCol.112616$gt<-as.factor(gsub("(AT[[:alnum:]]+)(\\.)([[:digit:]]+)","\\3",rownames(SOM.onlyCol.112616)))
SOM.onlyCol.112616$AGI<-as.factor(gsub("(AT[[:alnum:]]+)(\\.)([[:digit:]]+)","\\1",rownames(SOM.onlyCol.112616)))
head(SOM.onlyCol.112616)
## ATHB2
SOM.onlyCol.112616[rownames(SOM.onlyCol.112616)=="AT4G16780.1",]
save(SOM.onlyCol.112616,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","SOM.onlyCol.112616.Rdata"))
#### making complete table (for SOM data) ###
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","SOM.onlyCol.112616.Rdata") )

# replace gt num into gt name
### convert genotype number into actual genotype name
#SOM.onlyCol.112616$gt2<-""
SOM.onlyCol.112616$gt2<-vlookup(SOM.onlyCol.112616$gt,conversion.table,2)

# adding gene description
SOM.onlyCol.112616.clusters.description<-merge(SOM.onlyCol.112616[SOM.onlyCol.112616$gt==levels(SOM.onlyCol.112616$gt)[1],],TAIR10_gene_descriptions[!duplicated(TAIR10_gene_descriptions$AGI),c("AGI","Short_description","colE...col.F")],by="AGI")
write.csv(SOM.onlyCol.112616.clusters.description,file=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","TableS3.SOM.onlyCol.nonscaling.112616.clusters.description.csv")) #090116
# graph
SOM.onlyCol.112616.clusters.description<-read.csv("../../Nozue2016_SAStranscriptome_output/figs_tables/TableS3.SOM.onlyCol.112616.clusters.description.csv",check.names=FALSE) # "check.names=F" to avoid "X1h" etc.
SOM.onlyCol.112616.clusters.description<-SOM.onlyCol.112616.clusters.description[,-1]
names(SOM.onlyCol.112616.clusters.description) 
SOM.onlyCol.112616.melt<-melt(SOM.onlyCol.112616.clusters.description[,c("1h","25h","4h","49h","ssom1.7.unit.classif","gt2","AGI")],id.var=c("gt2","AGI","ssom1.7.unit.classif"))
head(SOM.onlyCol.112616.melt)
# SOM.onlyCol.112616.melt$ssom1.7.unit.classif<-factor(SOM.onlyCol.112616.melt$ssom1.7.unit.classif,levels=c(****)) # use this factor for all heatmap. *** think order of levels after looking expression pattern

SOM.onlyCol.112616.melt$variable<-factor(SOM.onlyCol.112616.melt$variable,levels=c("49h","25h","4h","1h")) # Julin's request for manuscript fig (092215)
names(SOM.onlyCol.112616.melt)[3]<-"ssom1.7.unit.classif"
p <- ggplot(SOM.onlyCol.112616.melt,aes(x=variable,y=value,color=value))
p <- p + geom_hline(yintercept=0,colour="red",size=1) # adding red line at fold change = 0 (060515)
p <- p + geom_point(position="jitter",size=1.5,alpha=0.3) 
p <- p + geom_boxplot(outlier.size=0,aes(fill=factor(ssom1.7.unit.classif),alpha=0.8)) + theme_bw() +xlab("sun/shade treatment (hr)") + scale_colour_gradient2(low="green1",mid="gray",high="magenta")
p <- p + facet_grid(gt2~ssom1.7.unit.classif,space="free") #+ facet_wrap(ssom3.4unit.classif~.)  ,scale="free"
p <- p + theme(axis.text.x=element_text(angle=90),strip.text.y = element_text(angle=0),legend.position="none")
p <- p + labs(x="Time (hr)", y="Log2 fold change (shade/sun)") #
p <- p + coord_flip()
p 
ggsave("Fig2b_Coexpression_clusters_of_shade-responsive_genes_onlyCol_nonscaling_092617.pdf",width=11,height=5,unit="in",path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables") )# only Col DE genes
# Note: Col DE genes are mostly auxin related genes. Surprisingly JA related genes which were downregulated were not detected. 
```
## differentially expressed gene analysis (092617)
2a.DE at any time point (sun)
```{r eval=FALSE}
#system("mkdir ../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/genotype_specific5") # one time
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","samples.PLoSGenetics2015mutants.errrDE.Rdata")) # 100717
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","dge.PLoSGenetics2015mutants.errrDE.Rdata"))

SAS.expression.vst<-read.csv(file.path("..","..","Nozue2016_SAStranscriptome_data","input","kazu.SAS.expression.VST.092517.csv"),row.names=1) # new transformed data by DEseq2 (092617) # not found? (092817)
dim(SAS.expression.vst) #[1] 30533   287
Genotype<-levels(as.factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|25|49)(A.merged.bam)","\\2",names(SAS.expression.vst)))) # 13 genotype

for(i in Genotype[-1]) {# Genotype 17, 3,4,5,6,7,9 exceed 2000 genes. List up 5000 genes
  samples.nolow5.s<-samples.PLoSGenetics2015mutants.errrDE[samples.PLoSGenetics2015mutants.errrDE$genotype=="1"|samples.PLoSGenetics2015mutants.errrDE$genotype==i,] # 100717
  samples.nolow5.s<-samples.nolow5.s[samples.nolow5.s$trt=="H",] # only "sun" treatment
  samples.nolow5.s$genotype<-as.character(samples.nolow5.s$genotype)
  dge.nolow.s<-dge.PLoSGenetics2015mutants.errrDE[,colnames(dge.PLoSGenetics2015mutants.errrDE) %in% samples.nolow5.s$file]
  # eliminating low expressed genes in dge.nolow.s
  dge.nolow.s<-dge.nolow.s[rowSums(dge.nolow.s$counts > 5) >= 3,]
   # recalculate normalization (100717)
  dge.nolow.s<-calcNormFactors(dge.nolow.s)
  # design model 
  design.nolow.s.gt.batch3<- model.matrix(~samples.nolow5.s$genotype*samples.nolow5.s$time  + samples.nolow5.s$batch) # new model to detect genes with gt effects (100314)  
  colnames(design.nolow.s.gt.batch3) <- gsub("samples.nolow5.s$","",colnames(design.nolow.s.gt.batch3),fixed=TRUE) # nicer column names
  colnames(design.nolow.s.gt.batch3)
  print(design.nolow.s.gt.batch3)
  dge.nolow.s.batch3.glm <- estimateGLMCommonDisp(dge.nolow.s,design.nolow.s.gt.batch3, verbose=TRUE)
  dge.nolow.s.batch3.glm <- estimateGLMTrendedDisp(dge.nolow.s.batch3.glm,design.nolow.s.gt.batch3, verbose=TRUE)
  dge.nolow.s.batch3.glm<-estimateGLMTagwiseDisp(dge.nolow.s.batch3.glm,design.nolow.s.gt.batch3)
  #pdf(file=paste("BCV_MDS_with_genotype.wo16h_gt",i,".pdf",sep=""),width=11,height=8)
  #par(mfrow=c(1,2))
  #plotBCV(dge.nolow.s.batch3.glm,main=paste("genotype",i))
  #plotMDS(dge.nolow.s.batch3.glm,main=paste("genotype",i))
  #dev.off()  
  dge.nolow.s.batch3.glm.fit <- glmFit(dge.nolow.s.batch3.glm,design.nolow.s.gt.batch3)
  genotype.alltime.coef<-c(paste("genotype",i,sep=""),paste("genotype",i,":time4",sep=""),paste("genotype",i,":time25",sep=""),paste("genotype",i,":time49",sep=""))  
  dge.nolow.s.batch3.glm.lrt <- glmLRT(dge.nolow.s.batch3.glm.fit,coef=genotype.alltime.coef)
  DE.batch3.gt<-topTags(dge.nolow.s.batch3.glm.lrt,n=10000)$table[topTags(dge.nolow.s.batch3.glm.lrt,n=10000)$table$FDR<0.001,] # 0.05 for genotype_specific2 which gave me 5000 genes in hy5 (genotype3)
  dim(DE.batch3.gt) #
  save(DE.batch3.gt,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h","genotype_specific5",paste("DE.genotype.batch.sun.gt.wo16h.PLoSGenetics2015mutants.errrDE",i,".Rdata",sep=""))) 
  # save  dge.nolow.s.batch3.glm.lrt
    save(dge.nolow.s.batch3.glm.lrt,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h","genotype_specific5",paste("dge.nolow.s.batch3.glm.lrt.sun.PLoSGenetics2015mutants.errrDE",i,".Rdata",sep=""))) # added this line (try to use this object for coef, 061715)
}
```
2b.DE at any time point (shade)
```{r eval=FALSE}
for(i in Genotype[-1]) {# Genotype 17, 3,4,5,6,7,9 exceed 2000 genes. List up 5000 genes
  samples.nolow5.s<-samples.PLoSGenetics2015mutants.errrDE[samples.PLoSGenetics2015mutants.errrDE$genotype=="1"|samples.PLoSGenetics2015mutants.errrDE$genotype==i,] # 100717
  samples.nolow5.s<-samples.nolow5.s[samples.nolow5.s$trt=="L",] # only "shade" treatment
  samples.nolow5.s$genotype<-as.character(samples.nolow5.s$genotype)
  dge.nolow.s<-dge.PLoSGenetics2015mutants.errrDE[,colnames(dge.PLoSGenetics2015mutants.errrDE) %in% samples.nolow5.s$file] # 100717
  # eliminating low expressed genes in dge.nolow.s
  dge.nolow.s<-dge.nolow.s[rowSums(dge.nolow.s$counts > 5) >= 3,]
    # recalculate normalization (100717)
  dge.nolow.s<-calcNormFactors(dge.nolow.s)
  # design model 
  design.nolow.s.gt.batch3<- model.matrix(~samples.nolow5.s$genotype*samples.nolow5.s$time  + samples.nolow5.s$batch) # new model to detect genes with gt effects (100314)  
  colnames(design.nolow.s.gt.batch3) <- gsub("samples.nolow5.s$","",colnames(design.nolow.s.gt.batch3),fixed=TRUE) # nicer column names
  colnames(design.nolow.s.gt.batch3)
  print(design.nolow.s.gt.batch3)
  dge.nolow.s.batch3.glm <- estimateGLMCommonDisp(dge.nolow.s,design.nolow.s.gt.batch3, verbose=TRUE)
  dge.nolow.s.batch3.glm <- estimateGLMTrendedDisp(dge.nolow.s.batch3.glm,design.nolow.s.gt.batch3, verbose=TRUE)
  dge.nolow.s.batch3.glm<-estimateGLMTagwiseDisp(dge.nolow.s.batch3.glm,design.nolow.s.gt.batch3)
  #pdf(file=paste("BCV_MDS_with_genotype.wo16h_gt",i,".pdf",sep=""),width=11,height=8)
  #par(mfrow=c(1,2))
  #plotBCV(dge.nolow.s.batch3.glm,main=paste("genotype",i))
  #plotMDS(dge.nolow.s.batch3.glm,main=paste("genotype",i))
  #dev.off()  
  dge.nolow.s.batch3.glm.fit <- glmFit(dge.nolow.s.batch3.glm,design.nolow.s.gt.batch3)
  genotype.alltime.coef<-c(paste("genotype",i,sep=""),paste("genotype",i,":time4",sep=""),paste("genotype",i,":time25",sep=""),paste("genotype",i,":time49",sep=""))  
  dge.nolow.s.batch3.glm.lrt <- glmLRT(dge.nolow.s.batch3.glm.fit,coef=genotype.alltime.coef)
  
  DE.batch3.gt<-topTags(dge.nolow.s.batch3.glm.lrt,n=10000)$table[topTags(dge.nolow.s.batch3.glm.lrt,n=10000)$table$FDR<0.001,] # 0.05 for genotype_specific2 which gave me 5000 genes in hy5 (genotype3)
  dim(DE.batch3.gt) #
  save(DE.batch3.gt,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h","genotype_specific5",paste("DE.genotype.batch.shade.gt.wo16h.PLoSGenetics2015mutants.errrDE",i,".Rdata",sep="")))  
  save(dge.nolow.s.batch3.glm.lrt,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h","genotype_specific5",paste("dge.nolow.s.batch3.glm.lrt.shade.PLoSGenetics2015mutants.errrDE",i,".Rdata",sep=""))) # 100717
}
```
## overlap between Col modules and mutation affected genes
### summerize mutation affected genes
```{r eval=FALSE}
conversion.table # has to be made in function_
# sun
DE.genotype.batch.sun.files.kazu<-list.files(pattern="DE.genotype.batch.sun.gt.wo16h.PLoSGenetics2015mutants.errrDE",path="../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/genotype_specific5/") # needs to select "sun" object (060615)
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h","genotype_specific5",DE.genotype.batch.sun.files.kazu[1]))

DE.genotype.batch.sun.genes<-rownames(DE.batch3.gt)
for(xx in DE.genotype.batch.sun.files.kazu[-1]) {
  load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h","genotype_specific5",xx))
  DE.genotype.batch.sun.genes<-c(DE.genotype.batch.sun.genes,rownames(DE.batch3.gt))
}
# shade
DE.genotype.batch.shade.files.kazu<-list.files(pattern="DE.genotype.batch.shade.gt.wo16h.PLoSGenetics2015mutants.errrDE",path="../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/genotype_specific5/")
## load the first file
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h","genotype_specific5",DE.genotype.batch.shade.files.kazu[1]))
DE.genotype.batch.shade.genes<-rownames(DE.batch3.gt)
## load following files
for(xx in DE.genotype.batch.shade.files.kazu[-1]) {
  load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h","genotype_specific5",xx))
  DE.genotype.batch.shade.genes<-c(DE.genotype.batch.shade.genes,rownames(DE.batch3.gt))
}
# get unique gene list for making data frame
DE.genotype.batch.genes.unique<-unique(c(DE.genotype.batch.sun.genes,DE.genotype.batch.shade.genes))
length(DE.genotype.batch.genes.unique) #3356 > 3349
# convert to data frame version (092617)
DE.genotype.batch.genes.unique.DF<-data.frame(AGI=DE.genotype.batch.genes.unique)
# sun
for(x in DE.genotype.batch.sun.files.kazu) {
    load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h","genotype_specific5",x))
  temp<-as.integer(DE.genotype.batch.genes.unique %in% rownames(DE.batch3.gt)) 
  temp[temp==1]<-rep(paste(conversion.table[conversion.table==gsub("(DE.genotype.batch.sun.gt.wo16h.PLoSGenetics2015mutants.errrDE)([[:digit:]]+)(.Rdata)","\\2",x),2],"sun",sep="."),sum(temp))
  DE.genotype.batch.genes.unique.DF<-cbind(DE.genotype.batch.genes.unique.DF,temp)          
}
# shade
for(x in DE.genotype.batch.shade.files.kazu) {
  load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h","genotype_specific5",x))
  temp<-as.integer(DE.genotype.batch.genes.unique %in% rownames(DE.batch3.gt)) 
  temp[temp==1]<-rep(paste(conversion.table[conversion.table==gsub("(DE.genotype.batch.shade.gt.wo16h.PLoSGenetics2015mutants.errrDE)([[:digit:]]+)(.Rdata)","\\2",x),2],"shade",sep="."),sum(temp))
  DE.genotype.batch.genes.unique.DF<-cbind(DE.genotype.batch.genes.unique.DF,temp)                 
}
head(DE.genotype.batch.genes.unique.DF)
dim(DE.genotype.batch.genes.unique.DF)
# ## 
names(DE.genotype.batch.genes.unique.DF)
# 091016, 092617
DE.genotype.sun<-gsub("(DE.genotype.batch.sun.gt.wo16h.PLoSGenetics2015mutants.errrDE)([[:digit:]]+)(.Rdata)","\\2",DE.genotype.batch.sun.files.kazu) # 100717
DE.genotype.shade<-gsub("(DE.genotype.batch.shade.gt.wo16h.PLoSGenetics2015mutants.errrDE)([[:digit:]]+)(.Rdata)","\\2",DE.genotype.batch.shade.files.kazu) # 100717
names(DE.genotype.batch.genes.unique.DF)[-1]<-c(paste(vlookup(DE.genotype.sun,conversion.table,2),"sun",sep="."),paste(vlookup(DE.genotype.shade,conversion.table,2),"shade",sep=".")) # the first column is "AGI"
names(DE.genotype.batch.genes.unique.DF)
# the end of 092617 version

# remove "0"?
# remove splicing variant info in AGI
#DE.genotype.batch.genes.unique.DF[DE.genotype.batch.genes.unique.DF==0]<-""
DE.genotype.batch.genes.unique.DF$AGI<-gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",DE.genotype.batch.genes.unique.DF$AGI)
dim(DE.genotype.batch.genes.unique.DF) #3360 25 (100717)
save(DE.genotype.batch.genes.unique.DF,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","DE.genotype.batch.genes.unique.DF.Rdata")) 
```

# SOM clusters and mutant affected gfenes summary table for making overlapTable
```{r error=TRUE}
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","DE.genotype.batch.genes.unique.DF.Rdata"))
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","newdata.vst.PLoSGenetics2015mutants.errrDE.Rdata")) # 122214, 092617
newdata.vst.Col<-newdata.vst[newdata.vst$gt==1,]
# merge with SOM clusters
test<-merge(DE.genotype.batch.genes.unique.DF,newdata.vst.Col[,c("AGI","ssom3.4.unit.classif")],by="AGI",all=TRUE)
# making this with all AGI genes? (092717)
bias.DF<-data.frame(AGI=names(bias))
test2<-merge(bias.DF,test,by="AGI",all=TRUE)
## the end of adding bias (all genes used in GOseq)
DE.genotype.batch.genes.SOMclusters<-test2
DE.genotype.batch.genes.SOMclusters[is.na(DE.genotype.batch.genes.SOMclusters)]<-"0"
# adding gene description
head(TAIR10_gene_descriptions)
TAIR10_gene_descriptions$AGI<-gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",TAIR10_gene_descriptions$"Model_name")
DE.genotype.batch.genes.SOMclusters.description<-merge(DE.genotype.batch.genes.SOMclusters,TAIR10_gene_descriptions[!duplicated(TAIR10_gene_descriptions$AGI),c("AGI","Short_description","colE...col.F")],by="AGI")
write.csv(DE.genotype.batch.genes.SOMclusters.description,"../../Nozue2016_SAStranscriptome_output/figs_tables/Dataset3_Complete_list_of_shade-responsive_genes_and_SOM_gene_clusters_and_misexpressed_genes100717.csv")
```
# SOM vs mutant affected genes; making overlapTable
```{r}
test<-DE.genotype.batch.genes.SOMclusters
# overlap table has to be against all genes? I used all genes listed in DE.genotype.batch.genes.SOMclusters (071716)
for(i in 2:(dim(test)[2]-1)) { # the last column is "ssom3.4.unit.classif"
overlap.GOseq<-GOseq.overlapTable(test[,"ssom3.4.unit.classif"],test[,i],genes=test$AGI)
overlap.GOseq.melt<-melt(overlap.GOseq$pTable,id=1) #
overlap.GOseq.melt$count<-melt(overlap.GOseq$countTable,id=1)$value
overlap.GOseq.melt<-overlap.GOseq.melt[!overlap.GOseq.melt$Var2=="0",]
#overlap.GOseq.melt.summary<-rbind(overlap.GOseq.melt.summary,overlap.GOseq.melt)
save(overlap.GOseq.melt,file=file.path(homedir,"..","..","Nozue2016_SAStranscriptome_output","output",paste("overlap.GOseq.melt.SOM.mut.all.gene.PLoSGenetics2015mutants.errrDE.",i,".Rdata",sep="")))
}
```
# load all overlap.GOseq.melt files and combine them
```{r}
# 
#overlap.GOseq.melt.SOM.mut.files<-list.files(path=file.path("..","..","Nozue2016_SAStranscriptome_output","output"),pattern="overlap.GOseq.melt.SOM.mut.") # for 3000 genes as background (not preferable, 092717)
overlap.GOseq.melt.SOM.mut.files<-list.files(path=file.path("..","..","Nozue2016_SAStranscriptome_output","output"),pattern="overlap.GOseq.melt.SOM.mut.all.gene.PLoSGenetics2015mutants.errrDE") # for all gene as background (92717), .PLoSGenetics2015mutants.errrDE (100717)
overlap.GOseq.melt.summary.list2<-sapply(file.path("..","..","Nozue2016_SAStranscriptome_output","output",overlap.GOseq.melt.SOM.mut.files),function(x) mget(load(x))) # mget will return the value of the object(or objects) in a list. see https://stackoverflow.com/questions/29398630/load-data-frames-into-list
names(overlap.GOseq.melt.summary.list2)
overlap.GOseq.melt.summary<-do.call("rbind",overlap.GOseq.melt.summary.list2) 
head(overlap.GOseq.melt.summary)
rownames(overlap.GOseq.melt.summary)<-1:nrow(overlap.GOseq.melt.summary)
# format overlap.GOseq.melt.summary
overlap.GOseq.melt.summary[,"value"]<--log10(overlap.GOseq.melt.summary[,"value"])
overlap.GOseq.melt.summary[overlap.GOseq.melt.summary$value>10,"value"]<-10
overlap.GOseq.melt.summary[overlap.GOseq.melt.summary$count==0,"count"]<-""
#save(overlap.GOseq.melt.summary, file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","overlap.GOseq.melt.summary.SOM.DEgenes.092817.Rdata") )
save(overlap.GOseq.melt.summary, file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","overlap.GOseq.melt.summary.SOM.DEgenes.100717.Rdata") )

```
# drawing overlapTable heatmap (SOM vs mutant affected genes) (092717; 100717)
```{r}
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","overlap.GOseq.melt.summary.SOM.DEgenes.100717.Rdata"))
overlap.GOseq.melt.summary.s<-overlap.GOseq.melt.summary[!is.na(overlap.GOseq.melt.summary$Var2),]
overlap.GOseq.melt.summary.ss<-overlap.GOseq.melt.summary.s[!overlap.GOseq.melt.summary.s$Var1==0,]
overlap.GOseq.melt.summary.ss<-overlap.GOseq.melt.summary.ss[!overlap.GOseq.melt.summary.ss$Var2==0,]
# consistent order of SOM cluster (see Fig2)
#overlap.GOseq.melt.summary.ss$Var1<-factor(overlap.GOseq.melt.summary.ss$Var1,levels=c(3,2,12,6,4,9,7,1,5,10,8,11))
overlap.GOseq.melt.summary.ss$Var1<-factor(overlap.GOseq.melt.summary.ss$Var1,levels=c(3,2,6,4,12,9,7,1,5,10,8,11))
# 100717
overlap.GOseq.melt.summary.ss$Var2<-factor(overlap.GOseq.melt.summary.ss$Var2,levels=paste(rep(conversion.table[-1,"genotype"],2),rep(c("sun","shade"),each=12),sep=".")) 
# ggplot2
q4b<-ggplot(overlap.GOseq.melt.summary.ss,aes(y=factor(Var2),x=factor(Var1))) 
#q4b <- q4b + geom_tile(color="black",aes(fill=value))+scale_fill_gradient2(low=muted("green"), high=muted("magenta"),midpoint=2)  #+ ,guide = guide_legend(title="-log10\n p-value")) 
#q4b <- q4b + geom_tile(color="black",aes(fill=value))+scale_fill_gradient(low="white", high="magenta") 
q4b <- q4b + geom_tile(color="black",aes(fill=value))+scale_fill_gradient2(low="white", mid="white",high="magenta",midpoint=2)
#q4 <- q4 + theme(axis.text.x=element_text(angle=90),axis.title=element_blank())
q4b <- q4b + theme(axis.text.x=element_text(angle=90,hjust=1),
                 panel.grid.major = element_blank(),
                 panel.grid.minor = element_blank(),
                 panel.background = element_blank(),
                 axis.ticks = element_line(size=0,color="white"))
q4b <- q4b + geom_text(aes(label=count),size=5)
q4b <- q4b + labs(y="genotype specific genes",x="SOM clusters in Col",title="SOM clusters affected in SAS mutants",fill="-log10\n p-value") # chagne legend title (do no use gude_legend in scale_fill_gradient2)
ggsave(q4b,file="Fig4_Shade_avoidance_mutants_misexpress_specific_shade-response_SOM_cluster_100717.png",width=13,height=6,path="../../Nozue2016_SAStranscriptome_output/figs_tables") 
ggsave(q4b,file="Fig4_Shade_avoidance_mutants_misexpress_specific_shade-response_SOM_cluster_100717.pdf",width=13,height=6,path="../../Nozue2016_SAStranscriptome_output/figs_tables") 
q4b 
```
# Overlap analysis with custome categories (SOM)
## making summary table/dataset (100717)
```{r eval=FALSE}
#options(stringsAsFactors = FALSE) # for WGCNA
conversion.table
# custom categories
load(file.path("..","..","Nozue2016_SAStranscriptome_data","input","hormone.responsiveness6.Rdata")) #corrected version. (113016). file.path (092717)
## change list into data.frame
attributes(hormone.responsiveness6)
hormone.responsiveness6.DF<-data.frame(t(sapply(hormone.responsiveness6,c)))
names(hormone.responsiveness6.DF)<-attributes(hormone.responsiveness6)$category
## convert into hormone.responsiveness3.DF (omit "PCC1", "GRF1_3")
hormone.responsiveness6.DF<-hormone.responsiveness6.DF[,!names(hormone.responsiveness6.DF) %in% c("GRF1_3target","PCC1up","PCC1down")]
summary(hormone.responsiveness6.DF)
hormone.responsiveness6.DF$AGI<-rownames(hormone.responsiveness6.DF)
# SOM clusters
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","newdata.vst.PLoSGenetics2015mutants.errrDE.Rdata")) # 092617, 100717

newdata.vst.Col<-newdata.vst[newdata.vst$gt==1,] 
# merge with SOM clusters
test<-merge(hormone.responsiveness6.DF,newdata.vst.Col[,c("AGI","ssom3.4.unit.classif")],by="AGI",all=TRUE)
hormone.responsiveness.SOMclusters<-test
hormone.responsiveness.SOMclusters[is.na(hormone.responsiveness.SOMclusters)]<-"0"
#write.csv(hormone.responsiveness.SOMclusters,file=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","TableS4_hormone.responsiveness.SOMclusters100717.csv"))
hormone.responsiveness.SOMclusters.description<-merge(hormone.responsiveness.SOMclusters,TAIR10_gene_descriptions[!duplicated(TAIR10_gene_descriptions$AGI),c("AGI","Short_description","colE...col.F")],by="AGI")
write.csv(hormone.responsiveness.SOMclusters.description,file="../../Nozue2016_SAStranscriptome_output/figs_tables//Dataset2_Custom_enrichment_analysis_of_co-expressed_SOM_clusters_of_shade-responsive_genes100717.csv") 
```
## overlapTable (SOM vs custom category)
```{r}
test<-read.csv(file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Dataset2_Custom_enrichment_analysis_of_co-expressed_SOM_clusters_of_shade-responsive_genes100717.csv"),row.names=1) # 
#test<-test[,-1]
# overlap table 
for(i in 2:30) { # 3:30
  overlap.GOseq<-GOseq.overlapTable(test[,"ssom3.4.unit.classif"],test[,i],genes=test$AGI)
  overlap.GOseq.melt<-melt(overlap.GOseq$pTable,id=1) #
  overlap.GOseq.melt$count<-melt(overlap.GOseq$countTable,id=1)$value
  overlap.GOseq.melt<-overlap.GOseq.melt[!overlap.GOseq.melt$Var2=="0",]
  #overlap.melt.summary<-rbind(overlap.melt.summary,overlap.melt)
  save(overlap.GOseq.melt,file=file.path(homedir,"..","..","Nozue2016_SAStranscriptome_output","output",paste("overlap.GOseq.melt.SOM.customcat.PLoSGenetics2015mutants.errrDE.",i,".Rdata",sep="")))
}
```
##  load all overlap.GOseq.melt files and combine them (overlap.GOseq.melt.SOM.customcat)
```{r}
overlap.GOseq.melt.SOM.customcat.files<-list.files(path=file.path("..","..","Nozue2016_SAStranscriptome_output","output"),pattern="overlap.GOseq.melt.SOM.customcat.PLoSGenetics2015mutants.errrDE")

#overlap.GOseq.melt.summary.list<-sapply(file.path("..","..","Nozue2016_SAStranscriptome_output","output",overlap.GOseq.melt.SOM.mut.files),load)
# load(file.path("..","..","Nozue2016_SAStranscriptome_output","output",overlap.GOseq.melt.SOM.mut.files[1])) # works for one RData file
overlap.GOseq.melt.summary.list2<-sapply(file.path("..","..","Nozue2016_SAStranscriptome_output","output",overlap.GOseq.melt.SOM.customcat.files),function(x) mget(load(x))) # mget will return the value of the object(or objects) in a list. see https://stackoverflow.com/questions/29398630/load-data-frames-into-list
names(overlap.GOseq.melt.summary.list2)
overlap.GOseq.melt.summary<-do.call("rbind",overlap.GOseq.melt.summary.list2) 
head(overlap.GOseq.melt.summary) # make sure those are SOM.customcat files
rownames(overlap.GOseq.melt.summary)<-1:nrow(overlap.GOseq.melt.summary)
# format overlap.GOseq.melt.summary
overlap.GOseq.melt.summary[,"value"]<--log10(overlap.GOseq.melt.summary[,"value"])
overlap.GOseq.melt.summary[overlap.GOseq.melt.summary$value>10,"value"]<-10
overlap.GOseq.melt.summary[overlap.GOseq.melt.summary$count==0,"count"]<-""
save(overlap.GOseq.melt.summary, file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","overlap.GOseq.melt.summary.SOM.customcat.100717.Rdata") )
```
## plot graph (SOM vs custome category)
```{r}
#load("../../Nozue2016_SAStranscriptome_output/output/overlap.melt.summary.SOM.custom.GOseq.113016.Rdata")
#load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","overlap.GOseq.melt.summary.SOM.customcat.092817.Rdata"))
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","overlap.GOseq.melt.summary.SOM.customcat.100717.Rdata"))
# graph
overlap.GOseq.melt.summary.s<-overlap.GOseq.melt.summary[!overlap.GOseq.melt.summary$Var1==0,]
#overlap.GOseq.melt.summary.s$Var1<-factor(overlap.GOseq.melt.summary.s$Var1,levels=c(3,2,12,6,4,9,7,1,5,10,8,11)) # 092717
overlap.GOseq.melt.summary.s$Var1<-factor(overlap.GOseq.melt.summary.s$Var1,levels=c(3,2,6,4,12,9,7,1,5,10,8,11))
# 100717
overlap.GOseq.melt.summary.s<-overlap.GOseq.melt.summary.s[!overlap.GOseq.melt.summary.s$Var2=="",]
overlap.GOseq.melt.summary.s<-overlap.GOseq.melt.summary.s[!overlap.GOseq.melt.summary.s$Var2=="GRF1_3target",] # # remove "GRF1_3target" (052616)
overlap.GOseq.melt.summary.s$Var2<-factor(overlap.GOseq.melt.summary.s$Var2,levels=c("IAAup","IAAdown","BLup","BLdown","MJup","MJdown","GAup","GAdown","ABAup","ABAdown","ACCup","ACCdown","CKup","CKdown","SAup_2h","SAdown_2h","SARplus","SARminus","SAup_24h","SAdown_24h","PIFtarget","MYC234up","MYC234down","NPR1up","NPR1down", "NPR1up_uponSA","NPR1down_uponSA" ,"WRKY33up","WRKY33down")) # remove "GRF1_3target" (052616)
# switching x and y axis
overlap.GOseq.melt.summary.s$Var2<-factor(overlap.GOseq.melt.summary.s$Var2,levels=rev(levels(overlap.GOseq.melt.summary.s$Var2)))                                    
q4b<-ggplot(overlap.GOseq.melt.summary.s,aes(y=factor(Var2),x=factor(Var1))) 
q4b <- q4b + geom_tile(color="black",aes(fill=value))+scale_fill_gradient2(low="white", mid="white",midpoint=2,high="magenta")
#q4 <- q4 + theme(axis.text.x=element_text(angle=90),axis.title=element_blank())
q4b <- q4b + theme(axis.text.x=element_text(angle=90,hjust=1),
                 panel.grid.major = element_blank(),
                 panel.grid.minor = element_blank(),
                 panel.background = element_blank(),
                 axis.ticks = element_line(size=0,color="white"))
q4b <- q4b + geom_text(aes(label=count),size=5) 
q4b <- q4b + labs(y="custom categories",x="SOM clusters in Col",title="custom category enrichment",fill="-log10\n p-value")
ggsave(q4b,file="Fig3_Col.SOM.clusters.vs.my.category.table.s.2.GOseq.100717.pdf",width=13,height=6,path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables"))
ggsave(q4b,file="Fig3_Col.SOM.clusters.vs.my.category.table.s.2.GOseq.100717.png",width=13,height=6,path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables")) # 
```
# custom category heatmap (Fig. 5)
# Making list of custom catetory genes enriched in shade responsive genes (TableS7 for Fig5)
```{r}
## start from this line ## 052616
# fold chagne (shade responsiveness)
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","summary.vst.response.Rdata") )
dim(summary.vst.response) #[1] 30533    52
head(summary.vst.response)
# find overlap between shade responsive (in Col) and hormone responsive
#test<-read.csv(file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","TableS4_hormone.responsiveness.SOMclusters100717.csv") ,row.names=1) # needs to use updated one (092917)
test<-read.csv(file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Dataset2_Custom_enrichment_analysis_of_co-expressed_SOM_clusters_of_shade-responsive_genes100717.csv") ,row.names=1) # needs to use updated one (092917)

# reading overlapTable
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","overlap.GOseq.melt.summary.SOM.customcat.100717.Rdata")) 
overlap.GOseq.melt.summary.s<-overlap.GOseq.melt.summary[!overlap.GOseq.melt.summary$Var1==0,]
overlap.GOseq.melt.summary.s$Var1<-factor(overlap.GOseq.melt.summary.s$Var1,levels=c(3,2,6,4,12,9,7,1,5,10,8,11))
# 100717
overlap.GOseq.melt.summary.s<-overlap.GOseq.melt.summary.s[!overlap.GOseq.melt.summary.s$Var2=="",]
overlap.GOseq.melt.summary.s<-overlap.GOseq.melt.summary.s[!overlap.GOseq.melt.summary.s$Var2=="GRF1_3target",] # # remove "GRF1_3target" (052616)
overlap.GOseq.melt.summary.s<-overlap.GOseq.melt.summary.s[!overlap.GOseq.melt.summary.s$Var2=="PIFtarget",] # # remove "PIFtarget" (101017). No sense to discuss IAAup specific and PIFtarget specific genes.

overlap.GOseq.melt.summary.s$Var2<-factor(overlap.GOseq.melt.summary.s$Var2,levels=c("IAAup","IAAdown","BLup","BLdown","MJup","MJdown","GAup","GAdown","ABAup","ABAdown","ACCup","ACCdown","CKup","CKdown","SAup_2h","SAdown_2h","SARplus","SARminus","SAup_24h","SAdown_24h","PIFtarget","MYC234up","MYC234down","NPR1up","NPR1down", "NPR1up_uponSA","NPR1down_uponSA" ,"WRKY33up","WRKY33down")) # remove "GRF1_3target" (052616)

# switching x and y axis
overlap.GOseq.melt.summary.s$Var2<-factor(overlap.GOseq.melt.summary.s$Var2,levels=rev(levels(overlap.GOseq.melt.summary.s$Var2)))                                    
# select log10>2 & !Var1==0 
overlap.GOseq.melt.summary.s.significant<-overlap.GOseq.melt.summary.s[overlap.GOseq.melt.summary.s$value> -log10(0.05),] # pvalue<0.05
overlap.GOseq.melt.summary.s.significant$category.cluster<-with(overlap.GOseq.melt.summary.s.significant,paste(Var2,Var1,sep="_"))
test.wozero<-test[!test$ssom3.4.unit.classif=="0",1:31]
head(test.wozero)
test.wozero.melt<-melt(test.wozero,id.var=c("AGI","ssom3.4.unit.classif"))
head(test.wozero.melt)
test.wozero.melt$category.cluster<-with(test.wozero.melt,paste(variable,ssom3.4.unit.classif,sep="_"))
test.wozero.melt.hormone.significant<-test.wozero.melt[test.wozero.melt$category.cluster %in% overlap.GOseq.melt.summary.s.significant$category.cluster,]
head(test.wozero.melt.hormone.significant)
# # remove small group (less than 3)
# category.morethan3<-names(table(test.wozero.melt.hormone.significant$value)[table(test.wozero.melt.hormone.significant$value)>3])
#test.wozero.melt.hormone.significant<-test.wozero.melt.hormone.significant[test.wozero.melt.hormone.significant$value %in% category.morethan3,] # more than 3 genes
test.wozero.melt.hormone.significant$value<-as.character(test.wozero.melt.hormone.significant$value)
# test.wozero.melt.hormone.significant.s is my category marker gene list 
test.wozero.melt.hormone.significant.s<-test.wozero.melt.hormone.significant[,c("AGI","value")]
names(test.wozero.melt.hormone.significant.s)<-c("AGI","my.category")
test.wozero.melt.hormone.significant.s.summary<-table(test.wozero.melt.hormone.significant.s)[,-1]
table(test.wozero.melt.hormone.significant.s$my.category)

# use only hormone responsive genes and PIFtarget
# MYC234up is redundant with MJup
# for SA, use only SAup_24h which is redundant with SARplus, NPR1up_uponSA.
# NPR1up and down (npr1 vs Col with water control) are difficult to interpret (different from NPR1up/down_uponSA). So omit this time
# 
#test.wozero.melt.hormone.significant.ss<-test.wozero.melt.hormone.significant.s[!test.wozero.melt.hormone.significant.s$my.category %in% c("NPR1up","NPR1down","NPR1up_uponSA","SARplus","WRKY33up","MYC234up"),] # eliminate those categories
# how to look overlap between categories?
dim(test.wozero.melt.hormone.significant.s.summary) # 213 19
for(i in 2:18) {
  for(n in (i+1):19) {
category.overlapTable<-overlapTable(test.wozero.melt.hormone.significant.s.summary[,i],test.wozero.melt.hormone.significant.s.summary[,n])
category.overlapTable.pTable<-category.overlapTable$pTable
    save(category.overlapTable.pTable,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output",paste("category.overlapTable",colnames(test.wozero.melt.hormone.significant.s.summary)[i],colnames(test.wozero.melt.hormone.significant.s.summary)[n],"Rdata",sep=".")))
  }
}
# 
category.overlapTable.files<-list.files(pattern="category.overlapTable",path=file.path("..","..","Nozue2016_SAStranscriptome_output","output"))

category.overlapTable.files.list<-sapply(file.path("..","..","Nozue2016_SAStranscriptome_output","output",category.overlapTable.files),function(x) mget(load(x))) # mget 
names(category.overlapTable.files.list)<-gsub("(../../Nozue2016_SAStranscriptome_output/output/category.overlapTable.)([[:print:]]+)(.Rdata.category.overlap)","\\2",names(category.overlapTable.files.list))
#category.overlapTable.files.DF<-as.data.frame(category.overlapTable.files.list)
TF<-data.frame(comparison=rep("",length(names(category.overlapTable.files.list))),signif.overlap="")
TF[1,2]<-category.overlapTable.files.list[[1]][2,2]
TF[1,1]<-gsub("([[:print:]]+)(Table.pTable)","\\1",names(category.overlapTable.files.list)[1])

for(n in 2:length(names(category.overlapTable.files.list))) {
TF[n,2]<-category.overlapTable.files.list[[n]][2,2]
TF[n,1]<-gsub("([[:print:]]+)(Table.pTable)","\\1",names(category.overlapTable.files.list)[n])
}
TF[TF$signif.overlap<0.01,]
#                comparison       signif.overlap
# 51             BLup.IAAup  0.00689499648909803
# 88           IAAup.MJdown 0.000190847396048117
# 109         MJup.MYC234up 0.000206781784225688
# 136      NPR1down.SARplus 0.000123257188893522
# 138   NPR1down.WRKY33down  0.00131613250998435
# 140 NPR1up_uponSA.SARplus 0.000123257188893522
# 152    SARplus.WRKY33down  0.00196790148023975
##############
test.wozero.melt.hormone.significant.ss<-test.wozero.melt.hormone.significant.s[!test.wozero.melt.hormone.significant.s$my.category %in% c("BLup","MJdown","MYC234up","NPR1up","NPR1down","NPR1down_uponSA","NPR1up_uponSA","WRKY33up","WRKY33down","SAup_24h"),] # eliminate those categories

# remove my.category =="" (092917)
test.wozero.melt.hormone.significant.ss<-test.wozero.melt.hormone.significant.ss[!test.wozero.melt.hormone.significant.ss$my.category=="",]
table(test.wozero.melt.hormone.significant.ss$my.category)
 # ABAdown    ABAup    ACCup   BLdown     CKup  IAAdown    IAAup     MJup SARminus  SARplus 
 #       9        4        5        4        2        2       44       20        1        6 

#write.csv(test.wozero.melt.hormone.significant.ss,file=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","TableS6_test.wozero.melt.hormone.significant.ss.100717.csv")) # 
#### what genes do belong to multiple categories?
#reshape(test.wozero.melt.hormone.significant.s,idvar="AGI",timevar="my.category",direction="wide")
test.wozero.melt.hormone.significant.s.summary<-table(test.wozero.melt.hormone.significant.s)[,-1]
# specific to each category (fixed 101017)
test.wozero.melt.hormone.significant.s.summary.specific<-test.wozero.melt.hormone.significant.s.summary[rowSums(test.wozero.melt.hormone.significant.s.summary)==1,] # need to merge SA related categories?
test.wozero.melt.hormone.significant.s.summary.specific 
duplicated(rownames(test.wozero.melt.hormone.significant.s.summary.specific))
## 
reshape(test.wozero.melt.hormone.significant.ss,idvar="AGI",timevar="my.category",direction="wide") #  should be test.wozero.melt.hormone.significant.s.summary.specific
test.wozero.melt.hormone.significant.ss.summary<-table(test.wozero.melt.hormone.significant.s)[,-1]
test.wozero.melt.hormone.significant.ss.summary # table object

test.wozero.melt.hormone.significant.sss<-test.wozero.melt.hormone.significant.ss[test.wozero.melt.hormone.significant.ss$AGI %in% test.wozero.melt.hormone.significant.ss$AGI,] # 
test.wozero.melt.hormone.significant.sss<-test.wozero.melt.hormone.significant.ss[test.wozero.melt.hormone.significant.ss$AGI %in% rownames(test.wozero.melt.hormone.significant.s.summary.specific),] # revised. 101017
dim(test.wozero.melt.hormone.significant.sss) # 48 2
table(test.wozero.melt.hormone.significant.sss$my.category)
# ABAdown   ABAup   ACCup  BLdown    CKup IAAdown   IAAup    MJup SARplus 
#       3       1       2       2       1       1      25       9       2 
# needs to eliminate categories less than 3 genes
category.morethan3<-names(table(test.wozero.melt.hormone.significant.sss$my.category))[table(test.wozero.melt.hormone.significant.sss$my.category)>2]
test.wozero.melt.hormone.significant.ssss<-test.wozero.melt.hormone.significant.sss[test.wozero.melt.hormone.significant.sss$my.category %in% category.morethan3,]
dim(test.wozero.melt.hormone.significant.ssss) #37 2
table(test.wozero.melt.hormone.significant.sss$my.category)
table(test.wozero.melt.hormone.significant.ssss$my.category)
# ABAdown   IAAup    MJup 
#       3      25      9 
write.csv(test.wozero.melt.hormone.significant.ssss,file="../../Nozue2016_SAStranscriptome_output/figs_tables/TableS6_test.wozero.melt.hormone.significant.ssss.101017.csv") # This should be TableS6 (101017)
```
# visualize hormone pathway related genes (1213-1214,2015). Updated (052616-)
# new FDR version (071716)
# with new SOM clusters (111816)
# with new SOM clusters (092917)
# with new SOM clusters (100717)
```{r}
my.category.diff.heatmap.graph<-list()
#categories <- read.csv(file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","TableS7_test.wozero.melt.hormone.significant.sss.092917.csv"),row.names=1,as.is=TRUE) 
categories <- read.csv(file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","TableS6_test.wozero.melt.hormone.significant.ssss.101017.csv"),row.names=1,as.is=TRUE) #  re-normalize data after subsetting in edgeR DE analysis. specific to each category (101017)
summary(categories)
head(categories)
table(categories$my.category)
#categories$my.category<-factor(categories$my.category,levels=rev(c("IAAup","BLup","BLdown","PIFtarget","ACCup","ABAup","ABAdown","MJup","MJdown","SAup_24h"))) # 120116
categories$my.category<-factor(categories$my.category,levels=rev(c("IAAup","MJup","ABAdown"))) # 120116

# summary.vst.response.kazu
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","summary.vst.response.Rdata")) # response (092917)
summary.vst.response.kazu<-summary.vst.response
dim(summary.vst.response.kazu) # [1] 30533    52
head(summary.vst.response.kazu)
# for Fig5A, Col
my.category.heatmap.graph<-list()
for(i in c(1,17,19,3,4,5,14,10,6,13,7,8,9)) { # 
my.category.heatmap.graph[[i]]<-my.category.heatmap5(summary3.response=summary.vst.response.kazu,gt.num=i,gt2=conversion.table[conversion.table$num==i,2],newdata.Col.wGO.selected=categories,plotname="my.category.expression.pattern.092917v1.png",w=5,h=5*4/3, path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables"),save.plot=F,legend.fill=F) #  legend.fill=F to T for new Fig5 (all absolute value, previously called FigS2)
}
# summerize data in one data frame
FigS2.absolute.summary.table.melt.files<-list.files(path=file.path("..","..","Nozue2016_SAStranscriptome_output","output"),pattern="FigS2.absolute.summary.table.melt") # (101017)
FigS2.absolute.summary.table.melt.files<-grep(".Rdata",FigS2.absolute.summary.table.melt.files,value=T)

FigS2.absolute.summary.table.melt.list2<-sapply(file.path("..","..","Nozue2016_SAStranscriptome_output","output",FigS2.absolute.summary.table.melt.files),function(x) mget(load(x))) # mget will return the value of the object(or objects) in a list. see 
FigS2.absolute.summary.table.melt.DF<-do.call(cbind,FigS2.absolute.summary.table.melt.list2)
# remove redundant columns 
colnames(FigS2.absolute.summary.table.melt.DF)<-gsub("(../../Nozue2016_SAStranscriptome_output/output/FigS2.absolute.summary.table.melt)([[:digit:]]+)(.Rdata.summary.table.melt.)(my.category|variable|value)","\\2.\\4",colnames(FigS2.absolute.summary.table.melt.DF))
FigS2.absolute.summary.table.melt.DF<-FigS2.absolute.summary.table.melt.DF[,c("1.my.category","1.variable" ,grep("value",colnames(FigS2.absolute.summary.table.melt.DF),value=TRUE))]
write.csv(FigS2.absolute.summary.table.melt.DF,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","FigS2.absolute.summary.table.melt.DF.csv")) 

# calculate diff (using mean value; not used; 101017)
FigS2.absolute.summary.table.melt.DF.diff<-FigS2.absolute.summary.table.melt.DF[,-(1:3)]-FigS2.absolute.summary.table.melt.DF[,3]
## considering biological meaning of negative value in Col
FigS2.absolute.summary.table.melt.DF.diff[FigS2.absolute.summary.table.melt.DF[,3]<0,]<- -(FigS2.absolute.summary.table.melt.DF.diff[FigS2.absolute.summary.table.melt.DF[,3]<0,])
FigS2.absolute.summary.table.melt.DF.diff$my.category<-FigS2.absolute.summary.table.melt.DF$"1.my.category"
FigS2.absolute.summary.table.melt.DF.diff$time<-FigS2.absolute.summary.table.melt.DF$"1.variable"
colnames(FigS2.absolute.summary.table.melt.DF.diff)<-gsub(".value","",colnames(FigS2.absolute.summary.table.melt.DF.diff))
write.csv(FigS2.absolute.summary.table.melt.DF.diff,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","FigS2.absolute.summary.table.melt.DF.diff.csv")) # # I did not use this.
# drawing heatmap graph (difference): Fig5B 
### Is this difference correct? compare absolute data (100717)
### "my.category.diff.heatmap.graph" function calculate differences between Col and mut at gene by gene base.
### "FigS2.absolute.summary.table.melt.DF.diff.csv" calculate at category mean base.
# need to create diff data frame for (mut - Col)
# calculate max and min for heatmap color grade 
#summary.vst.response.kazu.diffCol<-as.data.frame(my.category.diff.data)  # under construction
max.value<-max(FigS2.absolute.summary.table.melt.DF.diff[,grep(".value",colnames(FigS2.absolute.summary.table.melt.DF.diff))])
min.value<-min(FigS2.absolute.summary.table.melt.DF.diff[,grep(".value",colnames(FigS2.absolute.summary.table.melt.DF.diff))])
max.min<-c(round(min.value,2),round(max.value,2))
# ## I decided not to use this version (10117)
# my.category.diff.heatmap.graph<-list()
# #  
# for(i in c(17,19,3,4,5,14,10,6,13,7,8,9)) { # c("aos","co_9","hy5","jar1","kat1_2","mida9_4","PAR1","phyb","pif3","pif45","spt_11","yucQ")
# my.category.diff.heatmap.graph[[i]]<-my.category.diff.heatmap1(summary3.response=summary.vst.response.kazu,gt.num=i,gt2=conversion.table[conversion.table$num==i,2],newdata.Col.wGO.selected=categories,plotname="my.category.expression.pattern.092917mutsv1.png",w=5,h=5*4/3, path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables"),save.plot=F,legend.fill=T,max.min.value=max.min) # make specialized for difference heatmap function
# } # this is not consistent with FigS2.absolute.summary.table.melt.DF.diff above (101017)
# summerize data in one data frame
Fig5B.diff.summary.table.melt.files<-list.files(path=file.path("..","..","Nozue2016_SAStranscriptome_output","output"),pattern="Fig5B.diff.summary.table.melt") # (101017)
Fig5B.diff.summary.table.melt.files<-grep(".Rdata",Fig5B.diff.summary.table.melt.files,value=TRUE)
Fig5B.diff.summary.table.melt.list2<-sapply(file.path("..","..","Nozue2016_SAStranscriptome_output","output",Fig5B.diff.summary.table.melt.files),function(x) mget(load(x))) # mget will return the value of the object(or objects) in a list. see 

Fig5B.diff.summary.table.melt.DF<-do.call(cbind,Fig5B.diff.summary.table.melt.list2)
# remove redundant columns 
colnames(Fig5B.diff.summary.table.melt.DF)<-gsub("(../../Nozue2016_SAStranscriptome_output/output/Fig5B.diff.summary.table.melt)([[:digit:]]+)(.Rdata.summary.table.melt.)(my.category|variable|value)","\\2.\\4",colnames(Fig5B.diff.summary.table.melt.DF))
Fig5B.diff.summary.table.melt.DF<-Fig5B.diff.summary.table.melt.DF[,c("10.my.category","10.variable" ,grep("value",colnames(Fig5B.diff.summary.table.melt.DF),value=TRUE))]
write.csv(Fig5B.diff.summary.table.melt.DF,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","Fig5B.diff.summary.table.melt.DF.csv")) # data used in Fig5B heatmap, which is quite different from FigS2.absolute.summary.table.melt.DF.diff.csv (calculated from absolute mean value). eg. phyB MJup. Why does this happen? 

# calculating mut - Col with individual gene adjusted biological meaning may not good.
# use subtraction of mean value 
my.category.diff.heatmap.graph.ver2<-list()
for(i in c(17,19,3,4,5,14,10,6,13,7,8,9)) { # c("aos","co_9","hy5","jar1","kat1_2","mida9_4","PAR1","phyb","pif3","pif45","spt_11","yucQ")
my.category.diff.heatmap.graph.ver2[[i]]<-
  my.category.diff.heatmap2(summary3.response.diff=FigS2.absolute.summary.table.melt.DF.diff,gt.num=i,gt2=conversion.table[conversion.table$num==i,2],newdata.Col.wGO.selected=categories,plotname="my.category.expression.pattern.092917mutsv1.png",w=5,h=5*4/3,path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables"),save.plot=F,legend.fill=T,max.min.value=max.min)
} # (FigS2.absolute.summary.table.melt.DF.diff (101117)

```

# significance calculation (needs to fix)
```{r}
# how to get summary.vst.response.kazu? (102617)
summary.vst.response.kazu$AGI <- row.names(summary.vst.response.kazu)
response.melt <- melt(summary.vst.response.kazu,id.vars= "AGI",variable.name = "sample", value.name = "log2.dif") # this value is not log transformed. This is just ratio (responsiveness to shade; 101117)
hist(response.melt$log2.dif)
response.melt$log2.dif<-log2(response.melt$log2.dif)
head(response.melt)
response.melt$gt <- as.factor(regmatches(response.melt$sample,regexpr("^[0-9]{1,2}",response.melt$sample)))
response.melt$time <- as.factor(regmatches(response.melt$sample,regexpr("[0-9]{1,2}hr",response.melt$sample)))
response.melt$rep <- as.factor(regmatches(response.melt$sample,regexpr("[A-Z]$",response.melt$sample)))
head(response.melt)
tail(response.melt)
manuscript.genotypes <- as.character(c(1,3,4,5,6,7,8,9,10,13,14,17,19) )
response.melt <- response.melt[response.melt$gt %in% manuscript.genotypes,]
response.melt <- droplevels(response.melt) # droplevels is useful.
summary(response.melt)
summary(categories)
table(categories$my.category)
# select genes in categories
response.melt.merge <- merge(response.melt,categories,by="AGI",all=FALSE)
response.melt.merge$gt <- relevel(response.melt.merge$gt,ref="1")
table(response.melt.merge$gt,response.melt.merge$time)
mutants <- levels(response.melt.merge$gt)[-1] #1 is Col 
category.names<- unique(response.melt.merge$my.category)
results <- lapply(mutants,function(gt) { 
  sapply(category.names,function(cat.name) {
    tmp <- response.melt.merge[(response.melt.merge$gt=="1" | response.melt.merge$gt==gt) & response.melt.merge$my.category==cat.name,]
    lmer1 <- lmer(log2.dif ~ gt*time + (1|AGI), data=tmp)
    tmp.results <- anova(lmer1)[, "Pr(>F)"]
    print(row.names(anova(lmer1)))
    names(tmp.results) <- row.names(anova(lmer1))
    min(tmp.results[-2]) # we do not care about time
    print(min(tmp.results[-2]))
  })
} )
names(results) <- mutants 
results
results.df <- ldply(results,.id="gt")
results.df
colnames(results.df)[-1]<-as.character(category.names)
results.df
```
# calclate adjusted p-value
```{r}
results.df.adj <- matrix(
  p.adjust(unlist(results.df[,-1]),"fdr",n=2*length(unlist(results.df[,-1]))), #multiply by 2 because we took the min value of gt and gt*time
  nrow=nrow(results.df))
results.df.adj <- cbind(as.character(results.df[,1]),as.data.frame(results.df.adj))
colnames(results.df.adj) <- colnames(results.df)
results.df.adj
#write.csv(results.df.adj,file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Fig5B_response_p_values.v3.csv"))
#write.csv(results.df.adj,file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Fig5B_response_p_values.v4.csv"))
write.csv(results.df.adj,file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Fig5B_response_p_values.v6.log2.csv")) #101017 three categories


```
# add p value in heatmap 
```{r}
response.p.values<-read.csv(file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Fig5B_response_p_values.v6.log2.csv"),row.names=1) # from modified Julin's script (copied from Kazu_test_v2.R) (100717)
#response.p.values<-read.csv(file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Fig5B_response_p_values.v5.csv"),row.names=1) # from modified Julin's script (copied from Kazu_test_v2.R) (100717)

response.p.values.melt<-melt(response.p.values,id="gt") #
my.categories<-grep("gt",colnames(response.p.values),invert=TRUE,value=TRUE)
##  my.category.diff.heatmap.graph (not used 101117)
# for(i in c(17,19,3,4,5,14,10,6,13,7,8,9)) {
# significant.position<-data.frame(category=my.categories,x=0,y=1:length(my.categories),symbol.gt="",symbol.gt.time="")
# response.p.values.melt.gt<-response.p.values.melt[response.p.values.melt$gt==i,]  
# significant.position$symbol.gt<-(as.numeric(as.character(vlookup(significant.position$category,response.p.values.melt.gt[,c(2,3)],2,F)))<0.01)*1 # significant level 
#   significant.position$symbol.gt<-gsub("1","*",significant.position$symbol.gt)
#   significant.position$symbol.gt<-gsub("0","",significant.position$symbol.gt)
# 
# # add significance to heatmap
# my.category.diff.heatmap.graph[[i]]<-my.category.diff.heatmap.graph[[i]] + geom_text(data=significant.position,aes(x=x-0.2,y=y,label=symbol.gt),size=20,color="red") #, remove lines for x and y axis. "gt" is red.
# rm(significant.position)
# }
# add significant to absolute value heatmap
for(i in c(17,19,3,4,5,14,10,6,13,7,8,9)) {
significant.position<-data.frame(category=my.categories,x=0,y=1:length(my.categories),symbol.gt="",symbol.gt.time="")
response.p.values.melt.gt<-response.p.values.melt[response.p.values.melt$gt==i,]  
significant.position$symbol.gt<-(as.numeric(as.character(vlookup(significant.position$category,response.p.values.melt.gt[,c(2,3)],2,F)))<0.01)*1 # significant level 
  significant.position$symbol.gt<-gsub("1","*",significant.position$symbol.gt)
  significant.position$symbol.gt<-gsub("0","",significant.position$symbol.gt)

# add significance to heatmap
my.category.heatmap.graph[[i]]<-my.category.heatmap.graph[[i]] + geom_text(data=significant.position,aes(x=x-0.2,y=y,label=symbol.gt),size=20,color="red") #, remove lines for x and y axis. "gt" is red.
rm(significant.position)
}

# 
# ### use geom_blank() for blank plot
# library(cowplot)
# my.category.heatmap.all <- plot_grid(my.category.heatmap.graph[[1]],ggplot()+geom_blank(),ggplot()+geom_blank(),my.category.diff.heatmap.graph[[17]],my.category.diff.heatmap.graph[[19]],my.category.diff.heatmap.graph[[3]],my.category.diff.heatmap.graph[[4]],my.category.diff.heatmap.graph[[5]],my.category.diff.heatmap.graph[[14]],my.category.diff.heatmap.graph[[10]],my.category.diff.heatmap.graph[[6]],my.category.diff.heatmap.graph[[13]],my.category.diff.heatmap.graph[[7]],my.category.diff.heatmap.graph[[8]],my.category.diff.heatmap.graph[[9]],
#                       labels=c("A","","","B","","","","","","","","","","",""),
#                       vjust = 1.1, hjust=-0.5,
#                       ncol=3,nrow=5,
#                       label_size = 50)
# #setwd("../../Nozue2016_SAStranscriptome_output/figs_tables/")
# save_plot(file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Fig5_all.my.category.diff.expression.pattern.v9.101017.FDR0.01.png"), my.category.heatmap.all,
#           ncol = 3, # we're saving a grid plot of 2 columns
#           nrow = 5, # and 2 rows
#           base_aspect_ratio = 2.5,
#           base_height=6,
#           limitsize = FALSE
# )
# Fig5B ver2: diff calculated from absolute mean value 
# for(i in c(17,19,3,4,5,14,10,6,13,7,8,9)) {
# significant.position<-data.frame(category=my.categories,x=0,y=1:length(my.categories),symbol.gt="",symbol.gt.time="")
# response.p.values.melt.gt<-response.p.values.melt[response.p.values.melt$gt==i,]  
# significant.position$symbol.gt<-(as.numeric(as.character(vlookup(significant.position$category,response.p.values.melt.gt[,c(2,3)],2,F)))<0.01)*1 # significant level 
#   significant.position$symbol.gt<-gsub("1","*",significant.position$symbol.gt)
#   significant.position$symbol.gt<-gsub("0","",significant.position$symbol.gt)
# 
# # add significance to heatmap
# my.category.diff.heatmap.graph.ver2[[i]]<-my.category.diff.heatmap.graph.ver2[[i]] + geom_text(data=significant.position,aes(x=x-0.2,y=y,label=symbol.gt),size=20,color="red") #, remove lines for x and y axis. "gt" is red.
# rm(significant.position)
# }
# library(cowplot)
# my.category.heatmap.all <- plot_grid(my.category.heatmap.graph[[1]],ggplot()+geom_blank(),ggplot()+geom_blank(),my.category.diff.heatmap.graph.ver2[[17]],my.category.diff.heatmap.graph.ver2[[19]],my.category.diff.heatmap.graph.ver2[[3]],my.category.diff.heatmap.graph.ver2[[4]],my.category.diff.heatmap.graph.ver2[[5]],my.category.diff.heatmap.graph.ver2[[14]],my.category.diff.heatmap.graph.ver2[[10]],my.category.diff.heatmap.graph.ver2[[6]],my.category.diff.heatmap.graph.ver2[[13]],my.category.diff.heatmap.graph.ver2[[7]],my.category.diff.heatmap.graph.ver2[[8]],my.category.diff.heatmap.graph.ver2[[9]],
#                       labels=c("A","","","B","","","","","","","","","","",""),
#                       vjust = 1.1, hjust=-0.5,
#                       ncol=3,nrow=5,
#                       label_size = 50)
# #setwd("../../Nozue2016_SAStranscriptome_output/figs_tables/")
# save_plot(file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Fig5_all.my.category.diff.expression.pattern.ver2.101117.FDR0.01.png"), my.category.heatmap.all,
#           ncol = 3, # we're saving a grid plot of 2 columns
#           nrow = 5, # and 2 rows
#           base_aspect_ratio = 2.5,
#           base_height=6,
#           limitsize = FALSE
# )



# supplement Fig2 (absolute value), which will become new Fig5 (101117)
# extract legend
# since Col with legend is different in size from others. I need to extract legend and plot it in different space.
# see http://stackoverflow.com/questions/12041042/how-to-plot-just-the-legends-in-ggplot2
library(gridExtra)
g_legend<-function(a.gplot){ 
  tmp <- ggplot_gtable(ggplot_build(a.gplot)) 
  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box") 
  legend <- tmp$grobs[[leg]] 
  return(legend)} 
legend <- g_legend(my.category.heatmap.graph[[1]]) 
#
my.category.heatmap.all <- plot_grid(my.category.heatmap.graph[[1]]+theme(legend.position="none"),legend ,ggplot()+geom_blank(),my.category.heatmap.graph[[17]],my.category.heatmap.graph[[19]],my.category.heatmap.graph[[3]],my.category.heatmap.graph[[4]],my.category.heatmap.graph[[5]],my.category.heatmap.graph[[14]],my.category.heatmap.graph[[10]],my.category.heatmap.graph[[6]],my.category.heatmap.graph[[13]],my.category.heatmap.graph[[7]],my.category.heatmap.graph[[8]],my.category.heatmap.graph[[9]],
                      labels=c("A","","","B","","","","","","","","","","",""),
                      vjust = 1.1, hjust=-0.5,
                      ncol=3,nrow=5,
                      label_size = 50)
#
#setwd("../../Nozue2016_SAStranscriptome_output/figs_tables/")
save_plot(file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Fig5_oldFig_S2_all.my.category.expression.pattern.v16.log2.signif.101117.png"), my.category.heatmap.all,
          ncol = 3, # we're saving a grid plot of 2 columns
          nrow = 5, # and 2 rows
          base_aspect_ratio = 2.5,
          base_height=6,
          limitsize = FALSE
)
```
# the end of custom category heatmap (shade responsiveness) (Fig. 5)

## co-expression analysis by WGCNA
```{r eval=FALSE}
# The following setting is important, do not omit.
options(stringsAsFactors = FALSE)
```
## Work in Whitney with multi-threds 
```{r eval=FALSE}
# Allow multi-threading within WGCNA. This helps speed up certain calculations.
# At present this call is necessary for the code to work.
# Any error here may be ignored but you may want to update WGCNA if you see one.
# Caution: skip this line if you run RStudio or other third-party R environments.
# See note above.
if(Sys.info()["nodename"]=="whitney") {
  enableWGCNAThreads(10) # in Whitney (Maloof lab server) 
  }
```
## Co-expressed gene module detection by WGCNA
```{r eval=FALSE}
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","SAS.expression.vst.s.kazu.largeCV.Rdata")) 
#load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","samples.nolow5.Rdata"))
Genotype<-levels(as.factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|25|49)(A.merged.bam)","\\2",names(SAS.expression.vst.s.kazu.largeCV)[-c(1,288)])))

for(x in Genotype) {
  SAS.expression.vst.s.kazu.largeCV.x<-SAS.expression.vst.s.kazu.largeCV[,gsub("([C|D|E])([[:digit:]]+)(H|L)(1|4|16|25|49)(A.merged.bam)","\\2",colnames(SAS.expression.vst.s.kazu.largeCV))==x]
    dim(SAS.expression.vst.s.kazu.largeCV.x) #[1] 7438     21
  datExpr <-t(SAS.expression.vst.s.kazu.largeCV.x)
  # Choose a set of soft-thresholding powers
  powers = c(c(1:9), seq(from = 2, to=20, by=10))
  sft = pickSoftThreshold(datExpr, powerVector = powers, verbose = 5)  
  # Plot the results:
  #sizeGrWindow(9, 5)
  pdf(paste("../../Nozue2016_SAStranscriptome_output/output/all",x,".largeCV.softthresholding.pdf",sep=""),width=10,height=8)
  par(mfrow = c(1,2));
  cex1 = 0.9;
  # Scale-free topology fit index as a function of the soft-thresholding power
  plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
       xlab="Soft Threshold (power)",ylab="Scale Free Topology Model Fit,signed R^2",type="n",
       main = paste("Scale independence"));
  text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
       labels=powers,cex=cex1,col="red");
  # this line corresponds to using an R^2 cut-off of h
  abline(h=0.90,col="red")
  # Mean connectivity as a function of the soft-thresholding power
  plot(sft$fitIndices[,1], sft$fitIndices[,5],
       xlab="Soft Threshold (power)",ylab="Mean Connectivity", type="n",
       main = paste("Mean connectivity"))
  text(sft$fitIndices[,1], sft$fitIndices[,5], labels=powers, cex=cex1,col="red")
  dev.off()
  # 
  net = blockwiseModules(datExpr, power = 6,
                         TOMType = "unsigned", minModuleSize = 20,
                         reassignThreshold = 0, mergeCutHeight = 0.25,
                         numericLabels = TRUE, pamRespectsDendro = FALSE,
                         saveTOMs = TRUE,
                         saveTOMFileBase = paste("SAS.expression.vst.s.kazu.largeCV",x,".TOM",sep=""),
                         verbose = 3)
  save(net,file=paste("../../Nozue2016_SAStranscriptome_output/output/SAS.expression.vst.s.kazu.largeCV",x,".Rdata",sep=""))  # confusing name dim(net) [1] 17757
  # open a graphics window
  pdf(file=paste("../../Nozue2016_SAStranscriptome_output/output/all",x,".largeCV.dendrogram.pdf",sep=""),width=10,height=8)
  # Convert labels to colors for plotting
  mergedColors = labels2colors(net$colors)
  # Plot the dendrogram and the module colors underneath
  plotDendroAndColors(net$dendrograms[[1]], mergedColors[net$blockGenes[[1]]],
                      "Module colors",
                      dendroLabels = FALSE, hang = 0.03,
                      addGuide = TRUE, guideHang = 0.05)
  dev.off()
  # save parameters  
  moduleLabels = net$colors
  moduleColors = labels2colors(net$colors)
  MEs = net$MEs
  geneTree = net$dendrograms[[1]]
  save(MEs, moduleLabels, moduleColors, geneTree,file = paste("../../Nozue2016_SAStranscriptome_output/output/all",x,".largeCV.RData",sep=""))
} # recalculating 082016. Recalculating genotype "3" due to "3" was missing (091616), (092817)

# new WGCNA modules are too many (110716)
# how many modules?
  table(net$colors);length(table(net$colors)) #
## recutBlockwiseTrees() is too divide modules further
## try mergeCloseMOdulesl with higher height (0.6) for all Genotype
## resultant "tan" module is SA/JA module but it is too big. (110916)
## try 0.55
##  cutHeight = 0.55, 27 modules, yellow = SA is too large
## try 0.4, 37 modules, gray60 that PCC1 belongs is slight large (594)
## try 0.35 (merge.net35) 44 modules (JA and SA GO terms were found in the same module; gray60)
### all PCC1, XBAT34, AIG1, ACD6 belong to "gray60"
## try 0.3 (merge.net30) 50 modules (JA (gray60) and SA (orange) GO terms were separated). PCC1 (  AT3G22231; orange, XBAT34 (AT4G14365, orange), but AIG1 (thistle2;  AT1G33960), ACD6 (?)
## try 0.32; all four genes belong to "orange"

load("../../Nozue2016_SAStranscriptome_output/output/SAS.expression.vst.s.kazu.largeCV.Rdata") #  where 
load("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/samples.nolow5.Rdata")
Genotype<-levels(as.factor(sub("(C|D|E)([[:digit:]]+)(L|H)(1|4|25|49)(A.merged.bam)","\\2",names(SAS.expression.vst.s.kazu.largeCV)[-c(1,288)])))

rm(net)
for(x in Genotype) {
  SAS.expression.vst.s.kazu.largeCV.x<-SAS.expression.vst.s.kazu.largeCV[,gsub("([C|D|E])([[:digit:]]+)(H|L)(1|4|16|25|49)(A.merged.bam)","\\2",colnames(SAS.expression.vst.s.kazu.largeCV))==x]
    dim(SAS.expression.vst.s.kazu.largeCV.x) #[1] 17757     21
  datExpr <-t(SAS.expression.vst.s.kazu.largeCV.x)
  load(paste("../../Nozue2016_SAStranscriptome_output/output/SAS.expression.vst.s.kazu.largeCV",x,".Rdata",sep=""))
# merge.net2<-mergeCloseModules(
#      # input data
#   exprData=datExpr, colors=net$colors,
#   # Optional starting eigengenes
#   MEs = NULL,
#   # Optional restriction to a subset of all sets
#   useSets = NULL,
#   # If missing data are present, impute them?
#   impute = TRUE,
#   # Input handling options
#   checkDataFormat = TRUE,
#   unassdColor = if (is.numeric(colors)) 0 else "grey",
#   # Options for eigengene network construction
#   corFnc = cor, corOptions = list(use = 'p'),
#   useAbs = FALSE,
#   # Options for constructing the consensus
#   equalizeQuantiles = FALSE,
#   quantileSummary = "mean",
#   consensusQuantile = 0,
#   # Merging options
#   cutHeight = 0.6,
#   iterate = TRUE,
#   # Output options
#   relabel = FALSE,
#   colorSeq = NULL,
#   getNewMEs = TRUE,
#   getNewUnassdME = TRUE,
#   # Options controlling behaviour of the function
#   trapErrors = FALSE,
#   verbose = 1, indent = 0)
# # save parameters  
#   moduleLabels = merge.net2$colors
#   moduleColors = labels2colors(merge.net2$colors)
#   MEs = merge.net2$MEs
#   geneTree = merge.net2$dendrograms[[1]]
#   save(MEs, moduleLabels, moduleColors, geneTree,file = paste("../../Nozue2016_SAStranscriptome_output/output/all",x,".largeCV.merge.net2.RData",sep="")) 
# rm(net)
# rm(merge.net2)
# } # resultant "tan" module is SA/JA module but it is too big. (110916)
merge.net32<-mergeCloseModules(
     # input data
  exprData=datExpr, colors=net$colors,
  # Optional starting eigengenes
  MEs = NULL,
  # Optional restriction to a subset of all sets
  useSets = NULL,
  # If missing data are present, impute them?
  impute = TRUE,
  # Input handling options
  checkDataFormat = TRUE,
  unassdColor = if (is.numeric(colors)) 0 else "grey",
  # Options for eigengene network construction
  corFnc = cor, corOptions = list(use = 'p'),
  useAbs = FALSE,
  # Options for constructing the consensus
  equalizeQuantiles = FALSE,
  quantileSummary = "mean",
  consensusQuantile = 0,
  # Merging options
  cutHeight = 0.32, # need to optimize 
  iterate = TRUE,
  # Output options
  relabel = FALSE,
  colorSeq = NULL,
  getNewMEs = TRUE,
  getNewUnassdME = TRUE,
  # Options controlling behaviour of the function
  trapErrors = FALSE,
  verbose = 1, indent = 0)
# save parameters  
  moduleLabels = merge.net32$colors
  moduleColors = labels2colors(merge.net32$colors)
  #MEs = merge.net32$MEs # null (120916)
  MEs = merge.net32$newMEs # new one? (120916)
# # switch "orange" and "salmon" (to keep "salomn" module name) (not applicable to 092817 version)
# moduleColors.temp<-moduleColors
# moduleColors.temp<-gsub("^orange","salmon",moduleColors)
# moduleColors.temp<-gsub("salmon4","orange4",moduleColors.temp)
# moduleColors<-moduleColors.temp
  geneTree = merge.net32$dendrograms[[1]]
  save(MEs, moduleLabels, moduleColors, geneTree,file = paste("../../Nozue2016_SAStranscriptome_output/output/all",x,".largeCV.merge.net32.RData",sep="")) # "salmon" fixed (121216)
rm(net)
rm(merge.net32)
} # resultant ***

# check new module numbers
moduleColors.files<-list.files(pattern=".largeCV.merge.net32.RData",path=file.path("..","..","Nozue2016_SAStranscriptome_output","output")) # for  merged modules by mergeCloseModules function above for Col (110816); newer (110916)
### use merged modules by mergeCloseModules function above for Col (22 modules; 110716; ** modules 092817)
#load("all1.largeCV.merge.net32.RData")
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","all1.largeCV.merge.net32.RData")) 
#load("/Volumes/data_work/Data8/NGS_related/Arabidopsis_analysis/Nozue2016_SAStranscriptome_finalfinal_newVST/Nozue2016_SAStranscriptome_output/output/all1.largeCV.merge.net32.RData")
########
moduleColors.allCol.largeCV<-moduleColors
table(moduleColors.allCol.largeCV);length(table(moduleColors.allCol.largeCV)) #
## check module character by GO ORA below
## back to single thred in whitney
if(Sys.info()["nodename"]=="whitney") {
  enableWGCNAThreads(2) # in Whitney (Maloof lab server) 
  }
```
## summary of WGCNA modules in mutants (092817 running with error due to using old SAS.expression.vst.s.kazu.largeCV.Rdata)
```{r eval=FALSE}
#setwd(homedir)
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","SAS.expression.vst.s.kazu.largeCV.Rdata")) # 7025 

moduleColors.files<-list.files(path=file.path("..","..","Nozue2016_SAStranscriptome_output","output"),pattern=".largeCV.merge.net32.RData") # for  merged modules by mergeCloseModules function above for Col (110816); newer (110916)
# 092717
### use merged modules by mergeCloseModules function above for Col (22 modules; 110716); (**modules; 092817)
## start with Genotype1
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","all1.largeCV.merge.net32.RData"))
moduleColors.allCol.largeCV<-moduleColors
table(moduleColors.allCol.largeCV)
temp<-moduleColors.allCol.largeCV
## adding more genotype
for(y in moduleColors.files[!moduleColors.files=="all1.largeCV.merge.net32.RData"]) {
  load(file.path("..","..","Nozue2016_SAStranscriptome_output","output",y))
  print(length(moduleColors)) # [1] 17757
  # switch "orange" and "salmon" (to keep "salomn" module name)
#moduleColors.temp<-moduleColors
#moduleColors.temp<-gsub("^orange","salmon",moduleColors)
#moduleColors.temp<-gsub("salmon4","orange4",moduleColors.temp)
#moduleColors<-moduleColors.temp
# the end of switch
  matched.moduleColors<-matchLabels(moduleColors,moduleColors.allCol.largeCV)
  temp<-cbind(temp, matched.moduleColors)
}   
length(moduleColors.files)
temp<-as.data.frame(temp)

names(temp)<-gsub("(all)([[:digit:]]+)(.largeCV.merge.net32.RData)","\\2",c("all1.largeCV.merge.net32.RData",moduleColors.files[!moduleColors.files=="all1.largeCV.merge.net32.RData"]))
SAS.expression.vst.s.kazu.largeCV.modules<-cbind(rownames(SAS.expression.vst.s.kazu.largeCV),temp)

names(SAS.expression.vst.s.kazu.largeCV.modules)[1]<-"AGI"
names(SAS.expression.vst.s.kazu.largeCV.modules)
table(SAS.expression.vst.s.kazu.largeCV.modules$"1")
# look modules of PCC1, XBAT34, AIG1 (PCC1 (  AT3G22231; orange, XBAT34 (AT4G14365, orange), but AIG1 (thistle2;  AT1G33960), ACD6 (AT4G14400, ?) > all became "salmon" in threshold 0.32
SAS.expression.vst.s.kazu.largeCV.modules[ gsub("([[:digit:]]+)(\\.)([[:digit:]]+)","\\1",SAS.expression.vst.s.kazu.largeCV.modules$AGI) %in% c("AT3G22231","AT4G14365","AT1G33960","AT4G14400"),]
# all of them belong to "purple" (092817)
# switch "purple" and "salmon" (092817)
SAS.expression.vst.s.kazu.largeCV.modules[SAS.expression.vst.s.kazu.largeCV.modules=="salmon"]<-"temp"
SAS.expression.vst.s.kazu.largeCV.modules[SAS.expression.vst.s.kazu.largeCV.modules=="purple"]<-"salmon"
SAS.expression.vst.s.kazu.largeCV.modules[SAS.expression.vst.s.kazu.largeCV.modules=="temp"]<-"purple"
# 
SAS.expression.vst.s.kazu.largeCV.modules[ gsub("([[:digit:]]+)(\\.)([[:digit:]]+)","\\1",SAS.expression.vst.s.kazu.largeCV.modules$AGI) %in% c("AT3G22231","AT4G14365","AT1G33960","AT4G14400"),] # OK
head(SAS.expression.vst.s.kazu.largeCV.modules)
table(SAS.expression.vst.s.kazu.largeCV.modules$"1") # OK purple and salmon switched

save(SAS.expression.vst.s.kazu.largeCV.modules,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","SAS.expression.vst.s.kazu.largeCV.modules092817.Rdata"))

```

# overlap between WGCNA Col modules and mutation affected genes 
```{r,eval=FALSE}
#setwd(homedir)
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","DE.genotype.batch.genes.unique.DF.Rdata")) #[1] 3350   25 > 3360 25 (100717)
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","SAS.expression.vst.s.kazu.largeCV.modules092817.Rdata"))

dim(SAS.expression.vst.s.kazu.largeCV.modules)# [1] 7025   14
table(SAS.expression.vst.s.kazu.largeCV.modules$"1") # OK purple and salmon switched
SAS.expression.vst.s.kazu.largeCV.modules[ gsub("([[:digit:]]+)(\\.)([[:digit:]]+)","\\1",SAS.expression.vst.s.kazu.largeCV.modules$AGI) %in% c("AT3G22231","AT4G14365","AT1G33960","AT4G14400"),]
# MYC2 (AT1G32640), MYC3 (AT5G46760), MYC4 (AT4G17880)
SAS.expression.vst.s.kazu.largeCV.modules[ gsub("([[:digit:]]+)(\\.)([[:digit:]]+)","\\1",SAS.expression.vst.s.kazu.largeCV.modules$AGI) %in% c("AT1G32640","AT5G46760","AT4G17880"),] # only MYC2
# six MYBs that regulate indolic glucosinolates (GSLs): MYB34 AT5G60890, MYB51 AT1G18570, MYB122 AT1G74080, MYB28 AT5G61420, MYB29 AT5G07690, MYB76      AT5G07700
SAS.expression.vst.s.kazu.largeCV.modules[ gsub("([[:digit:]]+)(\\.)([[:digit:]]+)","\\1",SAS.expression.vst.s.kazu.largeCV.modules$AGI) %in% c("AT5G60890","AT1G18570","AT1G74080","AT5G61420","AT5G07690","AT5G07700"),]

#library(WGCNA)
SAS.expression.vst.s.kazu.largeCV.modules$AGI<-gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",SAS.expression.vst.s.kazu.largeCV.modules$AGI)
# merge with WGCNA modules
test<-merge(DE.genotype.batch.genes.unique.DF,SAS.expression.vst.s.kazu.largeCV.modules[,c("AGI","1")],by="AGI",all=TRUE)
DE.genotype.batch.genes.WGCNAmodules<-test
DE.genotype.batch.genes.WGCNAmodules[is.na(DE.genotype.batch.genes.WGCNAmodules)]<-"0"
names(DE.genotype.batch.genes.WGCNAmodules)<-sub("1$","WGCNAmodules",names(DE.genotype.batch.genes.WGCNAmodules)) # 
table(DE.genotype.batch.genes.WGCNAmodules$WGCNAmodules) # "salmon" should be 155, "purple" should be 169
length(table(DE.genotype.batch.genes.WGCNAmodules$WGCNAmodules)) #
# adding gene description
DE.genotype.batch.genes.WGCNAmodules.description<-merge(DE.genotype.batch.genes.WGCNAmodules,TAIR10_gene_descriptions[!duplicated(TAIR10_gene_descriptions$AGI),c("AGI","Short_description","colE...col.F")],by="AGI")
#write.csv(DE.genotype.batch.genes.WGCNAmodules.description,file=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Dataset5_Complete_gene_list_of_genes_misexpressed_in_mutant_and_found_in_WGCNA_modules_visualized_in_Fig7_092817.csv")) #091616. Column names were screwed up (111616). Orange > "salmon", "salmon" > "orange4" (121316), 092817
write.csv(DE.genotype.batch.genes.WGCNAmodules.description,file=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Dataset5_Complete_gene_list_of_genes_misexpressed_in_mutant_and_WGCNA_modules_in_Fig7_100717.csv")) #091616. Column names were screwed up (111616). Orange > "salmon", "salmon" > "orange4" (121316), 092817

# in whitney
#test<-read.csv(file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Dataset5_Complete_gene_list_of_genes_misexpressed_in_mutant_and_found_in_WGCNA_modules_visualized_in_Fig7_092817.csv"),row.names=1) # does  work (092717)
test<-read.csv(file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Dataset5_Complete_gene_list_of_genes_misexpressed_in_mutant_and_WGCNA_modules_in_Fig7_100717.csv"),row.names=1) # does  work (100717)

test[ gsub("([[:digit:]]+)(\\.)([[:digit:]]+)","\\1",test$AGI) %in% c("AT3G22231","AT4G14365","AT1G33960","AT4G14400"),] # OK they are salmon

# myc2 (AT1G32640), myc3 (AT5G46760), myc4 (AT4G17880)
test[ gsub("([[:digit:]]+)(\\.)([[:digit:]]+)","\\1",test$AGI) %in% c("AT1G32640","AT5G46760","AT4G17880"),] # 

# six MYBs that regulate indolic glucosinolates (GSLs): MYB34 AT5G60890, MYB51 AT1G18570, MYB122 AT1G74080, MYB28 AT5G61420, MYB29 AT5G07690, MYB76      AT5G07700
test[ gsub("([[:digit:]]+)(\\.)([[:digit:]]+)","\\1",test$AGI) %in% c("AT5G60890","AT1G18570","AT1G74080","AT5G61420","AT5G07690","AT5G07700"),]

#test<-test[,-1]
dim(test) # 17985   28 > 8156   28 > 8150 28 (100717)
# overlap table (revsised into GOseq.overlapTable; 052616; 092817; 092917)
for(i in 2:25) { #
  overlap.GOseq<-GOseq.overlapTable(test[,"WGCNAmodules"],test[,i],genes=test$AGI)
  overlap.GOseq.melt<-melt(overlap.GOseq$pTable,id=1) #
  overlap.GOseq.melt$count<-melt(overlap.GOseq$countTable,id=1)$value
  overlap.GOseq.melt<-overlap.GOseq.melt[!overlap.GOseq.melt$Var2=="0",]
  #overlap.melt.summary<-rbind(overlap.melt.summary,overlap.melt)
  save(overlap.GOseq.melt,file=file.path(homedir,"..","..","Nozue2016_SAStranscriptome_output","output",paste("overlap.GOseq.melt.WGCNA.mut.PLoSGenetics2015mutants.errrDE.",i,".Rdata",sep="")))
}
```
# combine 
```{r}
overlap.GOseq.melt.WGCNA.mut.files<-list.files(path=file.path("..","..","Nozue2016_SAStranscriptome_output","output"),pattern="overlap.GOseq.melt.WGCNA.mut.PLoSGenetics2015mutants.errrDE")
#overlap.GOseq.melt.summary.list<-sapply(file.path("..","..","Nozue2016_SAStranscriptome_output","output",overlap.GOseq.melt.SOM.mut.files),load)
# load(file.path("..","..","Nozue2016_SAStranscriptome_output","output",overlap.GOseq.melt.SOM.mut.files[1])) # works for one RData file
overlap.GOseq.melt.summary.list2<-sapply(file.path("..","..","Nozue2016_SAStranscriptome_output","output",overlap.GOseq.melt.WGCNA.mut.files),function(x) mget(load(x))) # mget will return the value of the object(or objects) in a list. see https://stackoverflow.com/questions/29398630/load-data-frames-into-list
names(overlap.GOseq.melt.summary.list2)
overlap.GOseq.melt.summary<-do.call("rbind",overlap.GOseq.melt.summary.list2) 
head(overlap.GOseq.melt.summary) # make sure those are SOM.customcat files
rownames(overlap.GOseq.melt.summary)<-1:nrow(overlap.GOseq.melt.summary)
# format overlap.GOseq.melt.summary
overlap.GOseq.melt.summary[,"value"]<--log10(overlap.GOseq.melt.summary[,"value"])
overlap.GOseq.melt.summary[overlap.GOseq.melt.summary$value>10,"value"]<-10
overlap.GOseq.melt.summary[overlap.GOseq.melt.summary$count==0,"count"]<-""
save(overlap.GOseq.melt.summary, file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","overlap.GOseq.melt.summary.WGCNA.mut.092817.Rdata") )
save(overlap.GOseq.melt.summary, file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","overlap.GOseq.melt.summary.WGCNA.mut.100717.Rdata") )


#save(overlap.melt.summary,file="../Nozue2016_SAStranscriptome_output/output/overlap.melt.summary.WGCNA.mut.GOseq.Rdata") # 110816, 121316, 091817
```

# reading and cleaning overlap table and graph (WGCNA vs mut; 100817 updated)
```{r}
#load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","overlap.GOseq.melt.summary.WGCNA.mut.092817.Rdata"))
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","overlap.GOseq.melt.summary.WGCNA.mut.100717.Rdata"))

overlap.GOseq.melt.summary
overlap.GOseq.melt.summary.ss<-overlap.GOseq.melt.summary[!overlap.GOseq.melt.summary$Var1=="0",]
overlap.GOseq.melt.summary.ss$Var1<-as.factor(overlap.GOseq.melt.summary.ss$Var1)
overlap.GOseq.melt.summary.ss$Var2<-factor(overlap.GOseq.melt.summary.ss$Var2,levels=paste(rep(conversion.table[-1,"genotype"],2),rep(c("sun","shade"),each=12),sep=".")) # sort Var2 (muts)

table(overlap.GOseq.melt.summary.ss$Var1)
q4<-ggplot(overlap.GOseq.melt.summary.ss,aes(y=factor(Var1),x=factor(Var2))) 
#q4 <- q4 + geom_tile(color="black",aes(fill=value))+scale_fill_gradient2(low=muted("green"), high=muted("magenta"),midpoint=2) # this color coding is not good (061716)
#q4 <- q4 + geom_tile(color="black",aes(fill=value))+scale_fill_gradient(low="white", high="magenta") # Is this better? (061716)
q4 <- q4 + geom_tile(color="black",aes(fill=value))+scale_fill_gradient2(low="white", mid="white",high="magenta",midpoint=2) # Is this better? (061716)

#q4 <- q4 + theme(axis.text.x=element_text(angle=90),axis.title=element_blank())
#q4 <- q4 +theme(legend.position="none") # this removed all legend
q4 <- q4 + theme(axis.text.x=element_text(angle=90,hjust=1),
                 axis.ticks = element_line(size=0,color="white"),
                 axis.line=element_line(size=1)) # adjust axis line width
q4 <- q4 + geom_text(aes(label=count),size=4)
q4 <- q4 + labs(x="genotype specific genes",y="WGCNA modules in Col",title="modules affected in SAS mutants",fill="-log10\n pvalue")

#q4 <- q4 + theme(legend.title=element_blank())
#ggsave(q4,file="Col.WGCNA.modules.vs.GenotypeSpecific.genes.table.pdf",width=11,height=8)
# ggsave(q4,file="Col.WGCNA.modules.vs.GenotypeSpecific.genes.table.s.pdf",width=11,height=8,path="../../Nozue2016_SAStranscriptome_output/figs_tables")
# ggsave(q4,file="Col.WGCNA.modules.vs.GenotypeSpecific.genes.table.s.png",width=11,height=8,path="../../Nozue2016_SAStranscriptome_output/figs_tables")
# ggsave(q4,file="Fig7_Col.WGCNA.modules.vs.GenotypeSpecific.genes.table.s.GOseq111816.pdf",width=11,height=8,path="../../Nozue2016_SAStranscriptome_output/figs_tables")
# ggsave(q4,file="Fig7_Col.WGCNA.modules.vs.GenotypeSpecific.genes.table.s.GOseq111816.png",width=11,height=8,path="../../Nozue2016_SAStranscriptome_output/figs_tables")
# ggsave(q4,file="Fig7_Col.WGCNA.modules.vs.GenotypeSpecific.genes.table.s.GOseq092817.pdf",width=11,height=8,path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables"))
# ggsave(q4,file="Fig7_Col.WGCNA.modules.vs.GenotypeSpecific.genes.table.s.GOseq092817.png",width=11,height=8,path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables"))
ggsave(q4,file="Fig7_Col.WGCNA.modules.vs.GenotypeSpecific.genes.table.s.GOseq100717.pdf",width=11,height=8,path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables"))
ggsave(q4,file="Fig7_Col.WGCNA.modules.vs.GenotypeSpecific.genes.table.s.GOseq100717.png",width=11,height=8,path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables"))

q4 # magenta indicates pvalue is smaller than 0.01.
# #### only "salmon","pink" (=JA),"purple" (=auxin) for job application 2015 - 2016 ###
# overlap.melt.summary.ss.selected<-overlap.melt.summary.ss[overlap.melt.summary.ss$Var1 %in% c("salmon","purple","pink"),]
# q5<-ggplot(overlap.melt.summary.ss.selected,aes(y=factor(Var1),x=factor(Var2))) 
# q5 <- q5 + geom_tile(color="black",aes(fill=value))+scale_fill_gradient2(low=muted("green"), high=muted("magenta"),midpoint=2)
# q5 <- q5 + theme(axis.text.x=element_text(angle=90,hjust=1),
#                  panel.grid.major = element_blank(),
#                  panel.grid.minor = element_blank(),
#                  panel.background = element_blank(),
#                  axis.ticks = element_line(size=0,color="white"))
# q5 <- q5 + geom_text(aes(label=count),size=5,show_guide = FALSE)
# 
# q5 <- q5 + labs(x="genotype specific genes",y="WGCNA modules in Col",title="modules affected in SAS mutants",fill="-log10\n p-value")
# q5 <- q5 + theme(legend.title=element_blank())
# #ggsave(q5,file="Col.WGCNA.modules.vs.GenotypeSpecific.genes.table.s.selected.pdf",width=11,height=4,path="../../Nozue2016_SAStranscriptome_output/figs_tables")

```
# Number of Col genes in each module are: 
```{r eval=FALSE} 
table(SAS.expression.vst.s.kazu.largeCV.modules$"1")  # 28 modules
```
# The next question is what are these modules?  GOseq analysis of each module
```{r eval=FALSE}
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","SAS.expression.vst.s.kazu.largeCV.modules092817.Rdata")) # 092717
SAS.expression.vst.s.kazu.largeCV.modules$AGI<-gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",SAS.expression.vst.s.kazu.largeCV.modules$AGI)
modules<-levels(as.factor(SAS.expression.vst.s.kazu.largeCV.modules$"1"))
for(x in modules) {
    temp.GOseq<-GOseq.ORA(SAS.expression.vst.s.kazu.largeCV.modules[SAS.expression.vst.s.kazu.largeCV.modules$"1"==x,"AGI"])
  if(temp.GOseq=="no enriched GO") {next} else {
  temp.genes<-genes.in.enriched.category(enrich.result=temp.GOseq,gene.list=SAS.expression.vst.s.kazu.largeCV.modules[SAS.expression.vst.s.kazu.largeCV.modules$"1"==x,"AGI"],category.table=Atgo.list)
  save(temp.GOseq,temp.genes,file=paste("../../Nozue2016_SAStranscriptome_output/output/WGCNA.Col.module.",x,".GOseq092717.Rdata",sep=""))
}
  } # 110916, 121316, 091817,092817
```
## write summary of WGCNA.Col.module.GOseq (modified 092717; no need to change 100717)
```{r eval=FALSE}
WGCNA.Col.module.GOseq.files<-list.files(path=file.path("..","..","Nozue2016_SAStranscriptome_output","output"),pattern="^WGCNA.Col.module.")
WGCNA.Col.module.GOseq.files<-grep(".GOseq092717.Rdata",WGCNA.Col.module.GOseq.files,value=TRUE)

load(file.path("..","..","Nozue2016_SAStranscriptome_output","output",WGCNA.Col.module.GOseq.files[1])) # using file.path 091817, 092717

temp.GOseq$modules<-gsub("(WGCNA.Col.module.)([[:print:]]+)(.GOseq092717.Rdata)","\\2",WGCNA.Col.module.GOseq.files[1])
temp<-temp.GOseq
 
for(i in WGCNA.Col.module.GOseq.files[-1]) {
  load(file.path("..","..","Nozue2016_SAStranscriptome_output","output",i)) # 091817
  temp.GOseq$modules<-gsub("(WGCNA.Col.module.)([[:print:]]+)(.GOseq092717.Rdata)","\\2",i)
  temp<-rbind(temp,temp.GOseq)
}
WGCNA.Col.module.GOseq.summary<-temp
write.csv(WGCNA.Col.module.GOseq.summary,file=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","TableS7_GO_enrichment_analysis_of_WGCNA_modules092817.csv")) # Dataset 5. GO enrichment analysis of WGCNA gene co-expression modules (091716, salmon4<>orange 121316, rerun 091817; 092817)
```
## What genes are found in given module with given GO category? (no need to change 100717)
```{r eval=FALSE}
# ""GO:0009507" in "salmon" (chloroplast)
load("../../Nozue2016_SAStranscriptome_output/output/GOCC.WGCNA.Col.module.salmon.GOseq.Rdata") # I did not run GOCC
temp.genes.wdescrition<-merge(temp.genes,TAIR10_gene_descriptions,by="AGI")
dim(temp.genes)
temp.genes.wdescrition<-temp.genes.wdescrition[!duplicated(temp.genes.wdescrition$AGI),]
temp.genes.wdescrition[temp.genes.wdescrition$"GO:0009507"==1,c("AGI","Short_description")]
# GO:0009697, "salmon" (old version)
load("../Nozue2016_SAStranscriptome_output/output/WGCNA.Col.module.salmon.GOseq.Rdata")
temp.genes.wdescrition<-merge(temp.genes,TAIR10_gene_descriptions,by="AGI")
temp.genes.wdescrition<-temp.genes.wdescrition[!duplicated(temp.genes.wdescrition$AGI),]
#temp.genes.wdescrition[temp.genes.wdescrition$"GO:0009697"==1,c("AGI","Short_description")]
temp.genes.wdescrition[temp.genes.wdescrition$"GO:0009697"==1,c("AGI","Short_description","colE...col.F")]
```

## GOseq analysis of each cluster (cellular component) (no need to change, 100717)
```{r eval=FALSE}
#load("../Nozue2016_SAStranscriptome_output/output/SAS.expression.vst.s.kazu.largeCV.module121316.Rdata")
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","SAS.expression.vst.s.kazu.largeCV.modules092817.Rdata")) # 092717

#library(WGCNA)
SAS.expression.vst.s.kazu.largeCV.modules$AGI<-gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",SAS.expression.vst.s.kazu.largeCV.modules$AGI)
modules<-levels(as.factor(SAS.expression.vst.s.kazu.largeCV.modules$"1"))
#setwd(homedir)
#setwd("../Nozue2016_SAStranscriptome_output/output/")
# needs to fix for new GOseq.CC.ORA (did I fixed it?; 110816)
for(x in modules) { # 
  temp.GOseq<-GOseq.CC.ORA(SAS.expression.vst.s.kazu.largeCV.modules[SAS.expression.vst.s.kazu.largeCV.modules$"1"==x,"AGI"])
  if(temp.GOseq=="no enriched GO") {next} else {
  temp.GOseq<-GOseq.CC.ORA(SAS.expression.vst.s.kazu.largeCV.modules[SAS.expression.vst.s.kazu.largeCV.modules$"1"==x,"AGI"])
  temp.genes<-genes.in.enriched.category(enrich.result=temp.GOseq,gene.list=SAS.expression.vst.s.kazu.largeCV.modules[SAS.expression.vst.s.kazu.largeCV.modules$"1"==x,"AGI"],category.table=Atgo.list)
  save(temp.GOseq,temp.genes,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output",paste("GOCC.WGCNA.Col.module.",x,".GOseq.Rdata",sep="")))
}
} 
```
## write summary of WGCNA.Col.module.GOseq.CC  (no need to change, 100717)
```{r eval=FALSE}
GOCC.WGCNA.Col.module.GOseq.files<-list.files(pattern="GOCC.WGCNA.Col.module.",path=file.path("..","..","Nozue2016_SAStranscriptome_output","output"))
GOCC.WGCNA.Col.module.GOseq.files<-grep(".GOseq.Rdata",GOCC.WGCNA.Col.module.GOseq.files,value=TRUE)
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output",GOCC.WGCNA.Col.module.GOseq.files[1]))
temp.GOseq$modules<-gsub("(GOCC.WGCNA.Col.module.)([[:print:]]+)(.GOseq.Rdata)","\\2",GOCC.WGCNA.Col.module.GOseq.files[1])
temp<-temp.GOseq
 
for(i in GOCC.WGCNA.Col.module.GOseq.files[-1]) {
  #load(i)
  load(file.path("..","..","Nozue2016_SAStranscriptome_output","output",i))
  temp.GOseq$modules<-gsub("(GOCC.WGCNA.Col.module.)([[:print:]]+)(.GOseq.Rdata)","\\2",i)
  temp<-rbind(temp,temp.GOseq)
}
WGCNA.Col.module.GOseq.CC.summary<-temp
write.csv(WGCNA.Col.module.GOseq.CC.summary,file=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","WGCNA.Col.module.GOseq.CC.summary.csv")) # not used
```
# overlap analysis with custome categories (WGCNA modules): Fig6  (no need to change, 100717). using SAup_combined & SAdown_combined (110917)
```{r eval=FALSE}
# custom categories
load(file.path("..","..","Nozue2016_SAStranscriptome_data","input","hormone.responsiveness6.DF.s.Rdata")) # using SAup_combined & SAdown_combined
hormone.responsiveness6.DF.s
hormone.responsiveness6.DF.s$AGI
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","SAS.expression.vst.s.kazu.largeCV.modules092817.Rdata") ) #
SAS.expression.vst.s.kazu.largeCV.modules$AGI<-gsub("([[:print:]]+)(\\.[[:digit:]]+)","\\1",SAS.expression.vst.s.kazu.largeCV.modules$AGI)
test<-merge(hormone.responsiveness6.DF.s,SAS.expression.vst.s.kazu.largeCV.modules[,c("AGI","1")],by="AGI",all=TRUE)

# ########################
#load(file.path("..","..","Nozue2016_SAStranscriptome_data","input","hormone.responsiveness6b.Rdata")) # remove PCC1up and PCC1down from hormone.responsiveness6.Rdata and called hormone.responsiveness6b.Rdata
# ## change list into data.frame
# hormone.responsiveness6b.DF<-data.frame(t(sapply(hormone.responsiveness6b,c))) # modified for ver6. removing GRF1_3target, PCC1up, PCC1down
# names(hormone.responsiveness6b.DF)<-attributes(hormone.responsiveness6b)$category
# hormone.responsiveness6b.DF$AGI<-rownames(hormone.responsiveness6b.DF)
# 
# load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","SAS.expression.vst.s.kazu.largeCV.modules092817.Rdata") ) #
# 
# SAS.expression.vst.s.kazu.largeCV.modules$AGI<-gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",SAS.expression.vst.s.kazu.largeCV.modules$AGI) # remove .1, .2, etc
# 
# # merge with WGCNA modules
# test<-merge(hormone.responsiveness6b.DF,SAS.expression.vst.s.kazu.largeCV.modules[,c("AGI","1")],by="AGI",all=TRUE)
names(test)[names(test)=="1"]<-"WGCNAmodules"
i <- sapply(test, is.factor)
test[i]<-lapply(test[i], as.character)
class(test[,2])
  
hormone.responsiveness.WGCNAmodules<-test
hormone.responsiveness.WGCNAmodules[hormone.responsiveness.WGCNAmodules==""]<-"0"
hormone.responsiveness.WGCNAmodules[is.na(hormone.responsiveness.WGCNAmodules)]<-"0"
head(hormone.responsiveness.WGCNAmodules)

# write.csv(hormone.responsiveness.WGCNAmodules,file="../../Nozue2016_SAStranscriptome_output/figs_tables/TableS9_hormone.responsiveness.WGCNAmodules.csv") #
# Plant Cell version
# remove GRF1_3target (052616)
hormone.responsiveness.WGCNAmodules<-hormone.responsiveness.WGCNAmodules[,!names(hormone.responsiveness.WGCNAmodules)=="GRF1_3target"]

#write.csv(hormone.responsiveness.WGCNAmodules,file=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Supplemental_Dataset4_Custom_category_enrichment_analysis_of_WGCNA_gene_co-expression_modules092817.csv"))
write.csv(hormone.responsiveness.WGCNAmodules,file=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Supplemental_Dataset4_Custom_category_enrichment_analysis_of_WGCNA_gene_co-expression_modules110917.csv"))


# salmon genes and SA, npr1 responsiveness? (PCC1, XBAT34, AIG1, ACD6,SARD1,cbp60g)
hormone.responsiveness.WGCNAmodules[hormone.responsiveness.WGCNAmodules$WGCNAmodules %in% "salmon" & hormone.responsiveness.WGCNAmodules$AGI %in% c("AT3G22231","AT4G14365","AT1G33960","AT4G14400","AT5G26920","AT1G73805","AT5G26920"), c("AGI","IAAup","IAAdown","SARminus","SARplus","NPR1down","WRKY33down","WRKY33up","SAup_24h","SAdown_24h","NPR1up_uponSA","NPR1down_uponSA","WGCNAmodules")]

# salmon genes and SA, npr1 responsiveness? (PCC1, XBAT34, AIG1, ACD6,SARD1,cbp60g)
hormone.responsiveness.WGCNAmodules[hormone.responsiveness.WGCNAmodules$WGCNAmodules %in% "salmon" & hormone.responsiveness.WGCNAmodules$AGI %in% c("AT3G22231","AT4G14365","AT1G33960","AT4G14400","AT5G26920","AT1G73805","AT5G26920"), c("AGI","IAAup","IAAdown","SAup_combined","SAdown_combined","WGCNAmodules")]

# OK. They belong to "salmon" and not IAA responsive.
# IAA responsive & green
hormone.responsiveness.WGCNAmodules[hormone.responsiveness.WGCNAmodules$IAAup %in% "IAAup" & hormone.responsiveness.WGCNAmodules$WGCNAmodules=="green", c("AGI","IAAup","IAAdown","SAup_combined","SAdown_combined","WGCNAmodules")]

```
# overlapTable for WGCNA module vs custome categories (100417)  (no need to change, 100717)
```{r}
#hormone.responsiveness.WGCNAmodules<-read.csv(file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","TableS4_Custom_category_enrichment_analysis_of_WGCNA_gene_co-expression_modules092817.csv"),row.names=1)
hormone.responsiveness.WGCNAmodules<-read.csv(file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Supplemental_Dataset4_Custom_category_enrichment_analysis_of_WGCNA_gene_co-expression_modules110917.csv"),row.names=1)

hormone.responsiveness.WGCNAmodules[hormone.responsiveness.WGCNAmodules$WGCNAmodules %in% "salmon" & hormone.responsiveness.WGCNAmodules$AGI %in% c("AT3G22231","AT4G14365","AT1G33960","AT4G14400"), c("AGI","IAAup","IAAdown","SAup_combined","SAdown_combined","WGCNAmodules")]
# OK, Salmon module
test<-hormone.responsiveness.WGCNAmodules
for(i in 19:20) { # 2:32 for 092817.csv version. 19 and 20 are new (SAup_combined and SAdown_combined)
  overlap.GOseq<-GOseq.overlapTable(test[,"WGCNAmodules"],test[,i],genes=test$AGI) # all genes
  overlap.GOseq.melt<-melt(overlap.GOseq$pTable,id=1) #
  overlap.GOseq.melt$count<-melt(overlap.GOseq$countTable,id=1)$value
  overlap.GOseq.melt<-overlap.GOseq.melt[!overlap.GOseq.melt$Var2=="0",]
  #overlap.melt.summary<-rbind(overlap.melt.summary,overlap.melt)
  save(overlap.GOseq.melt,file=file.path(homedir,"..","..","Nozue2016_SAStranscriptome_output","output",paste("overlap.GOseq.melt.WGCNA.customcat.110917.",i,".Rdata",sep="")))
} # renamed
```
# combine files (WGCMA vs customcat): updated 110917 for SAup_combined and SAdown_combined
```{r}
# read files
overlap.GOseq.melt.WGCNA.customcat.files<-list.files(path=file.path("..","..","Nozue2016_SAStranscriptome_output","output"),pattern="overlap.GOseq.melt.WGCNA.customcat")
select.files<-gsub("(overlap.GOseq.melt.WGCNA.customcat)(\\.)([[:print:]]+)(\\.Rdata)","\\3",overlap.GOseq.melt.WGCNA.customcat.files)
overlap.GOseq.melt.WGCNA.customcat.files<-overlap.GOseq.melt.WGCNA.customcat.files[select.files %in% c("2","3","4","5","6","7","8","9","10","11","12","13","14","15","26","25","110917.19","110917.20")]
# select 2:18 from "overlap.GOseq.melt.WGCNA.customcat", 19:20 for "overlap.GOseq.melt.WGCNA.customcat".110917, 26, 25 for WRKY33up and down

# remove unnecessary files
overlap.GOseq.melt.WGCNA.customcat.files<-overlap.GOseq.melt.WGCNA.customcat.files[!overlap.GOseq.melt.WGCNA.customcat.files %in% c("overlap.GOseq.melt.WGCNA.customcat.27.Rdata","overlap.GOseq.melt.WGCNA.customcat.28.Rdata")]

overlap.GOseq.melt.summary.list2<-sapply(file.path("..","..","Nozue2016_SAStranscriptome_output","output",overlap.GOseq.melt.WGCNA.customcat.files),function(x) mget(load(x))) # mget will return the value of the object(or objects) in a list. see https://stackoverflow.com/questions/29398630/load-data-frames-into-list
names(overlap.GOseq.melt.summary.list2)
overlap.GOseq.melt.summary<-do.call("rbind",overlap.GOseq.melt.summary.list2) 
head(overlap.GOseq.melt.summary) # make sure those are SOM.customcat files
rownames(overlap.GOseq.melt.summary)<-1:nrow(overlap.GOseq.melt.summary)
# format overlap.GOseq.melt.summary
overlap.GOseq.melt.summary[,"value"]<--log10(overlap.GOseq.melt.summary[,"value"])
overlap.GOseq.melt.summary[overlap.GOseq.melt.summary$value>10,"value"]<-10
overlap.GOseq.melt.summary[overlap.GOseq.melt.summary$count==0,"count"]<-""

#save(overlap.GOseq.melt.summary, file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","overlap.GOseq.melt.summary.WGCNA.customcat.092817.Rdata") )
save(overlap.GOseq.melt.summary, file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","overlap.GOseq.melt.summary.WGCNA.customcat.110917.Rdata") )

```

## cleaning and plotting  (no need to change, 100717); (changed 110917 due to "SAup_combined" and "SAdown_combined")
```{r}
#load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","overlap.GOseq.melt.summary.WGCNA.customcat.092817.Rdata"))
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","overlap.GOseq.melt.summary.WGCNA.customcat.110917.Rdata"))

custom.category.overlap.melt.summary.s<-overlap.GOseq.melt.summary
table(custom.category.overlap.melt.summary.s$Var2)
custom.category.overlap.melt.summary.s<-custom.category.overlap.melt.summary.s[!custom.category.overlap.melt.summary.s$Var1=="0",]
custom.category.overlap.melt.summary.s<-custom.category.overlap.melt.summary.s[!custom.category.overlap.melt.summary.s$Var2=="0",]
custom.category.overlap.melt.summary.s<-custom.category.overlap.melt.summary.s[!is.na(custom.category.overlap.melt.summary.s$Var2),]
custom.category.overlap.melt.summary.s[is.na(custom.category.overlap.melt.summary.s$Var2),]
table(custom.category.overlap.melt.summary.s$Var2)
#custom.category.overlap.melt.summary.s$Var2<-factor(as.character(custom.category.overlap.melt.summary.s$Var2),levels=c("IAAup","IAAdown","BLup","BLdown","MJup","MJdown","GAup","GAdown","ABAup","ABAdown","ACCup","ACCdown","CKup","CKdown","SAup_2h","SAdown_2h","SARplus","SARminus","SAup_24h","SAdown_24h","PIFtarget","MYC234up","MYC234down","NPR1up","NPR1down", "NPR1up_uponSA","NPR1down_uponSA" ,"WRKY33up","WRKY33down")) # remove "GRF1_3target" (052616)
custom.category.overlap.melt.summary.s$Var2<-factor(as.character(custom.category.overlap.melt.summary.s$Var2),levels=c("IAAup","IAAdown","BLup","BLdown","MJup","MJdown","GAup","GAdown","ABAup","ABAdown","ACCup","ACCdown","CKup","CKdown","SAup_combined","SAdown_combined","MYC234up","MYC234down","WRKY33up","WRKY33down")) # (110917)

table(as.character(custom.category.overlap.melt.summary.s$Var2))
#save(custom.category.overlap.melt.summary.s,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","custom.category.v4.overlap.overlap.melt.summary.s.GOseq092817.Rdata"))
save(custom.category.overlap.melt.summary.s,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","custom.category.v4.overlap.overlap.melt.summary.s.GOseq110917.Rdata"))

```
# graph  (no need to change, 100717); chagned 110917 due to "SAup_combined" and "SAdown_combined"
```{r}
#load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","custom.category.v4.overlap.overlap.melt.summary.s.GOseq092817.Rdata")) # 09291717
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","custom.category.v4.overlap.overlap.melt.summary.s.GOseq110917.Rdata")) # 11091717

f6<-ggplot(custom.category.overlap.melt.summary.s,aes(y=factor(Var1),x=factor(Var2))) 
#q4 <- q4 + geom_tile(color="black",aes(fill=value))+scale_fill_gradient2(low=muted("green"), high=muted("magenta"),midpoint=2)
#q4 <- q4 + geom_tile(color="black",aes(fill=value))+scale_fill_gradient(low="white", high="magenta")
f6 <- f6 + geom_tile(color="black",aes(fill=value))+scale_fill_gradient2(low="white", mid="white",high="magenta",midpoint=2)

#q4 <- q4 + theme(axis.text.x=element_text(angle=90),axis.title=element_blank())
f6 <- f6 + theme(axis.text.x=element_text(angle=90,hjust=1),
                 panel.grid.major = element_blank(),
                 panel.grid.minor = element_blank(),
                 panel.background = element_blank(),
                 axis.ticks = element_line(size=0,color="white"),
                 axis.line=element_line(size=1)) # adjust axis line width
f6 <- f6 + geom_text(aes(label=count),size=4)
f6 <- f6 + labs(x="custom category",y="WGCNA modules in Col",title="custom category enrichment",fill="-log10\n pvalue")
#ggsave(q4,file="Col.WGCNA.modules.vs.GenotypeSpecific.genes.table.pdf",width=11,height=8)
# ggsave(q4,file="Fig6_Col.WGCNA.modules.vs.my.category.table.s.113016.pdf",width=11,height=8,path="../../Nozue2016_SAStranscriptome_output/figs_tables")
# ggsave(q4,file="Fig6_Col.WGCNA.modules.vs.my.category.table.s.113016.png",width=11,height=8,path="../../Nozue2016_SAStranscriptome_output/figs_tables")
# ggsave(f6,file="Fig6_Col.WGCNA.modules.vs.my.category.table.s.121316.pdf",width=11,height=8,path="Nozue2016_SAStranscriptome_output/figs_tables")
# ggsave(f6,file="Fig6_Col.WGCNA.modules.vs.my.category.table.s.121316.png",width=11,height=8,path="Nozue2016_SAStranscriptome_output/figs_tables")
# needs to work on sorting castum categories (092918)
# ggsave(f6,file="Fig6_Col.WGCNA.modules.vs.my.category.table.s.092817.pdf",width=11,height=8,path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables"))
# ggsave(f6,file="Fig6_Col.WGCNA.modules.vs.my.category.table.s.092817.png",width=11,height=8,path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables"))
ggsave(f6,file="Fig6_Col.WGCNA.modules.vs.my.category.table.s.110917.pdf",width=11,height=8,path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables"))
ggsave(f6,file="Fig6_Col.WGCNA.modules.vs.my.category.table.s.110917.png",width=11,height=8,path=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables"))

f6 # magenta indicates pvalue is smaller than 0.01.
```

# Checking expression pattern of gene of interests (absolute and shade responsiveness) (060515)
```{r}
rm(list=ls())
# absolute value
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","summary.vst.Rdata"))
summary.vst.kazu<-summary.vst[,gsub("([[:digit:]]+)(H|L)(1|4|16|25|49)","\\1",names(summary.vst)) %in% 1:19] 
# plot
p<-expression.pattern.graph3(summary.vst.kazu,"AT2G46970") # PIL1
p
p.Col<-expression.pattern.graph3(summary.vst.kazu[,c("1H1","1H25","1H4","1H49","1L1","1L25","1L4","1L49")],"AT2G46970") #PIL1, Col

# 100717
SAS.expression.vst<-read.csv(file.path("..","..","Nozue2016_SAStranscriptome_data","input","kazu.SAS.expression.VST.092517.csv"),row.names=1) # new transformed data by DEseq2 (092617) # not found? (092817)
SAS.expression.vst.Col<-SAS.expression.vst[,samples.PLoSGenetics2015mutants.errrDE[samples.PLoSGenetics2015mutants.errrDE$genotype=="1","file"]]
# remove splicing variant info in rownames
rownames(SAS.expression.vst.Col)<-gsub("([[:print:]]+)(.)([[:digit:]]+)","\\1",rownames(SAS.expression.vst.Col))
p.Col<-expression.pattern.graph3b(data.mean=summary.vst.kazu[,c("1H1","1H25","1H4","1H49","1L1","1L25","1L4","1L49")],data.expression=SAS.expression.vst.Col,target.genes="AT2G46970") #PIL1, Col

p.Col<-expression.pattern.graph3b(data.mean=summary.vst.kazu[,c("1H1","1H25","1H4","1H49","1L1","1L25","1L4","1L49")],data.expression=SAS.expression.vst.Col,target.genes=c("AT2G46970","AT4G16780","AT4G32280")) #Col, PIL1 ("AT2G46970"), ATHB2("AT4G16780"), IAA29 ("AT4G32280") # barplot and jitter overlapped. looks good

```


# GOseq analysis of misexpressed geens in phyB (sun) and public available phytochrome mutnats (exposed to long light condition) (060815) (100517)
```{r eval=FALSE}
# my timecourse RNAseq data
#setwd("/Volumes/Data8/NGS_related/Arabidopsis_analysis/SAS_muts_time_course_RNAseq2_mycomp_output")
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","edgeR","outlier_wo16h","genotype_specific5","DE.genotype.batch.sun.gt.wo16h6.Rdata"))# sun, phyB vs Col any timepoints
temp.GOseq<-GOseq.ORA(gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",rownames(DE.batch3.gt))) # JA and SA pathways enriched, no auxin?! as I expected.
#write.csv(temp.GOseq,file=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","TableS5_GOseq.phyB.sun.csv")) # no matter how PHYB regulate (either up or down)
# knowing up or down regulated genes
temp.genes<-genes.in.enriched.category(enrich.result=temp.GOseq,gene.list=gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",rownames(DE.batch3.gt)),category.table=Atgo.list) 

# predFC (061015)
predFC.phyB<-predFC(dge.nolow.s.batch3.glm,design.nolow.s.gt.batch3,prior.count=1,dispersion=0.05)
head(predFC.phyB[rownames(predFC.phyB) %in% rownames(DE.batch3.gt),]) #
tail(predFC.phyB[rownames(predFC.phyB) %in% rownames(DE.batch3.gt),]) #
hist(predFC.phyB[,1]) # I do not understand why there are -15.... 

# plot 
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","summary.vst.Rdata"))

# downregulated by PHYB (higher in phyB mutant)
## expression pattern
p.down<-expression.pattern.graph3(summary.vst,gsub("([[:print:]]+)(.[[:digit:]])","\\1",rownames(DE.batch3.gt[DE.batch3.gt[,1]>1,]))[1:10]) # downregulated in phyB at time=1h (logFC.genotype6<0)
p.down<-p.down + labs(title="phyB logFC.genotype6 <0 top 20")
#ggsave("phyB_down_expression_pattern.pdf",height=20,width=8,path="../../Nozue2016_SAStranscriptome_output/figs_tables")
# GOseq 
phyBdown.GOseq<-GOseq.ORA(gsub("([[:print:]]+)(.[[:digit:]])","\\1",rownames(DE.batch3.gt[DE.batch3.gt[,1]>1,]))) # JA and SA pathways enriched, no auxin?! as I expected.
#write.csv(phyBdown.GOseq,file="../../Nozue2016_SAStranscriptome_output/figs_tables/TableS1_phyBdown.GOseq.csv")

# upregulated by PHYB (lower in phyB mutant)
p.up<-expression.pattern.graph3(summary.vst,gsub("([[:print:]]+)(.[[:digit:]])","\\1",rownames(DE.batch3.gt[DE.batch3.gt[,1]< -1,]))[1:10]) # downregulated in phyB at time=1h (logFC.genotype6<0)
p.up
## GOseq
phyBup.GOseq<-GOseq.ORA(gsub("([[:print:]]+)(.[[:digit:]])","\\1",rownames(DE.batch3.gt[DE.batch3.gt[,1]< -1,]))) # JA and SA pathways enriched, no auxin?! as I expected.
#write.csv(phyBup.GOseq,file="../../Nozue2016_SAStranscriptome_output/figs_tables/TableS1_phyBup.GOseq.csv")
# GOseq up or down phyB combined table (TAble S1;100517)
phyBup.GOseq$updown<-"up"
phyBdown.GOseq$updown<-"down"
phyBupdown.GOseq<-rbind(phyBup.GOseq,phyBdown.GOseq)
write.csv(phyBupdown.GOseq,file=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","TableS4_phyBupdown.GOseq.100517.csv"))

#p<-expression.pattern.graph3(summary.vst.large.kazu,gsub("([[:print:]]+)(.[[:digit:]])","\\1",rownames(DE.batch3.gt[DE.batch3.gt[,"logCPM"]< -1,]))) # logCPM is "average log2-counts per million, the average taken over all libraries in y." (help file of glmLRT)

```
# phyABDE seedlings DE analysis (microarray)
```{r}
# high-order phytochrome muntat seedlings (microarray)
# download some cel file from E-GEOD-31587 - Expression data of Arabidopsis thaliana (Ler accession) phytochrome phyABCDE quintuple mutant and phyABDE quadruple mutant in response to red light, and their comparison to WT expression data (http://www.ebi.ac.uk/arrayexpress/experiments/E-GEOD-31587/samples/?query=phytochrome&sortby=organism&sortorder=ascending)
# under construction #
#setwd("/Volumes/Data8/E-GEOD-31587_RAW")
library(affy)
#Data <- ReadAffy(widget=T) # error
CEL.filename<-list.files(pattern=".CEL$",path=file.path("/Volumes","data_work","Data8","E-GEOD-31587_RAW"))
CEL.filename
Data <- ReadAffy(filenames=file.path("/Volumes","data_work","Data8","E-GEOD-31587_RAW",CEL.filename))
Data
eset <- rma(Data) # normalize chips 
affyData <- read.csv(file.path("/Volumes","data_work","Data6","bioconductor_R","affy_ATH1_array_elements-2010-12-20.csv")) #read TAIR annotation data from file (TAIR10)
affyData2<-affyData[,c("array_element_name","is_control","locus","description","chromosome","start","stop")]
names(affyData2) <- c("element","is.control","locus","description","chromosome","start","stop")#get rid of underlines
head(affyData2)
genes <- data.frame(element=featureNames(eset)) #get affy gene names from dataset
affyMatch <- merge(affyData2,genes,by.x="element",by.y=1)# reorder the annotation data to match chip data
exp.values <- data.frame(exprs(eset)) #put normalized intensities in a data frame
head(exp.values)
exp.values <- cbind(affyMatch,exp.values) #add annotation to same frame as expression values
head(exp.values)

write.csv(exp.values,file=file.path("/Volumes","data_work","Data8","E-GEOD-31587_RAW","exp.values.csv"))
######## RP analysis
library(RankProd)
exp.values<-read.csv(file.path("/Volumes","data_work","Data8","E-GEOD-31587_RAW","exp.values.csv"),row.names=1)
#exp.values<-exp.values[,-1]
names(exp.values)
names(exp.values)[c(8:12)] # should be six "CEL" file names
RP.E_GEOD_31587.phyABDE <- RP(exp.values[,c(8:12)], rep(c(1,0),c(3,2)), num.perm=100) #  0 (class 1) = Col, 1 (class 2)= phyABDE
topgene.RP.E_GEOD_31587.phyABDE<-topGene(RP.E_GEOD_31587.phyABDE,cutoff=0.01,gene.names=exp.values$element,logged=TRUE,logbase=2)
#  repressed genes in phyABDE
topgene.RP.E_GEOD_31587.phyABDE.T2<-as.data.frame(topgene.RP.E_GEOD_31587.phyABDE$Table2)
topgene.RP.E_GEOD_31587.phyABDE.T2$element<-rownames(topgene.RP.E_GEOD_31587.phyABDE.T2)
topgene.RP.E_GEOD_31587.phyABDE.T2.m<-merge(topgene.RP.E_GEOD_31587.phyABDE.T2, affyData2,by="element")
head(topgene.RP.E_GEOD_31587.phyABDE.T2.m)
topgene.RP.E_GEOD_31587.phyABDE.T2.m.exp.value<-merge(topgene.RP.E_GEOD_31587.phyABDE.T2.m, exp.values[,c(1,8:12)],by="element")
topgene.RP.E_GEOD_31587.phyABDE.T2.m.exp.value<=topgene.RP.E_GEOD_31587.phyABDE.T2.m.exp.value[order(topgene.RP.E_GEOD_31587.phyABDE.T2.m.exp.value$pfp),]

#  induced genes in phyABDE
topgene.RP.E_GEOD_31587.phyABDE.T1<-as.data.frame(topgene.RP.E_GEOD_31587.phyABDE$Table1)
topgene.RP.E_GEOD_31587.phyABDE.T1$element<-rownames(topgene.RP.E_GEOD_31587.phyABDE.T1)
topgene.RP.E_GEOD_31587.phyABDE.T1.m<-merge(topgene.RP.E_GEOD_31587.phyABDE.T1, affyData2,by="element")
head(topgene.RP.E_GEOD_31587.phyABDE.T1.m)
topgene.RP.E_GEOD_31587.phyABDE.T1.m.exp.value<-merge(topgene.RP.E_GEOD_31587.phyABDE.T1.m, exp.values[,c(1,8:12)],by="element")
topgene.RP.E_GEOD_31587.phyABDE.T1.m.exp.value<-topgene.RP.E_GEOD_31587.phyABDE.T1.m.exp.value[order(topgene.RP.E_GEOD_31587.phyABDE.T1.m.exp.value$pfp),]

# summary
topgene.RP.RP.E_GEOD_31587.phyABDE.summary<-rbind(topgene.RP.E_GEOD_31587.phyABDE.T1.m.exp.value,topgene.RP.E_GEOD_31587.phyABDE.T2.m.exp.value)
write.csv(topgene.RP.RP.E_GEOD_31587.phyABDE.summary,file=file("/Volumes","data_work","Data8","E-GEOD-31587_RAW","topgene.RP.RP.E_GEOD_31587.phyABDE.summary.csv"))
# GOseq
topgene.RP.RP.E_GEOD_31587.phyABDE.summary.locus<-topgene.RP.RP.E_GEOD_31587.phyABDE.summary[nchar(topgene.RP.RP.E_GEOD_31587.phyABDE.summary$locus)==9,"locus"]
phyABDE.seedlings.GOseq<-GOseq.ORA(topgene.RP.RP.E_GEOD_31587.phyABDE.summary.locus)
# write
write.csv(phyABDE.seedlings.GOseq,file=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","TableS1_phyABDE.GOseq.summary.csv"))
phyABDE.seedlings.GOseq.genes<-genes.in.enriched.category(enrich.result=temp.GOseq,gene.list=topgene.RP.RP.E_GEOD_31587.phyABDE.summary.locus,category.table=Atgo.list) #  no auxin?! as I expected.
```

# Does each WGCNA module contain shade-responsive genes? (011216; 100417)
```{r}
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","SAS.expression.vst.s.kazu.largeCV.modules092817.Rdata") ) #
SAS.expression.vst.s.kazu.largeCV.modules
SAS.expression.vst.s.kazu.largeCV.modules$AGI<-gsub("([[:print:]]+)(.[[:digit:]]+)","\\1",SAS.expression.vst.s.kazu.largeCV.modules$AGI)
# shade responsive genes 
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output","DGEsummary.edgeR.response.Rdata")) # this has all genotypes
head(DGEsummary.edgeR.response)
DGEsummary.edgeR.response$AGI<-gsub("([[:print:]]+)(.[[:digit:]]+)(.)([[:digit:]]+)","\\1",rownames(DGEsummary.edgeR.response))
DGEsummary.edgeR.response$gt<-gsub("([[:print:]]+)(.[[:digit:]]+)(.)([[:digit:]]+)","\\4",rownames(DGEsummary.edgeR.response))
Col.DGE<-DGEsummary.edgeR.response[DGEsummary.edgeR.response$gt=="1",]

# mark Col.DGE genes in SAS.expression.vst.s.kazu.largeCV.modules
SAS.expression.vst.s.kazu.largeCV.modules$ColDGE<-(SAS.expression.vst.s.kazu.largeCV.modules$AGI %in% Col.DGE$AGI)*1
# 
overlap.WGCNAmodule.ColDGE<-overlapTable(SAS.expression.vst.s.kazu.largeCV.modules$"1",SAS.expression.vst.s.kazu.largeCV.modules$ColDGE)
save(overlap.WGCNAmodule.ColDGE,file=file.path("..","..","Nozue2016_SAStranscriptome_output/output/overlap.WGCNAmodule.ColDGE.Rdata"))

# add WGCNAmodule
temp<-merge(Col.DGE,TAIR10_gene_descriptions[,c("AGI","colE...col.F")],by="AGI")
#SAS.expression.vst.s.kazu.largeCV.modules$AGI<-gsub("([[:print:]]+)(.[[:digit:]]+)(.)([[:digit:]]+)","\\1",SAS.expression.vst.s.kazu.largeCV.modules$AGI)
temp2<-merge(SAS.expression.vst.s.kazu.largeCV.modules[,c("AGI","1")],temp)
Col.DGE.descripiton<-temp2[!duplicated(temp2$AGI),]
colnames(Col.DGE.descripiton)<-gsub("1","WGCNAmodule",colnames(Col.DGE.descripiton))
write.csv(Col.DGE.descripiton,file=file.path("..","..","Nozue2016_SAStranscriptome_output","output","Col.DGE.descripiton.csv"))

```
# Are salmon genes modulated in Sessa (2007) 4d samples?
```{r}
Sessa.4d<-read.csv("../../Nozue2016_SAStranscriptome_data/input/sessa4days.csv")
Sessa.1h<-read.csv("../../Nozue2016_SAStranscriptome_data/input/sessa1hour.csv")
DE.genotype.batch.genes.WGCNAmodules<-read.csv("../../Nozue2016_SAStranscriptome_output/figs_tables/DE.genotype.batch.genes.WGCNAmodules.csv")
salmon<-DE.genotype.batch.genes.WGCNAmodules[DE.genotype.batch.genes.WGCNAmodules$X1=="salmon",]
salmon$AGI
dim(Sessa.4d[Sessa.4d$ID %in% salmon$AGI,])
Sessa.4d.salmon<-Sessa.4d[Sessa.4d$ID %in% salmon$AGI,]
write.csv(Sessa.4d.salmon,file="Sessa.4d.salmon.csv")
ConvictionsSessa.1h[Sessa.1h$ID %in% salmon$AGI,]
length(salmon$AGI)
Sessa.4d.s<-Sessa.4d[Sessa.4d$adj.P.Val<0.01,]
# Fisher's exact test (assuming total gene number is 25000)
Sessa.4d.salmon<-matrix(c(11,1941-11,112-11,25000-1941-(112-11)),nrow=2)
fisher.test(Sessa.4d.salmon) # p-value = 0.3768
```
# Salmon module genes are shade responsive in Col? (100717)
```{r}
#setwd(homedir)
DEfiles<-list.files(path=file.path("..","..","Nozue2016_SAStranscriptome_output","output"),pattern="DE.batch.anytime")
DEfiles<-grep("DE.batch.anytime_genotype_wo_outlier_wo16h.PLoSGenetics2015mutants.errrDE",DEfiles,value=TRUE)
#setwd("../../Nozue2016_SAStranscriptome_output/output/")
load(file.path("..","..","Nozue2016_SAStranscriptome_output","output",DEfiles[1])) 
rownames(DE.batch.anytime)<-paste(rownames(DE.batch.anytime),gsub("(DE.batch.anytime_genotype_wo_outlier_wo16h.PLoSGenetics2015mutants.errrDE)([[:print:]]+)(.Rdata)","\\2",DEfiles[1]),sep=".")
# salmon
#test<-read.csv("../Nozue2016_SAStranscriptome_output/figs_tables/Dataset7.csv")
DE.genotype.batch.genes.WGCNAmodules.description<-read.csv(file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Supplemental_Dataset4_Custom_category_enrichment_analysis_of_WGCNA_gene_co-expression_modules092817.csv"),row.names=1)
dim(DE.genotype.batch.genes.WGCNAmodules.description) # 17985   28 > 33602 31
salmon<-DE.genotype.batch.genes.WGCNAmodules.description[DE.genotype.batch.genes.WGCNAmodules.description$WGCNAmodules=="salmon",]
salmon$AGI
#
DE.batch.anytime[gsub("([[:print:]]+)(.)([[:digit:]]+)(.1)","\\1",rownames(DE.batch.anytime)) %in% salmon$AGI,] # none
# Salmon module genes do not have any shade responsive genes in Col...

```
# Salmon module genes vs 49h shade responsive genes in Col (old version; 100717s)
```{r}
DE.genotype.batch.genes.WGCNAmodules<-read.csv("../../Nozue2016_SAStranscriptome_output/figs_tables/DE.genotype.batch.genes.WGCNAmodules.csv")
salmon<-DE.genotype.batch.genes.WGCNAmodules[DE.genotype.batch.genes.WGCNAmodules$X1=="salmon",]
load("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/samples.nolow5.Rdata")
load("../../Nozue2016_SAStranscriptome_output/output/edgeR/outlier_wo16h/dge.nolow5.Rdata")
  samples.nolow.s<-samples.nolow5[samples.nolow5$genotype==1,]
  dge.nolow.s<-dge.nolow5[,samples.nolow5$genotype==1]
  dge.nolow.s<-dge.nolow.s[rowSums(dge.nolow.s$counts > 5) >= 3,] 
# subset only 49h
samples.nolow.s.49h<-samples.nolow.s[samples.nolow.s$time=="49",]
design.nolow.s.batch.fixedtime<- model.matrix(~samples.nolow.s.49h$trt + samples.nolow.s.49h$batch)
colnames(design.nolow.s.batch.fixedtime) <- gsub("samples.nolow.s.49h$","",colnames(design.nolow.s.batch.fixedtime),fixed=TRUE) # nicer column names
colnames(design.nolow.s.batch.fixedtime)
dge.nolow.s.fixedtime<-dge.nolow5[,samples.nolow5$file %in% samples.nolow.s.49h$file]
  print(design.nolow.s.batch.fixedtime)
  print(dge.nolow.s.fixedtime)
  dge.nolow.s.batch.fixedtime.glm <- estimateGLMCommonDisp(dge.nolow.s.fixedtime,design.nolow.s.batch.fixedtime, verbose=TRUE) 
  dge.nolow.s.batch.glm <- estimateGLMTrendedDisp(dge.nolow.s.batch.fixedtime.glm,design.nolow.s.batch.fixedtime, verbose=TRUE)
  dge.nolow.s.batch.fixedtime.glm.fit <- glmFit(dge.nolow.s.batch.fixedtime.glm,design.nolow.s.batch.fixedtime)
  dge.nolow.s.batch.fixedtime.glm.lrt <- glmLRT(dge.nolow.s.batch.fixedtime.glm.fit,coef="trtL")
  DE.batch.49h.Col<-topTags(dge.nolow.s.batch.fixedtime.glm.lrt,n=2000)$table[topTags(dge.nolow.s.batch.fixedtime.glm.lrt,n=2000)$table$FDR<0.05,]
# 
dim(DE.batch.49h.Col) # 278 5
DE.batch.49h.Col.salmon<-DE.batch.49h.Col[sub("([[:print:]]+)(.[[:digit:]]+)","\\1",rownames(DE.batch.49h.Col)) %in% salmon$AGI,]
setwd("../../Nozue2016_SAStranscriptome_output/output/")
write.csv(DE.batch.49h.Col.salmon,file="DE.batch.49h.Col.salmon.csv")
setwd("../../Nozue2016_SAStranscriptome_scripts/")
```
# appendix 1: differentially expressed geens used for ORA
```{r eval=FALSE}
# differential expression analysis with npr1-1
# http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE51626
# Singh M, Bag SK, Bhardwaj A, Ranjan A et al. Global nucleosome positioning 
# regulates salicylic acid mediated transcription in Arabidopsis thaliana. BMC Plant # Biol 2015 Jan 21;15:13. PMID: 25604550
# 24 h SA or mock treatment (thnaks for a Plant Cell reviewer)
###  The RNA of col-0 and npr1-1 mutant were isolated after 24hr water and SA treatment.The ATH1 gene chip was used for expression analysis.
###  GSM1249675  Col-0 water, biological replicate 1
###  GSM1249676  Col-0 water, biological replicate 2
###  GSM1249677  Col-0 SA, biological replicate 1
###  GSM1249678  Col-0 SA, biological replicate 2
###  GSM1249679	npr1-1 water, biological replicate 1
###  GSM1249680	npr1-1 water, biological replicate 2
###  GSM1249681	npr1-1 SA, biological replicate 1
###  GSM1249682	npr1-1 SA, biological replicate 2
setwd(hormedir)
setwd("../../Nozue2016_SAStranscriptome_data/input/GSE51626_RAW")
library(affy)
#Data <- ReadAffy(widget=T) # error
CEL.filename<-list.files(pattern=".CEL$")
CEL.filename
Data <- ReadAffy(filenames=CEL.filename)
Data
eset <- rma(Data) # normalize chips 
affyData <- read.csv("../affy_ATH1_array_elements-2010-12-20.csv") #read TAIR annotation data from file (TAIR10)
affyData2<-affyData[,c("array_element_name","is_control","locus","description","chromosome","start","stop")]
names(affyData2) <- c("element","is.control","locus","description","chromosome","start","stop")#get rid of underlines
head(affyData2)
genes <- data.frame(element=featureNames(eset)) #get affy gene names from dataset
affyMatch <- merge(affyData2,genes,by.x="element",by.y=1)# reorder the annotation data to match chip data
exp.values <- data.frame(exprs(eset)) #put normalized intensities in a data frame
head(exp.values)
exp.values <- cbind(affyMatch,exp.values) #add annotation to same frame as expression values
head(exp.values)
setwd(homedir)
write.csv(exp.values,file="../../Nozue2016_SAStranscriptome_data/input/GSE51626_RAW/exp.values.csv")
######## RP analysis
library(RankProd)
exp.values<-read.csv("../../Nozue2016_SAStranscriptome_data/input/GSE51626_RAW/exp.values.csv")
names(exp.values)
## Col (water) vs npr1 (water)
names(exp.values)[c(9:10,13:14)]
RP.GSE51626.npr1 <- RP(exp.values[,c(9:10,13:14)], rep(1:0,each=2), num.perm=100) # water control, 0 (class 1) = Col, 1 (class 2)= npr1-1 
topgene.RP.GSE51626.npr1<-topGene(RP.GSE51626.npr1,cutoff=0.01,gene.names=exp.values$element,logged=TRUE,logbase=2)
#  induced genes in npr1-1
topgene.RP.GSE51626.npr1.T2<-as.data.frame(topgene.RP.GSE51626.npr1$Table2)
topgene.RP.GSE51626.npr1.T2$element<-rownames(topgene.RP.GSE51626.npr1.T2)
topgene.RP.GSE51626.npr1.T2.m<-merge(topgene.RP.GSE51626.npr1.T2, affyData2,by="element")
head(topgene.RP.GSE51626.npr1.T2.m)
topgene.RP.GSE51626.npr1.T2.m.exp.value<-merge(topgene.RP.GSE51626.npr1.T2.m, exp.values[,c(2,9:10,13:14)],by="element")
topgene.RP.GSE51626.npr1.T2.m.exp.value<-topgene.RP.GSE51626.npr1.T2.m.exp.value[order(topgene.RP.GSE51626.npr1.T2.m.exp.value$pfp),]

#  repressed genes in npr1-1
topgene.RP.GSE51626.npr1.T1<-as.data.frame(topgene.RP.GSE51626.npr1$Table1)
topgene.RP.GSE51626.npr1.T1$element<-rownames(topgene.RP.GSE51626.npr1.T1)
topgene.RP.GSE51626.npr1.T1.m<-merge(topgene.RP.GSE51626.npr1.T1, affyData2,by="element")
head(topgene.RP.GSE51626.npr1.T1.m)
topgene.RP.GSE51626.npr1.T1.m.exp.value<-merge(topgene.RP.GSE51626.npr1.T1.m, exp.values[,c(2,9:10,13:14)],by="element")
topgene.RP.GSE51626.npr1.T1.m.exp.value<-topgene.RP.GSE51626.npr1.T1.m.exp.value[order(topgene.RP.GSE51626.npr1.T1.m.exp.value$pfp),]

# summary
topgene.RP.GSE51626.npr1.summary<-rbind(topgene.RP.GSE51626.npr1.T1.m.exp.value,topgene.RP.GSE51626.npr1.T2.m.exp.value)
write.csv(topgene.RP.GSE51626.npr1.summary,file="../../Nozue2016_SAStranscriptome_data/input/GSE51626_RAW/topgene.RP.GSE51626.npr1.summary.csv")

# npr1 responsive (24h SA treated) (110916)
names(exp.values)[c(11:12,15:16)] # 
RP.GSE51626.SA.npr1 <- RP(exp.values[,c(11:12,15:16)], rep(1:0,each=2), num.perm=100) # water control, 0 (class 1) = SA + Col, 1 (class 2)= SA + npr1
topgene.RP.GSE51626.SA.npr1<-topGene(RP.GSE51626.SA.npr1,cutoff=0.01,gene.names=exp.values$element,logged=TRUE,logbase=2)
#  induced genes by SA in Col
topgene.RP.GSE51626.SA.npr1.T2<-as.data.frame(topgene.RP.GSE51626.SA.npr1$Table2)
topgene.RP.GSE51626.SA.npr1.T2$element<-rownames(topgene.RP.GSE51626.SA.npr1.T2)
topgene.RP.GSE51626.SA.npr1.T2.m<-merge(topgene.RP.GSE51626.SA.npr1.T2, affyData2,by="element")
head(topgene.RP.GSE51626.SA.npr1.T2.m)
topgene.RP.GSE51626.SA.npr1.T2.m.exp.value<-merge(topgene.RP.GSE51626.SA.npr1.T2.m, exp.values[,c(2,11:12,15:16)],by="element")
topgene.RP.GSE51626.SA.npr1.T2.m.exp.value<-topgene.RP.GSE51626.SA.npr1.T2.m.exp.value[order(topgene.RP.GSE51626.SA.npr1.T2.m.exp.value$pfp),]

#  repressed genes by SA
topgene.RP.GSE51626.SA.npr1.T1<-as.data.frame(topgene.RP.GSE51626.SA.npr1$Table1)
topgene.RP.GSE51626.SA.npr1.T1$element<-rownames(topgene.RP.GSE51626.SA.npr1.T1)
topgene.RP.GSE51626.SA.npr1.T1.m<-merge(topgene.RP.GSE51626.SA.npr1.T1, affyData2,by="element")
head(topgene.RP.GSE51626.SA.npr1.T1.m)
topgene.RP.GSE51626.SA.npr1.T1.m.exp.value<-merge(topgene.RP.GSE51626.SA.npr1.T1.m, exp.values[,c(2,11:12,15:16)],by="element")
topgene.RP.GSE51626.SA.npr1.T1.m.exp.value<-topgene.RP.GSE51626.SA.npr1.T1.m.exp.value[order(topgene.RP.GSE51626.SA.npr1.T1.m.exp.value$pfp),]

# summary
topgene.RP.GSE51626.SA.npr1.summary<-rbind(topgene.RP.GSE51626.SA.npr1.T1.m.exp.value,topgene.RP.GSE51626.SA.npr1.T2.m.exp.value)
write.csv(topgene.RP.GSE51626.SA.npr1.summary,file="../../Nozue2016_SAStranscriptome_data/input/GSE51626_RAW/topgene.RP.GSE51626.SA.npr1.summary.csv")

# SA responsive
## Col (water) vs Col (SA)
names(exp.values)[c(9:12)]
RP.GSE51626.SA.Col <- RP(exp.values[,c(9:12)], rep(1:0,each=2), num.perm=100) # water control, 0 (class 1) = Col, 1 (class 2)= SA Col
topgene.RP.GSE51626.SA.Col<-topGene(RP.GSE51626.SA.Col,cutoff=0.01,gene.names=exp.values$element,logged=TRUE,logbase=2)
#  induced genes in npr1-1
topgene.RP.GSE51626.SA.Col.T2<-as.data.frame(topgene.RP.GSE51626.SA.Col$Table2)
topgene.RP.GSE51626.SA.Col.T2$element<-rownames(topgene.RP.GSE51626.SA.Col.T2)
topgene.RP.GSE51626.SA.Col.T2.m<-merge(topgene.RP.GSE51626.SA.Col.T2, affyData2,by="element")
head(topgene.RP.GSE51626.SA.Col.T2.m)
topgene.RP.GSE51626.SA.Col.T2.m.exp.value<-merge(topgene.RP.GSE51626.SA.Col.T2.m, exp.values[,c(2,9:12)],by="element")
topgene.RP.GSE51626.SA.Col.T2.m.exp.value<-topgene.RP.GSE51626.SA.Col.T2.m.exp.value[order(topgene.RP.GSE51626.SA.Col.T2.m.exp.value$pfp),]

#  SA repressed genes
topgene.RP.GSE51626.SA.Col.T1<-as.data.frame(topgene.RP.GSE51626.SA.Col$Table1)
topgene.RP.GSE51626.SA.Col.T1$element<-rownames(topgene.RP.GSE51626.SA.Col.T1)
topgene.RP.GSE51626.SA.Col.T1.m<-merge(topgene.RP.GSE51626.SA.Col.T1, affyData2,by="element")
head(topgene.RP.GSE51626.SA.Col.T1.m)
topgene.RP.GSE51626.SA.Col.T1.m.exp.value<-merge(topgene.RP.GSE51626.SA.Col.T1.m, exp.values[,c(2,9:12)],by="element")
topgene.RP.GSE51626.SA.Col.T1.m.exp.value<-topgene.RP.GSE51626.SA.Col.T1.m.exp.value[order(topgene.RP.GSE51626.SA.Col.T1.m.exp.value$pfp),]

# summary
topgene.RP.GSE51626.SA.Col.summary<-rbind(topgene.RP.GSE51626.SA.Col.T1.m.exp.value,topgene.RP.GSE51626.SA.Col.T2.m.exp.value)
write.csv(topgene.RP.GSE51626.SA.Col.summary,file="../../Nozue2016_SAStranscriptome_data/input/GSE51626_RAW/topgene.RP.GSE51626.SA.Col.summary.csv")

# VennDiagram (111116) see more in https://rstudio-pubs-static.s3.amazonaws.com/13301_6641d73cfac741a59c0a851feb99e98b.html. That is Ruijuan is using for her color VennDiagram.
library(VennDiagram)
setwd(homedir)
# reading DE gene list tables
GSE51626.Venn<-data.frame(element=affyData$array_element_name,AGI=affyData$locus,npr1=" ",SA.npr1=" ",SA.Col=" ")
topgene.RP.GSE51626.npr1.summary<-read.csv("../../Nozue2016_SAStranscriptome_data/input/GSE51626_RAW/topgene.RP.GSE51626.npr1.summary.csv");topgene.RP.GSE51626.npr1.summary<-topgene.RP.GSE51626.npr1.summary[,-1]
topgene.RP.GSE51626.SA.npr1.summary<-read.csv("../../Nozue2016_SAStranscriptome_data/input/GSE51626_RAW/topgene.RP.GSE51626.SA.npr1.summary.csv");topgene.RP.GSE51626.SA.npr1.summary<-topgene.RP.GSE51626.SA.npr1.summary[,-1]
topgene.RP.GSE51626.SA.Col.summary<-read.csv("../../Nozue2016_SAStranscriptome_data/input/GSE51626_RAW/topgene.RP.GSE51626.SA.Col.summary.csv");topgene.RP.GSE51626.SA.Col.summary<-topgene.RP.GSE51626.SA.Col.summary[,-1]

GSE51626.Venn[GSE51626.Venn$element %in% topgene.RP.GSE51626.npr1.summary$element,"npr1"]<-"npr1"
GSE51626.Venn[GSE51626.Venn$element %in% topgene.RP.GSE51626.SA.npr1.summary$element,"SA.npr1"]<-"SA.npr1"
GSE51626.Venn[GSE51626.Venn$element %in% topgene.RP.GSE51626.SA.Col.summary$element,"SA.Col"]<-"SA.Col"
# gene of interest in the VennDiagram  PCC1, XBAT34, AIG1 (PCC1 (  AT3G22231; orange, XBAT34 (AT4G14365, orange), but AIG1 (thistle2;  AT1G33960), ACD6 (AT4G14400, ?)
GSE51626.Venn[GSE51626.Venn$AGI %in% c("AT3G22231","AT4G14365","AT1G33960","AT4G14400"),] #
GSE51626.Venn[GSE51626.Venn$npr1=="npr1",]
# update hormone responsiveness data set (111116)

# no need to make data frame! Use use element name for each set
setwd("../../Nozue2016_SAStranscriptome_output/output/")
venn.plot<-venn.diagram(list(npr1=topgene.RP.GSE51626.npr1.summary$element,SA.npr1=topgene.RP.GSE51626.SA.npr1.summary$element,SA.Col=topgene.RP.GSE51626.SA.Col.summary$element),filename="Venn_npr1.png")
# up and down
setwd("/Volumes/data_work/Data8/NGS_related/Arabidopsis_analysis") # for temporal system
group_comparison.Ver5<-read.csv("group_comparison.Ver5.csv") 
# target categories
categories<-c("IAAup","IAAdown","BLup","BLdown","GAinga1.3UP","GAinga1.3DOWN","BLup","BLdown","MJup","MJdown","ABAup","ABAdown","ACCup","ACCdown","CKup","CKdown","SA_UP","SA_DOWN",
              "SARplus","SARminus","GRF1_3target","PIFtarget","MYC234_up","MYC234_down","NPR1_up","NPR1_down","WRKY33up","WRKY33down","PCC1up","PCC1down","SAup_24h","SAdown_24h","NPR1up_uponSA","NPR1down_uponSA")
# rename categories
group_comparison.Ver5.selected<-group_comparison.Ver5[,names(group_comparison.Ver5[-c(1:3),]) %in% categories]
#names(group_comparison.Ver5.selected)<-c("ABAup","ABAdown","ACCup","ACCdown","BLup","BLdown","IAAup","IAAdown","MJup","MJdown","CKup","CKdown","GAup","GAdown",
                                          "SAup_2h","SAdown_2h","SARminus","SARplus","GRF1_3target","PIFtarget","MYC234up","MYC234down","NPR1up","NPR1down","WRKY33up","WRKY33down","PCC1up","PCC1down","SAup_24h","SAdown_24h","NPR1up_uponSA","NPR1down_uponSA")
# "WRKY33up" and "WRKY33donw" are reversed! (correct group_comparison.Ver5.csv)
names(group_comparison.Ver5.selected)<-gsub("MYC234_","MYC234",names(group_comparison.Ver5.selected))
names(group_comparison.Ver5.selected)<-gsub("GAinga1.3","GA",names(group_comparison.Ver5.selected))
names(group_comparison.Ver5.selected)<-gsub("SA_UP","SAup_2h",names(group_comparison.Ver5.selected))
names(group_comparison.Ver5.selected)<-gsub("SA_DOWN","SAdown_2h",names(group_comparison.Ver5.selected))
names(group_comparison.Ver5.selected)<-gsub("UP","up",names(group_comparison.Ver5.selected))
names(group_comparison.Ver5.selected)<-gsub("DOWN","down",names(group_comparison.Ver5.selected))
names(group_comparison.Ver5.selected)<-gsub("NPR1_","NPR1",names(group_comparison.Ver5.selected))
names(group_comparison.Ver5.selected)
setwd(homedir)
setwd("../../Nozue2016_SAStranscriptome_output/output/")

temp.list<-list(NPR1.up=group_comparison.Ver5.selected$NPR1up[-(1:3)],NPR1.down=group_comparison.Ver5.selected$NPR1down[-(1:3)],NPR1up_uponSA=group_comparison.Ver5.selected$NPR1up_uponSA[-(1:3)],NPR1down_uponSA=group_comparison.Ver5.selected$NPR1down_uponSA[-(1:3)])
temp.list2<-lapply(temp.list,as.character)
venn.plot2<-venn.diagram(temp.list2,filename="Venn_npr2.png")
######


# ### other SA microarray
# http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE3984
# 3w to 4w old plants were sprayed with mock or SA with 0.01% Silwet. Leaves were collected 2h after treatment.
# Supplementary data files not provided
# use SOFT file (http://www2.warwick.ac.uk/fac/sci/moac/people/students/peter_cock/r/geo/)
source("http://www.bioconductor.org/biocLite.R")
setwd("/Volumes/Data8/")
biocLite("GEOquery")
library(Biobase)
library(GEOquery)
#Download GDS file, put it in the current directory, and load it:
#gds3984 <- getGEO('GSE3984', destdir=".")

#Or, open an existing GDS file (even if its compressed):
gse3984 <- getGEO(filename='GSE3984_family.soft.gz')
Meta(gds3984)
eset <- GDS2eSet(gds3984, do.log2=TRUE)
# error
# see  "4.4  Converting GSE to an ExpressionSet" in vignet of GEOquery
gsmplatforms <- lapply(GSMList(gse3984),function(x) {Meta(x)$platform})
Table(GSMList(gse3984)[[1]])[1:5,]
probesets <- Table(GPLList(gse3984)[[1]])$ID
data.matrix <- do.call('cbind',lapply(GSMList(gse3984),function(x) 
{tab <- Table(x) 
 mymatch <- match(probesets,tab$ID_REF) 
 return(tab$VALUE[mymatch])
}))
data.matrix <- apply(data.matrix,2,function(x) {as.numeric(as.character(x))})
hist(data.matrix) # log2 transformed data
#data.matrix <- log2(data.matrix) # is this already transformed?
data.matrix[1:5,]
rownames(data.matrix) <- probesets
colnames(data.matrix) <- names(GSMList(gse3984))
pdata <- data.frame(samples=names(GSMList(gse3984)))
rownames(pdata) <- names(GSMList(gse3984))
pheno <- as(pdata,"AnnotatedDataFrame")
eset2 <- new('ExpressionSet',exprs=data.matrix,phenoData=pheno)
eset2
genes <- data.frame(element=featureNames(eset2)) #get affy gene names from dataset
affyMatch <- merge(affyData2,genes,by.x="element",by.y=1)# reorder the annotation data to match chip data
exp.values <- data.frame(exprs(eset2)) #put normalized intensities in a data frame
head(exp.values)
exp.values <- cbind(affyMatch,exp.values) #add annotation to same frame as expression values
head(exp.values)
# GSM90867  2-hour mock-treated leaf material biological replicate 1
# GSM90868  2-hour mock-treated leaf material biological replicate 2
# GSM90869  2-hour mock-treated leaf material biological replicate 3
# GSM90870	leaf material harvested 2 hours after spraying of 1 mM salicylic acid biological replicate 1
# GSM90871	leaf material harvested 2 hours after spraying of 1 mM salicylic acid biological replicate 2
# GSM90872	leaf material harvested 2 hours after spraying of 1 mM salicylic acid biological replicate 3
# RP
library(RankProd)
RP.ThibaudNissen.SA <- RP(exp.values[,c(8:13)], rep(1:0,each=3), num.perm=100) # 0 (class 1) = mock, 1 (class 2)= SA
#  induced genes
topgene.RP.ThibaudNissen.SA<-topGene(RP.ThibaudNissen.SA,cutoff=0.01,gene.names=rownames(exp.values),logged=TRUE,logbase=2)
topgene.RP.ThibaudNissen.SA.T2<-as.data.frame(topgene.RP.ThibaudNissen.SA$Table2)
topgene.RP.ThibaudNissen.SA.T2$element<-rownames(topgene.RP.ThibaudNissen.SA.T2)
topgene.RP.ThibaudNissen.SA.T2.m<-merge(topgene.RP.ThibaudNissen.SA.T2, affyData2,by="element")
head(topgene.RP.ThibaudNissen.SA.T2.m)
#  repressed genes
topgene.RP.ThibaudNissen.SA.T1<-as.data.frame(topgene.RP.ThibaudNissen.SA$Table1)
topgene.RP.ThibaudNissen.SA.T1$element<-rownames(topgene.RP.ThibaudNissen.SA.T1)
topgene.RP.ThibaudNissen.SA.T1.m<-merge(topgene.RP.ThibaudNissen.SA.T1, affyData2,by="element")
head(topgene.RP.ThibaudNissen.SA.T1.m)

## limma (see limmar user guide, 091114)
design <- cbind(mock=c(1,1,1,0,0,0),SA=c(0,0,0,1,1,1))
# or
Group <- factor(c("mock","mock","mock","SA","SA","SA"), levels=c("mock","SA"))
design <- model.matrix(~0+Group)                
colnames(design) <- c("mock","SA")

fit <- lmFit(exp.values[,c(8:13)], design)
cont.matrix <- makeContrasts(mockvsSA=SA-mock, levels=design)
fit2 <- contrasts.fit(fit, cont.matrix)
fit2 <- eBayes(fit2,0.01) # 0.01 is used in GEO2R (http://www.ncbi.nlm.nih.gov/geo/geo2r/?acc=GSE3984)
topgene.limma.ThibaudNissen.SA<-topTable(fit2, adjust="fdr",sort.by="B",number=500) # this is different from GEO2R output ...
topgene.limma.ThibaudNissen.SA<-topgene.limma.ThibaudNissen.SA[topgene.limma.ThibaudNissen.SA$adj.P.Val<0.05,]

# overlap between RP and limma
topgene.RP.ThibaudNissen.SA.T1.m.limma<-topgene.RP.ThibaudNissen.SA.T1.m[topgene.RP.ThibaudNissen.SA.T1.m$element %in% rownames(topgene.limma.ThibaudNissen.SA),]
topgene.RP.ThibaudNissen.SA.T2.m.limma<-topgene.RP.ThibaudNissen.SA.T2.m[topgene.RP.ThibaudNissen.SA.T2.m$element %in% rownames(topgene.limma.ThibaudNissen.SA),]
setwd("/Volumes/Data8/NGS_related/Arabidopsis_analysis")
write.csv(topgene.RP.ThibaudNissen.SA.T1.m.limma,file="topgene.RP.ThibaudNissen.SA.T1.m.limma.csv")
write.csv(topgene.RP.ThibaudNissen.SA.T2.m.limma,file="topgene.RP.ThibaudNissen.SA.T2.m.limma.csv")
```
# appendix 2 summary of custom category table
```{r}
group_comparison.Ver5<-read.csv(file.path("","Volumes","data_work","Data8","NGS_related","Arabidopsis_analysis","group_comparison.Ver5.csv")) #
# target categories (renewd 111417)
categories<-c("IAAup","IAAdown","BLup","BLdown","GAinga1.3UP","GAinga1.3DOWN","BLup","BLdown","MJup","MJdown","ABAup","ABAdown","ACCup","ACCdown","CKup","CKdown","SA_UP","SA_DOWN",
              "SARplus","SARminus","PIFtarget","MYC234_up","MYC234_down","NPR1_up","NPR1_down","WRKY33up","WRKY33down","SAup_24h","SAdown_24h","NPR1up_uponSA","NPR1down_uponSA")

# rename categories
group_comparison.Ver5.selected<-group_comparison.Ver5[,names(group_comparison.Ver5[-c(1:3),]) %in% categories]
names(group_comparison.Ver5.selected)<-gsub("MYC234_","MYC234",names(group_comparison.Ver5.selected))
names(group_comparison.Ver5.selected)<-gsub("GAinga1.3","GA",names(group_comparison.Ver5.selected))
names(group_comparison.Ver5.selected)<-gsub("SA_UP","SAup_2h",names(group_comparison.Ver5.selected))
names(group_comparison.Ver5.selected)<-gsub("SA_DOWN","SAdown_2h",names(group_comparison.Ver5.selected))
names(group_comparison.Ver5.selected)<-gsub("UP","up",names(group_comparison.Ver5.selected))
names(group_comparison.Ver5.selected)<-gsub("DOWN","down",names(group_comparison.Ver5.selected))
names(group_comparison.Ver5.selected)<-gsub("NPR1_","NPR1",names(group_comparison.Ver5.selected))
names(group_comparison.Ver5.selected)

dummy.cat3<-as.data.frame(matrix(nrow=length(names(bias)),ncol=length(names(group_comparison.Ver5.selected))))

dummy.cat3<-as.data.frame(matrix(nrow=length(names(bias)),ncol=1))
rownames(dummy.cat3)<-names(bias) # use "bias" object from function_RNAseq_time_course.R 
for(i in names(group_comparison.Ver5.selected)) {# 
  print(i)
  x.locus<-group_comparison.Ver5.selected[-c(1:3),names(group_comparison.Ver5.selected)==i]
  # clean up x.locus
  x.locus<-x.locus[!x.locus==""]  
  # only single AGI name is extracted (no probe naem, no multiple AGI name, such as "AT1G23900 AT1G23940") 
  dummy.cat3[rownames(dummy.cat3) %in% toupper(x.locus),i]<-rep(i,sum(as.integer(rownames(dummy.cat3) %in% toupper(x.locus))))
}
head(dummy.cat3)
dummy.cat3<-dummy.cat3[,-1]
# clean up <NA>
dummy.cat3[is.na(dummy.cat3)]<-""
# 
### convert into list
temp<-list()
for(i in 1:dim(dummy.cat3)[1]) {
  temp[[i]]<-paste(dummy.cat3[i,])
}
names(temp)<-rownames(dummy.cat3)   
hormone.responsiveness6b<-temp
head(hormone.responsiveness6b)
table(hormone.responsiveness6b[1:10])
attributes(hormone.responsiveness6b)$category<-names(group_comparison.Ver5.selected)
## change list into data.frame
attributes(hormone.responsiveness6b)
hormone.responsiveness6.DF<-data.frame(t(sapply(hormone.responsiveness6b,c)))
names(hormone.responsiveness6.DF)<-attributes(hormone.responsiveness6b)$category
summary(hormone.responsiveness6.DF)
hormone.responsiveness6.DF$AGI<-rownames(hormone.responsiveness6.DF)
# should I combine SA-related categories? (111417)
hormone.responsiveness6.DF$SAup_combined<-with(hormone.responsiveness6.DF,paste(SAup_2h,SARplus,NPR1down,SAup_24h,NPR1up_uponSA),sep="")
hormone.responsiveness6.DF$SAup_combined[!hormone.responsiveness6.DF$SAup_combined=="    "]<-"SAup_combined"
hormone.responsiveness6.DF$SAup_combined[hormone.responsiveness6.DF$SAup_combined=="    "]<-""
table(hormone.responsiveness6.DF$SAup_combined)
# SAdown_combined
hormone.responsiveness6.DF$SAdown_combined<-with(hormone.responsiveness6.DF,paste(SAdown_2h,SARminus,NPR1up,SAdown_24h,NPR1down_uponSA),sep="")
hormone.responsiveness6.DF$SAdown_combined[!hormone.responsiveness6.DF$SAdown_combined=="    "]<-"SAdown_combined"
hormone.responsiveness6.DF$SAdown_combined[hormone.responsiveness6.DF$SAdown_combined=="    "]<-""
table(hormone.responsiveness6.DF$SAdown_combined)
# 
hormone.responsiveness6.DF.s<-hormone.responsiveness6.DF[,c("ABAup","ABAdown","ACCup","ACCdown","BLup","BLdown","IAAup","IAAdown","MJup","MJdown","CKup","CKdown","GAup","GAdown","PIFtarget","MYC234up","MYC234down","SAup_combined","SAdown_combined","WRKY33up","WRKY33down","AGI")] # 110617
summary(hormone.responsiveness6.DF.s)
lapply(hormone.responsiveness6.DF.s[,grep("AGI",colnames(hormone.responsiveness6.DF.s),invert=T)],table)
## for Supplemental Table 5. Supplemental Datasets for custom categories (110817)
save(hormone.responsiveness6.DF.s,file=file.path("..","..","Nozue2016_SAStranscriptome_data","input","hormone.responsiveness6.DF.s.Rdata"))
# save hormone.responsiveness6.DF.s in csv file format as Table S5
hormone.responsiveness6.DF.s.list<-as.list(hormone.responsiveness6.DF.s)
hormone.responsiveness6.DF.s.list2<-lapply(hormone.responsiveness6.DF.s.list,function(x) hormone.responsiveness6.DF.s.list[["AGI"]][!x==""])
hormone.responsiveness6.DF.s.list2.DF<-plyr::ldply(hormone.responsiveness6.DF.s.list2, rbind) # works
hormone.responsiveness6.DF.s.list2.DF2<-as.data.frame(t(hormone.responsiveness6.DF.s.list2.DF[1:21,])) # eliminate "AGI" row (22)
hormone.responsiveness6.DF.s.list2.DF2[is.na(hormone.responsiveness6.DF.s.list2.DF2)]<-""
hormone.responsiveness6.DF.s.list2.DF2<-hormone.responsiveness6.DF.s.list2.DF2[,-1]
write.csv(hormone.responsiveness6.DF.s.list2.DF2,file=file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Supplemental_Dataset_4_custom_categories.csv")) # remove 1st column and 1st row in excel.
```
# appendix 3  "Should I filter probesets or genes for WGCNA analysis?
*Probesets or genes may be filtered by mean expression or variance (or their robust analogs such as median and median absolute deviation, MAD) since low-expressed or non-varying genes usually represent noise. Whether it is better to filter by mean expression or variance is a matter of debate; both have advantages and disadvantages, but more importantly, they tend to filter out similar sets of genes since mean and variance are usually related.
*We do not recommend filtering genes by differential expression. WGCNA is designed to be an unsupervised analysis method that clusters genes based on their expression profiles. Filtering genes by differential expression will lead to a set of correlated genes that will essentially form a single (or a few highly correlated) modules. It also completely invalidates the scale-free topology assumption, so choosing soft thresholding power by scale-free topology fit will fail. "
### http://labs.genetics.ucla.edu/horvath/CoexpressionNetwork/Rpackages/WGCNA/faq.html
#### it is OK to filter genes by variance (= largeCV genes in my scripts)
# misc
```{r}
# chloroplast localized genes in orange module (112916)
Atgo.CC.orange<-Atgo.list[names(Atgo.list) %in% SAS.expression.vst.s.kazu.largeCV.modules[SAS.expression.vst.s.kazu.largeCV.modules$"1"=="orange","AGI"]] #"GO:0009507" is chloroplast
Atgo.CC.orange.chl<-Atgo.CC.orange[grep("GO:0009507",Atgo.CC.orange)]
Atgo.CC.orange.chl.description<-TAIR10_gene_descriptions[TAIR10_gene_descriptions$AGI %in% names(Atgo.CC.orange.chl),]
write.csv(Atgo.CC.orange.chl.description,file="Atgo.CC.orange.chl.description.csv")
```
# fold chagne (shade responsiveness) of each module (not used; 09**17)
```{r}
load("../Nozue2016_SAStranscriptome_output/output/summary.vst.response.Rdata") # response. needs to fix colnames (060515)
summary.vst.response.kazu<-summary.vst.response[,gsub("([[:digit:]]+)(_)(1|4|16|25|49)(hrA)","\\1",colnames(summary.vst.response)) %in% 1:19]
dim(summary.vst.response.kazu)
head(summary.vst.response.kazu)
q<-expression.pattern.graph2b(summary.vst.response.kazu,"AT2G46970") # PIL1
# XBATs (120716)
XBATs<-c("AT2G28840","AT5G57740","AT5G07270","AT4G14365","AT3G23280") # XBAT31, 32, 33, 34, 35
# plot
setwd("/Volumes/data_work/Data8/XBAT34")
p<-expression.pattern.graph3(summary.vst.kazu,XBATs,plot.order=c("AT2G28840","AT5G57740","AT5G07270","AT4G14365","AT3G23280")) 
p
ggsave(p, file="XBATs_absolute.expression.pdf")
# shade response
q<-expression.pattern.graph2b(summary.vst.response.kazu,XBATs,plot.order=c("AT2G28840","AT5G57740","AT5G07270","AT4G14365","AT3G23280"))
q
ggsave(q, file="XBATs_shaderesponsiveness.pdf")
# salmon module Col absolute value
# salmon
#test<-read.csv("../Nozue2016_SAStranscriptome_output/figs_tables/Dataset7.csv")
DE.genotype.batch.genes.WGCNAmodules.description<-read.csv(file.path("..","..","Nozue2016_SAStranscriptome_output","figs_tables","Supplemental_Dataset4_Custom_category_enrichment_analysis_of_WGCNA_gene_co-expression_modules092817.csv"),row.names=1)
#DE.genotype.batch.genes.WGCNAmodules.description<-test[,-1]
dim(DE.genotype.batch.genes.WGCNAmodules.description) # 17985   28 > [1] 33602    31

salmon<-DE.genotype.batch.genes.WGCNAmodules.description[DE.genotype.batch.genes.WGCNAmodules.description$WGCNAmodules=="salmon",]
salmon$AGI
lightgreen<-DE.genotype.batch.genes.WGCNAmodules.description[DE.genotype.batch.genes.WGCNAmodules.description$WGCNAmodules=="lightgreen",]
lightgreen$AGI
purple<-DE.genotype.batch.genes.WGCNAmodules.description[DE.genotype.batch.genes.WGCNAmodules.description$WGCNAmodules=="purple",]
plum1<-DE.genotype.batch.genes.WGCNAmodules.description[DE.genotype.batch.genes.WGCNAmodules.description$WGCNAmodules=="plum1",]

purple$AGI
# graph
data.cpm<-summary.vst.kazu
target.gene<-lightgreen$AGI
p.Col.lightgreen<-expression.pattern.graph5(summary.vst.kazu[,c("1H1","1H25","1H4","1H49","1L1","1L25","1L4","1L49")],lightgreen$AGI) #lightgreen, Col. 

p.Col.salmon<-expression.pattern.graph5(summary.vst.kazu[,c("1H1","1H25","1H4","1H49","1L1","1L25","1L4","1L49")],salmon$AGI) #lightgreen, Col. 

p.Col.purple<-expression.pattern.graph5(summary.vst.kazu[,c("1H1","1H25","1H4","1H49","1L1","1L25","1L4","1L49")],purple$AGI) #purple, Col

p.Col.purple.b<-expression.pattern.graph5b(summary.vst.kazu[,c("1H1","1H25","1H4","1H49","1L1","1L25","1L4","1L49")],purple$AGI) #purple, Col
p.Col.purple.abs<-expression.pattern.graph5.abs(summary.vst.kazu[,c("1H1","1H25","1H4","1H49","1L1","1L25","1L4","1L49")],purple$AGI) #purple, Col

p.Col.shade.responsive.abs<-expression.pattern.graph5.abs(summary.vst.kazu[,c("1H1","1H25","1H4","1H49","1L1","1L25","1L4","1L49")],c("AT1G04240","AT1G04310","AT1G18400","AT1G27730","AT1G29440","AT1G29450","AT1G29460","AT1G29490","AT1G29500","AT1G29510","AT1G30700","AT1G52830","AT1G65920","AT1G70940","AT2G18010","AT2G22810","AT2G26710","AT2G39030","AT3G02170","AT3G03830","AT3G03840","AT3G15540","AT3G23030","AT3G42800","AT3G58120")) #Col, a part of IAAup. hard to see the pattern

p.Col.shade.responsive.FC<-expression.pattern.graph5.FC(summary.vst.response.kazu,target.genes=plum1$AGI[1:10]) # this is good.

p.Col.plum1<-expression.pattern.graph5(summary.vst.kazu[,c("1H1","1H25","1H4","1H49","1L1","1L25","1L4","1L49")],plum1$AGI) #plum1, Col

p.Col.plum1.b<-expression.pattern.graph5b(summary.vst.kazu[,c("1H1","1H25","1H4","1H49","1L1","1L25","1L4","1L49")],plum1$AGI) #plum1, Col

p.salmon<-expression.pattern.graph5(summary.vst.kazu,salmon$AGI) #lightgreen, Col. 
p.lightgreen<-expression.pattern.graph5(summary.vst.kazu,lightgreen$AGI) #lightgreen, Col. 
p.purple<-expression.pattern.graph5(summary.vst.kazu,purple$AGI) #purple, Col. 
p.plum1<-expression.pattern.graph5(summary.vst.kazu,plum1$AGI) #plum1, Col. 

q<-expression.pattern.graph2b(summary.vst.response.kazu,target.genes=plum1$AGI[1:10])
p.plum1<-expression.pattern.graph5b(summary.vst.kazu,plum1$AGI) #plum1, Col. 

```

# misc
```{r}
# SA related genes are shade-responsive? (071115)
load("../../Nozue2016_SAStranscriptome_output/output/newdata.vst.Rdata") 
newdata.vst$gt2<-""
for(i in 1:27){
  newdata.vst[newdata.vst$gt==conversion.table[i,1],"gt2"]<-as.character(conversion.table[i,2])
}
newdata.vst[newdata.vst$AGI %in% c("AT3G48090","AT4G35580","AT1G74710","AT5G08330","AT3G52430","AT3G26830","AT5G26920","AT2G46970"),c("1h","4h","25h","49h","ssom3.4.unit.classif","gt2","AGI")] 
# NTL9 (AT4G35580), CHE (AT5G08330), ICS1(AT1G74710, SID2), PAD4 (AT3G52430), PAD3 (AT3G26830), CBP60g (AT5G26920), PIL1 (AT2G46970) as a positive control for this script. 

```

# session info
```{r}
sessionInfo()
```
